diff --git a/drivers/net/ethernet/airoha/airoha_eth.c b/drivers/net/ethernet/airoha/airoha_eth.c
index f61bcce6803f..8542e971e658 100644
--- a/drivers/net/ethernet/airoha/airoha_eth.c
+++ b/drivers/net/ethernet/airoha/airoha_eth.c
@@ -266,6 +266,12 @@ static int airoha_fe_set_pse_oq_rsv(struct airoha_eth *eth,
 	airoha_fe_rmw(eth, REG_FE_PSE_BUF_SET, PSE_ALLRSV_MASK,
 		      FIELD_PREP(PSE_ALLRSV_MASK, all_rsv));
 
+	if (airoha_is(eth, en7523)) {
+		airoha_fe_wr(eth, REG_FE_PSE_BUF_SET, 0x2b4);
+		airoha_fe_wr(eth, REG_PSE_SHARE_USED_THD, 0x1e001f4);
+		return 0;
+	}
+
 	/* modify hthd */
 	tmp = airoha_fe_rr(eth, PSE_FQ_CFG);
 	fq_limit = FIELD_GET(PSE_FQ_LIMIT_MASK, tmp);
@@ -310,7 +316,8 @@ static void airoha_fe_pse_ports_init(struct airoha_eth *eth)
 		all_rsv += PSE_RSV_PAGES *
 			   pse_port_num_queues[FE_PSE_PORT_PPE2];
 	}
-	airoha_fe_set(eth, REG_FE_PSE_BUF_SET, all_rsv);
+	if (!airoha_is(eth, en7523))
+		airoha_fe_set(eth, REG_FE_PSE_BUF_SET, all_rsv);
 
 	/* CMD1 */
 	for (q = 0; q < pse_port_num_queues[FE_PSE_PORT_CDM1]; q++)
@@ -486,7 +493,7 @@ static int airoha_fe_init(struct airoha_eth *eth)
 	/* set PSE Page as 128B */
 	airoha_fe_rmw(eth, REG_FE_DMA_GLO_CFG,
 		      FE_DMA_GLO_L2_SPACE_MASK | FE_DMA_GLO_PG_SZ_MASK,
-		      FIELD_PREP(FE_DMA_GLO_L2_SPACE_MASK, 2) |
+		      FIELD_PREP(FE_DMA_GLO_L2_SPACE_MASK, airoha_is(eth, en7523) ? 3 : 2) |
 		      FE_DMA_GLO_PG_SZ_MASK);
 	airoha_fe_wr(eth, REG_FE_RST_GLO_CFG,
 		     FE_RST_CORE_MASK | FE_RST_GDM3_MBI_ARB_MASK |
@@ -530,7 +537,14 @@ static int airoha_fe_init(struct airoha_eth *eth)
 	airoha_fe_crsn_qsel_init(eth);
 
 	airoha_fe_clear(eth, REG_FE_CPORT_CFG, FE_CPORT_QUEUE_XFC_MASK);
-	airoha_fe_set(eth, REG_FE_CPORT_CFG, FE_CPORT_PORT_XFC_MASK);
+	
+	if (!airoha_is(eth, en7523))
+		airoha_fe_set(eth, REG_FE_CPORT_CFG, FE_CPORT_PORT_XFC_MASK);
+	else {	
+		airoha_fe_set(eth, REG_FE_CPORT_CFG, FE_CPORT_PORT_XFC_MASK | FE_CPORT_DIS_FE2GSW_CRC);
+		airoha_fe_rmw(eth, REG_FE_CPORT_CFG, FE_CPORT_FE2SW_IPG,
+			      FIELD_PREP(FE_CPORT_FE2SW_IPG, 2));
+	}
 
 	/* default aging mode for mbi unlock issue */
 	airoha_fe_rmw(eth, REG_GDM_CHN_RLS(2),
@@ -541,6 +555,10 @@ static int airoha_fe_init(struct airoha_eth *eth)
 	/* disable IFC by default */
 	airoha_fe_clear(eth, REG_FE_CSR_IFC_CFG, FE_IFC_EN_MASK);
 
+	/* enable sp_tag generation */
+	if (airoha_is(eth, en7523))
+		airoha_fe_set(eth, GDM1_BASE_STAG_EN, CPORT_TX_STAG_EN | CPORT_RX_STAG_EN | GDM1_RX_LAN_SPORT);
+
 	/* enable 1:N vlan action, init vlan table */
 	airoha_fe_set(eth, REG_MC_VLAN_EN, MC_VLAN_EN_MASK);
 
@@ -591,20 +609,30 @@ static int airoha_qdma_fill_rx_queue(struct airoha_queue *q)
 	return nframes;
 }
 
+#define FIELD_GET_DYN(mask, val) \
+	(((val) & (mask)) >> __ffs(mask))
+
 static int airoha_qdma_get_gdm_port(struct airoha_eth *eth,
 				    struct airoha_qdma_desc *desc)
 {
 	u32 port, sport, msg1 = le32_to_cpu(desc->msg1);
 
-	sport = FIELD_GET(QDMA_ETH_RXMSG_SPORT_MASK, msg1);
+	sport = FIELD_GET_DYN(eth->soc->eth_rx_msg1->sport_mask, msg1);
 	switch (sport) {
 	case 0x18:
+		port = 3; /* GDM4 */
+		if (airoha_is(eth, en7523))
+			port = 2;
+		break;
+	case 0x19:
 		port = 3; /* GDM4 */
 		break;
 	case 0x16:
+	case 0x17:
 		port = 2; /* GDM3 */
 		break;
-	case 0x10 ... 0x14:
+	case 0x0:
+	case 0x8 ... 0x14:
 		port = 0; /* GDM1 */
 		break;
 	case 0x2 ... 0x4:
@@ -656,6 +684,7 @@ static int airoha_qdma_rx_process(struct airoha_queue *q, int budget)
 		port = eth->ports[p];
 		if (!q->skb) { /* first buffer */
 			q->skb = napi_build_skb(e->buf, q->buf_size);
+			mb();
 			if (!q->skb)
 				goto free_frag;
 
@@ -671,7 +700,7 @@ static int airoha_qdma_rx_process(struct airoha_queue *q, int budget)
 
 			if (nr_frags >= ARRAY_SIZE(shinfo->frags))
 				goto free_frag;
-
+			mb();
 			skb_add_rx_frag(q->skb, nr_frags, page,
 					e->buf - page_address(page), len,
 					q->buf_size);
@@ -706,7 +735,9 @@ static int airoha_qdma_rx_process(struct airoha_queue *q, int budget)
 					     false);
 
 		done++;
+		mb();
 		napi_gro_receive(&q->napi, q->skb);
+		mb();
 		q->skb = NULL;
 		continue;
 free_frag:
@@ -738,7 +769,7 @@ static int airoha_qdma_rx_napi_poll(struct napi_struct *napi, int budget)
 		int intr_reg = qid < RX_DONE_HIGH_OFFSET ? QDMA_INT_REG_IDX1
 							 : QDMA_INT_REG_IDX2;
 
-		for (i = 0; i < ARRAY_SIZE(qdma->irq_banks); i++) {
+		for (i = 0; i < AIROHA_MAX_NUM_IRQ_BANKS(qdma->eth->soc); i++) {
 			if (!(BIT(qid) & RX_IRQ_BANK_PIN_MASK(i)))
 				continue;
 
@@ -828,7 +859,7 @@ static int airoha_qdma_init_rx(struct airoha_qdma *qdma)
 {
 	int i;
 
-	for (i = 0; i < ARRAY_SIZE(qdma->q_rx); i++) {
+	for (i = 0; i < AIROHA_NUM_RX_RING(qdma->eth->soc); i++) {
 		int err;
 
 		if (!(RX_DONE_INT_MASK & BIT(i))) {
@@ -880,7 +911,7 @@ static int airoha_qdma_tx_napi_poll(struct napi_struct *napi, int budget)
 		done++;
 
 		qid = FIELD_GET(IRQ_RING_IDX_MASK, val);
-		if (qid >= ARRAY_SIZE(qdma->q_tx))
+		if (qid >= AIROHA_NUM_TX_RING(eth->soc))
 			continue;
 
 		q = &qdma->q_tx[qid];
@@ -979,7 +1010,8 @@ static int airoha_qdma_init_tx_queue(struct airoha_queue *q,
 	}
 
 	/* xmit ring drop default setting */
-	airoha_qdma_set(qdma, REG_TX_RING_BLOCKING(qid),
+	if (!airoha_is(eth, en7523))
+		airoha_qdma_set(qdma, REG_TX_RING_BLOCKING(qid),
 			TX_RING_IRQ_BLOCKING_TX_DROP_EN_MASK);
 
 	airoha_qdma_wr(qdma, REG_TX_RING_BASE(qid), dma_addr);
@@ -1029,7 +1061,7 @@ static int airoha_qdma_init_tx(struct airoha_qdma *qdma)
 			return err;
 	}
 
-	for (i = 0; i < ARRAY_SIZE(qdma->q_tx); i++) {
+	for (i = 0; i < AIROHA_NUM_TX_RING(qdma->eth->soc); i++) {
 		err = airoha_qdma_init_tx_queue(&qdma->q_tx[i], qdma,
 						TX_DSCP_NUM);
 		if (err)
@@ -1204,7 +1236,7 @@ static int airoha_qdma_hw_init(struct airoha_qdma *qdma)
 {
 	int i;
 
-	for (i = 0; i < ARRAY_SIZE(qdma->irq_banks); i++) {
+	for (i = 0; i < AIROHA_MAX_NUM_IRQ_BANKS(qdma->eth->soc); i++) {
 		/* clear pending irqs */
 		airoha_qdma_wr(qdma, REG_INT_STATUS(i), 0xffffffff);
 		/* setup rx irqs */
@@ -1223,8 +1255,17 @@ static int airoha_qdma_hw_init(struct airoha_qdma *qdma)
 	airoha_qdma_irq_enable(&qdma->irq_banks[0], QDMA_INT_REG_IDX4,
 			       TX_COHERENT_HIGH_INT_MASK);
 
+	if (airoha_is(qdma->eth, en7523)) {
+		airoha_qdma_wr(qdma, 0x30, 0x7C000000);
+		airoha_qdma_wr(qdma, 0x34, 0x7C007C00);
+		airoha_qdma_wr(qdma, 0x38, 0x00200000);
+		airoha_qdma_wr(qdma, 0x3C, 0x00200020);
+		airoha_qdma_wr(qdma, 0x40, 0x00000030);
+		airoha_qdma_wr(qdma, 0x6C, 0x00000000);
+	}
+
 	/* setup irq binding */
-	for (i = 0; i < ARRAY_SIZE(qdma->q_tx); i++) {
+	for (i = 0; i < AIROHA_NUM_TX_RING(qdma->eth->soc); i++) {
 		if (!qdma->q_tx[i].ndesc)
 			continue;
 
@@ -1234,6 +1275,15 @@ static int airoha_qdma_hw_init(struct airoha_qdma *qdma)
 		else
 			airoha_qdma_clear(qdma, REG_TX_RING_BLOCKING(i),
 					  TX_RING_IRQ_BLOCKING_CFG_MASK);
+
+		if (airoha_is(qdma->eth, en7523)) {
+			if (i == 0)
+				airoha_qdma_set(qdma, REG_TX_RING_BLOCKING(i),
+						TX_RING_IRQ_BLOCKING_TX_DROP_EN_MASK);
+			else
+				airoha_qdma_clear(qdma, REG_TX_RING_BLOCKING(i),
+						TX_RING_IRQ_BLOCKING_TX_DROP_EN_MASK);
+		}
 	}
 
 	airoha_qdma_wr(qdma, REG_QDMA_GLOBAL_CFG,
@@ -1249,7 +1299,7 @@ static int airoha_qdma_hw_init(struct airoha_qdma *qdma)
 	airoha_qdma_init_qos(qdma);
 
 	/* disable qdma rx delay interrupt */
-	for (i = 0; i < ARRAY_SIZE(qdma->q_rx); i++) {
+	for (i = 0; i < AIROHA_NUM_RX_RING(qdma->eth->soc); i++) {
 		if (!qdma->q_rx[i].ndesc)
 			continue;
 
@@ -1293,7 +1343,7 @@ static irqreturn_t airoha_irq_handler(int irq, void *dev_instance)
 		rx_intr_mask |= (rx_intr2 << 16);
 	}
 
-	for (i = 0; rx_intr_mask && i < ARRAY_SIZE(qdma->q_rx); i++) {
+	for (i = 0; rx_intr_mask && i < AIROHA_NUM_RX_RING(qdma->eth->soc); i++) {
 		if (!qdma->q_rx[i].ndesc)
 			continue;
 
@@ -1320,8 +1370,10 @@ static int airoha_qdma_init_irq_banks(struct platform_device *pdev,
 {
 	struct airoha_eth *eth = qdma->eth;
 	int i, id = qdma - &eth->qdma[0];
+	qdma->irq_banks = devm_kzalloc(&pdev->dev,
+		sizeof(*qdma->irq_banks) * AIROHA_MAX_NUM_IRQ_BANKS(eth->soc), GFP_KERNEL);
 
-	for (i = 0; i < ARRAY_SIZE(qdma->irq_banks); i++) {
+	for (i = 0; i < AIROHA_MAX_NUM_IRQ_BANKS(eth->soc); i++) {
 		struct airoha_irq_bank *irq_bank = &qdma->irq_banks[i];
 		int err, irq_index = 4 * id + i;
 		const char *name;
@@ -1413,6 +1465,12 @@ static int airoha_hw_init(struct platform_device *pdev,
 		return err;
 
 	for (i = 0; i < ARRAY_SIZE(eth->qdma); i++) {
+		eth->qdma[i].q_tx = devm_kzalloc(&pdev->dev,
+			sizeof(*eth->qdma[i].q_tx) * AIROHA_NUM_TX_RING(eth->soc),
+			GFP_KERNEL);
+		eth->qdma[i].q_rx = devm_kzalloc(&pdev->dev,
+			sizeof(*eth->qdma[i].q_rx) * AIROHA_NUM_RX_RING(eth->soc),
+			GFP_KERNEL);
 		err = airoha_qdma_init(pdev, eth, &eth->qdma[i]);
 		if (err)
 			return err;
@@ -1431,7 +1489,7 @@ static void airoha_hw_cleanup(struct airoha_qdma *qdma)
 {
 	int i;
 
-	for (i = 0; i < ARRAY_SIZE(qdma->q_rx); i++) {
+	for (i = 0; i < AIROHA_NUM_RX_RING(qdma->eth->soc); i++) {
 		if (!qdma->q_rx[i].ndesc)
 			continue;
 
@@ -1444,7 +1502,7 @@ static void airoha_hw_cleanup(struct airoha_qdma *qdma)
 	for (i = 0; i < ARRAY_SIZE(qdma->q_tx_irq); i++)
 		netif_napi_del(&qdma->q_tx_irq[i].napi);
 
-	for (i = 0; i < ARRAY_SIZE(qdma->q_tx); i++) {
+	for (i = 0; i < AIROHA_NUM_TX_RING(qdma->eth->soc); i++) {
 		if (!qdma->q_tx[i].ndesc)
 			continue;
 
@@ -1459,7 +1517,7 @@ static void airoha_qdma_start_napi(struct airoha_qdma *qdma)
 	for (i = 0; i < ARRAY_SIZE(qdma->q_tx_irq); i++)
 		napi_enable(&qdma->q_tx_irq[i].napi);
 
-	for (i = 0; i < ARRAY_SIZE(qdma->q_rx); i++) {
+	for (i = 0; i < AIROHA_NUM_RX_RING(qdma->eth->soc); i++) {
 		if (!qdma->q_rx[i].ndesc)
 			continue;
 
@@ -1474,7 +1532,7 @@ static void airoha_qdma_stop_napi(struct airoha_qdma *qdma)
 	for (i = 0; i < ARRAY_SIZE(qdma->q_tx_irq); i++)
 		napi_disable(&qdma->q_tx_irq[i].napi);
 
-	for (i = 0; i < ARRAY_SIZE(qdma->q_rx); i++) {
+	for (i = 0; i < AIROHA_NUM_RX_RING(qdma->eth->soc); i++) {
 		if (!qdma->q_rx[i].ndesc)
 			continue;
 
@@ -1682,7 +1740,7 @@ static int airoha_dev_stop(struct net_device *dev)
 	if (err)
 		return err;
 
-	for (i = 0; i < ARRAY_SIZE(qdma->q_tx); i++)
+	for (i = 0; i < AIROHA_NUM_TX_RING(qdma->eth->soc); i++)
 		netdev_tx_reset_subqueue(dev, i);
 
 	if (atomic_dec_and_test(&qdma->users)) {
@@ -1690,7 +1748,7 @@ static int airoha_dev_stop(struct net_device *dev)
 				  GLOBAL_CFG_TX_DMA_EN_MASK |
 				  GLOBAL_CFG_RX_DMA_EN_MASK);
 
-		for (i = 0; i < ARRAY_SIZE(qdma->q_tx); i++) {
+		for (i = 0; i < AIROHA_NUM_TX_RING(qdma->eth->soc); i++) {
 			if (!qdma->q_tx[i].ndesc)
 				continue;
 
@@ -1738,7 +1796,7 @@ static int airhoha_set_gdm2_loopback(struct airoha_gdm_port *port)
 	airoha_fe_wr(eth, REG_GDM_TXCHN_EN(2), 0xffffffff);
 	airoha_fe_wr(eth, REG_GDM_RXCHN_EN(2), 0xffff);
 
-	chan = port->id == AIROHA_GDM3_IDX ? airoha_is_7581(eth) ? 4 : 3 : 0;
+	chan = port->id == AIROHA_GDM3_IDX ? airoha_is(eth, an7581, en7523) ? 4 : 3 : 0;
 	airoha_fe_rmw(eth, REG_GDM_LPBK_CFG(2),
 		      LPBK_CHAN_MASK | LPBK_MODE_MASK | LPBK_EN_MASK,
 		      FIELD_PREP(LPBK_CHAN_MASK, chan) |
@@ -1754,7 +1812,7 @@ static int airhoha_set_gdm2_loopback(struct airoha_gdm_port *port)
 	airoha_fe_clear(eth, REG_FE_IFC_PORT_EN, BIT(2));
 
 	/* XXX: handle XSI_USB_PORT and XSI_PCE1_PORT */
-	nbq = port->id == AIROHA_GDM3_IDX && airoha_is_7581(eth) ? 4 : 0;
+	nbq = port->id == AIROHA_GDM3_IDX && airoha_is(eth, an7581, en7523) ? 4 : 0;
 	src_port = eth->soc->ops.get_src_port_id(port, nbq);
 	if (src_port < 0)
 		return src_port;
@@ -1768,7 +1826,7 @@ static int airhoha_set_gdm2_loopback(struct airoha_gdm_port *port)
 		      SP_CPORT_MASK(val),
 		      FE_PSE_PORT_CDM2 << __ffs(SP_CPORT_MASK(val)));
 
-	if (port->id != AIROHA_GDM3_IDX && airoha_is_7581(eth))
+	if (port->id != AIROHA_GDM3_IDX && airoha_is(eth, an7581, en7523))
 		airoha_fe_rmw(eth, REG_SRC_PORT_FC_MAP6,
 			      FC_ID_OF_SRC_PORT24_MASK,
 			      FIELD_PREP(FC_ID_OF_SRC_PORT24_MASK, 2));
@@ -1929,14 +1987,18 @@ static int airoha_get_fe_port(struct airoha_gdm_port *port)
 	struct airoha_eth *eth = qdma->eth;
 
 	switch (eth->soc->version) {
+	case 0x7523:
 	case 0x7583:
-		return port->id == AIROHA_GDM3_IDX ? FE_PSE_PORT_GDM3
-						   : port->id;
+		if (port->id == AIROHA_GDM3_IDX)
+			return FE_PSE_PORT_GDM3;
+		break;
 	case 0x7581:
 	default:
-		return port->id == AIROHA_GDM4_IDX ? FE_PSE_PORT_GDM4
-						   : port->id;
+		if (port->id == AIROHA_GDM4_IDX)
+			return FE_PSE_PORT_GDM4;
+		break;
 	}
+	return port->id;
 }
 
 static netdev_tx_t airoha_dev_xmit(struct sk_buff *skb,
@@ -1954,7 +2016,10 @@ static netdev_tx_t airoha_dev_xmit(struct sk_buff *skb,
 	u16 index;
 	u8 fport;
 
-	qid = skb_get_queue_mapping(skb) % ARRAY_SIZE(qdma->q_tx);
+	if (skb->len > 2048)
+		goto error;
+
+	qid = skb_get_queue_mapping(skb) % AIROHA_NUM_TX_RING(qdma->eth->soc);
 	tag = airoha_get_dsa_tag(skb, dev);
 
 	msg0 = FIELD_PREP(QDMA_ETH_TXMSG_CHAN_MASK,
@@ -1962,7 +2027,7 @@ static netdev_tx_t airoha_dev_xmit(struct sk_buff *skb,
 	       FIELD_PREP(QDMA_ETH_TXMSG_QUEUE_MASK,
 			  qid % AIROHA_NUM_QOS_QUEUES) |
 	       FIELD_PREP(QDMA_ETH_TXMSG_SP_TAG_MASK, tag);
-	if (skb->ip_summed == CHECKSUM_PARTIAL)
+	if (!airoha_is(qdma->eth, en7523) && skb->ip_summed == CHECKSUM_PARTIAL)
 		msg0 |= FIELD_PREP(QDMA_ETH_TXMSG_TCO_MASK, 1) |
 			FIELD_PREP(QDMA_ETH_TXMSG_UCO_MASK, 1) |
 			FIELD_PREP(QDMA_ETH_TXMSG_ICO_MASK, 1);
@@ -1977,7 +2042,8 @@ static netdev_tx_t airoha_dev_xmit(struct sk_buff *skb,
 			__be16 csum = cpu_to_be16(skb_shinfo(skb)->gso_size);
 
 			tcp_hdr(skb)->check = (__force __sum16)csum;
-			msg0 |= FIELD_PREP(QDMA_ETH_TXMSG_TSO_MASK, 1);
+			if (!airoha_is(qdma->eth, en7523))
+				msg0 |= FIELD_PREP(QDMA_ETH_TXMSG_TSO_MASK, 1);
 		}
 	}
 
@@ -2031,6 +2097,11 @@ static netdev_tx_t airoha_dev_xmit(struct sk_buff *skb,
 		val = FIELD_PREP(QDMA_DESC_LEN_MASK, len);
 		if (i < nr_frags - 1)
 			val |= FIELD_PREP(QDMA_DESC_MORE_MASK, 1);
+		else if (len < 512)
+			val = FIELD_PREP(QDMA_DESC_LEN_MASK, len+4);
+		else
+			val = FIELD_PREP(QDMA_DESC_LEN_MASK, len+4);
+
 		WRITE_ONCE(desc->ctrl, cpu_to_le32(val));
 		WRITE_ONCE(desc->addr, cpu_to_le32(addr));
 		val = FIELD_PREP(QDMA_DESC_NEXT_ID_MASK, index);
@@ -2153,7 +2224,7 @@ static int airoha_qdma_set_chan_tx_sched(struct airoha_gdm_port *port,
 {
 	int i;
 
-	for (i = 0; i < AIROHA_NUM_TX_RING; i++)
+	for (i = 0; i < AIROHA_NUM_TX_RING(port->soc); i++)
 		airoha_qdma_clear(port->qdma, REG_QUEUE_CLOSE_CFG(channel),
 				  TXQ_DISABLE_CHAN_QUEUE_MASK(channel, i));
 
@@ -2557,7 +2628,7 @@ static int airoha_tc_htb_alloc_leaf_queue(struct airoha_gdm_port *port,
 	}
 
 	set_bit(channel, port->qos_sq_bmap);
-	opt->qid = AIROHA_NUM_TX_RING + channel;
+	opt->qid = AIROHA_NUM_TX_RING(port->soc) + channel;
 
 	return 0;
 }
@@ -2569,7 +2640,7 @@ static int airoha_qdma_set_rx_meter(struct airoha_gdm_port *port,
 	struct airoha_qdma *qdma = port->qdma;
 	int i;
 
-	for (i = 0; i < ARRAY_SIZE(qdma->q_rx); i++) {
+	for (i = 0; i < AIROHA_NUM_RX_RING(qdma->eth->soc); i++) {
 		int err;
 
 		if (!qdma->q_rx[i].ndesc)
@@ -2781,7 +2852,7 @@ static int airoha_tc_get_htb_get_leaf_queue(struct airoha_gdm_port *port,
 		return -EINVAL;
 	}
 
-	opt->qid = AIROHA_NUM_TX_RING + channel;
+	opt->qid = AIROHA_NUM_TX_RING(port->soc) + channel;
 
 	return 0;
 }
@@ -3031,8 +3102,8 @@ static int airoha_alloc_gdm_port(struct airoha_eth *eth,
 	}
 
 	dev = devm_alloc_etherdev_mqs(eth->dev, sizeof(*port),
-				      AIROHA_NUM_NETDEV_TX_RINGS,
-				      AIROHA_NUM_RX_RING);
+				      AIROHA_NUM_NETDEV_TX_RINGS(eth->soc),
+				      AIROHA_NUM_RX_RING(eth->soc));
 	if (!dev) {
 		dev_err(eth->dev, "alloc_etherdev failed\n");
 		return -ENOMEM;
@@ -3043,9 +3114,9 @@ static int airoha_alloc_gdm_port(struct airoha_eth *eth,
 	dev->ethtool_ops = &airoha_ethtool_ops;
 	dev->max_mtu = AIROHA_MAX_MTU;
 	dev->watchdog_timeo = 5 * HZ;
-	dev->hw_features = NETIF_F_IP_CSUM | NETIF_F_RXCSUM |
+	dev->hw_features = /* NETIF_F_IP_CSUM  |*/ NETIF_F_RXCSUM |
 			   NETIF_F_TSO6 | NETIF_F_IPV6_CSUM |
-			   NETIF_F_SG | NETIF_F_TSO |
+			   NETIF_F_SG | /*NETIF_F_TSO |*/
 			   NETIF_F_HW_TC;
 	dev->features |= dev->hw_features;
 	dev->vlan_features = dev->hw_features;
@@ -3054,7 +3125,7 @@ static int airoha_alloc_gdm_port(struct airoha_eth *eth,
 	SET_NETDEV_DEV(dev, eth->dev);
 
 	/* reserve hw queues for HTB offloading */
-	err = netif_set_real_num_tx_queues(dev, AIROHA_NUM_TX_RING);
+	err = netif_set_real_num_tx_queues(dev, AIROHA_NUM_TX_RING(eth->soc));
 	if (err)
 		return err;
 
@@ -3256,6 +3327,18 @@ static const char * const en7581_xsi_rsts_names[] = {
 	"xfp-mac",
 };
 
+struct airoha_qdma_eth_rx_msg1 default_qdma_eth_rxmsg = {
+	.dei_mask = BIT(31),
+	.ip6_mask = BIT(30),
+	.ip4_mask = BIT(29),
+	.ip4f_mask = BIT(28),
+	.l4_valid_mask = BIT(27),
+	.l4f_mask = BIT(26),
+	.sport_mask = GENMASK(25, 21),
+	.crsn_mask = GENMASK(20, 16),
+	.ppe_entry_mask = GENMASK(15, 0),
+};
+
 static int airoha_en7581_get_src_port_id(struct airoha_gdm_port *port, int nbq)
 {
 	switch (port->id) {
@@ -3308,29 +3391,96 @@ static int airoha_an7583_get_src_port_id(struct airoha_gdm_port *port, int nbq)
 	return -EINVAL;
 }
 
+static const char * const en7523_xsi_rsts_names[] = {
+	"xsi-mac",
+	"hsi0-mac",
+	"hsi1-mac",
+	"hsi-mac",
+};
+
+static struct airoha_qdma_eth_rx_msg1 en7523_qdma_eth_rxmsg = {
+	.dei_mask = BIT(30),
+	.ip6_mask = BIT(29),
+	.ip4_mask = BIT(28),
+	.ip4f_mask = BIT(27),
+	.l4_valid_mask = BIT(26),
+	.l4f_mask = BIT(25),
+	.sport_mask = GENMASK(24, 20),
+	.crsn_mask = GENMASK(19, 16),
+	.ppe_entry_mask = GENMASK(15, 0),
+};
+
+static int airoha_en7523_get_src_port_id(struct airoha_gdm_port *port, int nbq)
+{
+	switch (port->id) {
+	case 2:
+		if (!nbq)
+			return HSGMII_LAN_7523_AEWAN_SRCPORT;
+		break;
+	case 3:
+		if (nbq == 4)
+			return HSGMII_LAN_7523_PCIE0_SRCPORT;
+		if (nbq == 5)
+			return HSGMII_LAN_7523_PCIE1_SRCPORT;
+		if (nbq == 1)
+			return HSGMII_LAN_7523_USB_SRCPORT;
+		break;
+	default:
+		break;
+	}
+
+	return -EINVAL;
+}
+
 static const struct airoha_eth_soc_data en7581_soc_data = {
-	.version = 0x7581,
+	.version = an7581,
 	.xsi_rsts_names = en7581_xsi_rsts_names,
 	.num_xsi_rsts = ARRAY_SIZE(en7581_xsi_rsts_names),
+	.rx_ring = 32,
+	.tx_ring = 32,
 	.num_ppe = 2,
+	.irq_banks = 4,
+	.ppe_sram_entries = (4 * 1024),
+	.eth_rx_msg1 = &default_qdma_eth_rxmsg,
 	.ops = {
 		.get_src_port_id = airoha_en7581_get_src_port_id,
 	},
 };
 
 static const struct airoha_eth_soc_data an7583_soc_data = {
-	.version = 0x7583,
+	.version = an7583,
 	.xsi_rsts_names = an7583_xsi_rsts_names,
 	.num_xsi_rsts = ARRAY_SIZE(an7583_xsi_rsts_names),
+	.rx_ring = 32,
+	.tx_ring = 32,
 	.num_ppe = 1,
+	.irq_banks = 4,
+	.ppe_sram_entries = (4 * 1024),
+	.eth_rx_msg1 = &default_qdma_eth_rxmsg,
 	.ops = {
 		.get_src_port_id = airoha_an7583_get_src_port_id,
 	},
 };
 
+static const struct airoha_eth_soc_data en7523_soc_data = {
+	.version = en7523,
+	.xsi_rsts_names = en7523_xsi_rsts_names,
+	.num_xsi_rsts = ARRAY_SIZE(en7523_xsi_rsts_names),
+	.rx_ring = 16,
+	.tx_ring = 8,
+	.num_ppe = 1,
+	.irq_banks = 2,
+	.ppe_sram_entries = (16 * 1024),
+	.eth_rx_msg1 = &en7523_qdma_eth_rxmsg,
+	.ops = {
+		.get_src_port_id = airoha_en7523_get_src_port_id,
+	},
+};
+
 static const struct of_device_id of_airoha_match[] = {
 	{ .compatible = "airoha,en7581-eth", .data = &en7581_soc_data },
 	{ .compatible = "airoha,an7583-eth", .data = &an7583_soc_data },
+	{ .compatible = "airoha,en7523-eth", .data = &en7523_soc_data },
 	{ /* sentinel */ }
 };
 MODULE_DEVICE_TABLE(of, of_airoha_match);
diff --git a/drivers/net/ethernet/airoha/airoha_eth.h b/drivers/net/ethernet/airoha/airoha_eth.h
index 2514216c3468..a70ad1ae2187 100644
--- a/drivers/net/ethernet/airoha/airoha_eth.h
+++ b/drivers/net/ethernet/airoha/airoha_eth.h
@@ -18,16 +18,16 @@
 
 #define AIROHA_MAX_NUM_GDM_PORTS	4
 #define AIROHA_MAX_NUM_QDMA		2
-#define AIROHA_MAX_NUM_IRQ_BANKS	4
+#define AIROHA_MAX_NUM_IRQ_BANKS(_soc)	(_soc->irq_banks)
 #define AIROHA_MAX_DSA_PORTS		7
 #define AIROHA_MAX_NUM_RSTS		3
 #define AIROHA_MAX_MTU			9216
 #define AIROHA_MAX_PACKET_SIZE		2048
 #define AIROHA_NUM_QOS_CHANNELS		4
 #define AIROHA_NUM_QOS_QUEUES		8
-#define AIROHA_NUM_TX_RING		32
-#define AIROHA_NUM_RX_RING		32
-#define AIROHA_NUM_NETDEV_TX_RINGS	(AIROHA_NUM_TX_RING + \
+#define AIROHA_NUM_TX_RING(_soc)	_soc->tx_ring
+#define AIROHA_NUM_RX_RING(_soc)	_soc->rx_ring
+#define AIROHA_NUM_NETDEV_TX_RINGS(_soc) (AIROHA_NUM_TX_RING(_soc) + \
 					 AIROHA_NUM_QOS_CHANNELS)
 #define AIROHA_FE_MC_MAX_VLAN_TABLE	64
 #define AIROHA_FE_MC_MAX_VLAN_PORT	16
@@ -37,6 +37,8 @@
 #define TX_DSCP_NUM			1024
 #define RX_DSCP_NUM(_n)			\
 	((_n) ==  2 ? 128 :		\
+	 (_n) ==  1 ? 256 :		\
+	 (_n) ==  5 ? 256 :		\
 	 (_n) == 11 ? 128 :		\
 	 (_n) == 15 ? 128 :		\
 	 (_n) ==  0 ? 1024 : 16)
@@ -47,7 +49,6 @@
 #define QDMA_METER_IDX(_n)		((_n) & 0xff)
 #define QDMA_METER_GROUP(_n)		(((_n) >> 8) & 0x3)
 
-#define PPE_SRAM_NUM_ENTRIES		(8 * 1024)
 #define PPE_STATS_NUM_ENTRIES		(4 * 1024)
 #define PPE_DRAM_NUM_ENTRIES		(16 * 1024)
 #define PPE_ENTRY_SIZE			80
@@ -63,6 +64,8 @@ enum {
 	QDMA_INT_REG_IDX2,
 	QDMA_INT_REG_IDX3,
 	QDMA_INT_REG_IDX4,
+	QDMA_INT_REG_IDX5,
+	QDMA_INT_REG_IDX6,
 	QDMA_INT_REG_MAX
 };
 
@@ -79,6 +82,13 @@ enum {
 	HSGMII_LAN_7583_USB_SRCPORT,
 };
 
+enum {
+	HSGMII_LAN_7523_PCIE0_SRCPORT	= 0x16,
+	HSGMII_LAN_7523_PCIE1_SRCPORT,
+	HSGMII_LAN_7523_USB_SRCPORT,
+	HSGMII_LAN_7523_AEWAN_SRCPORT	= 0xffff,
+};
+
 enum {
 	XSI_PCIE0_VIP_PORT_MASK	= BIT(22),
 	XSI_PCIE1_VIP_PORT_MASK	= BIT(23),
@@ -523,17 +533,18 @@ struct airoha_qdma {
 
 	atomic_t users;
 
-	struct airoha_irq_bank irq_banks[AIROHA_MAX_NUM_IRQ_BANKS];
+	struct airoha_irq_bank *irq_banks;
 
 	struct airoha_tx_irq_queue q_tx_irq[AIROHA_NUM_TX_IRQ];
 
-	struct airoha_queue q_tx[AIROHA_NUM_TX_RING];
-	struct airoha_queue q_rx[AIROHA_NUM_RX_RING];
+	struct airoha_queue *q_tx;
+	struct airoha_queue *q_rx;
 };
 
 struct airoha_gdm_port {
 	struct airoha_qdma *qdma;
 	struct net_device *dev;
+	struct airoha_eth_soc_data *soc;
 	int id;
 
 #if defined(CONFIG_PCS_AIROHA)
@@ -577,8 +588,12 @@ struct airoha_ppe {
 struct airoha_eth_soc_data {
 	u16 version;
 	const char * const *xsi_rsts_names;
+	int tx_ring, rx_ring;
 	int num_xsi_rsts;
 	int num_ppe;
+	int irq_banks;
+	int ppe_sram_entries;
+	struct airoha_qdma_eth_rx_msg1 *eth_rx_msg1;
 	struct {
 		int (*get_src_port_id)(struct airoha_gdm_port *port, int nbq);
 	} ops;
@@ -641,16 +656,21 @@ static inline bool airhoa_is_lan_gdm_port(struct airoha_gdm_port *port)
 	return port->id == 1;
 }
 
-static inline bool airoha_is_7581(struct airoha_eth *eth)
-{
-	return eth->soc->version == 0x7581;
-}
+enum airoha_ids {
+	an7581 = 0x7581,
+	an7583 = 0x7583,
+	en7523 = 0x7523,
+};
 
-static inline bool airoha_is_7583(struct airoha_eth *eth)
-{
-	return eth->soc->version == 0x7583;
+static inline bool _airoha_is(struct airoha_eth *eth, const enum airoha_ids ids[]) {
+	for (int i = 0; ids[i] != 0; i++)
+		if (ids[i] == eth->soc->version)
+			return true;
+	return false;
 }
 
+#define airoha_is(eth, ...) _airoha_is(eth, (const enum airoha_ids[]){__VA_ARGS__, 0})
+
 bool airoha_is_valid_gdm_port(struct airoha_eth *eth,
 			      struct airoha_gdm_port *port);
 
diff --git a/drivers/net/ethernet/airoha/airoha_npu.c b/drivers/net/ethernet/airoha/airoha_npu.c
index c163e5cbee86..ddec065a8144 100644
--- a/drivers/net/ethernet/airoha/airoha_npu.c
+++ b/drivers/net/ethernet/airoha/airoha_npu.c
@@ -14,6 +14,8 @@
 
 #include "airoha_eth.h"
 
+#define NPU_EN7523_FIRMWARE_DATA		"airoha/en7523_npu_data.bin"
+#define NPU_EN7523_FIRMWARE_RV32		"airoha/en7523_npu_rv32.bin"
 #define NPU_EN7581_FIRMWARE_DATA		"airoha/en7581_npu_data.bin"
 #define NPU_EN7581_FIRMWARE_RV32		"airoha/en7581_npu_rv32.bin"
 #define NPU_AN7583_FIRMWARE_DATA		"airoha/an7583_npu_data.bin"
@@ -24,14 +26,10 @@
 
 #define REG_NPU_LOCAL_SRAM		0x0
 
-#define NPU_PC_BASE_ADDR		0x305000
-#define REG_PC_DBG(_n)			(0x305000 + ((_n) * 0x100))
-
-#define NPU_CLUSTER_BASE_ADDR		0x306000
-
-#define REG_CR_BOOT_TRIGGER		(NPU_CLUSTER_BASE_ADDR + 0x000)
-#define REG_CR_BOOT_CONFIG		(NPU_CLUSTER_BASE_ADDR + 0x004)
-#define REG_CR_BOOT_BASE(_n)		(NPU_CLUSTER_BASE_ADDR + 0x020 + ((_n) << 2))
+#define REG_PC_DBG(_soc, _n)		((_soc->pc_base_addr) + ((_n) * 0x100))
+#define REG_CR_BOOT_TRIGGER(_soc)	((_soc->cluster_base_addr) + 0x000)
+#define REG_CR_BOOT_CONFIG(_soc)	((_soc->cluster_base_addr) + 0x004)
+#define REG_CR_BOOT_BASE(_soc, _n)	((_soc->cluster_base_addr) + 0x020 + ((_n) << 2))
 
 #define NPU_MBOX_BASE_ADDR		0x30c000
 
@@ -111,6 +109,11 @@ struct airoha_npu_fw {
 };
 
 struct airoha_npu_soc_data {
+	bool use_memremap;
+	bool extra_cores;
+	int num_cores;
+	int cluster_base_addr;
+	int pc_base_addr;
 	struct airoha_npu_fw fw_rv32;
 	struct airoha_npu_fw fw_data;
 };
@@ -229,7 +232,11 @@ static int airoha_npu_run_firmware(struct device *dev, void __iomem *base,
 	if (!soc)
 		return -EINVAL;
 
-	addr = devm_ioremap(dev, rmem->base, rmem->size);
+	if (soc->use_memremap)
+	  addr = devm_memremap(dev, rmem->base, rmem->size, MEMREMAP_WB);
+	else
+	  addr = devm_ioremap(dev, rmem->base, rmem->size);
+
 	if (IS_ERR(addr))
 		return PTR_ERR(addr);
 
@@ -274,7 +281,7 @@ static void airoha_npu_wdt_work(struct work_struct *work)
 		return;
 
 	c = core - &npu->cores[0];
-	regmap_bulk_read(npu->regmap, REG_PC_DBG(c), val, ARRAY_SIZE(val));
+	regmap_bulk_read(npu->regmap, REG_PC_DBG(npu->soc_data, c), val, ARRAY_SIZE(val));
 	snprintf(dump, NPU_DUMP_SIZE, "PC: %08x SP: %08x LR: %08x\n",
 		 val[0], val[1], val[2]);
 
@@ -613,7 +620,45 @@ void airoha_npu_put(struct airoha_npu *npu)
 }
 EXPORT_SYMBOL_GPL(airoha_npu_put);
 
+static void airoha_npu_restart_core(struct airoha_npu *npu, int core) {
+	u32 val;
+	if (core == 0) {
+		dev_info(npu->dev, "enabling NPU %d core\n", core);
+		regmap_write(npu->regmap, REG_CR_BOOT_CONFIG(npu->soc_data), 0x1);
+		regmap_write(npu->regmap, REG_CR_BOOT_TRIGGER(npu->soc_data), 0x1);
+	} else {
+		regmap_read(npu->regmap, REG_CR_BOOT_CONFIG(npu->soc_data), &val);
+		val |= (0x1 << core);    // add new enable bit for CoreX
+		val |= (0x0100 << core); // add reboot bit for CoreX
+		dev_info(npu->dev, "rebooting NPU core %d with val 0x%x\n", core, val);
+		regmap_write(npu->regmap, REG_CR_BOOT_CONFIG(npu->soc_data), val);
+		regmap_write(npu->regmap, REG_CR_BOOT_TRIGGER(npu->soc_data), 0x2);
+	}
+	msleep(100);
+}
+
+static const struct airoha_npu_soc_data en7523_npu_soc_data = {
+	.use_memremap = true,
+	.extra_cores = false,
+	.num_cores = 4,
+	.cluster_base_addr = 0x308000,
+	.pc_base_addr = 0x308800,
+	.fw_rv32 = {
+		.name = NPU_EN7523_FIRMWARE_RV32,
+		.max_size = NPU_EN7581_FIRMWARE_RV32_MAX_SIZE,
+	},
+	.fw_data = {
+		.name = NPU_EN7523_FIRMWARE_DATA,
+		.max_size = NPU_EN7581_FIRMWARE_DATA_MAX_SIZE,
+	},
+};
+
 static const struct airoha_npu_soc_data en7581_npu_soc_data = {
+	.use_memremap = false,
+	.extra_cores = true,
+	.num_cores = 8,
+	.cluster_base_addr = 0x306000,
+	.pc_base_addr = 0x305000,
 	.fw_rv32 = {
 		.name = NPU_EN7581_FIRMWARE_RV32,
 		.max_size = NPU_EN7581_FIRMWARE_RV32_MAX_SIZE,
@@ -625,6 +670,11 @@ static const struct airoha_npu_soc_data en7581_npu_soc_data = {
 };
 
 static const struct airoha_npu_soc_data an7583_npu_soc_data = {
+	.use_memremap = false,
+	.extra_cores = true,
+	.num_cores = 8,
+	.cluster_base_addr = 0x306000,
+	.pc_base_addr = 0x305000,
 	.fw_rv32 = {
 		.name = NPU_AN7583_FIRMWARE_RV32,
 		.max_size = NPU_EN7581_FIRMWARE_RV32_MAX_SIZE,
@@ -636,6 +686,7 @@ static const struct airoha_npu_soc_data an7583_npu_soc_data = {
 };
 
 static const struct of_device_id of_airoha_npu_match[] = {
+	{ .compatible = "airoha,en7523-npu", .data = &en7523_npu_soc_data },
 	{ .compatible = "airoha,en7581-npu", .data = &en7581_npu_soc_data },
 	{ .compatible = "airoha,an7583-npu", .data = &an7583_npu_soc_data },
 	{ /* sentinel */ }
@@ -656,7 +707,7 @@ static int airoha_npu_probe(struct platform_device *pdev)
 	struct reserved_mem *rmem;
 	struct airoha_npu *npu;
 	struct device_node *np;
-	void __iomem *base;
+	void __iomem *base, *sram_base;
 	int i, irq, err;
 
 	base = devm_platform_ioremap_resource(pdev, 0);
@@ -667,6 +718,10 @@ static int airoha_npu_probe(struct platform_device *pdev)
 	if (!npu)
 		return -ENOMEM;
 
+	npu->soc_data = (struct airoha_npu_soc_data*)of_device_get_match_data(dev);
+	if (!npu->soc_data)
+		return -ENODEV;
+
 	npu->dev = dev;
 	npu->ops.ppe_init = airoha_npu_ppe_init;
 	npu->ops.ppe_deinit = airoha_npu_ppe_deinit;
@@ -682,6 +737,13 @@ static int airoha_npu_probe(struct platform_device *pdev)
 	npu->ops.wlan_enable_irq = airoha_npu_wlan_irq_enable;
 	npu->ops.wlan_disable_irq = airoha_npu_wlan_irq_disable;
 
+	/* default SRAM: 256K or 384K */
+	npu->sram_size = 0x40000;
+	sram_base = platform_get_resource(pdev, IORESOURCE_MEM, 1);
+	if (!!sram_base)
+		npu->sram_size = resource_size(sram_base);
+	dev_info(dev, "SRAM Size: 0x%02x\n", npu->sram_size);
+
 	npu->regmap = devm_regmap_init_mmio(dev, base, &regmap_config);
 	if (IS_ERR(npu->regmap))
 		return PTR_ERR(npu->regmap);
@@ -705,27 +767,37 @@ static int airoha_npu_probe(struct platform_device *pdev)
 	if (err)
 		return err;
 
-	for (i = 0; i < ARRAY_SIZE(npu->cores); i++) {
+	npu->cores = devm_kzalloc(&pdev->dev,
+				npu->soc_data->num_cores * sizeof(*npu->cores),
+				GFP_KERNEL);
+	if (!npu->cores)
+		return -ENOMEM;
+
+	for (i = 0; i < npu->soc_data->num_cores; i++) {
 		struct airoha_npu_core *core = &npu->cores[i];
 
 		spin_lock_init(&core->lock);
 		core->npu = npu;
 
-		irq = platform_get_irq(pdev, i + 1);
-		if (irq < 0)
-			return irq;
+		if (npu->soc_data->extra_cores) {
+			irq = platform_get_irq(pdev, i + 1);
+			if (irq < 0)
+				return irq;
+
+			err = devm_request_irq(dev, irq, airoha_npu_wdt_handler,
+					       IRQF_SHARED, "airoha-npu-wdt", core);
+			if (err)
+				return err;
+		}
 
-		err = devm_request_irq(dev, irq, airoha_npu_wdt_handler,
-				       IRQF_SHARED, "airoha-npu-wdt", core);
-		if (err)
-			return err;
 
 		INIT_WORK(&core->wdt_work, airoha_npu_wdt_work);
 	}
 
 	/* wlan IRQ lines */
 	for (i = 0; i < ARRAY_SIZE(npu->irqs); i++) {
-		irq = platform_get_irq(pdev, i + ARRAY_SIZE(npu->cores) + 1);
+		irq = platform_get_irq(pdev, i + 1 +
+			(npu->soc_data->extra_cores ? npu->soc_data->num_cores : 0));
 		if (irq < 0)
 			return irq;
 
@@ -741,21 +813,17 @@ static int airoha_npu_probe(struct platform_device *pdev)
 		return dev_err_probe(dev, err, "failed to run npu firmware\n");
 
 	regmap_write(npu->regmap, REG_CR_NPU_MIB(10),
-		     rmem->base + NPU_EN7581_FIRMWARE_RV32_MAX_SIZE);
-	regmap_write(npu->regmap, REG_CR_NPU_MIB(11), 0x40000); /* SRAM 256K */
+		     rmem->base + npu->soc_data->fw_rv32.max_size);
+	regmap_write(npu->regmap, REG_CR_NPU_MIB(11), npu->sram_size); /* SRAM 256K */
 	regmap_write(npu->regmap, REG_CR_NPU_MIB(12), 0);
 	regmap_write(npu->regmap, REG_CR_NPU_MIB(21), 1);
 	msleep(100);
 
-	/* setting booting address */
-	for (i = 0; i < NPU_NUM_CORES; i++)
-		regmap_write(npu->regmap, REG_CR_BOOT_BASE(i), rmem->base);
-	usleep_range(1000, 2000);
-
-	/* enable NPU cores */
-	regmap_write(npu->regmap, REG_CR_BOOT_CONFIG, 0xff);
-	regmap_write(npu->regmap, REG_CR_BOOT_TRIGGER, 0x1);
-	msleep(100);
+	/* setting booting address and enable NPU cores */
+	for (i = 0; i < npu->soc_data->num_cores; i++) {
+		regmap_write(npu->regmap, REG_CR_BOOT_BASE(npu->soc_data, i), rmem->base);
+		airoha_npu_restart_core(npu, i);
+	}
 
 	platform_set_drvdata(pdev, npu);
 
@@ -767,7 +835,7 @@ static void airoha_npu_remove(struct platform_device *pdev)
 	struct airoha_npu *npu = platform_get_drvdata(pdev);
 	int i;
 
-	for (i = 0; i < ARRAY_SIZE(npu->cores); i++)
+	for (i = 0; i < npu->soc_data->num_cores; i++)
 		cancel_work_sync(&npu->cores[i].wdt_work);
 }
 
@@ -781,6 +849,8 @@ static struct platform_driver airoha_npu_driver = {
 };
 module_platform_driver(airoha_npu_driver);
 
+MODULE_FIRMWARE(NPU_EN7523_FIRMWARE_DATA);
+MODULE_FIRMWARE(NPU_EN7523_FIRMWARE_RV32);
 MODULE_FIRMWARE(NPU_EN7581_FIRMWARE_DATA);
 MODULE_FIRMWARE(NPU_EN7581_FIRMWARE_RV32);
 MODULE_FIRMWARE(NPU_AN7583_FIRMWARE_DATA);
diff --git a/drivers/net/ethernet/airoha/airoha_ppe.c b/drivers/net/ethernet/airoha/airoha_ppe.c
index 0caabb0c3aa0..64635900f8ad 100644
--- a/drivers/net/ethernet/airoha/airoha_ppe.c
+++ b/drivers/net/ethernet/airoha/airoha_ppe.c
@@ -37,7 +37,7 @@ static int airoha_ppe_get_num_stats_entries(struct airoha_ppe *ppe)
 	if (!IS_ENABLED(CONFIG_NET_AIROHA_FLOW_STATS))
 		return -EOPNOTSUPP;
 
-	if (airoha_is_7583(ppe->eth))
+	if (airoha_is(ppe->eth, an7583))
 		return -EOPNOTSUPP;
 
 	return PPE_STATS_NUM_ENTRIES;
@@ -60,7 +60,7 @@ static u32 airoha_ppe_get_total_sram_num_entries(struct airoha_ppe *ppe)
 {
 	struct airoha_eth *eth = ppe->eth;
 
-	return PPE_SRAM_NUM_ENTRIES * eth->soc->num_ppe;
+	return ppe->eth->soc->ppe_sram_entries * eth->soc->num_ppe;
 }
 
 u32 airoha_ppe_get_total_num_entries(struct airoha_ppe *ppe)
@@ -87,7 +87,7 @@ static u32 airoha_ppe_get_timestamp(struct airoha_ppe *ppe)
 
 static void airoha_ppe_hw_init(struct airoha_ppe *ppe)
 {
-	u32 sram_ppe_num_data_entries = PPE_SRAM_NUM_ENTRIES, sram_num_entries;
+	u32 sram_ppe_num_data_entries = ppe->eth->soc->ppe_sram_entries, sram_num_entries;
 	u32 sram_tb_size, dram_num_entries;
 	struct airoha_eth *eth = ppe->eth;
 	int i, sram_num_stats_entries;
@@ -608,7 +608,7 @@ airoha_ppe_foe_get_entry_locked(struct airoha_ppe *ppe, u32 hash)
 
 	if (hash < sram_num_entries) {
 		u32 *hwe = ppe->foe + hash * sizeof(struct airoha_foe_entry);
-		bool ppe2 = hash >= PPE_SRAM_NUM_ENTRIES;
+		bool ppe2 = hash >= ppe->eth->soc->ppe_sram_entries;
 		struct airoha_eth *eth = ppe->eth;
 		u32 val;
 		int i;
@@ -663,7 +663,7 @@ static bool airoha_ppe_foe_compare_entry(struct airoha_flow_table_entry *e,
 static int airoha_ppe_foe_commit_sram_entry(struct airoha_ppe *ppe, u32 hash)
 {
 	struct airoha_foe_entry *hwe = ppe->foe + hash * sizeof(*hwe);
-	bool ppe2 = hash >= PPE_SRAM_NUM_ENTRIES;
+	bool ppe2 = hash >= ppe->eth->soc->ppe_sram_entries;
 	u32 *ptr = (u32 *)hwe, val;
 	int i;
 
diff --git a/drivers/net/ethernet/airoha/airoha_regs.h b/drivers/net/ethernet/airoha/airoha_regs.h
index 9a05432b353a..283676c28614 100644
--- a/drivers/net/ethernet/airoha/airoha_regs.h
+++ b/drivers/net/ethernet/airoha/airoha_regs.h
@@ -138,6 +138,11 @@
 #define GDM_INGRESS_FC_EN_MASK		BIT(1)
 #define GDM_STAG_EN_MASK		BIT(0)
 
+#define GDM1_BASE_STAG_EN		(GDM1_BASE + 0x10)
+#define CPORT_TX_STAG_EN		BIT(2)
+#define CPORT_RX_STAG_EN		BIT(1)
+#define GDM1_RX_LAN_SPORT		BIT(0)
+
 #define REG_GDM_LEN_CFG(_n)		(GDM_BASE(_n) + 0x14)
 #define GDM_SHORT_LEN_MASK		GENMASK(13, 0)
 #define GDM_LONG_LEN_MASK		GENMASK(29, 16)
@@ -160,9 +165,13 @@
 #define REG_GDM_RXCHN_EN(_n)		(GDM_BASE(_n) + 0x28)
 
 #define REG_FE_CPORT_CFG		(GDM1_BASE + 0x40)
+#define FE_CPORT_DIS_FE2GSW_CRC		BIT(31)
+#define FE_CPORT_DIS_GSW2FE_CRC		BIT(30)
 #define FE_CPORT_PAD			BIT(26)
 #define FE_CPORT_PORT_XFC_MASK		BIT(25)
 #define FE_CPORT_QUEUE_XFC_MASK		BIT(24)
+#define FE_CPORT_FE2SW_IPG		GENMASK(15, 8)
+#define FE_CPORT_SW2FE_IPG		GENMASK(7, 0)
 
 #define REG_FE_GDM_MIB_CLEAR(_n)	(GDM_BASE(_n) + 0xf0)
 #define FE_GDM_MIB_RX_CLEAR_MASK	BIT(1)
@@ -436,7 +445,9 @@
 	 ((_n) == 1) ? 0x0024 : 0x0020)
 
 #define REG_INT_ENABLE(_b, _n)		\
-	(((_n) == 4) ? 0x0750 + ((_b) << 5) :	\
+	(((_n) == 6) ? 0x0034 + ((_b) << 5) :	\
+	 ((_n) == 5) ? 0x0030 + ((_b) << 5) :	\
+	 ((_n) == 4) ? 0x0750 + ((_b) << 5) :	\
 	 ((_n) == 3) ? 0x0744 + ((_b) << 5) :	\
 	 ((_n) == 2) ? 0x0740 + ((_b) << 5) :	\
 	 ((_n) == 1) ? 0x002c + ((_b) << 3) :	\
@@ -475,6 +486,13 @@
 #define IRQ0_FULL_INT_MASK		BIT(1)
 #define IRQ0_INT_MASK			BIT(0)
 
+/* EN7523 QDMA_CSR_INT_ENABLE5 30 */
+#define EN7523_RX_COHERENT_LOW_INT_MASK				\
+	(RX14_COHERENT_INT_MASK |	\
+	 RX13_COHERENT_INT_MASK | RX12_COHERENT_INT_MASK |	\
+	 RX11_COHERENT_INT_MASK | RX10_COHERENT_INT_MASK)	
+
+
 #define RX_COHERENT_LOW_INT_MASK				\
 	(RX15_COHERENT_INT_MASK | RX14_COHERENT_INT_MASK |	\
 	 RX13_COHERENT_INT_MASK | RX12_COHERENT_INT_MASK |	\
@@ -496,11 +514,11 @@
 	 TX1_COHERENT_INT_MASK | TX0_COHERENT_INT_MASK)
 
 #define TX_DONE_INT_MASK(_n)					\
-	((_n) ? IRQ1_INT_MASK | IRQ1_FULL_INT_MASK		\
+	((_n) ? 0		\
 	      : IRQ0_INT_MASK | IRQ0_FULL_INT_MASK)
 
 #define INT_TX_MASK						\
-	(IRQ1_INT_MASK | IRQ1_FULL_INT_MASK |			\
+	(0 |			\
 	 IRQ0_INT_MASK | IRQ0_FULL_INT_MASK)
 
 /* QDMA_CSR_INT_ENABLE2 */
@@ -537,6 +555,13 @@
 #define RX1_DONE_INT_MASK		BIT(1)
 #define RX0_DONE_INT_MASK		BIT(0)
 
+
+/* EN7523 QDMA_CSR_INT_ENABLE6 34 */
+#define EN7523_RX_NO_CPU_DSCP_LOW_INT_MASK					\
+	(RX14_NO_CPU_DSCP_INT_MASK |	\
+	 RX13_NO_CPU_DSCP_INT_MASK | RX12_NO_CPU_DSCP_INT_MASK |	\
+	 RX11_NO_CPU_DSCP_INT_MASK | RX10_NO_CPU_DSCP_INT_MASK)
+
 #define RX_NO_CPU_DSCP_LOW_INT_MASK					\
 	(RX15_NO_CPU_DSCP_INT_MASK | RX14_NO_CPU_DSCP_INT_MASK |	\
 	 RX13_NO_CPU_DSCP_INT_MASK | RX12_NO_CPU_DSCP_INT_MASK |	\
@@ -890,15 +915,17 @@
 /* RX MSG0 */
 #define QDMA_ETH_RXMSG_SPTAG		GENMASK(21, 14)
 /* RX MSG1 */
-#define QDMA_ETH_RXMSG_DEI_MASK		BIT(31)
-#define QDMA_ETH_RXMSG_IP6_MASK		BIT(30)
-#define QDMA_ETH_RXMSG_IP4_MASK		BIT(29)
-#define QDMA_ETH_RXMSG_IP4F_MASK	BIT(28)
-#define QDMA_ETH_RXMSG_L4_VALID_MASK	BIT(27)
-#define QDMA_ETH_RXMSG_L4F_MASK		BIT(26)
-#define QDMA_ETH_RXMSG_SPORT_MASK	GENMASK(25, 21)
-#define QDMA_ETH_RXMSG_CRSN_MASK	GENMASK(20, 16)
-#define QDMA_ETH_RXMSG_PPE_ENTRY_MASK	GENMASK(15, 0)
+struct airoha_qdma_eth_rx_msg1 {
+	unsigned int dei_mask;
+	unsigned int ip6_mask;
+	unsigned int ip4_mask;
+	unsigned int ip4f_mask;
+	unsigned int l4_valid_mask;
+	unsigned int l4f_mask;
+	unsigned int sport_mask;
+	unsigned int crsn_mask;
+	unsigned int ppe_entry_mask;
+};
 
 struct airoha_qdma_desc {
 	__le32 rsv;
diff --git a/drivers/net/ethernet/mediatek/Kconfig b/drivers/net/ethernet/mediatek/Kconfig
index 6f35d4f7cf30..e06ccd727e88 100644
--- a/drivers/net/ethernet/mediatek/Kconfig
+++ b/drivers/net/ethernet/mediatek/Kconfig
@@ -8,7 +8,7 @@ config NET_VENDOR_MEDIATEK
 if NET_VENDOR_MEDIATEK
 
 config NET_MEDIATEK_SOC_WED
-	depends on ARCH_MEDIATEK || COMPILE_TEST
+	depends on ARCH_MEDIATEK || ARCH_AIROHA || COMPILE_TEST
 	def_bool NET_MEDIATEK_SOC != n
 
 config NET_MEDIATEK_SOC
diff --git a/drivers/net/phy/mediatek/Kconfig b/drivers/net/phy/mediatek/Kconfig
index 4308002bb82c..c28f50f65300 100644
--- a/drivers/net/phy/mediatek/Kconfig
+++ b/drivers/net/phy/mediatek/Kconfig
@@ -15,7 +15,6 @@ config MEDIATEK_GE_PHY
 
 config MEDIATEK_GE_SOC_PHY
 	tristate "MediaTek SoC Ethernet PHYs"
-	depends on ARM64 || COMPILE_TEST
 	depends on ARCH_AIROHA || (ARCH_MEDIATEK && NVMEM_MTK_EFUSE) || \
 		   COMPILE_TEST
 	select MTK_NET_PHYLIB
diff --git a/include/linux/soc/airoha/airoha_offload.h b/include/linux/soc/airoha/airoha_offload.h
index 4d23cbb7d407..0bbc635c25b1 100644
--- a/include/linux/soc/airoha/airoha_offload.h
+++ b/include/linux/soc/airoha/airoha_offload.h
@@ -65,7 +65,6 @@ static inline void airoha_ppe_dev_check_skb(struct airoha_ppe_dev *dev,
 }
 #endif
 
-#define NPU_NUM_CORES		8
 #define NPU_NUM_IRQ		6
 #define NPU_RX0_DESC_NUM	512
 #define NPU_RX1_DESC_NUM	512
@@ -167,13 +166,16 @@ struct airoha_npu {
 #if (IS_BUILTIN(CONFIG_NET_AIROHA_NPU) || IS_MODULE(CONFIG_NET_AIROHA_NPU))
 	struct device *dev;
 	struct regmap *regmap;
+	uint sram_size;
 
+	struct airoha_npu_soc_data *soc_data;
+	
 	struct airoha_npu_core {
 		struct airoha_npu *npu;
 		/* protect concurrent npu memory accesses */
 		spinlock_t lock;
 		struct work_struct wdt_work;
-	} cores[NPU_NUM_CORES];
+	} *cores;
 
 	int irqs[NPU_NUM_IRQ];
 
