Index: linux-3.12.5/arch/arm/boot/dts/Makefile
===================================================================
--- linux-3.12.5.orig/arch/arm/boot/dts/Makefile	2014-01-07 20:20:35.112013217 +0100
+++ linux-3.12.5/arch/arm/boot/dts/Makefile	2014-01-07 20:20:35.124013454 +0100
@@ -232,6 +232,7 @@
 	sun4i-a10-mini-xplus.dtb \
 	sun4i-a10-hackberry.dtb \
 	sun4i-a10-pcduino.dtb \
+	sun4i-a10-olinuxino-lime.dtb \
 	sun5i-a10s-olinuxino-micro.dtb \
 	sun5i-a13-olinuxino.dtb \
 	sun5i-a13-olinuxino-micro.dtb \
