--- a/drivers/pci/probe.c
+++ b/drivers/pci/probe.c
@@ -2794,7 +2794,8 @@ static void pcie_write_mrrs(struct pci_d
 		pci_err(dev, "MRRS was unable to be configured with a safe value.  If problems are experienced, try running with pci=pcie_bus_safe\n");
 }
 
-static int pcie_bus_configure_set(struct pci_dev *dev, void *data)
+int pcie_bus_configure_set(struct pci_dev *dev, void *data);
+int pcie_bus_configure_set(struct pci_dev *dev, void *data)
 {
 	int mps, orig_mps;
 
