--- a/drivers/pci/quirks.c
+++ b/drivers/pci/quirks.c
@@ -3121,24 +3121,6 @@ DECLARE_PCI_FIXUP_FINAL(PCI_VENDOR_ID_AT
 			quirk_msi_intx_disable_qca_bug);
 DECLARE_PCI_FIXUP_FINAL(PCI_VENDOR_ID_ATTANSIC, 0xe091,
 			quirk_msi_intx_disable_qca_bug);
-
-/*
- * Amazon's Annapurna Labs 1c36:0031 Root Ports don't support MSI-X, so it
- * should be disabled on platforms where the device (mistakenly) advertises it.
- *
- * Notice that this quirk also disables MSI (which may work, but hasn't been
- * tested), since currently there is no standard way to disable only MSI-X.
- *
- * The 0031 device id is reused for other non Root Port device types,
- * therefore the quirk is registered for the PCI_CLASS_BRIDGE_PCI class.
- */
-static void quirk_al_msi_disable(struct pci_dev *dev)
-{
-	dev->no_msi = 1;
-	pci_warn(dev, "Disabling MSI/MSI-X\n");
-}
-DECLARE_PCI_FIXUP_CLASS_FINAL(PCI_VENDOR_ID_AMAZON_ANNAPURNA_LABS, 0x0031,
-			      PCI_CLASS_BRIDGE_PCI, 8, quirk_al_msi_disable);
 #endif /* CONFIG_PCI_MSI */
 
 /*
