--- a/arch/mips/ath79/Kconfig
+++ b/arch/mips/ath79/Kconfig
@@ -71,7 +71,7 @@ config ATH79_MACH_AP121
 	  Atheros AP121 reference board.
 
 config ATH79_MACH_AP136
-	bool "Atheros AP136 reference board"
+	bool "Atheros AP136/AP135 reference board"
 	select SOC_QCA955X
 	select ATH79_DEV_GPIO_BUTTONS
 	select ATH79_DEV_LEDS_GPIO
@@ -80,7 +80,7 @@ config ATH79_MACH_AP136
 	select ATH79_DEV_WMAC
 	help
 	  Say 'Y' here if you want your kernel to support the
-	  Atheros AP136 reference board.
+	  Atheros AP136 or AP135 reference boards.
 
 config ATH79_MACH_AP81
 	bool "Atheros AP81 reference board"
--- a/arch/mips/ath79/mach-ap136.c
+++ b/arch/mips/ath79/mach-ap136.c
@@ -237,3 +237,23 @@ static void __init ap136_020_setup(void)
 MIPS_MACHINE(ATH79_MACH_AP136_020, "AP136-020",
 	     "Atheros AP136-020 reference board",
 	     ap136_020_setup);
+
+/*
+ * AP135-020 is similar to AP136-020, any future AP135 specific init
+ * code can be added here.
+ */
+static void __init ap135_020_setup(void)
+{
+	ap136_leds_gpio[0].name = "ap135:green:status";
+	ap136_leds_gpio[1].name = "ap135:red:status";
+	ap136_leds_gpio[2].name = "ap135:green:wps";
+	ap136_leds_gpio[3].name = "ap135:red:wps";
+	ap136_leds_gpio[4].name = "ap135:red:wlan-2g";
+	ap136_leds_gpio[5].name = "ap135:red:usb";
+
+	ap136_020_setup();
+}
+
+MIPS_MACHINE(ATH79_MACH_AP135_020, "AP135-020",
+	     "Atheros AP135-020 reference board",
+	     ap135_020_setup);
--- a/arch/mips/ath79/machtypes.h
+++ b/arch/mips/ath79/machtypes.h
@@ -24,6 +24,7 @@ enum ath79_mach_type {
 	ATH79_MACH_AP113,		/* Atheros AP113 reference board */
 	ATH79_MACH_AP121,		/* Atheros AP121 reference board */
 	ATH79_MACH_AP121_MINI,		/* Atheros AP121-MINI reference board */
+	ATH79_MACH_AP135_020,		/* Atheros AP135-020 reference board */
 	ATH79_MACH_AP136_010,		/* Atheros AP136-010 reference board */
 	ATH79_MACH_AP136_020,		/* Atheros AP136-020 reference board */
 	ATH79_MACH_AP81,		/* Atheros AP81 reference board */
