--- a/drivers/usb/host/ehci-omap.c
+++ b/drivers/usb/host/ehci-omap.c
@@ -96,11 +96,10 @@ static void omap_ehci_soft_phy_reset(int
 }
 #endif
 
-inline int __board_usb_init(void)
+static int board_usb_init(void)
 {
 	return 0;
 }
-int board_usb_init(void) __attribute__((weak, alias("__board_usb_init")));
 
 #if defined(CONFIG_OMAP_EHCI_PHY1_RESET_GPIO) || \
 	defined(CONFIG_OMAP_EHCI_PHY2_RESET_GPIO) || \
