Index: qc-usb-0.6.5/qc-memory.c
===================================================================
--- qc-usb-0.6.5.orig/qc-memory.c	2007-06-04 13:22:52.398889440 +0200
+++ qc-usb-0.6.5/qc-memory.c	2007-06-04 13:22:52.469878648 +0200
@@ -27,7 +27,7 @@
  */
 /* }}} */
 /* {{{ [fold] Includes */
-#include <linux/config.h>
+#include <linux/autoconf.h>
 #include <linux/version.h>
 
 #ifdef CONFIG_SMP
Index: qc-usb-0.6.5/quickcam.h
===================================================================
--- qc-usb-0.6.5.orig/quickcam.h	2007-06-04 13:22:52.404888528 +0200
+++ qc-usb-0.6.5/quickcam.h	2007-06-04 13:22:52.470878496 +0200
@@ -76,7 +76,7 @@
 
 #ifdef __KERNEL__
 
-#include <linux/config.h>
+#include <linux/autoconf.h>
 #include <linux/version.h>
 
 #ifdef CONFIG_SMP
