--- a/arch/mips/ath79/Kconfig
+++ b/arch/mips/ath79/Kconfig
@@ -585,12 +585,13 @@ config ATH79_MACH_TL_WR841N_V1
 	select ATH79_DEV_M25P80
 
 config ATH79_MACH_TL_WR841N_V8
-	bool "TP-LINK TL-WR841N/ND v8 support"
+	bool "TP-LINK TL-WR841N/ND v8/TL-MR3420 v2 support"
 	select SOC_AR934X
 	select ATH79_DEV_ETH
 	select ATH79_DEV_GPIO_BUTTONS
 	select ATH79_DEV_LEDS_GPIO
 	select ATH79_DEV_M25P80
+	select ATH79_DEV_USB
 	select ATH79_DEV_WMAC
 
 config ATH79_MACH_TL_WR941ND
--- a/arch/mips/ath79/machtypes.h
+++ b/arch/mips/ath79/machtypes.h
@@ -84,6 +84,7 @@ enum ath79_mach_type {
 	ATH79_MACH_TL_MR3220,		/* TP-LINK TL-MR3220 */
 	ATH79_MACH_TL_MR3220_V2,	/* TP-LINK TL-MR3220 v2 */
 	ATH79_MACH_TL_MR3420,		/* TP-LINK TL-MR3420 */
+	ATH79_MACH_TL_MR3420_V2,	/* TP-LINK TL-MR3420 v2 */
 	ATH79_MACH_TL_WA7510N_V1,	/* TP-LINK TL-WA7510N v1*/
 	ATH79_MACH_TL_WA901ND,		/* TP-LINK TL-WA901ND */
 	ATH79_MACH_TL_WA901ND_V2,	/* TP-LINK TL-WA901ND v2 */
