--- a/arch/mips/ath79/machtypes.h
+++ b/arch/mips/ath79/machtypes.h
@@ -117,6 +117,7 @@ enum ath79_mach_type {
 	ATH79_MACH_WHR_HP_G300N,	/* Buffalo WHR-HP-G300N */
 	ATH79_MACH_WHR_HP_GN,		/* Buffalo WHR-HP-GN */
 	ATH79_MACH_WLAE_AG300N,		/* Buffalo WLAE-AG300N */
+	ATH79_MACH_WNDAP360,		/* NETGEAR WNDAP360 */
 	ATH79_MACH_WNDR3700,		/* NETGEAR WNDR3700/WNDR3800/WNDRMAC */
 	ATH79_MACH_WNDR4300,		/* NETGEAR WNDR4300 */
 	ATH79_MACH_WNR2000,		/* NETGEAR WNR2000 */
--- a/arch/mips/ath79/Kconfig
+++ b/arch/mips/ath79/Kconfig
@@ -378,6 +378,16 @@ config ATH79_MACH_RB2011
 	select ATH79_DEV_WMAC
 	select ATH79_ROUTERBOOT
 
+config ATH79_MACH_WNDAP360
+	bool "NETGEAR WNDAP360 board support"
+	select SOC_AR71XX
+	select ATH79_DEV_AP9X_PCI if PCI
+	select ATH79_DEV_ETH
+	select ATH79_DEV_GPIO_BUTTONS
+	select ATH79_DEV_LEDS_GPIO
+	select ATH79_DEV_M25P80
+	select ATH79_DEV_USB
+
 config ATH79_MACH_WNDR3700
 	bool "NETGEAR WNDR3700 board support"
 	select SOC_AR71XX
--- a/arch/mips/ath79/Makefile
+++ b/arch/mips/ath79/Makefile
@@ -98,6 +98,7 @@ obj-$(CONFIG_ATH79_MACH_UBNT)		+= mach-u
 obj-$(CONFIG_ATH79_MACH_UBNT_XM)	+= mach-ubnt-xm.o
 obj-$(CONFIG_ATH79_MACH_WHR_HP_G300N)	+= mach-whr-hp-g300n.o
 obj-$(CONFIG_ATH79_MACH_WLAE_AG300N)	+= mach-wlae-ag300n.o
+obj-$(CONFIG_ATH79_MACH_WNDAP360)	+= mach-wndap360.o
 obj-$(CONFIG_ATH79_MACH_WNDR3700)	+= mach-wndr3700.o
 obj-$(CONFIG_ATH79_MACH_WNDR4300)	+= mach-wndr4300.o
 obj-$(CONFIG_ATH79_MACH_WNR2000)	+= mach-wnr2000.o
@@ -111,4 +112,3 @@ obj-$(CONFIG_ATH79_MACH_WZR_HP_G300NH2)
 obj-$(CONFIG_ATH79_MACH_WZR_HP_AG300H)	+= mach-wzr-hp-ag300h.o
 obj-$(CONFIG_ATH79_MACH_WZR_HP_G450H)	+= mach-wzr-hp-g450h.o
 obj-$(CONFIG_ATH79_MACH_ZCN_1523H)	+= mach-zcn-1523h.o
-
