--- a/configure.ac
+++ b/configure.ac
@@ -94,7 +94,7 @@ AC_SUBST(with_libreadline)
 # Checks for header files.
 AC_HEADER_STDC
 AC_HEADER_ASSERT
-AC_CHECK_HEADERS([arpa/inet.h fcntl.h inttypes.h libintl.h limits.h malloc.h \
+AC_CHECK_HEADERS([arpa/inet.h fcntl.h inttypes.h limits.h malloc.h \
 		  netdb.h netinet/in.h netinet/ip.h netinet/ip6.h \
 		  netinet/tcp.h netinet/udp.h netinet/ip_icmp.h \
 		  stddef.h stdint.h stdlib.h string.h unistd.h], ,
