--- a/libstdc++-v3/include/ext/rope
+++ b/libstdc++-v3/include/ext/rope
@@ -58,6 +58,9 @@
 #include <bits/allocator.h>
 #include <ext/hash_fun.h>
 
+/* cope w/ index defined as macro, SuSv3 proposal */
+#undef index
+
 # ifdef __GC
 #   define __GC_CONST const
 # else
--- a/libstdc++-v3/include/ext/ropeimpl.h
+++ b/libstdc++-v3/include/ext/ropeimpl.h
@@ -54,6 +54,9 @@
 #include <ext/memory> // For uninitialized_copy_n
 #include <ext/numeric> // For power
 
+/* cope w/ index defined as macro, SuSv3 proposal */
+#undef index
+
 _GLIBCXX_BEGIN_NAMESPACE(__gnu_cxx)
 
   using std::size_t;
