Index: linux-3.10.34/drivers/mtd/nand/xway_nand.c
===================================================================
--- linux-3.10.34.orig/drivers/mtd/nand/xway_nand.c	2014-03-28 11:51:49.790953906 +0000
+++ linux-3.10.34/drivers/mtd/nand/xway_nand.c	2014-03-28 15:45:08.196369790 +0000
@@ -80,13 +80,16 @@
 
 static void xway_select_chip(struct mtd_info *mtd, int chip)
 {
+	static unsigned long csflags;
 
 	switch (chip) {
 	case -1:
 		ltq_ebu_w32_mask(NAND_CON_CE, 0, EBU_NAND_CON);
 		ltq_ebu_w32_mask(NAND_CON_NANDM, 0, EBU_NAND_CON);
+		spin_unlock_irqrestore(&ebu_lock, csflags);
 		break;
 	case 0:
+		spin_lock_irqsave(&ebu_lock, csflags);
 		ltq_ebu_w32_mask(0, NAND_CON_NANDM, EBU_NAND_CON);
 		ltq_ebu_w32_mask(0, NAND_CON_CE, EBU_NAND_CON);
 		break;
