--- a/Makefile	2010-06-14 14:17:11.000000000 +1000
+++ b/Makefile	2010-06-14 14:17:54.000000000 +1000
@@ -13,6 +13,11 @@
 KDIR	?= /lib/modules/$(shell uname -r)/build
 PWD	:= $(shell pwd)
 
+soloscli/soloscli:
+	$(MAKE) -C soloscli
+
+install:
+
 default:
 	$(MAKE) -C $(KDIR) M=$(PWD) modules
 
@@ -20,5 +25,6 @@
 	rm -f *.o *.ko *.mod.c .*.cmd
 	rm -rf .tmp_versions
 	rm -f Module.symvers
+	$(MAKE) -C soloscli clean
 
 endif
