diff -ruN olsrd-0.4.9-old/Makefile olsrd-0.4.9-new/Makefile
--- olsrd-0.4.9-old/Makefile	2005-05-20 17:37:26.000000000 +0200
+++ olsrd-0.4.9-new/Makefile	2005-05-20 17:42:20.000000000 +0200
@@ -75,7 +75,7 @@
 OFLAGS ?=	-O2 -g
 CFLAGS ?=	$(CCWARNINGS) $(OFLAGS) #-DDEBUG #-pg #-march=i686
 LIBS =		-lm -ldl #-pg
-MAKEDEPEND = 	makedepend -f $(DEPFILE) $(DEFINES) -Y $(INCLUDES) $(SRCS) >/dev/null 2>&1
+MAKEDEPEND = 	$(CC) -M $(DEFINES) $(INCLUDES) $(SRCS) >> $(DEPFILE)
 
 all:	 cfgparser olsrd
 install: install_olsrd
diff -ruN olsrd-0.4.9-old/lib/dot_draw/Makefile olsrd-0.4.9-new/lib/dot_draw/Makefile
--- olsrd-0.4.9-old/lib/dot_draw/Makefile	2005-04-01 23:53:01.000000000 +0200
+++ olsrd-0.4.9-new/lib/dot_draw/Makefile	2005-05-20 17:42:52.000000000 +0200
@@ -64,8 +64,7 @@
 
 EXTRA_OBJS =	# nothing
 
-MAKEDEPEND = 	makedepend -f $(DEPFILE) $(CFLAGS_ADD) -Y $(INCLUDES) \
-		$(SRCS) >/dev/null 2>&1
+MAKEDEPEND = 	$(CC) -M $(CFLAGS_ADD) $(INCLUDES) $(SRCS) >> $(DEPFILE) 
 
 all:		all2
 install:	install2
diff -ruN olsrd-0.4.9-old/lib/nameservice/Makefile olsrd-0.4.9-new/lib/nameservice/Makefile
--- olsrd-0.4.9-old/lib/nameservice/Makefile	2005-03-01 22:41:34.000000000 +0100
+++ olsrd-0.4.9-new/lib/nameservice/Makefile	2005-05-20 17:42:57.000000000 +0200
@@ -98,8 +98,7 @@
 
 EXTRA_OBJS =	# nothing
 
-MAKEDEPEND = 	makedepend -f $(DEPFILE) $(CFLAGS_ADD) -Y $(INCLUDES) \
-		$(SRCS) >/dev/null 2>&1
+MAKEDEPEND = 	$(CC) -M $(CFLAGS_ADD) $(INCLUDES) $(SRCS) >> $(DEPFILE)
 
 all:		all2
 install:	install2
diff -ruN olsrd-0.4.9-old/src/cfgparser/Makefile olsrd-0.4.9-new/src/cfgparser/Makefile
--- olsrd-0.4.9-old/src/cfgparser/Makefile	2005-03-21 03:17:36.000000000 +0100
+++ olsrd-0.4.9-new/src/cfgparser/Makefile	2005-05-20 17:43:50.000000000 +0200
@@ -56,7 +56,7 @@
 
 ifeq ($(OS), linux)
 PIPETONULL =	>/dev/null 2>&1
-DEPFLAGS +=	-Dlinux -Y
+DEPFLAGS +=	-Dlinux
 endif
 
 ifeq ($(OS), fbsd)
@@ -129,7 +129,7 @@
 endif
 endif
 
-MAKEDEPEND = 	makedepend -f $(DEPFILE) $(DEPFLAGS) $(INCLUDES) $(SRCS) $(PIPETONULL)
+MAKEDEPEND = 	$(CC) -M $(DEPFLAGS) $(INCLUDES) $(SRCS) >> $(DEPFILE)
 
 OBJS =		olsrd_conf.o oparse.o oscan.o cfgfile_gen.o $(PORT_OBJS)
 LIBS +=		$(PORT_LIBS)
