
	Copyright (C) 2006 Markus Wigge

Index: libpcap-1.8.1/Makefile.in
===================================================================
--- libpcap-1.8.1.orig/Makefile.in
+++ libpcap-1.8.1/Makefile.in
@@ -57,7 +57,7 @@ LN_S = @LN_S@
 MKDEP = @MKDEP@
 CCOPT = @V_CCOPT@
 INCLS = -I. @V_INCLS@
-DEFS = -DBUILDING_PCAP @DEFS@ @V_DEFS@
+DEFS = -DBUILDING_PCAP -D_BSD_SOURCE @DEFS@ @V_DEFS@
 ADDLOBJS = @ADDLOBJS@
 ADDLARCHIVEOBJS = @ADDLARCHIVEOBJS@
 LIBS = @LIBS@
