--- a/pcap-int.h
+++ b/pcap-int.h
@@ -209,6 +209,7 @@ struct pcap_opt {
 	char	*source;
 	int	promisc;
 	int	rfmon;
+	int 	proto;      /* protocol for packet socket (linux) */
 	int	tstamp_type;
 };
 
--- a/pcap-linux.c
+++ b/pcap-linux.c
@@ -363,7 +363,7 @@ static int	iface_get_id(int fd, const ch
 static int	iface_get_mtu(int fd, const char *device, char *ebuf);
 static int 	iface_get_arptype(int fd, const char *device, char *ebuf);
 #ifdef HAVE_PF_PACKET_SOCKETS
-static int 	iface_bind(int fd, int ifindex, char *ebuf);
+static int 	iface_bind(int fd, int ifindex, char *ebuf, unsigned short proto);
 #ifdef IW_MODE_MONITOR
 static int	has_wext(int sock_fd, const char *device, char *ebuf);
 #endif /* IW_MODE_MONITOR */
@@ -980,7 +980,7 @@ pcap_can_set_rfmon_linux(pcap_t *handle)
 	 * (We assume that if we have Wireless Extensions support
 	 * we also have PF_PACKET support.)
 	 */
-	sock_fd = socket(PF_PACKET, SOCK_RAW, htons(ETH_P_ALL));
+	sock_fd = socket(PF_PACKET, SOCK_RAW, p->opt.proto);
 	if (sock_fd == -1) {
 		(void)snprintf(handle->errbuf, PCAP_ERRBUF_SIZE,
 		    "socket: %s", pcap_strerror(errno));
@@ -1266,6 +1266,9 @@ pcap_activate_linux(pcap_t *handle)
 	handle->read_op = pcap_read_linux;
 	handle->stats_op = pcap_stats_linux;
 
+	if (handle->opt.proto < 0)
+		handle->opt.proto = (int) htons(ETH_P_ALL);
+
 	/*
 	 * The "any" device is a special device which causes us not
 	 * to bind to a particular device and thus to look at all
@@ -2897,8 +2900,8 @@ activate_new(pcap_t *handle)
 	 * try a SOCK_RAW socket for the raw interface.
 	 */
 	sock_fd = is_any_device ?
-		socket(PF_PACKET, SOCK_DGRAM, htons(ETH_P_ALL)) :
-		socket(PF_PACKET, SOCK_RAW, htons(ETH_P_ALL));
+		socket(PF_PACKET, SOCK_DGRAM, handle->opt.proto) :
+		socket(PF_PACKET, SOCK_RAW, handle->opt.proto);
 
 	if (sock_fd == -1) {
 		if (errno == EINVAL || errno == EAFNOSUPPORT) {
@@ -3015,7 +3018,7 @@ activate_new(pcap_t *handle)
 				return PCAP_ERROR;
 			}
 			sock_fd = socket(PF_PACKET, SOCK_DGRAM,
-			    htons(ETH_P_ALL));
+			    handle->opt.proto);
 			if (sock_fd == -1) {
 				snprintf(handle->errbuf, PCAP_ERRBUF_SIZE,
 				    "socket: %s", pcap_strerror(errno));
@@ -3078,7 +3081,7 @@ activate_new(pcap_t *handle)
 		}
 
 		if ((err = iface_bind(sock_fd, handle->md.ifindex,
-		    handle->errbuf)) != 1) {
+		    handle->errbuf, handle->opt.proto)) != 1) {
 		    	close(sock_fd);
 			if (err < 0)
 				return err;
@@ -4149,7 +4152,7 @@ iface_get_id(int fd, const char *device,
  *  or a PCAP_ERROR_ value on a hard error.
  */
 static int
-iface_bind(int fd, int ifindex, char *ebuf)
+iface_bind(int fd, int ifindex, char *ebuf, unsigned short proto)
 {
 	struct sockaddr_ll	sll;
 	int			err;
@@ -4158,7 +4161,7 @@ iface_bind(int fd, int ifindex, char *eb
 	memset(&sll, 0, sizeof(sll));
 	sll.sll_family		= AF_PACKET;
 	sll.sll_ifindex		= ifindex;
-	sll.sll_protocol	= htons(ETH_P_ALL);
+	sll.sll_protocol	= proto;
 
 	if (bind(fd, (struct sockaddr *) &sll, sizeof(sll)) == -1) {
 		if (errno == ENETDOWN) {
@@ -5040,7 +5043,7 @@ activate_old(pcap_t *handle)
 
 	/* Open the socket */
 
-	handle->fd = socket(PF_INET, SOCK_PACKET, htons(ETH_P_ALL));
+	handle->fd = socket(PF_INET, SOCK_PACKET, handle->opt.proto);
 	if (handle->fd == -1) {
 		snprintf(handle->errbuf, PCAP_ERRBUF_SIZE,
 			 "socket: %s", pcap_strerror(errno));
--- a/pcap.c
+++ b/pcap.c
@@ -309,6 +309,7 @@ pcap_create_common(const char *source, c
 	pcap_set_snaplen(p, 65535);	/* max packet size */
 	p->opt.promisc = 0;
 	p->opt.buffer_size = 0;
+	p->opt.proto = -1;
 	p->opt.tstamp_type = -1;	/* default to not setting time stamp type */
 	return (p);
 }
@@ -405,6 +406,15 @@ pcap_set_buffer_size(pcap_t *p, int buff
 }
 
 int
+pcap_set_protocol(pcap_t *p, unsigned short proto)
+{
+	if (pcap_check_activated(p))
+		return PCAP_ERROR_ACTIVATED;
+	p->opt.proto = proto;
+	return 0;
+}
+
+int
 pcap_activate(pcap_t *p)
 {
 	int status;
--- a/pcap/pcap.h
+++ b/pcap/pcap.h
@@ -68,6 +68,7 @@ extern "C" {
 #define PCAP_VERSION_MINOR 4
 
 #define PCAP_ERRBUF_SIZE 256
+#define HAS_PROTO_EXTENSION
 
 /*
  * Compatibility for systems that have a bpf.h that
@@ -280,6 +281,7 @@ int	pcap_set_rfmon(pcap_t *, int);
 int	pcap_set_timeout(pcap_t *, int);
 int	pcap_set_tstamp_type(pcap_t *, int);
 int	pcap_set_buffer_size(pcap_t *, int);
+int	pcap_set_protocol(pcap_t *, unsigned short);
 int	pcap_activate(pcap_t *);
 
 int	pcap_list_tstamp_types(pcap_t *, int **);
