From cc4fe7fb2f4e0ab21a33292830480d50c9d73cac Mon Sep 17 00:00:00 2001
From: Andy Green <andy@openmoko.com>
Date: Fri, 25 Jul 2008 23:06:17 +0100
Subject: [PATCH] change-remove-kernel-charging-led-drive.patch

Signed-off-by: Andy Green <andy@openmoko.com>
---
 arch/arm/mach-s3c2440/mach-gta02.c |    5 -----
 1 files changed, 0 insertions(+), 5 deletions(-)

diff --git a/arch/arm/mach-s3c2440/mach-gta02.c b/arch/arm/mach-s3c2440/mach-gta02.c
index e66498b..686291b 100644
--- a/arch/arm/mach-s3c2440/mach-gta02.c
+++ b/arch/arm/mach-s3c2440/mach-gta02.c
@@ -485,12 +485,7 @@ static int pmu_callback(struct device *dev, unsigned int feature,
 			pcf50633_charge_enable(pcf50633_global, 0);
 			break;
 		case PMU_EVT_CHARGER_IDLE:
-			/* printk(KERN_ERR"PMU_EVT_CHARGER_IDLE\n"); */
-			neo1973_gpb_setpin(GTA02_GPIO_AUX_LED, 0);
-			break;
 		case PMU_EVT_CHARGER_ACTIVE:
-			/* printk(KERN_ERR"PMU_EVT_CHARGER_ACTIVE\n"); */
-			neo1973_gpb_setpin(GTA02_GPIO_AUX_LED, 1);
 			break;
 		default:
 			break;
-- 
1.5.6.3

