Index: backports-4.19.161-1/net/mac80211/rx.c
===================================================================
--- backports-4.19.161-1.orig/net/mac80211/rx.c
+++ backports-4.19.161-1/net/mac80211/rx.c
@@ -2621,6 +2621,24 @@ ieee80211_rx_h_amsdu(struct ieee80211_rx
 
 	if (is_multicast_ether_addr(hdr->addr1))
 		return RX_DROP_UNUSABLE;
+	
+	if (rx->key) {
+		/*
+		 * We should not receive A-MSDUs on pre-HT connections,
+		 * and HT connections cannot use old ciphers. Thus drop
+		 * them, as in those cases we couldn't even have SPP
+		 * A-MSDUs or such.
+		 */
+		switch (rx->key->conf.cipher) {
+		case WLAN_CIPHER_SUITE_WEP40:
+		case WLAN_CIPHER_SUITE_WEP104:
+		case WLAN_CIPHER_SUITE_TKIP:
+			return RX_DROP_UNUSABLE;
+		default:
+			break;
+		}
+	}
+	
 
 	return __ieee80211_rx_h_amsdu(rx, 0);
 }
