Index: linux-2.6.26/init/main.c
===================================================================
--- linux-2.6.26.orig/init/main.c	2008-12-29 19:55:54.000000000 +0100
+++ linux-2.6.26/init/main.c	2008-12-29 20:14:45.000000000 +0100
@@ -827,6 +827,7 @@ static int noinline init_post(void)
 		printk(KERN_WARNING "Failed to execute %s.  Attempting "
 					"defaults...\n", execute_command);
 	}
+	run_init_process("/etc/preinit");
 	run_init_process("/sbin/init");
 	run_init_process("/etc/init");
 	run_init_process("/bin/init");
