--- a/squashfs-tools/Makefile
+++ b/squashfs-tools/Makefile
@@ -22,7 +22,7 @@
 # Obviously, you must select at least one of the available gzip, xz, lzo,
 # lz4, zstd or lzma (deprecated) compression types.
 #
-GZIP_SUPPORT = 1
+#GZIP_SUPPORT = 1
 
 ########### Building XZ support #############
 #
@@ -92,7 +92,7 @@
 # in Mksquashfs.  Obviously the compression algorithm must have been
 # selected to be built
 #
-COMP_DEFAULT = gzip
+#COMP_DEFAULT = gzip
 
 
 ###############################################
@@ -104,7 +104,7 @@
 # If your C library or build/target environment doesn't support XATTRs then
 # comment out the next line to build Mksquashfs and Unsquashfs without XATTR
 # support
-XATTR_SUPPORT = 1
+#XATTR_SUPPORT = 1
 
 # Select whether you wish xattrs to be stored by Mksquashfs and extracted
 # by Unsquashfs by default.  If selected users can disable xattr support by
