--- a/squashfs-tools/Makefile
+++ b/squashfs-tools/Makefile
@@ -463,8 +463,8 @@
 .PHONY: install
 install: mksquashfs unsquashfs
 	mkdir -p $(INSTALL_DIR)
-	cp mksquashfs $(INSTALL_DIR)
-	cp unsquashfs $(INSTALL_DIR)
-	ln -fs unsquashfs $(INSTALL_DIR)/sqfscat
-	ln -fs mksquashfs $(INSTALL_DIR)/sqfstar
+	cp mksquashfs $(INSTALL_DIR)/mksquashfs4
+	cp unsquashfs $(INSTALL_DIR)/unsquashfs4
+	ln -fs unsquashfs4 $(INSTALL_DIR)/sqfscat4
+	ln -fs mksquashfs4 $(INSTALL_DIR)/sqfstar4
 	../generate-manpages/install-manpages.sh $(shell pwd)/.. "$(INSTALL_MANPAGES_DIR)" "$(USE_PREBUILT_MANPAGES)"
