diff -ruN pcre-5.0-old/pcre-config.in pcre-5.0-new/pcre-config.in
--- pcre-5.0-old/pcre-config.in	2004-09-13 16:20:00.000000000 +0200
+++ pcre-5.0-new/pcre-config.in	2005-09-04 16:34:39.000000000 +0200
@@ -18,6 +18,10 @@
   libR=" -R@libdir@"
   ;;
 esac
+libS=
+if test @libdir@ != /usr/lib ; then
+  libS=-L@libdir@
+fi
 
 while test $# -gt 0; do
   case "$1" in
@@ -52,10 +56,10 @@
       echo $includes
       ;;
     --libs-posix)
-      echo -L@libdir@$libR -lpcreposix -lpcre
+      echo $libS$libR -lpcreposix -lpcre
       ;;
     --libs)
-      echo -L@libdir@$libR -lpcre
+      echo $libS$libR -lpcre
       ;;
     *)
       echo "${usage}" 1>&2
