Index: linux-2.6.25.4/drivers/Kconfig
===================================================================
--- linux-2.6.25.4.orig/drivers/Kconfig
+++ linux-2.6.25.4/drivers/Kconfig
@@ -97,4 +97,6 @@ source "drivers/dca/Kconfig"
 source "drivers/auxdisplay/Kconfig"
 
 source "drivers/uio/Kconfig"
+
+source "drivers/vlynq/Kconfig"
 endmenu
Index: linux-2.6.25.4/drivers/Makefile
===================================================================
--- linux-2.6.25.4.orig/drivers/Makefile
+++ linux-2.6.25.4/drivers/Makefile
@@ -91,5 +91,6 @@ obj-$(CONFIG_DCA)		+= dca/
 obj-$(CONFIG_HID)		+= hid/
 obj-$(CONFIG_PPC_PS3)		+= ps3/
 obj-$(CONFIG_OF)		+= of/
+obj-$(CONFIG_VLYNQ)		+= vlynq/
 obj-$(CONFIG_SSB)		+= ssb/
 obj-$(CONFIG_VIRTIO)		+= virtio/
