Index: linux-2.6.23/drivers/mtd/chips/cfi_cmdset_0002.c
===================================================================
--- linux-2.6.23.orig/drivers/mtd/chips/cfi_cmdset_0002.c	2007-11-02 23:02:29.000000000 +0100
+++ linux-2.6.23/drivers/mtd/chips/cfi_cmdset_0002.c	2007-11-02 23:07:15.000000000 +0100
@@ -1007,7 +1007,9 @@
 	int ret = 0;
 	map_word oldd;
 	int retry_cnt = 0;
-
+#ifdef CONFIG_IFXMIPS
+	adr ^= 2;
+#endif
 	adr += chip->start;
 
 	spin_lock(chip->mutex);
