--- a/Embedded/src/GbE/iegbe_ethtool.c
+++ b/Embedded/src/GbE/iegbe_ethtool.c
@@ -944,7 +944,8 @@ iegbe_intr_test(struct iegbe_adapter *ad
     *data = 0;
 
     /* Hook up test interrupt handler just for this test */
-     if(!request_irq(irq, &iegbe_test_intr, 0, netdev->name, netdev)) {
+     if(!request_irq(irq, &iegbe_test_intr, IRQF_PROBE_SHARED, netdev->name,
+			netdev)) {
          shared_int = FALSE;
      } else if(request_irq(irq, &iegbe_test_intr, IRQF_SHARED,
                   netdev->name, netdev)){
