--- a/Embedded/src/1588/1588.c
+++ b/Embedded/src/1588/1588.c
@@ -72,6 +72,7 @@
  *
  ****************************************************************************/
 
+#include <linux/sched.h>
 #include "1588.h"
 
 MODULE_AUTHOR("Intel(R) Corporation");
--- a/Embedded/src/CAN/can_main.c
+++ b/Embedded/src/CAN/can_main.c
@@ -68,6 +68,7 @@
  *
  **************************************************************************/
 
+#include <linux/sched.h>
 #include "can_main.h"
 #include "can_ioctl.h"
 #include <linux/fs.h>
--- a/Embedded/src/WDT/iwdt.c
+++ b/Embedded/src/WDT/iwdt.c
@@ -137,6 +137,7 @@
 #include <linux/watchdog.h>
 #include <linux/miscdevice.h>
 #include <linux/interrupt.h>
+#include <linux/sched.h>
 #include "iwdt.h"
 
 MODULE_AUTHOR("Intel(R) Corporation");
