--- elfutils-0.160/libelf/libelfP.h.old	2014-11-06 01:47:55.420116984 -0200
+++ elfutils-0.160/libelf/libelfP.h	2014-11-06 01:48:33.368116980 -0200
@@ -43,6 +43,9 @@
 #include <string.h>
 
+#ifdef _ /* fix libintl-stub */
+#undef _
+#endif
 /* gettext helper macros.  */
 #define _(Str) dgettext ("elfutils", Str)
 
 
--- elfutils-0.160/libdw/libdwP.h.old	2014-11-06 02:27:12.628116703 -0200
+++ elfutils-0.160/libdw/libdwP.h	2014-11-06 02:27:39.816116700 -0200
@@ -36,7 +36,9 @@
 #include <libdw.h>
 #include <dwarf.h>
 
-
+#ifdef _ /* fix libintl-stub */
+#undef _
+#endif
 /* gettext helper macros.  */
 #define _(Str) dgettext ("elfutils", Str)
 
--- elfutils-0.160/libdwfl/libdwflP.h.old	2014-11-06 02:31:20.112116673 -0200
+++ elfutils-0.160/libdwfl/libdwflP.h	2014-11-06 02:33:00.548116661 -0200
@@ -46,6 +46,9 @@
 
 typedef struct Dwfl_Process Dwfl_Process;
 
+#ifdef _ /* fix libintl-stub */
+#undef _
+#endif
 /* gettext helper macros.  */
 #define _(Str) dgettext ("elfutils", Str)
 
--- elfutils-0.160/libasm/libasmP.h.old	2014-11-06 02:31:32.064116672 -0200
+++ elfutils-0.160/libasm/libasmP.h	2014-11-06 02:33:08.656116660 -0200
@@ -33,6 +33,9 @@
 
 #include <libasm.h>
 
+#ifdef _ /* fix libintl-stub */
+#undef _
+#endif
 /* gettext helper macros.  */
 #define _(Str) dgettext ("elfutils", Str)
 
