--- elfutils-0.160/Makefile.in.old	2014-11-06 03:56:04.828116067 -0200
+++ elfutils-0.160/Makefile.in	2014-11-06 03:57:03.380116060 -0200
@@ -359,8 +359,7 @@
 pkginclude_HEADERS = version.h
 
 # Add doc back when we have some real content.
-SUBDIRS = config m4 lib libelf libebl libdwelf libdwfl libdw libcpu libasm \
-	  backends src po tests
+SUBDIRS = config m4 lib libelf libebl libdwelf libdwfl libdw libasm
 
 EXTRA_DIST = elfutils.spec GPG-KEY NOTES CONTRIBUTING \
 	     COPYING COPYING-GPLV2 COPYING-LGPLV3
--- elfutils-0.160/Makefile.am.old	2014-11-06 03:58:13.012116051 -0200
+++ elfutils-0.160/Makefile.am	2014-11-06 03:58:25.948116050 -0200
@@ -23,8 +23,7 @@
 pkginclude_HEADERS = version.h
 
 # Add doc back when we have some real content.
-SUBDIRS = config m4 lib libelf libebl libdwelf libdwfl libdw libcpu libasm \
-	  backends src po tests
+SUBDIRS = config m4 lib libelf libebl libdwelf libdwfl libdw libasm
 
 EXTRA_DIST = elfutils.spec GPG-KEY NOTES CONTRIBUTING \
 	     COPYING COPYING-GPLV2 COPYING-LGPLV3
