--- a/drivers/net/wireless/ralink/rt2x00/rt2800lib.c
+++ b/drivers/net/wireless/ralink/rt2x00/rt2800lib.c
@@ -6305,7 +6305,8 @@ static int rt2800_init_registers(struct
 		reg = rt2800_register_read(rt2x00dev, US_CYC_CNT);
 		rt2x00_set_field32(&reg, US_CYC_CNT_CLOCK_CYCLE, 125);
 		rt2800_register_write(rt2x00dev, US_CYC_CNT, reg);
-	} else if (rt2x00_is_soc(rt2x00dev)) {
+	} else if (rt2x00_is_soc(rt2x00dev) &&
+		   !rt2x00_rt(rt2x00dev, RT6352)) {
 		struct clk *clk = clk_get_sys("bus", NULL);
 		int rate;
 
