From 1b6941ae603f2885e6cf729119ef753deb7eb835 Mon Sep 17 00:00:00 2001
From: John Crispin <blogic@openwrt.org>
Date: Thu, 6 Dec 2012 19:59:53 +0100
Subject: [PATCH 123/123] USB: fix roothub for IFXHCD

---
 drivers/usb/core/hub.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-3.7-rc8/drivers/usb/core/hub.c
===================================================================
--- linux-3.7-rc8.orig/drivers/usb/core/hub.c	2012-12-03 20:22:37.000000000 +0100
+++ linux-3.7-rc8/drivers/usb/core/hub.c	2012-12-10 23:37:16.658956109 +0100
@@ -3839,7 +3839,7 @@
 		udev->ttport = hdev->ttport;
 	} else if (udev->speed != USB_SPEED_HIGH
 			&& hdev->speed == USB_SPEED_HIGH) {
-		if (!hub->tt.hub) {
+		if (hdev->parent && !hub->tt.hub) {
 			dev_err(&udev->dev, "parent hub has no TT\n");
 			retval = -EINVAL;
 			goto fail;
Index: linux-3.7-rc8/arch/mips/lantiq/Kconfig
===================================================================
--- linux-3.7-rc8.orig/arch/mips/lantiq/Kconfig	2012-12-10 23:37:16.622956108 +0100
+++ linux-3.7-rc8/arch/mips/lantiq/Kconfig	2012-12-11 13:58:10.816251053 +0100
@@ -3,6 +3,7 @@
 config SOC_TYPE_XWAY
 	bool
 	select PINCTRL_XWAY
+	select USB_ARCH_HAS_HCD
 	default n
 
 choice
