--- a/Embedded/src/CAN/can_main.c
+++ b/Embedded/src/CAN/can_main.c
@@ -654,7 +654,7 @@ int can_dev_io(struct inode *inode, stru
 /*****************************************************************************
  * Interrupt handler.
  *****************************************************************************/
-irqreturn_t can_irq_handler(int irq, void *dev_id, struct pt_regs *regs)
+irqreturn_t can_irq_handler(int irq, void *dev_id)
 {
     can_os_t *can_os = (can_os_t *) dev_id;
     unsigned int int_status;
--- a/Embedded/src/CAN/can_main.h
+++ b/Embedded/src/CAN/can_main.h
@@ -165,8 +165,7 @@ int can_dev_io(
 
 irqreturn_t can_irq_handler(
     int irq, 
-    void *dev_id, 
-    struct pt_regs *regs);
+    void *dev_id);
 
 void can_tasklet(
     unsigned long arg
