From ff4470f274b61cebaeb1586f2f462ff66b8041cb Mon Sep 17 00:00:00 2001
From: John Crispin <blogic@openwrt.org>
Date: Tue, 20 Mar 2012 08:22:11 +0100
Subject: [PATCH 66/70] MIPS: lantiq: dont always register asc0

---
 arch/mips/lantiq/xway/prom.c |    6 ++----
 1 files changed, 2 insertions(+), 4 deletions(-)

--- a/arch/mips/lantiq/xway/prom.c
+++ b/arch/mips/lantiq/xway/prom.c
@@ -101,12 +101,10 @@ void __init ltq_soc_detect(struct ltq_so
 
 void __init ltq_soc_setup(void)
 {
-	if (ltq_is_ase()) {
+	if (ltq_is_ase())
 		ltq_register_ase_asc();
-	} else {
-		ltq_register_asc(0);
+	else
 		ltq_register_asc(1);
-	}
 	ltq_register_gpio();
 	ltq_register_wdt();
 }
