Index: linux-3.2.9/arch/mips/lantiq/irq.c
===================================================================
--- linux-3.2.9.orig/arch/mips/lantiq/irq.c	2012-03-08 17:17:13.937473031 +0100
+++ linux-3.2.9/arch/mips/lantiq/irq.c	2012-03-08 17:32:57.389513360 +0100
@@ -10,6 +10,7 @@
 #include <linux/interrupt.h>
 #include <linux/ioport.h>
 #include <linux/sched.h>
+#include <linux/module.h>
 
 #include <asm/bootinfo.h>
 #include <asm/irq_cpu.h>
@@ -111,6 +112,7 @@
 	ltq_icu_w32(ltq_icu_r32(ier) & ~(1 << irq_nr), ier);
 	ltq_icu_w32((1 << irq_nr), isr);
 }
+EXPORT_SYMBOL(ltq_mask_and_ack_irq);
 
 static void ltq_ack_irq(struct irq_data *d)
 {
Index: linux-3.2.9/arch/mips/mm/cache.c
===================================================================
--- linux-3.2.9.orig/arch/mips/mm/cache.c	2012-03-01 01:32:49.000000000 +0100
+++ linux-3.2.9/arch/mips/mm/cache.c	2012-03-08 17:17:14.193473042 +0100
@@ -57,6 +57,8 @@
 void (*_dma_cache_inv)(unsigned long start, unsigned long size);
 
 EXPORT_SYMBOL(_dma_cache_wback_inv);
+EXPORT_SYMBOL(_dma_cache_wback);
+EXPORT_SYMBOL(_dma_cache_inv);
 
 #endif /* CONFIG_DMA_NONCOHERENT */
 
Index: linux-3.2.9/net/atm/proc.c
===================================================================
--- linux-3.2.9.orig/net/atm/proc.c	2012-03-01 01:32:49.000000000 +0100
+++ linux-3.2.9/net/atm/proc.c	2012-03-08 17:17:14.193473042 +0100
@@ -154,7 +154,7 @@
 static void pvc_info(struct seq_file *seq, struct atm_vcc *vcc)
 {
 	static const char *const class_name[] = {
-		"off", "UBR", "CBR", "VBR", "ABR"};
+		"off","UBR","CBR","NTR-VBR","ABR","ANY","RT-VBR","UBR+","GFR"};
 	static const char *const aal_name[] = {
 		"---",	"1",	"2",	"3/4",	/*  0- 3 */
 		"???",	"5",	"???",	"???",	/*  4- 7 */
Index: linux-3.2.9/net/atm/common.c
===================================================================
--- linux-3.2.9.orig/net/atm/common.c	2012-03-01 01:32:49.000000000 +0100
+++ linux-3.2.9/net/atm/common.c	2012-03-08 17:17:14.193473042 +0100
@@ -62,11 +62,17 @@
 	write_unlock_irq(&vcc_sklist_lock);
 }
 
+struct sk_buff* (*ifx_atm_alloc_tx)(struct atm_vcc *, unsigned int) = NULL;
+EXPORT_SYMBOL(ifx_atm_alloc_tx);
+
 static struct sk_buff *alloc_tx(struct atm_vcc *vcc, unsigned int size)
 {
 	struct sk_buff *skb;
 	struct sock *sk = sk_atm(vcc);
 
+	if (ifx_atm_alloc_tx != NULL)
+		return ifx_atm_alloc_tx(vcc, size);
+
 	if (sk_wmem_alloc_get(sk) && !atm_may_send(vcc, size)) {
 		pr_debug("Sorry: wmem_alloc = %d, size = %d, sndbuf = %d\n",
 			 sk_wmem_alloc_get(sk), size, sk->sk_sndbuf);
