--- a/drivers/mtd/devices/m25p80.c
+++ b/drivers/mtd/devices/m25p80.c
@@ -922,6 +922,16 @@ static int __devinit m25p_probe(struct s
 					part_probes, &parts, 0);
 		}
 
+#ifdef CONFIG_MTD_MYLOADER_PARTS
+		if (nr_parts <= 0) {
+			static const char *part_probes[]
+					= { "MyLoader", NULL, };
+
+			nr_parts = parse_mtd_partitions(&flash->mtd,
+					part_probes, &parts, 0);
+		}
+#endif
+
 		if (nr_parts <= 0 && data && data->parts) {
 			parts = data->parts;
 			nr_parts = data->nr_parts;
