--- a/drivers/net/Kconfig
+++ b/drivers/net/Kconfig
@@ -479,6 +479,10 @@ config MIPS_AU1X00_ENET
 	  If you have an Alchemy Semi AU1X00 based system
 	  say Y.  Otherwise, say N.
 
+config ADM5120_ENET
+	tristate "ADM5120 Ethernet switch support"
+	depends on ADM5120
+
 config SGI_IOC3_ETH
 	bool "SGI IOC3 Ethernet"
 	depends on PCI && SGI_IP27
--- a/drivers/net/Makefile
+++ b/drivers/net/Makefile
@@ -207,6 +207,7 @@ obj-$(CONFIG_SC92031) += sc92031.o
 # This is also a 82596 and should probably be merged
 obj-$(CONFIG_LP486E) += lp486e.o
 
+obj-$(CONFIG_ADM5120_ENET) += adm5120sw.o
 obj-$(CONFIG_ETH16I) += eth16i.o
 obj-$(CONFIG_ZORRO8390) += zorro8390.o 8390.o
 obj-$(CONFIG_HPLANCE) += hplance.o 7990.o
