Index: uclibc-0.9.29/libc/sysdeps/linux/arm/bits/uClibc_arch_features.h
===================================================================
--- uclibc-0.9.29.orig/libc/sysdeps/linux/arm/bits/uClibc_arch_features.h
+++ uclibc-0.9.29/libc/sysdeps/linux/arm/bits/uClibc_arch_features.h
@@ -38,4 +38,8 @@
 /* define if target supports IEEE signed zero floats */
 #define __UCLIBC_HAVE_SIGNED_ZERO__
 
+#ifdef __ARM_EABI__
+# define __UCLIBC_TRUNCATE64_HAS_4_ARGS__
+#endif
+
 #endif /* _BITS_UCLIBC_ARCH_FEATURES_H */
