diff -urN uClibc-0.9.29.old/ldso/ldso/ldso.c uClibc-0.9.29/ldso/ldso/ldso.c
--- uClibc-0.9.29.old/ldso/ldso/ldso.c	2008-09-25 01:02:42.000000000 +0200
+++ uClibc-0.9.29/ldso/ldso/ldso.c	2008-09-24 23:55:11.000000000 +0200
@@ -286,15 +286,20 @@
 			_dl_debug_early("calling mprotect on the application program\n");
 			/* Now cover the application program. */
 			if (app_tpnt->dynamic_info[DT_TEXTREL]) {
-				ppnt = (ElfW(Phdr) *) auxvt[AT_PHDR].a_un.a_val;
-				for (i = 0; i < auxvt[AT_PHNUM].a_un.a_val; i++, ppnt++) {
-					if (ppnt->p_type == PT_LOAD && !(ppnt->p_flags & PF_W))
-						_dl_mprotect((void *) (DL_RELOC_ADDR(app_tpnt->loadaddr, ppnt->p_vaddr) & PAGE_ALIGN),
-							     ((ppnt->p_vaddr + app_tpnt->loadaddr) & ADDR_ALIGN) +
-							     (unsigned long) ppnt->p_filesz,
+				ElfW(Phdr) *ppnt_inner = (ElfW(Phdr) *) auxvt[AT_PHDR].a_un.a_val;
+				for (i = 0; i < auxvt[AT_PHNUM].a_un.a_val; i++, ppnt_inner++) {
+					if (ppnt_inner->p_type == PT_LOAD && !(ppnt_inner->p_flags & PF_W))
+						_dl_mprotect((void *) (DL_RELOC_ADDR(app_tpnt->loadaddr, ppnt_inner->p_vaddr) & PAGE_ALIGN),
+							     ((ppnt_inner->p_vaddr + app_tpnt->loadaddr) & ADDR_ALIGN) +
+							     (unsigned long) ppnt_inner->p_filesz,
 							     PROT_READ | PROT_WRITE | PROT_EXEC);
 				}
 			}
+#else
+			if (app_tpnt->dynamic_info[DT_TEXTREL]) {
+				_dl_dprintf(_dl_debug_file, "Can't modify application's text section; use the GCC option -fPIE for position-independent executables.\n");
+				_dl_exit(1);
+}
 #endif
 
 #ifndef ALLOW_ZERO_PLTGOT
