diff -ur uClibc.old/libc/sysdeps/linux/powerpc/bits/kernel_types.h uClibc.dev/libc/sysdeps/linux/powerpc/bits/kernel_types.h
--- uClibc.old/libc/sysdeps/linux/powerpc/bits/kernel_types.h	2005-08-18 00:49:41.000000000 +0200
+++ uClibc.dev/libc/sysdeps/linux/powerpc/bits/kernel_types.h	2006-11-30 19:26:22.000000000 +0100
@@ -4,7 +4,8 @@
  * our private content, and not the kernel header, will win.
  *  -Erik
  */
-#if ! defined _PPC_POSIX_TYPES_H && ! defined _PPC64_POSIX_TYPES_H
+#if ! defined _PPC_POSIX_TYPES_H && ! defined _PPC64_POSIX_TYPES_H && !defined _ASM_POWERPC_POSIX_TYPES_H
+#define _ASM_POWERPC_POSIX_TYPES_H
 #define _PPC_POSIX_TYPES_H
 #define _PPC64_POSIX_TYPES_H
 
