--- a/Rules.mak
+++ b/Rules.mak
@@ -498,7 +498,8 @@ CFLAGS += -I$(KERNEL_HEADERS)
 
 # Sigh, some stupid versions of gcc can't seem to cope with '-iwithprefix include'
 #CFLAGS+=-iwithprefix include
-CFLAGS+=-isystem $(shell $(CC) -print-file-name=include)
+CC_IPREFIX:=$(shell $(CC) -print-file-name=include)
+CFLAGS+=-isystem $(CC_IPREFIX) -isystem $(CC_IPREFIX)-fixed
 
 ifneq ($(DOASSERTS),y)
 CFLAGS+=-DNDEBUG
