--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -534,6 +534,40 @@ static struct board_info __initdata boar
 	},
 };
 
+static struct board_info __initdata board_CPVA502plus = {
+	.name				= "CPVA502+",
+	.expected_cpu_id		= 0x6348,
+
+	.has_uart0			= 1,
+	.has_enet0			= 1,
+	.has_enet1			= 1,
+	.has_pci			= 1,
+
+	.enet0 = {
+		.has_phy		= 1,
+		.use_internal_phy	= 1,
+	},
+	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
+	},
+
+	.leds = {
+		{
+			.name		= "CPVA502+:green:phone",
+			.gpio		= 0,
+			.active_low	= 1,
+		},
+		{
+			.name		= "CPVA502+:amber:link",
+			.gpio		= 5,
+			.active_low	= 1,
+		},
+	},
+
+	.ephy_reset_gpio		= 4,
+	.ephy_reset_gpio_flags		= GPIOF_INIT_HIGH,
+};
 
 static struct board_info __initdata board_ct536_ct5621 = {
 	.name				= "CT536_CT5621",
@@ -2246,6 +2280,7 @@ static const struct board_info __initcon
 	&board_V2110,
 	&board_ct536_ct5621,
 	&board_96348A_122,
+	&board_CPVA502plus,
 #endif
 
 #ifdef CONFIG_BCM63XX_CPU_6358
