--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -794,6 +794,8 @@ static struct board_info __initdata boar
 	.has_uart0			= 1,
 	.has_enet0			= 1,
 	.enet0 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -838,6 +840,8 @@ static struct board_info __initdata boar
 	.has_uart0			= 1,
 	.has_enet0			= 1,
 	.enet0 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1128,6 +1132,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1202,6 +1208,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1483,6 +1491,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1662,6 +1672,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1738,6 +1750,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1761,6 +1775,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1782,6 +1798,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -2127,6 +2145,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -2181,6 +2201,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -2323,6 +2345,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -2442,6 +2466,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
