--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -792,6 +792,8 @@ static struct board_info __initdata boar
 	.has_uart0			= 1,
 	.has_enet0			= 1,
 	.enet0 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -836,6 +838,8 @@ static struct board_info __initdata boar
 	.has_uart0			= 1,
 	.has_enet0			= 1,
 	.enet0 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1126,6 +1130,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1200,6 +1206,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1481,6 +1489,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1660,6 +1670,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1736,6 +1748,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1759,6 +1773,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1780,6 +1796,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -2125,6 +2143,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -2179,6 +2199,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -2321,6 +2343,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -2440,6 +2464,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
