--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -13,6 +13,7 @@
 #include <linux/gpio_keys.h>
 #include <linux/input.h>
 #include <linux/pci_ids.h>
+#include <linux/platform_data/b53.h>
 #include <linux/platform_device.h>
 #include <linux/spi/spi.h>
 #include <linux/spi/spi_gpio.h>
@@ -4584,6 +4585,117 @@ static struct board_info __initdata boar
 	.has_ohci0 = 1,
 	.has_ehci0 = 1,
 };
+
+static struct b53_platform_data WAP5813n_b53_pdata = {
+	.alias = "eth0",
+};
+
+static struct spi_board_info WAP5813n_spi_devices[] = {
+	{
+		.modalias = "b53-switch",
+		.max_speed_hz = 781000,
+		.bus_num = 0,
+		.chip_select = 0,
+		.platform_data = &WAP5813n_b53_pdata,
+	}
+};
+
+static struct sprom_fixup __initdata wap5813n_fixups[] = {
+	{ .offset = 97, .value = 0xfeed },
+	{ .offset = 98, .value = 0x15d1 },
+	{ .offset = 99, .value = 0xfb0d },
+	{ .offset = 113, .value = 0xfef7 },
+	{ .offset = 114, .value = 0x15f7 },
+	{ .offset = 115, .value = 0xfb1a },
+};
+
+static struct board_info __initdata board_WAP5813n = {
+	.name				= "96369R-1231N",
+	.expected_cpu_id		= 0x6368,
+
+	.has_uart0			= 1,
+	.has_pci			= 1,
+	.use_fallback_sprom		= 1,
+	.has_ohci0			= 1,
+	.has_ehci0			= 1,
+
+	.has_enetsw			= 1,
+	.enetsw = {
+		.used_ports = {
+			[4] = {
+				.used = 1,
+				.phy_id = 0xff,
+				.bypass_link = 1,
+				.force_speed = 1000,
+				.force_duplex_full = 1,
+				.name = "RGMII",
+			},
+		},
+	},
+
+	.leds = {
+		{
+			.name		= "WAP-5813n:green:power",
+			.gpio		= 22,
+			.default_trigger = "default-on",
+		},
+		{
+			.name		= "WAP-5813n:red:power",
+			.gpio		= 24,
+		},
+		{
+			.name		= "WAP-5813n:green:inet",
+			.gpio		= 5,
+		},
+		{
+			.name		= "WAP-5813n:red:inet",
+			.gpio		= 31,
+		},
+		{
+			.name		= "WAP-5813n:green:wps",
+			.gpio		= 23,
+			.active_low	= 1,
+		},
+	},
+
+	.buttons = {
+		{
+			.desc		= "wlan",
+			.gpio		= 32,
+			.active_low	= 1,
+			.type		= EV_KEY,
+			.code		= KEY_WLAN,
+			.debounce_interval = BCM963XX_KEYS_DEBOUNCE_INTERVAL,
+		},
+		{
+			.desc		= "reset",
+			.gpio		= 34,
+			.active_low	= 1,
+			.type		= EV_KEY,
+			.code		= KEY_RESTART,
+			.debounce_interval = BCM963XX_KEYS_DEBOUNCE_INTERVAL,
+		},
+		{
+			.desc		= "wps",
+			.gpio		= 35,
+			.active_low	= 1,
+			.type		= EV_KEY,
+			.code		= KEY_WPS_BUTTON,
+			.debounce_interval = BCM963XX_KEYS_DEBOUNCE_INTERVAL,
+		},
+	},
+
+	.fallback_sprom = {
+		.type 				= SPROM_BCM43222,
+		.pci_bus			= 0,
+		.pci_dev			= 1,
+		.board_fixups			= wap5813n_fixups,
+		.num_board_fixups		= ARRAY_SIZE(wap5813n_fixups),
+	},
+
+	.spis = WAP5813n_spi_devices,
+	.num_spis = ARRAY_SIZE(WAP5813n_spi_devices),
+};
 #endif /* CONFIG_BCM63XX_CPU_6368 */
 
 /*
@@ -4967,6 +5079,7 @@ static const struct board_info __initcon
 #ifdef CONFIG_BCM63XX_CPU_6368
 	&board_96368mvwg,
 	&board_96368mvngr,
+	&board_WAP5813n,
 #endif
 #ifdef CONFIG_BCM63XX_CPU_63268
 	&board_963268bu_p300,
