--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -4486,6 +4486,89 @@ static struct board_info __initdata boar
 	.has_ehci0 = 1,
 };
 
+static struct board_info __initdata board_P870HW51A_V2 = {
+	.name				= "P870HW-51a_v2",
+	.expected_cpu_id		= 0x6368,
+
+	.has_uart0			= 1,
+	.has_pci			= 1,
+	.has_ohci0			= 1,
+	.has_ehci0			= 1,
+
+	.has_enetsw			= 1,
+	.enetsw = {
+		.used_ports = {
+			[0] = {
+				.used	= 1,
+				.phy_id	= 1,
+				.name	= "port1",
+			},
+			[1] = {
+				.used	= 1,
+				.phy_id	= 2,
+				.name	= "port2",
+			},
+			[2] = {
+				.used	= 1,
+				.phy_id	= 3,
+				.name	= "port3",
+			},
+			[3] = {
+				.used	= 1,
+				.phy_id	= 4,
+				.name	= "port4",
+			},
+		},
+	},
+
+	.leds = {
+		{
+			.name		= "P870HW-51a:green:power",
+			.gpio		= 0,
+			.default_trigger = "default-on",
+		},
+		{
+			.name		= "P870HW-51a:green:dsl",
+			.gpio		= 2,
+			.active_low	= 1,
+		},
+		{
+			.name		= "P870HW-51a:green:inet",
+			.gpio		= 22,
+			.active_low	= 1,
+		},
+		{
+			.name		= "P870HW-51a:orange:wps",
+			.gpio		= 24,
+			.active_low	= 1,
+		},
+		{
+			.name		= "P870HW-51a:red:inet",
+			.gpio		= 33,
+			.active_low	= 1,
+		},
+	},
+
+	.buttons = {
+		{
+			.desc		= "reset",
+			.gpio		= 34,
+			.type		= EV_KEY,
+			.code		= KEY_RESTART,
+			.debounce_interval = BCM963XX_KEYS_DEBOUNCE_INTERVAL,
+			.active_low	= 1,
+		},
+		{
+			.desc		= "wps",
+			.gpio		= 35,
+			.type		= EV_KEY,
+			.code		= KEY_WPS_BUTTON,
+			.debounce_interval = BCM963XX_KEYS_DEBOUNCE_INTERVAL,
+			.active_low	= 1,
+		},
+	},
+};
+
 static struct board_info __initdata board_VR3025u = {
 	.name					= "96368M-1541N",
 	.expected_cpu_id		= 0x6368,
@@ -5145,6 +5228,7 @@ static const struct board_info __initcon
 #ifdef CONFIG_BCM63XX_CPU_6368
 	&board_96368mvwg,
 	&board_96368mvngr,
+	&board_P870HW51A_V2,
 	&board_VR3025u,
 	&board_VR3025un,
 	&board_WAP5813n,
