Index: gcc-4.4.2/gcc/config.gcc
===================================================================
--- gcc-4.4.2.orig/gcc/config.gcc	2009-10-21 08:24:59.000000000 +0200
+++ gcc-4.4.2/gcc/config.gcc	2009-10-21 08:24:59.000000000 +0200
@@ -1516,6 +1516,7 @@
 	if test x$sjlj != x1; then
 	    tmake_file="$tmake_file m68k/t-slibgcc-elf-ver"
 	fi
+	tmake_file="m68k/t-floatlib m68k/t-m68kbare m68k/t-m68kelf"
 	;;
 m68k-*-rtems*)
 	default_m68k_cpu=68020
