--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -589,6 +589,8 @@ static struct board_info __initdata boar
 	.has_uart0			= 1,
 	.has_enet0			= 1,
 	.enet0 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -632,6 +634,8 @@ static struct board_info __initdata boar
 	.has_uart0			= 1,
 	.has_enet0			= 1,
 	.enet0 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -917,6 +921,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -989,6 +995,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1218,6 +1226,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1393,6 +1403,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1467,6 +1479,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1488,6 +1502,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1507,6 +1523,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1882,6 +1900,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1934,6 +1954,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -2074,6 +2096,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -2191,6 +2215,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
