diff -ruN asterisk-1.2.0-old/channels/Makefile asterisk-1.2.0-new/channels/Makefile
--- asterisk-1.2.0-old/channels/Makefile	2005-12-04 04:48:40.000000000 +0100
+++ asterisk-1.2.0-new/channels/Makefile	2005-12-04 23:30:19.000000000 +0100
@@ -155,6 +155,9 @@
 
 #CFLAGS+=$(shell [ -f $(ZAPDIR)/libzap.a ] && echo "-I$(ZAPDIR)")
 
+CFLAGS += $(EXTRA_CFLAGS)
+CHANNEL_LIBS += $(EXTRA_CHAN_MODULES)
+
 all: depend $(CHANNEL_LIBS) 
 
 clean:
@@ -162,7 +165,7 @@
 	rm -f busy.h ringtone.h gentone gentone-ulaw
 
 %.so : %.o
-	$(CC) $(SOLINK) -o $@ ${CYGSOLINK} $< ${CYGSOLIB} ${LIBS}
+	$(CC) $(SOLINK) -o $@ ${CYGSOLINK} $< ${CYGSOLIB}
 
 ifneq ($(wildcard .depend),)
   include .depend
