Index: linux-2.6.38-rc6/drivers/cbus/retu-wdt.c
===================================================================
--- linux-2.6.38-rc6.orig/drivers/cbus/retu-wdt.c	2011-03-04 15:46:21.312816961 +0100
+++ linux-2.6.38-rc6/drivers/cbus/retu-wdt.c	2011-03-04 15:46:54.658384947 +0100
@@ -124,9 +124,9 @@ static int retu_wdt_release(struct inode
 	struct retu_wdt_dev *wdev = file->private_data;
 
 #ifndef CONFIG_WATCHDOG_NOWAYOUT
-	retu_wdt_ping_enable(retu_wdt);
+	retu_wdt_ping_enable(wdev);
 #endif
-	clear_bit(0, &retu_wdt->users);
+	clear_bit(0, &wdev->users);
 
 	return 0;
 }
