---
 arch/arm/mach-omap2/board-n8x0-lcd.c |   18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

--- a/arch/arm/mach-omap2/board-n8x0-lcd.c
+++ b/arch/arm/mach-omap2/board-n8x0-lcd.c
@@ -34,8 +34,26 @@ static void mipid_shutdown(struct mipid_
 	}
 }
 
+static int n8x0_get_backlight_level(struct mipid_platform_data *pdata)
+{
+	return tahvo_get_backlight_level();
+}
+
+static int n8x0_get_max_backlight_level(struct mipid_platform_data *pdata)
+{
+	return tahvo_get_max_backlight_level();
+}
+
+static void n8x0_set_backlight_level(struct mipid_platform_data *pdata, int level)
+{
+	tahvo_set_backlight_level(level);
+}
+
 struct mipid_platform_data n8x0_mipid_platform_data = {
 	.shutdown = mipid_shutdown,
+	.get_bklight_level = n8x0_get_backlight_level,
+	.set_bklight_level = n8x0_set_backlight_level,
+	.get_bklight_max = n8x0_get_max_backlight_level,
 };
 
 void __init n8x0_mipid_init(void)
