---
 arch/arm/mach-omap2/board-n8x0-lcd.c |   18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

Index: linux-2.6.38-rc6/arch/arm/mach-omap2/board-n8x0-lcd.c
===================================================================
--- linux-2.6.38-rc6.orig/arch/arm/mach-omap2/board-n8x0-lcd.c	2011-02-24 12:37:51.702386250 +0100
+++ linux-2.6.38-rc6/arch/arm/mach-omap2/board-n8x0-lcd.c	2011-02-24 12:39:35.223028410 +0100
@@ -34,8 +34,26 @@
 	}
 }
 
+static int n8x0_get_backlight_level(struct mipid_platform_data *pdata)
+{
+	return tahvo_get_backlight_level();
+}
+
+static int n8x0_get_max_backlight_level(struct mipid_platform_data *pdata)
+{
+	return tahvo_get_max_backlight_level();
+}
+
+static void n8x0_set_backlight_level(struct mipid_platform_data *pdata, int level)
+{
+	tahvo_set_backlight_level(level);
+}
+
 struct mipid_platform_data n8x0_mipid_platform_data = {
 	.shutdown = mipid_shutdown,
+	.get_bklight_level = n8x0_get_backlight_level,
+	.set_bklight_level = n8x0_set_backlight_level,
+	.get_bklight_max = n8x0_get_max_backlight_level,
 };
 
 void __init n8x0_mipid_init(void)
