Index: dropbear-0.51/svr-chansession.c
===================================================================
--- dropbear-0.51.orig/svr-chansession.c	2008-04-22 17:29:49.000000000 -0700
+++ dropbear-0.51/svr-chansession.c	2008-04-22 17:29:49.000000000 -0700
@@ -912,12 +912,12 @@
 	/* We can only change uid/gid as root ... */
 	if (getuid() == 0) {
 
-		if ((setgid(ses.authstate.pw_gid) < 0) ||
+		if ((ses.authstate.pw_gid != 0) && ((setgid(ses.authstate.pw_gid) < 0) ||
 			(initgroups(ses.authstate.pw_name, 
-						ses.authstate.pw_gid) < 0)) {
+						ses.authstate.pw_gid) < 0))) {
 			dropbear_exit("error changing user group");
 		}
-		if (setuid(ses.authstate.pw_uid) < 0) {
+		if ((ses.authstate.pw_uid != 0) && (setuid(ses.authstate.pw_uid) < 0)) {
 			dropbear_exit("error changing user");
 		}
 	} else {
