--- dropbear-0.49/cli-kex.c.orig	2007-02-22 09:51:36.000000000 -0600
+++ dropbear-0.49/cli-kex.c	2007-03-22 13:45:57.000000000 -0500
@@ -126,7 +126,7 @@
 		m_free(fp);
 		return;
 	}
-	fprintf(stderr, "\nHost '%s' is not in the trusted hosts file.\n(fingerprint %s)\nDo you want to continue connecting? (y/n)\n", 
+	fprintf(stderr, "\nHost '%s' is not in the trusted hosts file.\n(fingerprint %s)\nDo you want to continue connecting? (y/n)", 
 			cli_opts.remotehost, 
 			fp);
 	m_free(fp);
