--- a/cli-runopts.c
+++ b/cli-runopts.c
@@ -287,6 +287,10 @@ void cli_getopts(int argc, char ** argv)
 					debug_trace = 1;
 					break;
 #endif
+				case 'o':
+					next = &dummy;
+				case 'x':
+					break;
 				case 'F':
 				case 'e':
 				case 'c':
@@ -298,7 +302,6 @@ void cli_getopts(int argc, char ** argv)
 #ifndef ENABLE_CLI_LOCALTCPFWD
 				case 'L':
 #endif
-				case 'o':
 				case 'b':
 					next = &dummy;
 				default:
