Index: dropbear-0.52/cli-runopts.c
===================================================================
--- dropbear-0.52.orig/cli-runopts.c	2008-04-22 17:29:49.000000000 -0700
+++ dropbear-0.52/cli-runopts.c	2008-04-22 17:29:50.000000000 -0700
@@ -271,6 +271,10 @@
 					debug_trace = 1;
 					break;
 #endif
+				case 'o':
+					next = &dummy;
+				case 'x':
+					break;
 				case 'F':
 				case 'e':
 				case 'c':
@@ -282,7 +286,6 @@
 #ifndef ENABLE_CLI_LOCALTCPFWD
 				case 'L':
 #endif
-				case 'o':
 				case 'b':
 					next = &dummy;
 				default:
