diff -urN dropbear-0.45.old/options.h dropbear-0.45/options.h
--- dropbear-0.45.old/options.h	2005-03-14 17:12:22.000000000 +0100
+++ dropbear-0.45/options.h	2005-03-14 17:13:49.000000000 +0100
@@ -143,7 +143,7 @@
  * however significantly reduce the security of your ssh connections
  * if the PRNG state becomes guessable - make sure you know what you are
  * doing if you change this. */
-#define DROPBEAR_RANDOM_DEV "/dev/random"
+#define DROPBEAR_RANDOM_DEV "/dev/urandom"
 
 /* prngd must be manually set up to produce output */
 /*#define DROPBEAR_PRNGD_SOCKET "/var/run/dropbear-rng"*/
