diff -ur dropbear.old/svr-authpam.c dropbear.dev/svr-authpam.c
--- dropbear.old/svr-authpam.c	2006-03-12 05:52:52.000000000 +0100
+++ dropbear.dev/svr-authpam.c	2007-02-27 02:55:41.382234344 +0100
@@ -31,14 +31,13 @@
 #include "dbutil.h"
 #include "auth.h"
 
+#ifdef ENABLE_SVR_PAM_AUTH
 #if defined(HAVE_SECURITY_PAM_APPL_H)
 #include <security/pam_appl.h>
 #elif defined (HAVE_PAM_PAM_APPL_H)
 #include <pam/pam_appl.h>
 #endif
 
-#ifdef ENABLE_SVR_PAM_AUTH
-
 struct UserDataS {
 	char* user;
 	char* passwd;
