diff -urN dropbear-0.44.old/options.h dropbear-0.44/options.h
--- dropbear-0.44.old/options.h	2005-03-02 23:08:07.000000000 +0100
+++ dropbear-0.44/options.h	2005-03-03 01:12:21.000000000 +0100
@@ -139,7 +139,7 @@
  * will prevent Dropbear from blocking on the device. This could
  * however significantly reduce the security of your ssh connections
  * if the PRNG state becomes simpler. */
-#define DROPBEAR_RANDOM_DEV "/dev/random"
+#define DROPBEAR_RANDOM_DEV "/dev/urandom"
 
 /* prngd must be manually set up to produce output */
 /*#define DROPBEAR_PRNGD_SOCKET "/var/run/dropbear-rng"*/
