diff -ur ulogd-1.23/extensions/ulogd_SYSLOG.c ulogd-1.23-owrt/extensions/ulogd_SYSLOG.c
--- ulogd-1.23/extensions/ulogd_SYSLOG.c	2005-02-12 22:17:41.000000000 +0100
+++ ulogd-1.23-owrt/extensions/ulogd_SYSLOG.c	2006-01-29 19:58:43.000000000 +0100
@@ -136,7 +136,7 @@
 	.name = "syslog", 
 	.init = &syslog_init,
 	.fini = &syslog_fini,
-	.output &_output_syslog
+	.output = &_output_syslog
 };
 
 
diff -ur ulogd-1.23/include/ulogd/ulogd.h ulogd-1.23-owrt/include/ulogd/ulogd.h
--- ulogd-1.23/include/ulogd/ulogd.h	2004-04-25 00:40:54.000000000 +0200
+++ ulogd-1.23-owrt/include/ulogd/ulogd.h	2006-01-29 19:58:40.000000000 +0100
@@ -56,8 +56,6 @@
 #define ULOGD_ERROR	7	/* error condition, requires user action */
 #define ULOGD_FATAL	8	/* fatal, program aborted */
 
-extern FILE *logfile;
-
 typedef struct ulog_iret {
 	/* next interpreter return (key) in the global list */
 	struct ulog_iret *next;
