--- a/arch/mips/alchemy/mtx-1/init.c
+++ b/arch/mips/alchemy/mtx-1/init.c
@@ -33,6 +33,7 @@
 
 #include <asm/bootinfo.h>
 #include <asm/mach-au1x00/au1000.h>
+#include <asm/string.h>
 
 #include <prom.h>
 
@@ -50,7 +51,7 @@ void __init prom_init(void)
 	prom_argv = (char **)fw_arg1;
 	prom_envp = (char **)fw_arg2;
 
-	prom_init_cmdline();
+	strcpy(arcs_cmdline, CONFIG_CMDLINE);
 
 	memsize_str = prom_getenv("memsize");
 	if (!memsize_str)
