--- a/libc/sysdeps/linux/common/bits/termios.h
+++ b/libc/sysdeps/linux/common/bits/termios.h
@@ -156,7 +156,6 @@ struct termios
 #endif
 #define  B57600   0010001
 #define  B115200  0010002
-#if 0 /* limited on uClibc, keep in sync w/ cfsetspeed.c */
 #define  B230400  0010003
 #define  B460800  0010004
 #define  B500000  0010005
@@ -171,9 +170,6 @@ struct termios
 #define  B3500000 0010016
 #define  B4000000 0010017
 #define __MAX_BAUD B4000000
-#else
-#define __MAX_BAUD B115200
-#endif
 #ifdef __USE_MISC
 # define CIBAUD	  002003600000		/* input baud rate (not used) */
 # define CMSPAR   010000000000		/* mark or space (stick) parity */
