Index: linux-2.6.24.7/drivers/leds/Kconfig
===================================================================
--- linux-2.6.24.7.orig/drivers/leds/Kconfig
+++ linux-2.6.24.7/drivers/leds/Kconfig
@@ -146,4 +146,8 @@ config LEDS_TRIGGER_HEARTBEAT
 	  load average.
 	  If unsure, say Y.
 
+config LEDS_TRIGGER_MORSE
+	tristate "LED Morse Trigger"
+	depends on LEDS_TRIGGERS
+
 endif # NEW_LEDS
Index: linux-2.6.24.7/drivers/leds/Makefile
===================================================================
--- linux-2.6.24.7.orig/drivers/leds/Makefile
+++ linux-2.6.24.7/drivers/leds/Makefile
@@ -24,3 +24,4 @@ obj-$(CONFIG_LEDS_CM_X270)              
 obj-$(CONFIG_LEDS_TRIGGER_TIMER)	+= ledtrig-timer.o
 obj-$(CONFIG_LEDS_TRIGGER_IDE_DISK)	+= ledtrig-ide-disk.o
 obj-$(CONFIG_LEDS_TRIGGER_HEARTBEAT)	+= ledtrig-heartbeat.o
+obj-$(CONFIG_LEDS_TRIGGER_MORSE)	+= ledtrig-morse.o
