Index: linux-2.6.23.14/arch/arm/Kconfig
===================================================================
--- linux-2.6.23.14.orig/arch/arm/Kconfig	2008-01-26 02:46:18.000000000 +0100
+++ linux-2.6.23.14/arch/arm/Kconfig	2008-01-26 02:46:33.000000000 +0100
@@ -1039,6 +1039,8 @@
 
 source "drivers/leds/Kconfig"
 
+source "drivers/gpio/Kconfig"
+
 source "drivers/media/Kconfig"
 
 source "drivers/video/Kconfig"
Index: linux-2.6.23.14/drivers/Kconfig
===================================================================
--- linux-2.6.23.14.orig/drivers/Kconfig	2008-01-26 02:45:44.000000000 +0100
+++ linux-2.6.23.14/drivers/Kconfig	2008-01-26 02:46:05.000000000 +0100
@@ -74,6 +74,8 @@
 
 source "drivers/leds/Kconfig"
 
+source "drivers/gpio/Kconfig"
+
 source "drivers/infiniband/Kconfig"
 
 source "drivers/edac/Kconfig"
Index: linux-2.6.23.14/drivers/Makefile
===================================================================
--- linux-2.6.23.14.orig/drivers/Makefile	2008-01-26 02:44:13.000000000 +0100
+++ linux-2.6.23.14/drivers/Makefile	2008-01-26 02:44:42.000000000 +0100
@@ -91,3 +91,4 @@
 obj-$(CONFIG_OF)		+= of/
 obj-$(CONFIG_SSB)		+= ssb/
 obj-$(CONFIG_VIRTIO)		+= virtio/
+obj-$(CONFIG_GPIO_DEVICE)	+= gpio/
