--- a/crypto/Kconfig
+++ b/crypto/Kconfig
@@ -20,7 +20,7 @@ menuconfig CRYPTO
 if CRYPTO
 
 config CRYPTO_ALGAPI
-	tristate
+	tristate "ALG API"
 	help
 	  This option provides the API for cryptographic algorithms.
 
@@ -29,15 +29,15 @@ config CRYPTO_ABLKCIPHER
 	select CRYPTO_BLKCIPHER
 
 config CRYPTO_AEAD
-	tristate
+	tristate "AEAD"
 	select CRYPTO_ALGAPI
 
 config CRYPTO_BLKCIPHER
-	tristate
+	tristate "Block cipher"
 	select CRYPTO_ALGAPI
 
 config CRYPTO_HASH
-	tristate
+	tristate "HASH"
 	select CRYPTO_ALGAPI
 
 config CRYPTO_MANAGER
