--- a/arch/arm/boot/dts/sun8i-h2-plus-orangepi-r1.dts
+++ b/arch/arm/boot/dts/sun8i-h2-plus-orangepi-r1.dts
@@ -65,6 +65,7 @@
 
 	aliases {
 		ethernet1 = &rtl8189etv;
+		spi0 = &spi0;
 	};
 };
 
@@ -73,6 +74,22 @@
 
 	flash@0 {
 		compatible = "mxicy,mx25l12805d", "jedec,spi-nor";
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "u-boot";
+				reg = <0x000000 0x080000>;
+			};
+
+			partition@80000 {
+				compatible = "denx,fit";
+				label = "firmware";
+				reg = <0x080000 0xF80000>;
+			};
+		};
 	};
 };

