Index: ppp-2.4.5/pppd/plugins/pppoatm/pppoatm.c
===================================================================
--- ppp-2.4.5.orig/pppd/plugins/pppoatm/pppoatm.c	2013-11-11 14:36:41.708298496 +0100
+++ ppp-2.4.5/pppd/plugins/pppoatm/pppoatm.c	2013-11-11 14:39:03.556301877 +0100
@@ -133,8 +133,6 @@
 	int fd;
 	struct atm_qos qos;
 
-	system ("/sbin/modprobe -q pppoatm");
-
 	if (!device_got_set)
 		no_device_given_pppoatm();
 	fd = socket(AF_ATMPVC, SOCK_DGRAM, 0);
