From 888b33e84082fe72d60d528c05c885c0c9d70bc4 Mon Sep 17 00:00:00 2001
From: Daniel Schwierzeck <daniel.schwierzeck@gmail.com>
Date: Tue, 6 Nov 2012 21:19:43 +0100
Subject: MIPS: add SPI flash init hook

Signed-off-by: Daniel Schwierzeck <daniel.schwierzeck@gmail.com>

--- a/arch/mips/lib/board.c
+++ b/arch/mips/lib/board.c
@@ -32,6 +32,7 @@
 #include <nand.h>
 #include <onenand_uboot.h>
 #include <spi.h>
+#include <spi_flash.h>
 
 #ifdef CONFIG_BITBANGMII
 #include <miiphy.h>
@@ -312,6 +313,16 @@ void board_init_r(gd_t *id, ulong dest_a
 	onenand_init();
 #endif
 
+#ifdef CONFIG_CMD_SPI
+	puts("SPI:   ");
+	spi_init();		/* go init the SPI */
+	puts("ready\n");
+#endif
+
+#if defined(CONFIG_SPI_FLASH)
+	spi_flash_init();
+#endif
+
 	/* relocate environment function pointers etc. */
 	env_relocate();
 
@@ -335,12 +346,6 @@ void board_init_r(gd_t *id, ulong dest_a
 	/* Initialize from environment */
 	load_addr = getenv_ulong("loadaddr", 16, load_addr);
 
-#ifdef CONFIG_CMD_SPI
-	puts("SPI:   ");
-	spi_init();		/* go init the SPI */
-	puts("ready\n");
-#endif
-
 #if defined(CONFIG_MISC_INIT_R)
 	/* miscellaneous platform dependent initialisations */
 	misc_init_r();
