From 7afbe4633773905ef94a8404510fb5a459926000 Mon Sep 17 00:00:00 2001
From: Daniel Schwierzeck <daniel.schwierzeck@gmail.com>
Date: Mon, 19 Aug 2013 18:11:57 +0200
Subject: MIPS: lantiq: easy80920: add openwrt-lantiq-common.h

Signed-off-by: Daniel Schwierzeck <daniel.schwierzeck@gmail.com>

diff --git a/include/configs/easy80920.h b/include/configs/easy80920.h
index 56ceedb..6f5ac03 100644
--- a/include/configs/easy80920.h
+++ b/include/configs/easy80920.h
@@ -84,13 +84,13 @@
 #define CONFIG_CONSOLE_ASC		1
 #define CONFIG_CONSOLE_DEV		"ttyLTQ1"
 
-/* Commands */
-#define CONFIG_CMD_PING
-
 /* Pull in default board configs for Lantiq XWAY VRX200 */
 #include <asm/lantiq/config.h>
 #include <asm/arch/config.h>
 
+/* Pull in default OpenWrt configs for Lantiq SoC */
+#include "openwrt-lantiq-common.h"
+
 #define CONFIG_ENV_UPDATE_UBOOT_NOR					\
 	"update-uboot-nor=run load-uboot-norspl-lzo write-uboot-nor\0"
 
-- 
1.8.3.2

