diff -ruN freeradius-1.0.5-old/src/modules/rlm_sql/drivers/rules.mak freeradius-1.0.5-new/src/modules/rlm_sql/drivers/rules.mak
--- freeradius-1.0.5-old/src/modules/rlm_sql/drivers/rules.mak	2003-06-05 22:16:54.000000000 +0200
+++ freeradius-1.0.5-new/src/modules/rlm_sql/drivers/rules.mak	2005-10-03 05:33:39.000000000 +0200
@@ -67,7 +67,7 @@
 #
 #######################################################################
 $(TARGET).a: $(STATIC_OBJS)
-	$(LIBTOOL) --mode=link $(LD) -module -static $(CFLAGS) $(RLM_SQL_CFLAGS) $^ -o $@ 
+	$(LIBTOOL) --mode=link $(LD) -module -static $(CFLAGS) $(RLM_SQL_CFLAGS) $(LDFLAGS) $^ -o $@ 
 
 #
 #  If the module is in the list of static modules, then the "dynamic"
@@ -93,7 +93,7 @@
 $(TARGET).la: $(DYNAMIC_OBJS)
 	$(LIBTOOL) --mode=link $(CC) -release $(RADIUSD_VERSION) \
 	-module $(LINK_MODE) $(CFLAGS) \
-	$(RLM_SQL_CFLAGS) -o $@ -rpath $(libdir) $^ $(RLM_SQL_LIBS)
+	$(RLM_SQL_CFLAGS) $(LDFLAGS) -o $@ -rpath $(libdir) $^ $(RLM_SQL_LIBS)
 
 #######################################################################
 #
diff -ruN freeradius-1.0.5-old/src/modules/rules.mak freeradius-1.0.5-new/src/modules/rules.mak
--- freeradius-1.0.5-old/src/modules/rules.mak	2005-07-31 14:48:01.000000000 +0200
+++ freeradius-1.0.5-new/src/modules/rules.mak	2005-10-03 05:35:05.000000000 +0200
@@ -71,7 +71,7 @@
 #######################################################################
 $(TARGET).a: $(STATIC_OBJS)
 	$(LIBTOOL) --mode=link $(LD) \
-	-module -static $(CFLAGS) $(RLM_CFLAGS) $^ -o $@ 
+	-module -static $(CFLAGS) $(RLM_CFLAGS) $(LDFLAGS) $(RLM_LDFLAGS) $^ -o $@ 
 
 #
 #  If the module is in the list of static modules, then the "dynamic"
@@ -96,7 +96,7 @@
 
 $(TARGET).la: $(DYNAMIC_OBJS)
 	$(LIBTOOL) --mode=link $(CC) -release $(RADIUSD_VERSION) \
-	-module $(LINK_MODE) $(LDFLAGS) $(RLM_LDFLAGS) \
+	-module $(LINK_MODE) $(CFLAGS) $(RLM_CFLAGS) $(LDFLAGS) $(RLM_LDFLAGS) \
 	-o $@ -rpath $(libdir) $^ $(RLM_LIBS) $(LIBS)
 
 #######################################################################
