diff -ruN freeradius-1.1.1-old/Makefile freeradius-1.1.1-new/Makefile
--- freeradius-1.1.1-old/Makefile	2005-04-11 23:18:27.000000000 +0200
+++ freeradius-1.1.1-new/Makefile	2006-05-22 08:27:44.000000000 +0200
@@ -38,6 +38,7 @@
 install:
 	$(INSTALL) -d -m 755	$(R)$(sbindir)
 	$(INSTALL) -d -m 755	$(R)$(bindir)
+	$(INSTALL) -d -m 755	$(R)$(libdir)
 	$(INSTALL) -d -m 755	$(R)$(raddbdir)
 	$(INSTALL) -d -m 755	$(R)$(mandir)
 	$(INSTALL) -d -m 755	$(R)$(RUNDIR)
