diff -Nur yaffs2_android.orig/yaffs2/yaffs_guts.h yaffs2_android/yaffs2/yaffs_guts.h
--- yaffs2_android.orig/yaffs2/yaffs_guts.h	Sun Feb 15 16:23:50 2009
+++ yaffs2_android/yaffs2/yaffs_guts.h	Tue Mar  5 15:45:21 2013
@@ -823,6 +823,10 @@
 int yaffs_SetAttributes(yaffs_Object * obj, struct iattr *attr);
 int yaffs_GetAttributes(yaffs_Object * obj, struct iattr *attr);
 
+#if defined(__OpenBSD__)
+#define loff_t off_t
+#endif
+
 /* File operations */
 int yaffs_ReadDataFromFile(yaffs_Object * obj, __u8 * buffer, loff_t offset,
 			   int nBytes);
