Index: linux-2.6.23.17/arch/arm/Kconfig
===================================================================
--- linux-2.6.23.17.orig/arch/arm/Kconfig
+++ linux-2.6.23.17/arch/arm/Kconfig
@@ -1039,6 +1039,8 @@ source "drivers/mfd/Kconfig"
 
 source "drivers/leds/Kconfig"
 
+source "drivers/gpio/Kconfig"
+
 source "drivers/media/Kconfig"
 
 source "drivers/video/Kconfig"
Index: linux-2.6.23.17/drivers/Kconfig
===================================================================
--- linux-2.6.23.17.orig/drivers/Kconfig
+++ linux-2.6.23.17/drivers/Kconfig
@@ -76,6 +76,8 @@ source "drivers/mmc/Kconfig"
 
 source "drivers/leds/Kconfig"
 
+source "drivers/gpio/Kconfig"
+
 source "drivers/infiniband/Kconfig"
 
 source "drivers/edac/Kconfig"
Index: linux-2.6.23.17/drivers/Makefile
===================================================================
--- linux-2.6.23.17.orig/drivers/Makefile
+++ linux-2.6.23.17/drivers/Makefile
@@ -89,3 +89,4 @@ obj-$(CONFIG_DMA_ENGINE)	+= dma/
 obj-$(CONFIG_HID)		+= hid/
 obj-$(CONFIG_PPC_PS3)		+= ps3/
 obj-$(CONFIG_OF)		+= of/
+obj-$(CONFIG_GPIO_DEVICE)	+= gpio/
