Index: aodv-uu-0.9.3/lnx/Makefile
===================================================================
--- aodv-uu-0.9.3.orig/lnx/Makefile	2007-06-04 13:22:19.859836128 +0200
+++ aodv-uu-0.9.3/lnx/Makefile	2007-06-04 13:22:19.932825032 +0200
@@ -30,7 +30,7 @@
 KCFLAGS=-Wall -Wno-strict-aliasing -O2 $(KDEFS) $(KINC) $(XDEFS)
 KCFLAGS_ARM=-Wall -O2 -D__KERNEL__ -DMODULE -nostdinc $(shell $(ARM_CC) -print-search-dirs | sed -ne 's/install: \(.*\)/-I \1include/gp') -I$(KERNEL_INC)
 KCFLAGS_MIPS=-Wall -mips2 -O2 -fno-pic -mno-abicalls -mlong-calls -G0 -msoft-float -D__KERNEL__ -DMODULE -nostdinc $(shell $(MIPS_CC) -print-search-dirs | sed -ne 's/install: \(.*\)/-I \1include/gp') -I$(KERNEL_INC) $(XDEFS)
-
+endif
 .PHONY: clean clean-2.4 clean-2.6 indent default
 
 # Check for kernel version
@@ -89,5 +89,3 @@
 	etags *.c *.h
 indent:
 	indent -kr -i8 -ts8 -sob -l80 -ss -ncs *.c *.h 
-endif
-
Index: aodv-uu-0.9.3/Makefile
===================================================================
--- aodv-uu-0.9.3.orig/Makefile	2007-06-04 13:22:19.866835064 +0200
+++ aodv-uu-0.9.3/Makefile	2007-06-04 13:22:19.932825032 +0200
@@ -87,7 +87,7 @@
 mips: aodvd-mips kaodv-mips
 
 endian.h:
-	$(CC) $(CFLAGS) -o endian endian.c
+	gcc -o endian endian.c
 	./endian > endian.h
 
 $(OBJS): %.o: %.c Makefile
@@ -143,19 +143,18 @@
 	@makedepend -Y./ -- $(DEFS) -- $(SRC) &>/dev/null
 	@makedepend -a -Y./ -- $(KDEFS) kaodv.c &>/dev/null
 
-install: default
-	install -s -m 755 aodvd /usr/sbin/aodvd
-	@if [ ! -d /lib/modules/$(KERNEL)/aodv ]; then \
-		mkdir /lib/modules/$(KERNEL)/aodv; \
-	fi
-
-	@echo "Installing kernel module in /lib/modules/$(KERNEL)/aodv/";
-	@if [ -f ./kaodv.ko ]; then \
-		install -m 644 kaodv.ko /lib/modules/$(KERNEL)/aodv/kaodv.ko; \
-	else \
-		install -m 644 kaodv.o /lib/modules/$(KERNEL)/aodv/kaodv.o; \
-	fi
-	/sbin/depmod -a
+install: install-aodvd install-kaodv
+
+install-aodvd:
+	@echo "Installing aodv in $(DESTDIR)/usr/sbin";
+	install -d -m0755 $(DESTDIR)/usr/sbin/
+	install -m0755 aodvd $(DESTDIR)/usr/sbin/aodvd
+
+install-kaodv:
+	@echo "Installing kernel module in $(DESTDIR)/lib/modules/$(KERNEL)/";
+	install -d -m0644 $(DESTDIR)/lib/modules/$(KERNEL)/
+	install -m 644 lnx/kaodv.$(KMOD_SUFFIX) $(DESTDIR)/lib/modules/$(KERNEL)/kaodv.$(KMOD_SUFFIX);
+
 uninstall:
 	rm -f /usr/sbin/aodvd
 	rm -rf /lib/modules/$(KERNEL)/aodv
