Index: aodv-uu-0.9.3/lnx/kaodv-netlink.c
===================================================================
--- aodv-uu-0.9.3.orig/lnx/kaodv-netlink.c	2007-06-04 13:22:19.810843576 +0200
+++ aodv-uu-0.9.3/lnx/kaodv-netlink.c	2007-06-04 13:22:20.326765144 +0200
@@ -284,7 +284,11 @@
 /* 		RCV_SKB_FAIL(-EINVAL); */
 
 #ifdef KERNEL26
+# if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,19)	
+	if (security_netlink_recv(skb, CAP_NET_ADMIN))
+# else	
 	if (security_netlink_recv(skb))
+# endif
 		RCV_SKB_FAIL(-EPERM);
 #endif
 	//write_lock_bh(&queue_lock);
