Index: aodv-uu-0.9.3/main.c
===================================================================
--- aodv-uu-0.9.3.orig/main.c	2007-06-04 13:22:19.782847832 +0200
+++ aodv-uu-0.9.3/main.c	2007-06-04 13:22:20.513736720 +0200
@@ -26,6 +26,8 @@
 #include <sys/types.h>
 #include <sys/socket.h>
 #include <sys/stat.h>
+#include <linux/types.h>
+#include <linux/if.h>
 #include <linux/sockios.h>
 #include <linux/wireless.h>
 #include <getopt.h>
Index: aodv-uu-0.9.3/nl.c
===================================================================
--- aodv-uu-0.9.3.orig/nl.c	2007-06-04 13:22:19.789846768 +0200
+++ aodv-uu-0.9.3/nl.c	2007-06-04 13:22:20.513736720 +0200
@@ -33,6 +33,10 @@
 #include <netinet/in.h>
 #include <arpa/inet.h>
 #include <linux/rtnetlink.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,19)
+#include <linux/if_addr.h>
+#endif
 
 #include "defs.h"
 #include "lnx/kaodv-netlink.h"
