diff -urN linux-2.6.22.1.old/drivers/mtd/devices/at91_dataflash.c linux-2.6.22.1/drivers/mtd/devices/at91_dataflash.c
--- linux-2.6.22.1.old/drivers/mtd/devices/at91_dataflash.c	2007-07-29 07:18:10.000000000 +0200
+++ linux-2.6.22.1/drivers/mtd/devices/at91_dataflash.c	2007-07-29 07:32:52.000000000 +0200
@@ -161,12 +161,12 @@
 		.mask_flags	= MTD_WRITEABLE,	/* read-only */
 	},
 	{
-		.name		= "kernel",
+		.name		= "knlroot",
 		.offset		= MTDPART_OFS_NXTBLK,
-		.size		= 5 * 32 * 8 * 1056,	/* 5 sectors */
+		.size		= 0x320400,
 	},
 	{
-		.name		= "filesystem",
+		.name		= "jffs2",
 		.offset		= MTDPART_OFS_NXTBLK,
 		.size		= MTDPART_SIZ_FULL,	/* rest = 26 sectors */
 	}
