Index: alsa-driver-1.0.14rc1/include/adriver.h
===================================================================
--- alsa-driver-1.0.14rc1.orig/include/adriver.h	2007-06-04 13:22:44.887031416 +0200
+++ alsa-driver-1.0.14rc1/include/adriver.h	2007-06-04 13:22:44.956020928 +0200
@@ -514,7 +514,7 @@
 #endif /* < 2.6.0 */
 
 /* workarounds for USB API */
-#if defined(SND_NEED_USB_WRAPPER) && (defined(CONFIG_USB) || defined(CONFIG_USB_MODULE))
+#if 1
 
 #include <linux/usb.h>
 
@@ -1360,4 +1360,5 @@
 #endif
 #endif
 
+#undef info /* used in several structs */
 #endif /* __SOUND_LOCAL_DRIVER_H */
