Index: alsa-driver-1.0.14rc1/alsa-kernel/core/sound.c
===================================================================
--- alsa-driver-1.0.14rc1.orig/alsa-kernel/core/sound.c	2007-07-11 12:15:11.408530571 +0200
+++ alsa-driver-1.0.14rc1/alsa-kernel/core/sound.c	2007-07-11 12:15:26.421386105 +0200
@@ -171,7 +171,7 @@
 	return err;
 }
 
-static struct file_operations snd_fops =
+struct file_operations snd_fops =
 {
 	.owner =	THIS_MODULE,
 	.open =		snd_open
