diff -urN linux-2.6.21.1.old/Makefile linux-2.6.21.1.dev/Makefile
--- linux-2.6.21.1.old/Makefile	2007-04-27 23:49:26.000000000 +0200
+++ linux-2.6.21.1.dev/Makefile	2007-05-26 19:14:22.967804016 +0200
@@ -507,6 +507,9 @@
 NOSTDINC_FLAGS += -nostdinc -isystem $(shell $(CC) -print-file-name=include)
 CHECKFLAGS     += $(NOSTDINC_FLAGS)
 
+# improve gcc optimization
+CFLAGS += $(call cc-option,-funit-at-a-time,)
+
 # warn about C99 declaration after statement
 CFLAGS += $(call cc-option,-Wdeclaration-after-statement,)
 
