From fd05f127909bbf4f6983a4de2dcb611947488dee Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Sun, 12 Aug 2012 17:48:50 +0100
Subject: [PATCH] Set prefix on-link bit in RAs

---
 CHANGELOG  |    3 +++
 src/radv.c |    4 ++--
 2 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/CHANGELOG b/CHANGELOG
index 8828694..e1daeef 100644
--- a/CHANGELOG
+++ b/CHANGELOG
@@ -45,6 +45,9 @@ version 2.63
 	    uk.org.thekelleys.dnsmasq. Thanks to Mathieu
 	    Trudel-Lapierre for the patch. 
 
+	    Set the "prefix on-link" bit in Router
+	    Advertisements. Thanks to Gui Iribarren for the patch.
+
 
 version 2.62
             Update German translation. Thanks to Conrad Kostecki.
diff --git a/src/radv.c b/src/radv.c
index a3e691a..d1f5268 100644
--- a/src/radv.c
+++ b/src/radv.c
@@ -412,8 +412,8 @@ static int add_prefixes(struct in6_addr *local,  int prefix,
 		  opt->type = ICMP6_OPT_PREFIX;
 		  opt->len = 4;
 		  opt->prefix_len = prefix;
-		  /* autonomous only if we're not doing dhcp */
-		  opt->flags = do_slaac ? 0x40 : 0x00;
+		  /* autonomous only if we're not doing dhcp, always set "on-link" */
+		  opt->flags = do_slaac ? 0xC0 : 0x80;
 		  opt->valid_lifetime = htonl(time);
 		  opt->preferred_lifetime = htonl(deprecate ? 0 : time);
 		  opt->reserved = 0; 
-- 
1.7.10.GIT

