From 065e5bb0b1c5b9b4ce124ecf35cb2c51feda70e6 Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Fri, 1 Mar 2019 14:38:51 +0000
Subject: [PATCH 36/57] More /etc/hosts linecount fixing.

Signed-off-by: Kevin Darbyshire-Bryant <ldir@darbyshire-bryant.me.uk>
---
 src/cache.c | 1 +
 1 file changed, 1 insertion(+)

--- a/src/cache.c
+++ b/src/cache.c
@@ -1126,6 +1126,7 @@ int read_hostsfile(char *filename, unsig
 	  my_syslog(LOG_ERR, _("bad address at %s line %d"), filename, lineno); 
 	  while (atnl == 0)
 	    atnl = gettok(f, token);
+	  lineno += atnl;
 	  continue;
 	}
       
