From 1da81f7e23962e160b3b776479631d2a6e7f4e49 Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Thu, 28 Mar 2019 13:51:11 +0000
Subject: [PATCH 47/57] CHANGELOG typo fix.

Signed-off-by: Kevin Darbyshire-Bryant <ldir@darbyshire-bryant.me.uk>
---
 CHANGELOG | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/CHANGELOG
+++ b/CHANGELOG
@@ -37,7 +37,7 @@ version 2.81
 	Add --shared-network config. This enables allocation of addresses
 	the DHCP server in subnets where the server (or relay) doesn't
 	have an interface on the network in that subnet. Many thanks to
-	plank.de for sponsoring this feature.
+	kamp.de for sponsoring this feature.
 
 	
 version 2.80
