Index: dnsmasq-2.66/src/ipset.c
===================================================================
--- dnsmasq-2.66.orig/src/ipset.c	2013-04-17 14:52:49.000000000 +0200
+++ dnsmasq-2.66/src/ipset.c	2013-04-24 08:25:05.905712707 +0200
@@ -110,7 +110,7 @@
       return -1;
     }
   
-  memset(buffer, 0, sizeof(buffer));
+  memset(buffer, 0, BUFF_SZ);
 
   nlh = (struct nlmsghdr *)buffer;
   nlh->nlmsg_len = NL_ALIGN(sizeof(struct nlmsghdr));
