--- linux-2.6.22.4.orig/drivers/i2c/busses/Kconfig	2007-08-21 06:33:06.000000000 +0200
+++ linux-2.6.22.4/drivers/i2c/busses/Kconfig	2007-10-09 12:53:13.000000000 +0200
@@ -125,6 +125,40 @@
 	  This is a very simple bitbanging I2C driver utilizing the
 	  arch-neutral GPIO API to control the SCL and SDA lines.
 
+config I2C_GPIO_CUSTOM
+	tristate "Custom GPIO-based I2C device"
+	depends on GENERIC_GPIO
+	select I2C_GPIO
+	help
+	  This is an I2C driver to register a custom i2c-gpio device.
+
+	  This support is also available as a module.  If so, the module
+	  will be called i2c-gpio-dev.
+
+config I2C_GPIO_CUSTOM_SDA
+	int "Custom GPIO pin for SDA"
+	depends on I2C_GPIO_CUSTOM
+	default "0"
+	help
+	  Enter the GPIO pin number used for the SDA signal.  This value can
+	  also be specified with a module parameter.
+
+config I2C_GPIO_CUSTOM_SCL
+	int "Custom GPIO pin for SCL"
+	depends on I2C_GPIO_CUSTOM
+	default "1"
+	help
+	  Enter the GPIO pin number used for the SCL signal.  This value can
+	  also be specified with a module parameter.
+
+config I2C_GPIO_CUSTOM_DEVICE_ID
+	int "Custom GPIO device id"
+	depends on I2C_GPIO_CUSTOM
+	default	"0"
+	help
+	  Enter the number used for the device id of the custom i2c-gpio device.
+	  This value can also be specified with a module parameter.
+
 config I2C_HYDRA
 	tristate "CHRP Apple Hydra Mac I/O I2C interface"
 	depends on PCI && PPC_CHRP && EXPERIMENTAL
--- linux-2.6.22.4.orig/drivers/i2c/busses/Makefile	2007-08-21 06:33:06.000000000 +0200
+++ linux-2.6.22.4/drivers/i2c/busses/Makefile	2007-10-09 12:07:45.000000000 +0200
@@ -13,6 +13,7 @@
 obj-$(CONFIG_I2C_BLACKFIN_TWI)	+= i2c-bfin-twi.o
 obj-$(CONFIG_I2C_ELEKTOR)	+= i2c-elektor.o
 obj-$(CONFIG_I2C_GPIO)		+= i2c-gpio.o
+obj-$(CONFIG_I2C_GPIO_CUSTOM)	+= i2c-gpio-custom.o
 obj-$(CONFIG_I2C_HYDRA)		+= i2c-hydra.o
 obj-$(CONFIG_I2C_I801)		+= i2c-i801.o
 obj-$(CONFIG_I2C_I810)		+= i2c-i810.o
