Index: gcc-4.2.3/gcc/config/avr32/avr32.opt
===================================================================
--- gcc-4.2.3.orig/gcc/config/avr32/avr32.opt	2008-05-21 13:45:54.165287871 +0200
+++ gcc-4.2.3/gcc/config/avr32/avr32.opt	2008-05-21 13:45:57.194288771 +0200
@@ -71,3 +71,6 @@
 Target Report Var(avr32_imm_in_const_pool) Init(-1)
 Put large immediates in constant pool. This is enabled by default for archs with insn-cache.
 
+mno-pic
+Target Report RejectNegative Mask(NO_PIC)
+Do not generate position-independent code.
Index: gcc-4.2.3/gcc/config/avr32/avr32.c
===================================================================
--- gcc-4.2.3.orig/gcc/config/avr32/avr32.c	2008-05-21 13:45:54.145288116 +0200
+++ gcc-4.2.3/gcc/config/avr32/avr32.c	2008-05-21 13:45:57.202289511 +0200
@@ -230,6 +230,9 @@
           avr32_imm_in_const_pool = 0;
       }
     
+    if (TARGET_NO_PIC)
+      flag_pic = 0;
+
     avr32_add_gc_roots ();
   }
 
