diff -ruN ipset-2.2.3/ipset.c ipset-2.2.3.new/ipset.c
--- ipset-2.2.3/ipset.c	2005-04-05 02:56:02.000000000 -0500
+++ ipset-2.2.3.new/ipset.c	2007-05-31 14:05:13.763087750 -0500
@@ -21,7 +21,7 @@
 #include <stdarg.h>
 #include <netdb.h>
 #include <dlfcn.h>
-#include <asm/bitops.h>
+#include <linux/bitops.h>
 
 #include "ipset.h"
 
diff -ruN ipset-2.2.3/ipset_iphash.c ipset-2.2.3.new/ipset_iphash.c
--- ipset-2.2.3/ipset_iphash.c	2005-01-19 07:38:57.000000000 -0600
+++ ipset-2.2.3.new/ipset_iphash.c	2007-05-31 14:05:32.212240750 -0500
@@ -25,7 +25,7 @@
 #include <sys/types.h>
 #include <netinet/in.h>
 #include <arpa/inet.h>
-#include <asm/bitops.h>
+#include <linux/bitops.h>
 #include <asm/types.h>
 
 #include <linux/netfilter_ipv4/ip_set_iphash.h>
diff -ruN ipset-2.2.3/ipset_ipmap.c ipset-2.2.3.new/ipset_ipmap.c
--- ipset-2.2.3/ipset_ipmap.c	2005-01-19 07:38:57.000000000 -0600
+++ ipset-2.2.3.new/ipset_ipmap.c	2007-05-31 14:05:47.969225500 -0500
@@ -22,7 +22,7 @@
 #include <sys/socket.h>
 #include <netinet/in.h>
 #include <arpa/inet.h>
-#include <asm/bitops.h>
+#include <linux/bitops.h>
 
 #include <linux/netfilter_ipv4/ip_set_ipmap.h>
 #include "ipset.h"
diff -ruN ipset-2.2.3/ipset_macipmap.c ipset-2.2.3.new/ipset_macipmap.c
--- ipset-2.2.3/ipset_macipmap.c	2005-01-19 07:38:57.000000000 -0600
+++ ipset-2.2.3.new/ipset_macipmap.c	2007-05-31 14:06:17.627079000 -0500
@@ -24,7 +24,7 @@
 #include <sys/socket.h>
 #include <netinet/in.h>
 #include <arpa/inet.h>
-#include <asm/bitops.h>
+#include <linux/bitops.h>
 #include <linux/if_ether.h>
 
 #include <linux/netfilter_ipv4/ip_set_macipmap.h>
diff -ruN ipset-2.2.3/ipset_nethash.c ipset-2.2.3.new/ipset_nethash.c
--- ipset-2.2.3/ipset_nethash.c	2005-05-09 00:44:25.000000000 -0500
+++ ipset-2.2.3.new/ipset_nethash.c	2007-05-31 14:06:43.392689250 -0500
@@ -25,7 +25,7 @@
 #include <sys/types.h>
 #include <netinet/in.h>
 #include <arpa/inet.h>
-#include <asm/bitops.h>
+#include <linux/bitops.h>
 #include <asm/types.h>
 
 #include <linux/netfilter_ipv4/ip_set_nethash.h>
diff -ruN ipset-2.2.3/ipset_portmap.c ipset-2.2.3.new/ipset_portmap.c
--- ipset-2.2.3/ipset_portmap.c	2005-01-19 07:38:57.000000000 -0600
+++ ipset-2.2.3.new/ipset_portmap.c	2007-05-31 14:06:58.521634750 -0500
@@ -21,7 +21,7 @@
 #include <sys/socket.h>
 #include <netinet/in.h>
 #include <arpa/inet.h>
-#include <asm/bitops.h>
+#include <linux/bitops.h>
 
 #include <linux/netfilter_ipv4/ip_set_portmap.h>
 #include "ipset.h"
