--- a/drivers/mtd/chips/jedec_probe.c
+++ b/drivers/mtd/chips/jedec_probe.c
@@ -105,6 +105,7 @@
 #define MX29LV040C	0x004F
 #define MX29LV160T	0x22C4
 #define MX29LV160B	0x2249
+#define MX29LV640T	0x227e
 #define MX29F040	0x00A4
 #define MX29F016	0x00AD
 #define MX29F002T	0x00B0
@@ -1162,6 +1163,19 @@ static const struct amd_flash_info jedec
 			ERASEINFO(0x04000,1)
 		}
 	}, {
+		.mfr_id		= CFI_MFR_MACRONIX,
+		.dev_id		= MX29LV640T,
+		.name		= "Macronix MX29LV640T",
+		.devtypes	= CFI_DEVICETYPE_X16|CFI_DEVICETYPE_X8,
+		.uaddr		= MTD_UADDR_0x0AAA_0x0555,
+		.dev_size	= SIZE_8MiB,
+		.cmd_set	= P_ID_AMD_STD,
+		.nr_regions	= 2,
+		.regions	= {
+			ERASEINFO(0x10000,127),
+			ERASEINFO(0x02000,8)
+		}
+	}, {
 		.mfr_id		= CFI_MFR_NEC,
 		.dev_id		= UPD29F064115,
 		.name		= "NEC uPD29F064115",
