--- a/configure.ac
+++ b/configure.ac
@@ -62,7 +62,5 @@ AX_CFLAGS_GCC_OPTION([-Wformat])
 AX_CFLAGS_GCC_OPTION([-Wformat-security])
 AX_CFLAGS_GCC_OPTION([-Wcast-align])
 AX_CFLAGS_GCC_OPTION([-Winline])
-AX_CFLAGS_GCC_OPTION([-fstack-protector])
-AX_CFLAGS_GCC_OPTION([-fstack-protector-strong])
 AX_CFLAGS_GCC_OPTION([-fno-omit-frame-pointer])
 AX_CFLAGS_GCC_OPTION([-D_FORTIFY_SOURCE=2])
 AX_CFLAGS_GCC_OPTION([-Wno-unused-parameter])
