From 6fa3166b901b1f3a88e52c232c09df48c430303d Mon Sep 17 00:00:00 2001
From: popcornmix <popcornmix@gmail.com>
Date: Tue, 11 Feb 2014 17:03:51 +0000
Subject: [PATCH 166/196] bcm2708: Allow disk activity led gpio to be specified

---
 arch/arm/mach-bcm2708/bcm2708.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/arch/arm/mach-bcm2708/bcm2708.c b/arch/arm/mach-bcm2708/bcm2708.c
index 993379c..caa9f48 100644
--- a/arch/arm/mach-bcm2708/bcm2708.c
+++ b/arch/arm/mach-bcm2708/bcm2708.c
@@ -83,6 +83,8 @@
 /* command line parameters */
 static unsigned boardrev, serial;
 static unsigned uart_clock;
+static unsigned disk_led_gpio = 16;
+static unsigned disk_led_active_low = 1;
 static unsigned reboot_part = 0;
 static unsigned w1_gpio_pin = W1_GPIO;
 
@@ -965,7 +967,9 @@ static struct platform_device bcm2708_led_device = {
 
 static void __init bcm2708_init_led(void)
 {
-	platform_device_register(&bcm2708_led_device);
+  bcm2708_leds[0].gpio = disk_led_gpio;
+  bcm2708_leds[0].active_low = disk_led_active_low;
+  platform_device_register(&bcm2708_led_device);
 }
 #else
 static inline void bcm2708_init_led(void)
@@ -1004,5 +1008,7 @@ MACHINE_END
 module_param(boardrev, uint, 0644);
 module_param(serial, uint, 0644);
 module_param(uart_clock, uint, 0644);
+module_param(disk_led_gpio, uint, 0644);
+module_param(disk_led_active_low, uint, 0644);
 module_param(reboot_part, uint, 0644);
 module_param(w1_gpio_pin, uint, 0644);
-- 
1.9.1

