--- a/arch/powerpc/boot/dts/mpc8548cds.dts
+++ b/arch/powerpc/boot/dts/mpc8548cds.dts
@@ -308,21 +308,21 @@
 			};
 		};
 
-		serial0: serial@4500 {
-			cell-index = <0>;
+		serial0: serial@4600 {
+			cell-index = <1>;
 			device_type = "serial";
 			compatible = "ns16550";
-			reg = <0x4500 0x100>;	// reg base, size
+			reg = <0x4600 0x100>;	// reg base, size
 			clock-frequency = <0>;	// should we fill in in uboot?
 			interrupts = <42 2>;
 			interrupt-parent = <&mpic>;
 		};
 
-		serial1: serial@4600 {
-			cell-index = <1>;
+		serial1: serial@4500 {
+			cell-index = <0>;
 			device_type = "serial";
 			compatible = "ns16550";
-			reg = <0x4600 0x100>;	// reg base, size
+			reg = <0x4500 0x100>;	// reg base, size
 			clock-frequency = <0>;	// should we fill in in uboot?
 			interrupts = <42 2>;
 			interrupt-parent = <&mpic>;
@@ -564,4 +564,7 @@
 				  0x0 0x100000>;
 		};
 	};
+	chosen {
+		linux,stdout-path = "/soc8548@e0000000/serial@4600";
+	};
 };
