diff -urN vncrepeater-0.12/Makefile vncrepeater-0.12.new/Makefile
--- vncrepeater-0.12/Makefile	2006-03-31 07:40:14.000000000 +0200
+++ vncrepeater-0.12.new/Makefile	2006-04-11 10:26:06.345879104 +0200
@@ -1,21 +1,21 @@
-CFLAGS=-Wall
+CFLAGS=-Wall 
 repeater: repeater.o repeaterproc.o openbsd_stringfuncs.o iniparser.o readini.o
-	g++ $(CFLAGS) -o repeater repeater.o repeaterproc.o openbsd_stringfuncs.o iniparser.o readini.o
+	$(CC) $(CFLAGS) -o repeater repeater.o repeaterproc.o openbsd_stringfuncs.o iniparser.o readini.o -nodefaultlibs -L$(STAGING_DIR)/usr/lib -L$(STAGING_DIR)/lib -luClibc++
 
 repeater.o: repeater.cpp
-	g++ $(CFLAGS) -c repeater.cpp
+	$(CC) $(CFLAGS) -c repeater.cpp
 
 repeaterproc.o: repeaterproc.cpp
-	g++ $(CFLAGS) -c repeaterproc.cpp
+	$(CC) $(CFLAGS) -c repeaterproc.cpp
 
 openbsd_stringfuncs.o: openbsd_stringfuncs.cpp
-	g++ $(CFLAGS) -c openbsd_stringfuncs.cpp
+	$(CC) $(CFLAGS) -c openbsd_stringfuncs.cpp
 
 iniparser.o: iniparser.cpp
-	g++ $(CFLAGS) -c iniparser.cpp
+	$(CC) $(CFLAGS) -c iniparser.cpp
 
 readini.o: readini.cpp
-	g++ $(CFLAGS) -c readini.cpp
+	$(CC) $(CFLAGS) -c readini.cpp
 
 clean:
 	rm -f *.o repeater
