--- a/drivers/watchdog/Kconfig
+++ b/drivers/watchdog/Kconfig
@@ -632,6 +632,18 @@ config SBC_EPX_C3_WATCHDOG
 
 # MIPS Architecture
 
+config ADM5120_WDT
+	tristate "Infineon ADM5120 SoC hardware watchdog"
+	depends on WATCHDOG && ADM5120
+	help
+	  This is a driver for hardware watchdog integrated in Infineon
+	  ADM5120 SoC. This watchdog simply watches your kernel to make sure
+	  it doesn't freeze, and if it does, it reboots your computer after a
+	  certain amount of time.
+
+	  To compile this driver as a module, choose M here: the module will be
+	  called adm5120_wdt.
+
 config INDYDOG
 	tristate "Indy/I2 Hardware Watchdog"
 	depends on SGI_HAS_INDYDOG
--- a/drivers/watchdog/Makefile
+++ b/drivers/watchdog/Makefile
@@ -91,6 +91,7 @@ obj-$(CONFIG_SBC_EPX_C3_WATCHDOG) += sbc
 # M68KNOMMU Architecture
 
 # MIPS Architecture
+obj-$(CONFIG_ADM5120_WDT) += adm5120_wdt.o
 obj-$(CONFIG_INDYDOG) += indydog.o
 obj-$(CONFIG_WDT_MTX1)	+= mtx-1_wdt.o
 obj-$(CONFIG_WDT_RM9K_GPI) += rm9k_wdt.o
