--- a/arch/arm/mach-gemini/board-wbd111.c
+++ b/arch/arm/mach-gemini/board-wbd111.c
@@ -22,9 +22,29 @@
 #include <asm/mach/arch.h>
 #include <asm/mach/time.h>
 
+#include <mach/gmac.h>
 
 #include "common.h"
 
+static struct mdio_gpio_platform_data wbd111_mdio = {
+	.mdc		= 22,
+	.mdio		= 21,
+	.phy_mask	= ~(1 << 1),
+};
+
+static struct platform_device wbd111_phy_device = {
+	.name	= "mdio-gpio",
+	.id	= 0,
+	.dev	= {
+		.platform_data = &wbd111_mdio,
+	},
+};
+
+static struct gemini_gmac_platform_data gmac_data = {
+	.bus_id[0] = "gpio-0:01",
+	.interface[0] = PHY_INTERFACE_MODE_MII,
+};
+
 static struct gpio_keys_button wbd111_keys[] = {
 	{
 		.code		= KEY_SETUP,
@@ -127,6 +147,8 @@ static void __init wbd111_init(void)
 	platform_device_register(&wbd111_keys_device);
 	platform_register_rtc();
 	platform_register_watchdog();
+	platform_device_register(&wbd111_phy_device);
+	platform_register_ethernet(&gmac_data);
 }
 
 MACHINE_START(WBD111, "Wiliboard WBD-111")
