diff -ruN poptop-old/Makefile.in poptop-new/Makefile.in
--- poptop-old/Makefile.in	2003-02-06 17:39:45.000000000 +0100
+++ poptop-new/Makefile.in	2004-04-04 15:50:17.000000000 +0200
@@ -74,8 +74,9 @@
 AUTOMAKE_OPTIONS = no-dependencies foreign
 
 INCLUDES = -I.
-CFLAGS = -O2 -fno-builtin -Wall -ansi -DSBINDIR='"$(sbindir)"'
-#CFLAGS = -O2 -fno-builtin -Wall -ansi -pedantic -Wmissing-prototypes -Werror -DSBINDIR='"$(sbindir)"'
+CCOPTS = -O2 -fno-builtin
+CFLAGS = $(CCOPTS) -Wall -ansi -DSBINDIR='"$(sbindir)"'
+#CFLAGS = $(CCOPTS) -Wall -ansi -pedantic -Wmissing-prototypes -Werror -DSBINDIR='"$(sbindir)"'
 
 man_MANS = pptpctrl.8 pptpd.8 pptpd.conf.5
 
