diff -ruN madwifi-ng-r1590-20060523/net80211/ieee80211_linux.c madwifi-ng-r1590-20060523-patch/net80211/ieee80211_linux.c
--- madwifi-ng-r1590-20060523/net80211/ieee80211_linux.c	2006-06-13 14:26:25 -04:00
+++ madwifi-ng-r1590-20060523-patch/net80211/ieee80211_linux.c	2006-06-13 14:26:41 -04:00
@@ -44,6 +44,7 @@
 #include <linux/etherdevice.h>
 #include <linux/if_vlan.h>
 #include <linux/vmalloc.h>
+#include <linux/slab.h>
 
 #include <net/iw_handler.h>
 #include <linux/wireless.h>
