--- a/init/main.c
+++ b/init/main.c
@@ -831,11 +831,11 @@ static noinline int init_post(void)
 	 * The Bourne shell can be used instead of init if we are
 	 * trying to recover a really broken machine.
 	 */
-	if (execute_command) {
-		run_init_process(execute_command);
-		printk(KERN_WARNING "Failed to execute %s.  Attempting "
-					"defaults...\n", execute_command);
-	}
+	//if (execute_command) {
+	//	run_init_process(execute_command);
+	//	printk(KERN_WARNING "Failed to execute %s.  Attempting "
+	//				"defaults...\n", execute_command);
+	//}
 	run_init_process("/etc/preinit");
 
 	panic("No init found.  Try passing init= option to kernel.");
