--- a/lib/Makefile
+++ b/lib/Makefile
@@ -36,6 +36,7 @@ CFLAGS   += -fPIC
 OUT_SLIB := $(PACKAGE_NAME).so.$(PACKAGE_VERSION)
 LN_SLIB  := $(PACKAGE_NAME).so
 LN1_SLIB := $(LN_SLIB).1
+OUT_ALIB := $(PACKAGE_NAME).a
 
 export
 
@@ -44,7 +45,7 @@ export
 
 all:
 	@echo "  MAKE $(OUT_SLIB)"; \
-	$(MAKE) $(OUT_SLIB)
+	$(MAKE) $(OUT_SLIB) $(OUT_ALIB)
 
 $(OUT_SLIB): ../Makefile.opts $(OBJ)
 	@echo "  LD $(OUT_SLIB)"; \
@@ -54,10 +55,15 @@ $(OUT_SLIB): ../Makefile.opts $(OBJ)
 	@echo "  LN $(LN1_SLIB) $(LN_SLIB)"; \
 	rm -f $(LN_SLIB) ; $(LN) -s $(LN1_SLIB) $(LN_SLIB)
 
+$(OUT_ALIB): ../Makefile.opts $(OBJ)
+	@echo "  AR $@"; \
+	$(AR) rcu $@ $(OBJ); \
+	$(RANLIB) $@
+
 clean:
 	@echo "  CLEAN lib"; \
 	$(RM) -f $(OBJ) $(OUT_SLIB) $(LN_SLIB) $(LN1_SLIB); \
-	$(RM) -f $(DEPS) $(OUT_SLIB) $(LN_SLIB) $(LN1_SLIB)
+	$(RM) -f $(DEPS) $(OUT_ALIB) $(LN_SLIB) $(LN1_SLIB)
 
 distclean:
 	@echo "  DISTCLEAN lib"; \
@@ -66,6 +72,7 @@ distclean:
 install:
 	mkdir -p $(DESTDIR)$(libdir)/
 	install -m 0644 $(OUT_SLIB) $(DESTDIR)$(libdir)
+	install -m 0644 $(OUT_ALIB) $(DESTDIR)$(libdir)
 	rm -f $(DESTDIR)$(libdir)/$(LN1_SLIB)
 	$(LN) -s $(OUT_SLIB) $(DESTDIR)$(libdir)/$(LN1_SLIB)
 	rm -f $(DESTDIR)$(libdir)/$(LN_SLIB)
