Index: linux-2.6.23.17/drivers/Kconfig
===================================================================
--- linux-2.6.23.17.orig/drivers/Kconfig
+++ linux-2.6.23.17/drivers/Kconfig
@@ -93,4 +93,6 @@ source "drivers/kvm/Kconfig"
 source "drivers/uio/Kconfig"
 
 source "drivers/lguest/Kconfig"
+
+source "drivers/vlynq/Kconfig"
 endmenu
Index: linux-2.6.23.17/drivers/Makefile
===================================================================
--- linux-2.6.23.17.orig/drivers/Makefile
+++ linux-2.6.23.17/drivers/Makefile
@@ -89,4 +89,5 @@ obj-$(CONFIG_DMA_ENGINE)	+= dma/
 obj-$(CONFIG_HID)		+= hid/
 obj-$(CONFIG_PPC_PS3)		+= ps3/
 obj-$(CONFIG_OF)		+= of/
+obj-$(CONFIG_VLYNQ)		+= vlynq/
 obj-$(CONFIG_GPIO_DEVICE)	+= gpio/
