--- a/arch/arm/mach-at91/board-flexibity.c
+++ b/arch/arm/mach-at91/board-flexibity.c
@@ -90,49 +90,49 @@ static struct gpio_led flexibity_leds[]
 		.name			= "usb1:green",
 		.gpio			= AT91_PIN_PA12,
 		.active_low		= 1,
-		.default_trigger	= "default-on",
+		.default_trigger	= "heartbeat",
 	},
 	{
 		.name			= "usb1:red",
 		.gpio			= AT91_PIN_PA13,
 		.active_low		= 1,
-		.default_trigger	= "default-on",
+		.default_trigger	= "heartbeat",
 	},
 	{
 		.name			= "usb2:green",
 		.gpio			= AT91_PIN_PB26,
 		.active_low		= 1,
-		.default_trigger	= "default-on",
+		.default_trigger	= "heartbeat",
 	},
 	{
 		.name			= "usb2:red",
 		.gpio			= AT91_PIN_PB27,
 		.active_low		= 1,
-		.default_trigger	= "default-on",
+		.default_trigger	= "heartbeat",
 	},
 	{
 		.name			= "usb3:green",
 		.gpio			= AT91_PIN_PC8,
 		.active_low		= 1,
-		.default_trigger	= "default-on",
+		.default_trigger	= "heartbeat",
 	},
 	{
 		.name			= "usb3:red",
 		.gpio			= AT91_PIN_PC6,
 		.active_low		= 1,
-		.default_trigger	= "default-on",
+		.default_trigger	= "heartbeat",
 	},
 	{
 		.name			= "usb4:green",
 		.gpio			= AT91_PIN_PB4,
 		.active_low		= 1,
-		.default_trigger	= "default-on",
+		.default_trigger	= "heartbeat",
 	},
 	{
 		.name			= "usb4:red",
 		.gpio			= AT91_PIN_PB5,
 		.active_low		= 1,
-		.default_trigger	= "default-on",
+		.default_trigger	= "heartbeat",
 	}
 };
 
