--- a/drivers/usb/host/ehci-ar71xx.c
+++ b/drivers/usb/host/ehci-ar71xx.c
@@ -79,7 +79,7 @@ static int ehci_ar71xx_probe(const struc
 	res = platform_get_resource(pdev, IORESOURCE_IRQ, 0);
 	if (!res) {
 		dev_dbg(&pdev->dev, "no IRQ specified for %s\n",
-			pdev->dev.bus_id);
+			dev_name(&pdev->dev));
 		return -ENODEV;
 	}
 	irq = res->start;
@@ -87,11 +87,11 @@ static int ehci_ar71xx_probe(const struc
 	res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
 	if (!res) {
 		dev_dbg(&pdev->dev, "no base address specified for %s\n",
-			pdev->dev.bus_id);
+			dev_name(&pdev->dev));
 		return -ENODEV;
 	}
 
-	hcd = usb_create_hcd(driver, &pdev->dev, pdev->dev.bus_id);
+	hcd = usb_create_hcd(driver, &pdev->dev, dev_name(&pdev->dev));
 	if (!hcd)
 		return -ENOMEM;
 
@@ -205,7 +205,7 @@ static int ehci_ar71xx_driver_probe(stru
 	pdata = pdev->dev.platform_data;
 	if (!pdata) {
 		dev_err(&pdev->dev, "no platform data specified for %s\n",
-			pdev->dev.bus_id);
+			dev_name(&pdev->dev));
 		return -ENODEV;
 	}
 
