diff -urN linux-2.6.19.old/include/asm-powerpc/segment.h linux-2.6.19.dev/include/asm-powerpc/segment.h
--- linux-2.6.19.old/include/asm-powerpc/segment.h	1970-01-01 01:00:00.000000000 +0100
+++ linux-2.6.19.dev/include/asm-powerpc/segment.h	2006-12-14 03:13:59.000000000 +0100
@@ -0,0 +1,6 @@
+#ifndef _ASM_SEGMENT_H
+#define _ASM_SEGMENT_H
+
+/* Only here because we have some old header files that expect it.. */
+
+#endif /* _ASM_SEGMENT_H */
