diff -urN linux-2.6.21.1.old/drivers/mtd/devices/block2mtd.c linux-2.6.21.1.dev/drivers/mtd/devices/block2mtd.c
--- linux-2.6.21.1.old/drivers/mtd/devices/block2mtd.c	2007-05-26 20:13:39.192175736 +0200
+++ linux-2.6.21.1.dev/drivers/mtd/devices/block2mtd.c	2007-05-26 21:07:17.921854448 +0200
@@ -41,7 +41,7 @@
 static LIST_HEAD(blkmtd_device_list);
 
 
-#define PAGE_READAHEAD 64
+#define PAGE_READAHEAD 0
 static void cache_readahead(struct address_space *mapping, int index)
 {
 	filler_t *filler = (filler_t*)mapping->a_ops->readpage;
