diff --exclude=.svn -pNur  glibc-ports-2.4/sysdeps/unix/sysv/linux/mips/nptl/fork.c glibc-ports-2.4/sysdeps/unix/sysv/linux/mips/nptl/fork.c
---  glibc-ports-2.4/sysdeps/unix/sysv/linux/mips/nptl/fork.c	2009-02-09 15:13:14.996030533 +0100
+++ glibc-ports-2.4/sysdeps/unix/sysv/linux/mips/nptl/fork.c	2009-02-09 15:13:34.402489340 +0100
@@ -1 +1,2 @@
-#include "../i386/fork.c"
+#include <sysdeps/unix/sysv/linux/i386/fork.c>
+
