Index: arptables-v0.0.3-2/include/linux/netfilter_arp.h
===================================================================
--- arptables-v0.0.3-2.orig/include/linux/netfilter_arp.h	2007-06-04 13:22:21.014660568 +0200
+++ arptables-v0.0.3-2/include/linux/netfilter_arp.h	2007-06-04 13:22:21.087649472 +0200
@@ -7,7 +7,7 @@
  * (C)2002 Rusty Russell IBM -- This code is GPL.
  */
 
-#include <linux/config.h>
+#include <linux/autoconf.h>
 #include <linux/netfilter.h>
 
 /* There is no PF_ARP. */
Index: arptables-v0.0.3-2/include/netfilter_arp.h
===================================================================
--- arptables-v0.0.3-2.orig/include/netfilter_arp.h	2007-06-04 13:22:21.021659504 +0200
+++ arptables-v0.0.3-2/include/netfilter_arp.h	2007-06-04 13:22:21.087649472 +0200
@@ -7,7 +7,7 @@
  * (C)2002 Rusty Russell IBM -- This code is GPL.
  */
 
-#include <linux/config.h>
+#include <linux/autoconf.h>
 #include <linux/netfilter.h>
 
 /* There is no PF_ARP. */
