diff -ur arptables.old/include/linux/netfilter_arp.h arptables.dev/include/linux/netfilter_arp.h
--- arptables.old/include/linux/netfilter_arp.h	2003-10-25 13:39:06.000000000 +0200
+++ arptables.dev/include/linux/netfilter_arp.h	2006-12-14 20:35:10.000000000 +0100
@@ -7,7 +7,7 @@
  * (C)2002 Rusty Russell IBM -- This code is GPL.
  */
 
-#include <linux/config.h>
+#include <linux/autoconf.h>
 #include <linux/netfilter.h>
 
 /* There is no PF_ARP. */
diff -ur arptables.old/include/netfilter_arp.h arptables.dev/include/netfilter_arp.h
--- arptables.old/include/netfilter_arp.h	2003-10-25 12:43:24.000000000 +0200
+++ arptables.dev/include/netfilter_arp.h	2006-12-14 20:34:57.000000000 +0100
@@ -7,7 +7,7 @@
  * (C)2002 Rusty Russell IBM -- This code is GPL.
  */
 
-#include <linux/config.h>
+#include <linux/autoconf.h>
 #include <linux/netfilter.h>
 
 /* There is no PF_ARP. */
