--- a/kernel/printk.c
+++ b/kernel/printk.c
@@ -44,6 +44,10 @@
 
 #include <asm/uaccess.h>
 
+#ifdef CONFIG_DEBUG_LL
+extern void printascii(char *);
+#endif /* CONFIG_DEBUG_LL */
+
 /*
  * Architectures can override it:
  */
@@ -900,6 +904,10 @@ asmlinkage int vprintk(const char *fmt,
 		}
 	}
 
+#ifdef CONFIG_DEBUG_LL
+	printascii(printk_buf);
+#endif
+
 	/*
 	 * Copy the output into log_buf. If the caller didn't provide
 	 * the appropriate log prefix, we insert them here
