From e825d46ed778699b2e96adbe7e947a2d3c0f092b Mon Sep 17 00:00:00 2001
From: Eric Anholt <eric@anholt.net>
Date: Mon, 25 Jan 2016 13:03:33 -0800
Subject: [PATCH] drm/vc4: Include vc4_drm.h in uapi in downstream build.

Signed-off-by: Eric Anholt <eric@anholt.net>
---
 include/uapi/drm/Kbuild | 1 +
 1 file changed, 1 insertion(+)

--- a/include/uapi/drm/Kbuild
+++ b/include/uapi/drm/Kbuild
@@ -14,6 +14,7 @@ header-y += radeon_drm.h
 header-y += savage_drm.h
 header-y += sis_drm.h
 header-y += tegra_drm.h
+header-y += vc4_drm.h
 header-y += via_drm.h
 header-y += vmwgfx_drm.h
 header-y += msm_drm.h
