diff -urN linux-2.6.19.2/drivers/mtd/chips/jedec_probe.c linux-2.6.19.2.new/drivers/mtd/chips/jedec_probe.c
--- linux-2.6.19.2/drivers/mtd/chips/jedec_probe.c	2007-01-10 20:10:37.000000000 +0100
+++ linux-2.6.19.2.new/drivers/mtd/chips/jedec_probe.c	2007-04-09 22:33:05.000000000 +0200
@@ -158,6 +158,7 @@
 #define SST49LF030A	0x001C
 #define SST49LF040A	0x0051
 #define SST49LF080A	0x005B
+#define SST39VF6402B   	0x236C
 
 /* Toshiba */
 #define TC58FVT160	0x00C2
@@ -1494,7 +1495,23 @@
                        ERASEINFO(0x1000,256),
                        ERASEINFO(0x1000,256)
                }
-
+	}, {
+		.mfr_id         = MANUFACTURER_SST,
+		.dev_id         = SST39VF6402B,
+		.name           = "SST 39VF6402B",
+		.uaddr          = {
+			[0] = MTD_UADDR_0x5555_0x2AAA,  /* x8 */
+			[1] = MTD_UADDR_0x5555_0x2AAA   /* x16 */
+		},
+		.DevSize        = SIZE_8MiB,
+		.CmdSet         = P_ID_AMD_STD,
+		.NumEraseRegions= 4,
+		.regions        = {
+			ERASEINFO(0x2000,256),
+			ERASEINFO(0x2000,256),
+			ERASEINFO(0x2000,256),
+			ERASEINFO(0x2000,256)
+		}
        }, {
 		.mfr_id		= MANUFACTURER_ST,	/* FIXME - CFI device? */
 		.dev_id		= M29W800DT,
