Index: ntfs-3g-1.0/src/Makefile.in
===================================================================
--- ntfs-3g-1.0.orig/src/Makefile.in	2007-06-04 13:23:57.444001080 +0200
+++ ntfs-3g-1.0/src/Makefile.in	2007-06-04 13:23:57.567982232 +0200
@@ -582,7 +582,7 @@
 install-exec-hook:
 	$(INSTALL) -d $(DESTDIR)/sbin
 	$(LN_S) -f $(bindir)/ntfs-3g $(DESTDIR)/sbin/mount.ntfs-3g
-	$(LDCONFIG)
+#	$(LDCONFIG)
 
 install-data-hook:
 	$(INSTALL) -d $(DESTDIR)$(man8dir)
