--- a/configure.in
+++ b/configure.in
@@ -64,7 +64,7 @@ dnl Set kernel build path
 AC_ARG_ENABLE(kernelbuild,
 	AS_HELP_STRING(--enable-kernelbuild=x,Set the target kernel build path (only for kernel 2.6.x)),
 	[
-		if test -e $enableval/include/linux/autoconf.h; then
+		if test -e $enableval/include/linux/autoconf.h -o -e $enableval/include/generated/autoconf.h; then
 			AC_SUBST([KERNEL_BUILD_PATH],[$enableval])
 		else
 			AC_MSG_ERROR([The kernel build directory is not valid or not configured!])
--- a/src/linux/ifxos_linux_thread_drv.c
+++ b/src/linux/ifxos_linux_thread_drv.c
@@ -34,7 +34,6 @@
 #include <linux/sched.h>
 #include <linux/version.h>
 #include <linux/completion.h>
-#include <linux/smp_lock.h>
 #include <linux/signal.h>
 
 
