--- ulogd-1.23/pcap/Makefile.in.orig	2004-03-23 23:36:04.000000000 +0100
+++ ulogd-1.23/pcap/Makefile.in	2005-05-16 00:14:13.000000000 +0200
@@ -14,10 +14,10 @@
 distrib:
 
 $(SHARED_LIBS): %.so: %_sh.o
-	$(LD) -shared -o $@ $< -lc -lpcap
+	$(LD) -shared $(PCAP_LDFLAGS) -o $@ $< -lc -lpcap
 
 %_sh.o: %.c
-	$(CC) $(SH_CFLAGS) -o $@ -c $<
+	$(CC) $(PCAP_CFLAGS) $(SH_CFLAGS) -o $@ -c $<
 	
 clean:
 	$(RM) $(SHARED_LIBS) *.o
