From a292696427853a269e28f42cb32f01942fd5d4c5 Mon Sep 17 00:00:00 2001
From: Eneas U de Queiroz <cote2004-github@yahoo.com>
Date: Thu, 27 Sep 2018 08:18:22 -0300
Subject: Do not build tests and fuzz directories

This shortens build time.

Signed-off-by: Eneas U de Queiroz <cote2004-github@yahoo.com>

diff --git a/Configure b/Configure
index 9080b4611c..1112f40226 100755
--- a/Configure
+++ b/Configure
@@ -306,7 +306,7 @@ my $default_ranlib;
 $config{fips}=0;
 
 # Top level directories to build
-$config{dirs} = [ "crypto", "ssl", "engines", "apps", "test", "util", "tools", "fuzz" ];
+$config{dirs} = [ "crypto", "ssl", "engines", "apps", "util", "tools" ];
 # crypto/ subdirectories to build
 $config{sdirs} = [
     "objects",
-- 
2.16.4

