--- a/Documentation/Configure.help
+++ b/Documentation/Configure.help
@@ -10687,6 +10687,20 @@ CONFIG_TUN
 
   If you don't know what to use this for, you don't need it.
 
+Allow Net Devices to contribute to /dev/random
+CONFIG_NET_RANDOM
+  If you say Y here, network device interrupts will contribute to the
+  kernel entropy pool. Normally, block devices and some other devices
+  feed the pool. Some systems, such as those that are headless or diskless,
+  need additional entropy sources.  Some people, however, feel that network
+  devices should not contribute to /dev/random because an external attacker
+  could observe incoming packets in an attempt to learn the entropy pool's
+  state.  If you say N, no network device will contribute entropy.
+
+  If you believe there is a chance of your network packets being observed
+  and you doubt the security of the entropy pool's one-way hash, do not
+  enable this.  If unsure, say N.
+
 Ethertap network tap (OBSOLETE)
 CONFIG_ETHERTAP
   If you say Y here (and have said Y to "Kernel/User network link
--- a/drivers/net/Config.in
+++ b/drivers/net/Config.in
@@ -8,6 +8,7 @@ tristate 'Dummy net driver support' CONF
 tristate 'Bonding driver support' CONFIG_BONDING
 tristate 'EQL (serial line load balancing) support' CONFIG_EQUALIZER
 tristate 'Universal TUN/TAP device driver support' CONFIG_TUN
+bool 'Allow Net Devices to contribute to /dev/random' CONFIG_NET_RANDOM
 if [ "$CONFIG_EXPERIMENTAL" = "y" ]; then
    tristate 'Ethertap network tap (OBSOLETE)' CONFIG_ETHERTAP
 fi
--- a/include/asm-alpha/signal.h
+++ b/include/asm-alpha/signal.h
@@ -121,8 +121,15 @@ typedef unsigned long sigset_t;
 #define SA_PROBE		SA_ONESHOT
 #define SA_SAMPLE_RANDOM	SA_RESTART
 #define SA_SHIRQ		0x40000000
+
+#ifdef CONFIG_NET_RANDOM
+#define SA_NET_RANDOM	SA_SAMPLE_RANDOM
+#else
+#define SA_NET_RANDOM	0
 #endif
 
+#endif /* __KERNEL__ */
+
 #define SIG_BLOCK          1	/* for blocking signals */
 #define SIG_UNBLOCK        2	/* for unblocking signals */
 #define SIG_SETMASK        3	/* for setting the signal mask */
--- a/include/asm-arm/signal.h
+++ b/include/asm-arm/signal.h
@@ -125,8 +125,15 @@ typedef unsigned long sigset_t;
 #define SA_PROBE		0x80000000
 #define SA_SAMPLE_RANDOM	0x10000000
 #define SA_SHIRQ		0x04000000
+
+#ifdef CONFIG_NET_RANDOM
+#define SA_NET_RANDOM	SA_SAMPLE_RANDOM
+#else
+#define SA_NET_RANDOM	0
 #endif
 
+#endif /* __KERNEL__ */
+
 #define SIG_BLOCK          0	/* for blocking signals */
 #define SIG_UNBLOCK        1	/* for unblocking signals */
 #define SIG_SETMASK        2	/* for setting the signal mask */
--- a/include/asm-cris/signal.h
+++ b/include/asm-cris/signal.h
@@ -120,8 +120,15 @@ typedef unsigned long sigset_t;
 #define SA_PROBE		SA_ONESHOT
 #define SA_SAMPLE_RANDOM	SA_RESTART
 #define SA_SHIRQ		0x04000000
+
+#ifdef CONFIG_NET_RANDOM
+#define SA_NET_RANDOM	SA_SAMPLE_RANDOM
+#else
+#define SA_NET_RANDOM	0
 #endif
 
+#endif /* __KERNEL__ */
+
 #define SIG_BLOCK          0	/* for blocking signals */
 #define SIG_UNBLOCK        1	/* for unblocking signals */
 #define SIG_SETMASK        2	/* for setting the signal mask */
--- a/include/asm-i386/signal.h
+++ b/include/asm-i386/signal.h
@@ -119,8 +119,15 @@ typedef unsigned long sigset_t;
 #define SA_PROBE		SA_ONESHOT
 #define SA_SAMPLE_RANDOM	SA_RESTART
 #define SA_SHIRQ		0x04000000
+
+#ifdef CONFIG_NET_RANDOM
+#define SA_NET_RANDOM	SA_SAMPLE_RANDOM
+#else
+#define SA_NET_RANDOM	0
 #endif
 
+#endif /* __KERNEL__ */
+
 #define SIG_BLOCK          0	/* for blocking signals */
 #define SIG_UNBLOCK        1	/* for unblocking signals */
 #define SIG_SETMASK        2	/* for setting the signal mask */
--- a/include/asm-ia64/signal.h
+++ b/include/asm-ia64/signal.h
@@ -117,6 +117,12 @@
 #define SA_SHIRQ		0x04000000
 #define SA_PERCPU_IRQ		0x02000000
 
+#ifdef CONFIG_NET_RANDOM
+#define SA_NET_RANDOM	SA_SAMPLE_RANDOM
+#else
+#define SA_NET_RANDOM	0
+#endif
+
 #endif /* __KERNEL__ */
 
 #define SIG_BLOCK          0	/* for blocking signals */
--- a/include/asm-m68k/signal.h
+++ b/include/asm-m68k/signal.h
@@ -116,8 +116,15 @@ typedef unsigned long sigset_t;
 #define SA_PROBE		SA_ONESHOT
 #define SA_SAMPLE_RANDOM	SA_RESTART
 #define SA_SHIRQ		0x04000000
+
+#ifdef CONFIG_NET_RANDOM
+#define SA_NET_RANDOM	SA_SAMPLE_RANDOM
+#else
+#define SA_NET_RANDOM	0
 #endif
 
+#endif /* __KERNEL__ */
+
 #define SIG_BLOCK          0	/* for blocking signals */
 #define SIG_UNBLOCK        1	/* for unblocking signals */
 #define SIG_SETMASK        2	/* for setting the signal mask */
--- a/include/asm-mips/signal.h
+++ b/include/asm-mips/signal.h
@@ -111,6 +111,12 @@ typedef unsigned long old_sigset_t;		/* 
 #define SA_SAMPLE_RANDOM	SA_RESTART
 #define SA_SHIRQ		0x02000000
 
+#ifdef CONFIG_NET_RANDOM
+#define SA_NET_RANDOM	SA_SAMPLE_RANDOM
+#else
+#define SA_NET_RANDOM	0
+#endif
+
 #endif /* __KERNEL__ */
 
 #define SIG_BLOCK	1	/* for blocking signals */
--- a/include/asm-mips64/signal.h
+++ b/include/asm-mips64/signal.h
@@ -119,6 +119,12 @@ typedef unsigned int old_sigset_t32;
 #define SA_SAMPLE_RANDOM	SA_RESTART
 #define SA_SHIRQ		0x02000000
 
+#ifdef CONFIG_NET_RANDOM
+#define SA_NET_RANDOM	SA_SAMPLE_RANDOM
+#else
+#define SA_NET_RANDOM	0
+#endif
+
 #endif /* __KERNEL__ */
 
 #define SIG_BLOCK	1	/* for blocking signals */
--- a/include/asm-parisc/signal.h
+++ b/include/asm-parisc/signal.h
@@ -100,6 +100,12 @@
 #define SA_SAMPLE_RANDOM	SA_RESTART
 #define SA_SHIRQ		0x04000000
 
+#ifdef CONFIG_NET_RANDOM
+#define SA_NET_RANDOM	SA_SAMPLE_RANDOM
+#else
+#define SA_NET_RANDOM	0
+#endif
+
 #endif /* __KERNEL__ */
 
 #define SIG_BLOCK          0	/* for blocking signals */
--- a/include/asm-ppc/signal.h
+++ b/include/asm-ppc/signal.h
@@ -111,6 +111,13 @@ typedef struct {
 #define SA_PROBE		SA_ONESHOT
 #define SA_SAMPLE_RANDOM	SA_RESTART
 #define SA_SHIRQ		0x04000000
+ 
+#ifdef CONFIG_NET_RANDOM
+#define SA_NET_RANDOM SA_SAMPLE_RANDOM
+#else
+#define SA_NET_RANDOM 0
+#endif
+  
 #endif /* __KERNEL__ */
 
 #define SIG_BLOCK          0	/* for blocking signals */
--- a/include/asm-s390/signal.h
+++ b/include/asm-s390/signal.h
@@ -129,8 +129,15 @@ typedef unsigned long sigset_t;
 #define SA_SHIRQ                0x04000000
 #define SA_DOPATHGROUP          0x00100000
 #define SA_FORCE                0x00200000
+
+#ifdef CONFIG_NET_RANDOM
+#define SA_NET_RANDOM		SA_SAMPLE_RANDOM
+#else
+#define SA_NET_RANDOM		0
 #endif
 
+#endif /* __KERNEL__ */
+
 #define SIG_BLOCK          0    /* for blocking signals */
 #define SIG_UNBLOCK        1    /* for unblocking signals */
 #define SIG_SETMASK        2    /* for setting the signal mask */
--- a/include/asm-s390x/signal.h
+++ b/include/asm-s390x/signal.h
@@ -129,8 +129,15 @@ typedef unsigned long sigset_t;
 #define SA_SHIRQ                0x04000000
 #define SA_DOPATHGROUP          0x00100000
 #define SA_FORCE                0x00200000
+
+#ifdef CONFIG_NET_RANDOM
+#define SA_NET_RANDOM		SA_SAMPLE_RANDOM
+#else
+#define SA_NET_RANDOM		0
 #endif
 
+#endif /* __KERNEL__ */
+
 #define SIG_BLOCK          0    /* for blocking signals */
 #define SIG_UNBLOCK        1    /* for unblocking signals */
 #define SIG_SETMASK        2    /* for setting the signal mask */
--- a/include/asm-sh/signal.h
+++ b/include/asm-sh/signal.h
@@ -107,8 +107,15 @@ typedef struct {
 #define SA_PROBE		SA_ONESHOT
 #define SA_SAMPLE_RANDOM	SA_RESTART
 #define SA_SHIRQ		0x04000000
+
+#ifdef CONFIG_NET_RANDOM
+#define SA_NET_RANDOM	SA_SAMPLE_RANDOM
+#else
+#define SA_NET_RANDOM	0
 #endif
 
+#endif /* __KERNEL__ */
+
 #define SIG_BLOCK          0	/* for blocking signals */
 #define SIG_UNBLOCK        1	/* for unblocking signals */
 #define SIG_SETMASK        2	/* for setting the signal mask */
--- a/include/asm-sparc/signal.h
+++ b/include/asm-sparc/signal.h
@@ -176,8 +176,15 @@ struct sigstack {
 #define SA_PROBE SA_ONESHOT
 #define SA_SAMPLE_RANDOM SA_RESTART
 #define SA_STATIC_ALLOC		0x80
+
+#ifdef CONFIG_NET_RANDOM
+#define SA_NET_RANDOM	SA_SAMPLE_RANDOM
+#else
+#define SA_NET_RANDOM	0
 #endif
 
+#endif /* __KERNEL__ */
+
 /* Type of a signal handler.  */
 #ifdef __KERNEL__
 typedef void (*__sighandler_t)(int, int, struct sigcontext *, char *);
--- a/include/asm-sparc64/signal.h
+++ b/include/asm-sparc64/signal.h
@@ -192,8 +192,15 @@ struct sigstack {
 #define SA_PROBE SA_ONESHOT
 #define SA_SAMPLE_RANDOM SA_RESTART
 #define SA_STATIC_ALLOC		0x80
+
+#ifdef CONFIG_NET_RANDOM
+#define SA_NET_RANDOM	SA_SAMPLE_RANDOM
+#else
+#define SA_NET_RANDOM	0
 #endif
 
+#endif /* __KERNEL__ */
+
 /* Type of a signal handler.  */
 #ifdef __KERNEL__
 typedef void (*__sighandler_t)(int, struct sigcontext *);
