--- a/include/linux/sockios.h
+++ b/include/linux/sockios.h
@@ -68,6 +68,11 @@
 
 #define SIOCGIFBR	0x8940		/* Bridging support		*/
 #define SIOCSIFBR	0x8941		/* Set bridging options 	*/
+#define SIOCBRADDBR     0x89a0          /* create new bridge device     */
+#define SIOCBRDELBR     0x89a1          /* remove bridge device         */
+#define SIOCBRADDIF     0x89a2          /* add interface to bridge      */
+#define SIOCBRDELIF     0x89a3          /* remove interface from bridge */
+
 
 #define SIOCGIFTXQLEN	0x8942		/* Get the tx queue length	*/
 #define SIOCSIFTXQLEN	0x8943		/* Set the tx queue length 	*/
