--- a/fs/Kconfig
+++ b/fs/Kconfig
@@ -97,6 +97,11 @@ config MANDATORY_FILE_LOCKING
 
 source "fs/crypto/Kconfig"
 
+config DIRECT_IO
+	bool "Enable O_DIRECT support" if EXPERT
+	depends on BLOCK
+	default y
+
 source "fs/notify/Kconfig"
 
 source "fs/quota/Kconfig"
--- a/fs/Makefile
+++ b/fs/Makefile
@@ -14,7 +14,8 @@ obj-y :=	open.o read_write.o file_table.
 		stack.o fs_struct.o statfs.o fs_pin.o nsfs.o
 
 ifeq ($(CONFIG_BLOCK),y)
-obj-y +=	buffer.o block_dev.o direct-io.o mpage.o
+obj-y +=	buffer.o block_dev.o mpage.o
+obj-$(CONFIG_DIRECT_IO) += direct-io.o
 else
 obj-y +=	no-block.o
 endif
--- a/include/linux/fs.h
+++ b/include/linux/fs.h
@@ -2852,6 +2852,7 @@ enum {
 	DIO_SKIP_DIO_COUNT = 0x08,
 };
 
+#ifdef CONFIG_DIRECT_IO
 void dio_end_io(struct bio *bio, int error);
 
 ssize_t __blockdev_direct_IO(struct kiocb *iocb, struct inode *inode,
@@ -2859,6 +2860,20 @@ ssize_t __blockdev_direct_IO(struct kioc
 			     get_block_t get_block,
 			     dio_iodone_t end_io, dio_submit_t submit_io,
 			     int flags);
+#else
+static inline void dio_end_io(struct bio *bio, int error)
+{
+}
+static inline
+ssize_t __blockdev_direct_IO(struct kiocb *iocb, struct inode *inode,
+			     struct block_device *bdev, struct iov_iter *iter,
+			     get_block_t get_block,
+			     dio_iodone_t end_io, dio_submit_t submit_io,
+			     int flags)
+{
+	return -EOPNOTSUPP;
+}
+#endif
 
 static inline ssize_t blockdev_direct_IO(struct kiocb *iocb,
 					 struct inode *inode,
--- a/fs/fcntl.c
+++ b/fs/fcntl.c
@@ -53,8 +53,10 @@ static int setfl(int fd, struct file * f
 
 	/* Pipe packetized mode is controlled by O_DIRECT flag */
 	if (!S_ISFIFO(filp->f_inode->i_mode) && (arg & O_DIRECT)) {
+#ifdef CONFIG_DIRECT_IO
 		if (!filp->f_mapping || !filp->f_mapping->a_ops ||
 			!filp->f_mapping->a_ops->direct_IO)
+#endif
 				return -EINVAL;
 	}
 
--- a/fs/open.c
+++ b/fs/open.c
@@ -690,7 +690,9 @@ int open_check_o_direct(struct file *f)
 {
 	/* NB: we're sure to have correct a_ops only after f_op->open */
 	if (f->f_flags & O_DIRECT) {
+#ifdef CONFIG_DIRECT_IO
 		if (!f->f_mapping->a_ops || !f->f_mapping->a_ops->direct_IO)
+#endif
 			return -EINVAL;
 	}
 	return 0;
