--- a/python/CMakeLists.txt
+++ b/python/CMakeLists.txt
@@ -3,7 +3,7 @@ option ( LINK_PYTHON_LIBRARY "Link again
 
 if ( PYTHON_BINDINGS )
   # workaround for cmake bug #0013449
-  if ( NOT DEFINED CMAKE_FIND_ROOT_PATH )
+  if ( NOT DEFINED CMAKE_FIND_ROOT_PATH OR CMAKE_VERSION VERSION_GREATER_EQUAL 3.0.0 )
     find_package ( SWIG )
   else ()
     find_program ( SWIG_EXECUTABLE NAMES swig2.0 swig )
