Index: automake-1.11.3/lib/am/python.am
===================================================================
--- automake-1.11.3.orig/lib/am/python.am	2012-01-31 12:41:18.000000000 +0100
+++ automake-1.11.3/lib/am/python.am	2012-06-07 16:42:53.488515898 +0200
@@ -23,8 +23,8 @@
 ## Installing.  ##
 ## ------------ ##
 
-if %?INSTALL%
 ?FIRST?am__py_compile = PYTHON=$(PYTHON) $(SHELL) $(py_compile)
+if %?INSTALL%
 am__installdirs += "$(DESTDIR)$(%NDIR%dir)"
 .PHONY install-%EXEC?exec:data%-am: install-%DIR%PYTHON
 install-%DIR%PYTHON: $(%DIR%_PYTHON)
