Index: linux-3.18-rc4/drivers/reset/Kconfig
===================================================================
--- linux-3.18-rc4.orig/drivers/reset/Kconfig
+++ linux-3.18-rc4/drivers/reset/Kconfig
@@ -12,4 +12,9 @@ menuconfig RESET_CONTROLLER
 
 	  If unsure, say no.
 
+config RESET_CONTROLLER_OXNAS
+	bool
+	select RESET_CONTROLLER
+
 source "drivers/reset/sti/Kconfig"
+
Index: linux-3.18-rc4/drivers/reset/Makefile
===================================================================
--- linux-3.18-rc4.orig/drivers/reset/Makefile
+++ linux-3.18-rc4/drivers/reset/Makefile
@@ -1,4 +1,5 @@
 obj-$(CONFIG_RESET_CONTROLLER) += core.o
+obj-$(CONFIG_RESET_CONTROLLER_OXNAS) += reset-ox820.o
 obj-$(CONFIG_ARCH_SOCFPGA) += reset-socfpga.o
 obj-$(CONFIG_ARCH_SUNXI) += reset-sunxi.o
 obj-$(CONFIG_ARCH_STI) += sti/
