diff -Nur fuse-2.5.2/kernel/Makefile.in fuse-2.5.2.patched/kernel/Makefile.in
--- fuse-2.5.2/kernel/Makefile.in	2006-02-02 18:04:52.000000000 +0100
+++ fuse-2.5.2.patched/kernel/Makefile.in	2006-03-28 20:17:13.000000000 +0200
@@ -54,23 +54,8 @@
 
 ifeq ($(majver), 2.4)
 
-CC = gcc
-LD = ld
-CFLAGS = -O2 -Wall -Wstrict-prototypes -fno-strict-aliasing -pipe @KERNELCFLAGS@
-CPPFLAGS = -I@kernelsrc@/include -I. -D__KERNEL__ -DMODULE -D_LOOSE_KERNEL_NAMES -DFUSE_VERSION=\"$(VERSION)\" @KERNELCPPFLAGS@
-
 fuse_objs = dev.o dir.o file.o inode.o compat/parser.o
 
-SUFFIXES = .c .o .s
-
-all-spec: fuse.o
-
-.c.o:
-	$(CC) $(CFLAGS) $(CPPFLAGS) -c $< -o $@
-
-fuse.o: $(fuse_objs)
-	$(LD) -r -o fuse.o $(fuse_objs)
-
 fuse_headers = fuse_i.h fuse_kernel.h
 
 dev.o: $(fuse_headers)
@@ -78,6 +63,18 @@
 file.o: $(fuse_headers)
 inode.o: $(fuse_headers)
 
+EXTRA_CFLAGS += -DFUSE_VERSION=\"$(VERSION)\"
+
+O_TARGET := fuse.o
+
+obj-y := $(fuse_objs) 
+obj-m := $(O_TARGET)
+
+-include $(TOPDIR)/Rules.make
+
+all-spec:
+	$(MAKE) -C @kernelsrc@ TOPDIR=@kernelsrc@ SUBDIRS=$(PWD) modules
+
 else
 
 EXTRA_CFLAGS += -DFUSE_VERSION=\"$(VERSION)\"
