--- fuse-2.7.3.orig/util/Makefile.in	2008-02-19 15:00:55.000000000 -0500
+++ fuse-2.7.3/util/Makefile.in	2008-03-17 14:14:10.000000000 -0400
@@ -528,7 +528,7 @@
 install-exec-hook:
 	-chown root $(DESTDIR)$(bindir)/fusermount
 	-chmod u+s $(DESTDIR)$(bindir)/fusermount
-	@if test ! -e $(DESTDIR)/dev/fuse; then \
+	-@if test ! -e $(DESTDIR)/dev/fuse; then \
 		$(mkdir_p) $(DESTDIR)/dev; \
 		echo "mknod $(DESTDIR)/dev/fuse -m 0666 c 10 229 || true"; \
 		mknod $(DESTDIR)/dev/fuse -m 0666 c 10 229 || true; \
