Index: fuse-2.6.5/kernel/file.c
===================================================================
--- fuse-2.6.5.orig/kernel/file.c	2007-06-23 13:03:50.000000000 +0200
+++ fuse-2.6.5/kernel/file.c	2007-06-23 13:03:51.000000000 +0200
@@ -781,6 +781,10 @@
 
 	if (cmd == F_GETLK) {
 		if (fc->no_lock) {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,22)
+			if (!posix_test_lock(file, fl))
+				fl->fl_type = F_UNLCK;
+#else
 #ifdef KERNEL_2_6_17_PLUS
 			if (!posix_test_lock(file, fl, fl))
 				fl->fl_type = F_UNLCK;
@@ -791,6 +795,7 @@
 			else
 				*fl = *cfl;
 #endif
+#endif
 			err = 0;
 		} else
 			err = fuse_getlk(file, fl);
