--- a/drivers/usb/Kconfig
+++ b/drivers/usb/Kconfig
@@ -164,4 +164,6 @@ source "drivers/usb/gadget/Kconfig"
 
 source "drivers/usb/otg/Kconfig"
 
+source "drivers/usb/dwc_otg/Kconfig"
+
 endif # USB_SUPPORT
--- a/drivers/usb/Makefile
+++ b/drivers/usb/Makefile
@@ -45,3 +45,4 @@ obj-$(CONFIG_EARLY_PRINTK_DBGP)	+= early
 
 obj-$(CONFIG_USB_ATM)		+= atm/
 obj-$(CONFIG_USB_SPEEDTOUCH)	+= atm/
+obj-$(CONFIG_DWC_OTG)		+= dwc_otg/
