--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -522,42 +522,48 @@ static struct board_info __initdata boar
 	.name				= "D-4P-W",
 	.expected_cpu_id		= 0x6348,
 
-	.has_enet0			= 1,
-	.has_enet1			= 1,
+	.has_enet1			= 1,	
 	.has_pci			= 1,
-
-	.enet0 = {
-		.has_phy		= 1,
-		.use_internal_phy	= 1,
-	},
+	.has_uart0 			= 1,
+	
 	.enet1 = {
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
-	},
-
-	.has_ohci0			= 1,
-	.has_ehci0			= 1,
+	},	
 
 	.leds = {
 		{
-			.name		= "ppp",
+			.name		= "power:green",
+			.gpio		= 0,
+			.active_low	= 1,
+		},
+		{
+			.name		= "status",
+			.gpio		= 3,
+			.active_low	= 1,			
+		},
+		{
+			.name		= "internet:green",
 			.gpio		= 4,
 			.active_low	= 1,
 		},
 		{
-			.name		= "ppp-fail",
+			.name		= "internet:red",
 			.gpio		= 5,
 			.active_low	= 1,
-		},
+		},		
+	},
+	
+	.buttons = {
 		{
-			.name		= "power",
-			.gpio		= 0,
+			.desc		= "reset",
+			.gpio		= 7,
 			.active_low	= 1,
-			.default_trigger = "default-on",
-
+			.type		= EV_KEY,
+			.code		= KEY_RESTART,
+			.threshold	= 3,
 		},
 	},
-
 };
 
 static struct board_info __initdata board_spw500v = {
