--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -573,6 +573,67 @@ static struct board_info __initdata boar
 
 };
 
+static struct board_info __initdata board_spw500v = {
+	.name				= "SPW500V",
+	.expected_cpu_id		= 0x6348,
+
+	.has_uart0			= 1,
+	.has_enet0			= 1,
+	.has_pci			= 1,
+
+	.enet0 = {
+		.has_phy		= 1,
+		.use_internal_phy	= 1,
+	},
+
+	.has_dsp			= 1,
+	.dsp = {
+		.gpio_rst		= 6,
+		.gpio_int		= 34,
+		.ext_irq		= 2,
+		.cs			= 2,
+	},
+
+	.leds = {
+		{
+			.name		= "power:red",
+			.gpio		= 1,
+			.active_low	= 1,
+		},
+		{
+			.name		= "power:green",
+			.gpio		= 0,
+			.active_low	= 1,
+			.default_trigger = "default-on",
+		},
+		{
+			.name		= "ppp",
+			.gpio		= 3,
+			.active_low	= 1,
+		},
+		{	.name		= "pstn",
+			.gpio		= 28,
+			.active_low	= 1,
+		},
+		{
+			.name		= "voip",
+			.gpio		= 32,
+			.active_low	= 1,
+		},
+	},
+
+	.buttons = {
+		{
+			.desc		= "reset",
+			.gpio		= 33,
+			.active_low	= 1,
+			.type		= EV_KEY,
+			.code		= KEY_RESTART,
+			.threshold	= 3,
+		},
+	},
+};
+
 #endif
 
 /*
@@ -1480,6 +1541,7 @@ static const struct board_info __initdat
 	&board_96348gw_a,
 	&board_rta1025w_16,
 	&board_96348_D4PW,
+	&board_spw500v,
 #endif
 
 #ifdef CONFIG_BCM63XX_CPU_6358
