diff -urN linux.old/arch/ppc/platforms/4xx/Kconfig linux.dev/arch/ppc/platforms/4xx/Kconfig
--- linux.old/arch/ppc/platforms/4xx/Kconfig	2006-11-29 22:57:37.000000000 +0100
+++ linux.dev/arch/ppc/platforms/4xx/Kconfig	2006-12-14 08:49:18.000000000 +0100
@@ -52,6 +52,12 @@
 	help
 	  This option enables support for the IBM PPC405GP evaluation board.
 
+config MAGICBOX
+	bool "MagicBox"
+	select WANT_EARLY_SERIAL
+	help
+	  This option enables support for the IBM PPC405EP evaluation board.
+
 config XILINX_ML300
 	bool "Xilinx-ML300"
 	help
@@ -173,7 +179,7 @@
 
 config IBM_OCP
 	bool
-	depends on ASH || BAMBOO || BUBINGA || CPCI405 || EBONY || EP405 || LUAN || YUCCA || OCOTEA || REDWOOD_5 || REDWOOD_6 || SYCAMORE || TAISHAN || WALNUT
+	depends on ASH || BAMBOO || BUBINGA || CPCI405 || EBONY || EP405 || LUAN || YUCCA || OCOTEA || REDWOOD_5 || REDWOOD_6 || SYCAMORE || TAISHAN || WALNUT || MAGICBOX
 	default y
 
 config IBM_EMAC4
@@ -183,7 +189,7 @@
 
 config BIOS_FIXUP
 	bool
-	depends on BUBINGA || EP405 || SYCAMORE || WALNUT || CPCI405
+	depends on BUBINGA || EP405 || SYCAMORE || WALNUT || CPCI405 || MAGICBOX
 	default y
 
 # OAK doesn't exist but wanted to keep this around for any future 403GCX boards
@@ -194,7 +200,7 @@
 
 config 405EP
 	bool
-	depends on BUBINGA
+	depends on BUBINGA || MAGICBOX
 	default y
 
 config 405GP
diff -urN linux.old/arch/ppc/platforms/4xx/Makefile linux.dev/arch/ppc/platforms/4xx/Makefile
--- linux.old/arch/ppc/platforms/4xx/Makefile	2006-11-29 22:57:37.000000000 +0100
+++ linux.dev/arch/ppc/platforms/4xx/Makefile	2006-12-14 08:48:56.000000000 +0100
@@ -13,6 +13,7 @@
 obj-$(CONFIG_REDWOOD_6)		+= redwood6.o
 obj-$(CONFIG_SYCAMORE)		+= sycamore.o
 obj-$(CONFIG_WALNUT)		+= walnut.o
+obj-$(CONFIG_MAGICBOX)		+= magicbox.o
 obj-$(CONFIG_XILINX_ML300)	+= xilinx_ml300.o
 obj-$(CONFIG_XILINX_ML403)	+= xilinx_ml403.o
 
diff -urN linux.old/include/asm-ppc/ibm4xx.h linux.dev/include/asm-ppc/ibm4xx.h
--- linux.old/include/asm-ppc/ibm4xx.h	2006-11-29 22:57:37.000000000 +0100
+++ linux.dev/include/asm-ppc/ibm4xx.h	2006-12-14 08:48:56.000000000 +0100
@@ -18,6 +18,10 @@
 
 #ifdef CONFIG_40x
 
+#if defined(CONFIG_MAGICBOX)
+#include <platforms/4xx/magicbox.h>
+#endif
+
 #if defined(CONFIG_BUBINGA)
 #include <platforms/4xx/bubinga.h>
 #endif
