diff -purN linux.old/drivers/ide/ide.c linux.dev/drivers/ide/ide.c
--- linux.old/drivers/ide/ide.c	2007-01-10 20:10:37.000000000 +0100
+++ linux.dev/drivers/ide/ide.c	2007-04-09 01:25:30.866800288 +0200
@@ -1783,6 +1783,7 @@ done:
 extern void __init pnpide_init(void);
 extern void __exit pnpide_exit(void);
 extern void __init h8300_ide_init(void);
+extern void __init ide_magicbox_init(void);
 
 /*
  * probe_for_hwifs() finds/initializes "known" IDE interfaces
@@ -1847,6 +1848,9 @@ static void __init probe_for_hwifs (void
 #ifdef CONFIG_H8300
 	h8300_ide_init();
 #endif
+#ifdef CONFIG_BLK_DEV_MAGICBOX_IDE
+	ide_magicbox_init();
+#endif
 }
 
 void ide_register_subdriver(ide_drive_t *drive, ide_driver_t *driver)
diff -purN linux.old/drivers/ide/Kconfig linux.dev/drivers/ide/Kconfig
--- linux.old/drivers/ide/Kconfig	2007-01-10 20:10:37.000000000 +0100
+++ linux.dev/drivers/ide/Kconfig	2007-04-09 01:15:27.570515264 +0200
@@ -925,6 +925,14 @@ config BLK_DEV_MPC8xx_IDE
 
 	  If unsure, say N.
 
+config BLK_DEV_MAGICBOX_IDE
+	bool "MagicBox 2.0 CF IDE support"
+	depends on 4xx && IDE
+	help
+	  This option provides support for IDE on MagicBox 2.0 boards.
+
+	  If unsure, say N.
+
 choice
 	prompt "Type of MPC8xx IDE interface"
 	depends on BLK_DEV_MPC8xx_IDE
diff -purN linux.old/drivers/ide/Makefile linux.dev/drivers/ide/Makefile
--- linux.old/drivers/ide/Makefile	2007-01-10 20:10:37.000000000 +0100
+++ linux.dev/drivers/ide/Makefile	2007-04-09 00:23:29.332559488 +0200
@@ -36,6 +36,7 @@ ide-core-$(CONFIG_BLK_DEV_Q40IDE)	+= leg
 # built-in only drivers from ppc/
 ide-core-$(CONFIG_BLK_DEV_MPC8xx_IDE)	+= ppc/mpc8xx.o
 ide-core-$(CONFIG_BLK_DEV_IDE_PMAC)	+= ppc/pmac.o
+ide-core-$(CONFIG_BLK_DEV_MAGICBOX_IDE)	+= ppc/magicbox_ide.o
 
 # built-in only drivers from h8300/
 ide-core-$(CONFIG_H8300)		+= h8300/ide-h8300.o
