--- p910nd-0.7/Makefile.orig	2005-09-25 13:54:28.465506888 +0200
+++ p910nd-0.7/Makefile	2005-09-25 13:54:46.928700056 +0200
@@ -2,9 +2,9 @@
 # below if you don't want to use libwrap (hosts.{allow,deny} access control)
 
 # If you don't have it in /var/log/subsys, uncomment and define
-#CFLAGS+=-DLOCKFILE_DIR=\"/var/log\"
+CFLAGS+=-DLOCKFILE_DIR=\"/tmp\"
 LIBWRAP=-lwrap
 
 p910nd:	p910nd.c
-#	$(CC) -Wall $(CFLAGS) -o $@ p910nd.c
-	$(CC) -Wall $(CFLAGS) -DUSE_LIBWRAP -o $@ p910nd.c $(LIBWRAP)
+	$(CC) -Wall $(CFLAGS) -o $@ p910nd.c
+#	$(CC) -Wall $(CFLAGS) -DUSE_LIBWRAP -o $@ p910nd.c $(LIBWRAP)
