--- a/gcc/config.gcc
+++ b/gcc/config.gcc
@@ -836,7 +836,7 @@ avr-*-*)
 	;;
 avr32*-*-linux*)
 	tm_file="dbxelf.h elfos.h linux.h avr32/linux-elf.h avr32/avr32.h "
-	tmake_file="t-linux avr32/t-avr32 avr32/t-elf"
+	tmake_file="t-linux avr32/t-avr32-linux"
 	extra_parts="crtbegin.o crtbeginS.o crtend.o crtendS.o"
 	extra_modes=avr32/avr32-modes.def
 	gnu_ld=yes
--- a/libgcc/config.host
+++ b/libgcc/config.host
@@ -240,6 +240,11 @@ arm-*-pe*)
 	;;
 arm*-*-kaos*)
 	;;
+avr32-*-linux*)
+	# No need to build crtbeginT.o on uClibc systems. Should probably be
+	# moved to the OS specific section above.
+	extra_parts="crtbegin.o crtbeginS.o crtend.o crtendS.o"
+	;;
 avr32-*-*)
 	;;
 avr-*-rtems*)
