--- a/gcc/config/avr32/lib1funcs.S
+++ b/gcc/config/avr32/lib1funcs.S
@@ -2866,7 +2866,7 @@ __truncdfsf_return_op1:
         /* NaN or inf */
         cbr     r12,31                  /* clear implicit bit */
         retne   -1                      /* Return NaN if mantissa not zero */
-	mov_imm	r12, 0xff000000
+	mov_imm	r12, 0x7f800000
         ret     r12                     /* Return inf */
  
 3:      /* Result is subnormal. Adjust it.*/
