--- a/fwcmd.c
+++ b/fwcmd.c
@@ -2393,8 +2393,6 @@ int mwl_fwcmd_check_ba(struct ieee80211_
 
 	if (pcmd->cmd_hdr.result != 0) {
 		mutex_unlock(&priv->fwcmd_mutex);
-		wiphy_err(hw->wiphy, "check ba result error %d\n",
-			  le16_to_cpu(pcmd->cmd_hdr.result));
 		return -EINVAL;
 	}
 
--- a/mac80211.c
+++ b/mac80211.c
@@ -644,8 +644,6 @@ static int mwl_mac80211_ampdu_action(str
 		spin_lock_bh(&priv->stream_lock);
 		if (rc) {
 			mwl_fwcmd_remove_stream(hw, stream);
-			wiphy_err(hw->wiphy,
-				  "ampdu start error code: %d\n", rc);
 			rc = -EPERM;
 			break;
 		}
