Index: ppp-2.4.3/pppd/plugins/rp-pppoe/discovery.c
===================================================================
--- ppp-2.4.3.orig/pppd/plugins/rp-pppoe/discovery.c	2007-06-04 13:22:08.458569384 +0200
+++ ppp-2.4.3/pppd/plugins/rp-pppoe/discovery.c	2007-06-04 13:22:09.225452800 +0200
@@ -365,8 +365,8 @@
 	if (!packetIsForMe(conn, &packet)) continue;
 
 	if (packet.code == CODE_PADO) {
-	    if (NOT_UNICAST(packet.ethHdr.h_source)) {
-		printErr("Ignoring PADO packet from non-unicast MAC address");
+	    if (BROADCAST(packet.ethHdr.h_source)) {
+		printErr("Ignoring PADO packet from broadcast MAC address");
 		continue;
 	    }
 	    parsePacket(&packet, parsePADOTags, &pc);
