Index: ppp-2.4.3/pppd/chap-new.c
===================================================================
--- ppp-2.4.3.orig/pppd/chap-new.c	2007-06-04 13:22:08.298593704 +0200
+++ ppp-2.4.3/pppd/chap-new.c	2007-06-04 13:22:11.199152752 +0200
@@ -57,6 +57,7 @@
 int chap_timeout_time = 3;
 int chap_max_transmits = 10;
 int chap_rechallenge_time = 0;
+int chapms_strip_domain = 0;
 
 /*
  * Command-line options.
@@ -68,6 +69,8 @@
 	  "Set max #xmits for challenge", OPT_PRIO },
 	{ "chap-interval", o_int, &chap_rechallenge_time,
 	  "Set interval for rechallenge", OPT_PRIO },
+	{ "chapms-strip-domain", o_bool, &chapms_strip_domain,
+	  "Strip the domain prefix before the Username", 1 },
 	{ NULL }
 };
 
@@ -338,6 +341,14 @@
 			/* Null terminate and clean remote name. */
 			slprintf(rname, sizeof(rname), "%.*v", len, name);
 			name = rname;
+
+			/* strip the MS domain name */
+			if (chapms_strip_domain && strrchr(rname, '\\')) {
+				char tmp[MAXNAMELEN+1];
+
+				strcpy(tmp, strrchr(rname, '\\') + 1);
+				strcpy(rname, tmp);
+			}
 		}
 
 		if (chap_verify_hook)
