Index: ppp-2.4.3/pppd/plugins/radius/Makefile.linux
===================================================================
--- ppp-2.4.3.orig/pppd/plugins/radius/Makefile.linux	2007-06-04 13:22:12.463960472 +0200
+++ ppp-2.4.3/pppd/plugins/radius/Makefile.linux	2007-06-04 13:22:12.705923688 +0200
@@ -12,7 +12,8 @@
 INSTALL	= install
 
 PLUGIN=radius.so radattr.so radrealms.so
-CFLAGS=-I. -I../.. -I../../../include -O2 -fPIC -DRC_LOG_FACILITY=LOG_DAEMON
+COPTS = -O2
+CFLAGS=-I. -I../.. -I../../../include $(COPTS) -fPIC -DRC_LOG_FACILITY=LOG_DAEMON
 
 # Uncomment the next line to include support for Microsoft's
 # MS-CHAP authentication protocol.
Index: ppp-2.4.3/pppdump/Makefile.linux
===================================================================
--- ppp-2.4.3.orig/pppdump/Makefile.linux	2007-06-04 13:22:12.468959712 +0200
+++ ppp-2.4.3/pppdump/Makefile.linux	2007-06-04 13:22:12.705923688 +0200
@@ -2,7 +2,8 @@
 BINDIR = $(DESTDIR)/sbin
 MANDIR = $(DESTDIR)/share/man/man8
 
-CFLAGS= -O -I../include/net
+COPTS = -O
+CFLAGS= $(COPTS) -I../include/net
 OBJS = pppdump.o bsd-comp.o deflate.o zlib.o
 
 INSTALL= install
