Index: ppp-2.4.3/pppd/plugins/rp-pppoe/plugin.c
===================================================================
--- ppp-2.4.3.orig/pppd/plugins/rp-pppoe/plugin.c	2007-06-04 13:22:09.417423616 +0200
+++ ppp-2.4.3/pppd/plugins/rp-pppoe/plugin.c	2007-06-04 13:22:12.953885992 +0200
@@ -282,9 +282,9 @@
     if (strlen(cmd) > 4 && !strncmp(cmd, "nic-", 4)) {
 	/* Strip off "nic-" */
 	cmd += 4;
-    } else if (strlen(cmd) < 4
-	       || (strncmp(cmd, "eth", 3) && strncmp(cmd, "nas", 3)
-		   && strncmp(cmd, "tap", 3) && strncmp(cmd, "br", 2))) {
+	} else if (strlen(cmd) < 3
+		|| (strncmp(cmd, "eth", 3) && strncmp(cmd, "nas", 3) && strncmp(cmd, "vlan", 4) && strncmp(cmd, "ath", 3)
+                   && strncmp(cmd, "tap", 3) && strncmp(cmd, "br", 2))) {
 	return 0;
     }
 
