Index: ppp-2.4.3/pppd/Makefile.linux
===================================================================
--- ppp-2.4.3.orig/pppd/Makefile.linux	2007-06-04 13:22:12.463960472 +0200
+++ ppp-2.4.3/pppd/Makefile.linux	2007-06-04 13:22:14.133706632 +0200
@@ -172,8 +172,8 @@
 
 ifdef FILTER
 ifneq ($(wildcard /usr/include/pcap-bpf.h),)
-LIBS    += -lpcap
-CFLAGS  += -DPPP_FILTER
+LIBS    += -lpcap -L$(STAGING_DIR)/usr/lib
+CFLAGS  += -DPPP_FILTER -I$(STAGING_DIR)/usr/include
 endif
 endif
 
