--- a/Makefile
+++ b/Makefile
@@ -275,10 +275,10 @@ $(U_BOOT_NAND):	$(NAND_SPL) $(obj)u-boot
 		cat nand_spl/u-boot-spl-16k.bin $(obj)u-boot.bin > $(obj)u-boot-nand.bin
 
 version:
-		@echo -n "#define U_BOOT_VERSION \"U-Boot " > $(VERSION_FILE); \
-		echo -n "$(U_BOOT_VERSION)" >> $(VERSION_FILE); \
-		echo -n $(shell $(CONFIG_SHELL) $(TOPDIR)/tools/setlocalversion \
-			 $(TOPDIR)) >> $(VERSION_FILE); \
+		@printf "#define U_BOOT_VERSION \"U-Boot " > $(VERSION_FILE); \
+		printf "$(U_BOOT_VERSION)" >> $(VERSION_FILE); \
+		printf "$(shell $(CONFIG_SHELL) $(TOPDIR)/tools/setlocalversion \
+			 $(TOPDIR))" >> $(VERSION_FILE); \
 		echo "\"" >> $(VERSION_FILE)
 
 gdbtools:
@@ -1593,10 +1593,10 @@ MPC8540EVAL_66_slave_config:      unconf
 	@mkdir -p $(obj)include
 	@echo "" >$(obj)include/config.h ; \
 	if [ "$(findstring _33_,$@)" ] ; then \
-		echo -n "... 33 MHz PCI" ; \
+		printf "... 33 MHz PCI" ; \
 	else \
 		echo "#define CONFIG_SYSCLK_66M" >>$(obj)include/config.h ; \
-		echo -n "... 66 MHz PCI" ; \
+		printf "... 66 MHz PCI" ; \
 	fi ; \
 	if [ "$(findstring _slave_,$@)" ] ; then \
 		echo "#define CONFIG_PCI_SLAVE" >>$(obj)include/config.h ; \
