--- linux-2.4.30/drivers/net/b44.h	2003-08-25 13:44:42.000000000 +0200
+++ openwrt/build_mipsel/linux-2.4.30/drivers/net/b44.h	2005-05-01 14:43:20.000000000 +0200
@@ -461,6 +461,8 @@
 };
 
 #define B44_MCAST_TABLE_SIZE	32
+#define B44_PHY_ADDR_NO_PHY	30
+#define B44_MDC_RATIO		5000000
 
 /* SW copy of device statistics, kept up to date by periodic timer
  * which probes HW values.  Must have same relative layout as HW
