--- a/fs/squashfs/inode.c
+++ b/fs/squashfs/inode.c
@@ -33,6 +33,7 @@
 #include <linux/fs.h>
 #include <linux/smp_lock.h>
 #include <linux/slab.h>
+#include <linux/exportfs.h>
 #include <linux/squashfs_fs_sb.h>
 #include <linux/squashfs_fs_i.h>
 #include <linux/buffer_head.h>
@@ -2125,7 +2126,7 @@ static void squashfs_destroy_inode(struc
 }
 
 
-static void init_once(void * foo, struct kmem_cache * cachep, unsigned long flags)
+static void init_once(struct kmem_cache * cachep, void *foo)
 {
 	struct squashfs_inode_info *ei = foo;
 
