--- a/fs/Kconfig
+++ b/fs/Kconfig
@@ -421,6 +421,7 @@ config FS_POSIX_ACL
 
 source "fs/xfs/Kconfig"
 source "fs/gfs2/Kconfig"
+source "fs/yaffs2/Kconfig"
 
 config OCFS2_FS
 	tristate "OCFS2 file system support"
--- a/fs/Makefile
+++ b/fs/Makefile
@@ -121,3 +121,4 @@ obj-$(CONFIG_HPPFS)		+= hppfs/
 obj-$(CONFIG_DEBUG_FS)		+= debugfs/
 obj-$(CONFIG_OCFS2_FS)		+= ocfs2/
 obj-$(CONFIG_GFS2_FS)           += gfs2/
+obj-$(CONFIG_YAFFS_FS)		+= yaffs2/
