--- a/fs/jffs2/erase.c
+++ b/fs/jffs2/erase.c
@@ -35,6 +35,8 @@ static void jffs2_erase_block(struct jff
 {
 	int ret;
 	uint32_t bad_offset;
+	static char s[]="|/-\\", *p=s;
+
 #ifdef __ECOS
        ret = jffs2_flash_erase(c, jeb);
        if (!ret) {
@@ -47,6 +49,11 @@ static void jffs2_erase_block(struct jff
 
 	D1(printk(KERN_DEBUG "jffs2_erase_block(): erase block %#08x (range %#08x-%#08x)\n",
 				jeb->offset, jeb->offset, jeb->offset + c->sector_size));
+
+	printk("%c\b", *p);
+	if (*++p==0)
+	p=s;
+
 	instr = kmalloc(sizeof(struct erase_info) + sizeof(struct erase_priv_struct), GFP_KERNEL);
 	if (!instr) {
 		printk(KERN_WARNING "kmalloc for struct erase_info in jffs2_erase_block failed. Refiling block for later\n");
