diff -ur old/src/Makefile dev/src/Makefile
--- old/src/Makefile	Sat Oct 14 17:49:54 2006
+++ dev/src/Makefile	Sat Oct 14 18:02:30 2006
@@ -25,12 +25,14 @@
 
 all:	libgcc_eh libsupc $(EXOBJS) $(ALLBIN)
 
+CP = cp -fPR
+
 install:
 	$(INSTALL) -d $(PREFIX)$(UCLIBCXX_RUNTIME_LIBDIR)
 ifneq ($(BUILD_ONLY_STATIC_LIB),y)
 	$(INSTALL) -m 755 $(SHARED_FULLNAME) \
 		$(PREFIX)$(UCLIBCXX_RUNTIME_LIBDIR)
-	cp -fa $(SHARED_MAJORNAME) $(LIBNAME).so $(PREFIX)$(UCLIBCXX_RUNTIME_LIBDIR)
+	$(CP) $(SHARED_MAJORNAME) $(LIBNAME).so $(PREFIX)$(UCLIBCXX_RUNTIME_LIBDIR)
 endif
 ifeq ($(BUILD_STATIC_LIB),y)
 	$(INSTALL) -m 644 $(LIBNAME).a $(PREFIX)$(UCLIBCXX_RUNTIME_LIBDIR)
