Index: linux-5.4.61/arch/mips/bcm63xx/boards/board_bcm963xx.c
===================================================================
--- linux-5.4.61.orig/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ linux-5.4.61/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -1321,6 +1321,49 @@ static struct board_info __initdata boar
 	},
 };
 
+static struct board_info __initdata board_TG582N = {
+	.name = "DANT-1",
+	.expected_cpu_id = 0x6328,
+
+	.has_pci = 1,
+	.has_ohci0 = 1,
+	.has_ehci0 = 1,
+	.num_usbh_ports = 1,
+
+	.has_enetsw = 1,
+	.enetsw = {
+		.used_ports = {
+			[0] = {
+				.used = 1,
+				.phy_id = 1,
+				.name = "Port 1",
+			},
+			[1] = {
+				.used = 1,
+				.phy_id = 2,
+				.name = "Port 2",
+			},
+			[2] = {
+				.used = 1,
+				.phy_id = 3,
+				.name = "Port 3",
+			},
+			[3] = {
+				.used = 1,
+				.phy_id = 4,
+				.name = "Port 4",
+			},
+		},
+	},
+
+	.use_fallback_sprom = 1,
+	.fallback_sprom = {
+		.type = SPROM_BCM43225,
+		.pci_bus = 1,
+		.pci_dev = 0,
+	},
+};
+
 static struct board_info __initdata board_96348sv = {
 	.name = "MAGIC",
 	.expected_cpu_id = 0x6348,
@@ -2835,6 +2878,7 @@ static const struct board_info __initcon
 	&board_dsl_274xb_f1,
 	&board_FAST2704V2,
 	&board_R5010UNV2,
+	&board_TG582N,
 #endif /* CONFIG_BCM63XX_CPU_6328 */
 #ifdef CONFIG_BCM63XX_CPU_6338
 	&board_96338gw,
@@ -2946,6 +2990,7 @@ static struct of_device_id const bcm963x
 	{ .compatible = "sagem,fast-2704-v2", .data = &board_FAST2704V2, },
 	{ .compatible = "sercomm,ad1018", .data = &board_AD1018, },
 	{ .compatible = "sercomm,ad1018-nor", .data = &board_AD1018, },
+	{ .compatible = "technicolor,tg582n", .data = &board_TG582N, },
 #endif /* CONFIG_BCM63XX_CPU_6328 */
 #ifdef CONFIG_BCM63XX_CPU_6338
 	{ .compatible = "brcm,bcm96338gw", .data = &board_96338gw, },
