Index: linux-2.6.23.12/arch/arm/mach-ixp4xx/fsg-setup.c
===================================================================
--- linux-2.6.23.12.orig/arch/arm/mach-ixp4xx/fsg-setup.c	2008-01-05 13:53:26.000000000 +1030
+++ linux-2.6.23.12/arch/arm/mach-ixp4xx/fsg-setup.c	2008-01-05 13:56:28.000000000 +1030
@@ -94,6 +94,31 @@
 	.resource		= fsg_uart_resources,
 };
 
+ /* Built-in 10/100 Ethernet MAC interfaces */
+ static struct eth_plat_info fsg_plat_eth[] = {
+         {
+                 .phy            = 5,
+                 .rxq            = 3,
+ 		.txreadyq	= 20,
+         }, {
+                 .phy            = 4,
+                 .rxq            = 4,
+ 		.txreadyq	= 21,
+         }
+ };
+ 
+ static struct platform_device fsg_eth[] = {
+         {
+                 .name                   = "ixp4xx_eth",
+                 .id                     = IXP4XX_ETH_NPEB,
+                 .dev.platform_data      = fsg_plat_eth,
+         }, {
+                 .name                   = "ixp4xx_eth",
+                 .id                     = IXP4XX_ETH_NPEC,
+                 .dev.platform_data      = fsg_plat_eth + 1,
+         }
+ };
+ 
 static struct platform_device fsg_leds = {
 	.name		= "fsg-led",
 	.id		= -1,
@@ -102,6 +127,8 @@
 static struct platform_device *fsg_devices[] __initdata = {
 	&fsg_i2c_controller,
 	&fsg_flash,
+	&fsg_eth[0],
+	&fsg_eth[1],
 	&fsg_leds,
 };
 
