diff -urN syslog-ng-1.6.9/Makefile.in syslog-ng-1.6.9.new/Makefile.in
--- syslog-ng-1.6.9/Makefile.in	2005-11-23 09:58:28.000000000 +0100
+++ syslog-ng-1.6.9.new/Makefile.in	2006-03-23 19:02:04.000000000 +0100
@@ -64,7 +64,7 @@
 INSTALL_STRIP_PROGRAM = @INSTALL_STRIP_PROGRAM@
 LDFLAGS = @LDFLAGS@
 LEX = @LEX@
-LEXLIB = @LEXLIB@
+LEXLIB =
 LEX_OUTPUT_ROOT = @LEX_OUTPUT_ROOT@
 LIBOBJS = @LIBOBJS@
 LIBOL_DIR = @LIBOL_DIR@
diff -urN syslog-ng-1.6.9/configure syslog-ng-1.6.9.new/configure
--- syslog-ng-1.6.9/configure	2005-11-23 09:58:29.000000000 +0100
+++ syslog-ng-1.6.9.new/configure	2006-03-23 21:56:43.000000000 +0100
@@ -3160,7 +3160,7 @@
   echo $ECHO_N "(cached) $ECHO_C" >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lfl  $LIBS"
+LIBS="$LIBS"
 cat >conftest.$ac_ext <<_ACEOF
 /* confdefs.h.  */
 _ACEOF
diff -urN syslog-ng-1.6.9/src/Makefile.in syslog-ng-1.6.9.new/src/Makefile.in
--- syslog-ng-1.6.9/src/Makefile.in	2005-11-23 09:58:37.000000000 +0100
+++ syslog-ng-1.6.9.new/src/Makefile.in	2006-03-23 19:02:27.000000000 +0100
@@ -64,7 +64,7 @@
 INSTALL_STRIP_PROGRAM = @INSTALL_STRIP_PROGRAM@
 LDFLAGS = @LDFLAGS@
 LEX = @LEX@
-LEXLIB = @LEXLIB@
+LEXLIB =
 LEX_OUTPUT_ROOT = @LEX_OUTPUT_ROOT@
 LIBOBJS = @LIBOBJS@
 LIBOL_DIR = @LIBOL_DIR@
diff -urN syslog-ng-1.6.9/src/cfg-lex.c syslog-ng-1.6.9.new/src/cfg-lex.c
--- syslog-ng-1.6.9/src/cfg-lex.c	2005-11-15 19:41:46.000000000 +0100
+++ syslog-ng-1.6.9.new/src/cfg-lex.c	2006-03-23 19:03:25.000000000 +0100
@@ -666,7 +666,9 @@
 #ifdef __cplusplus
 extern "C" int yywrap (void );
 #else
-extern int yywrap (void );
+int yywrap (void ) {
+	return 1;
+}
 #endif
 #endif
 
