--- a/drivers/tty/serial/Kconfig
+++ b/drivers/tty/serial/Kconfig
@@ -1192,6 +1192,25 @@ config SERIAL_BCM63XX_CONSOLE
 	  If you have enabled the serial port on the bcm63xx CPU
 	  you can make it the console by answering Y to this option.
 
+config SERIAL_ADM8668
+	tristate "ADM8668 serial port support"
+	select SERIAL_CORE
+	depends on ADM8668
+	help
+	  If you have an adm8668 CPU, you can enable its onboard
+	  serial port by enabling this options.
+
+          To compile this driver as a module, choose M here: the
+          module will be called adm8668_uart.
+
+config SERIAL_ADM8668_CONSOLE
+	bool "Console on adm8668 serial port"
+	depends on SERIAL_ADM8668=y
+	select SERIAL_CORE_CONSOLE
+	help
+	  If you have enabled the serial port on the adm8668 CPU
+	  you can make it the console by answering Y to this option.
+
 config SERIAL_GRLIB_GAISLER_APBUART
 	tristate "GRLIB APBUART serial support"
 	depends on OF && SPARC
--- a/include/linux/serial_core.h
+++ b/include/linux/serial_core.h
@@ -211,6 +211,9 @@
 #define PORT_AR933X	99
 
 
+/* ADM8668 UART */
+#define	PORT_ADM8668	100
+
 #ifdef __KERNEL__
 
 #include <linux/compiler.h>
