diff -Naurp linux-2.6.16.7-generic-patched/net/ipv4/af_inet.c linux-2.6.16.7-patched/net/ipv4/af_inet.c
--- linux-2.6.16.7-generic-patched/net/ipv4/af_inet.c	2006-04-17 23:53:25.000000000 +0200
+++ linux-2.6.16.7-patched/net/ipv4/af_inet.c	2006-07-05 15:33:47.000000000 +0200
@@ -940,7 +940,9 @@ void inet_register_protosw(struct inet_p
 out:
 	spin_unlock_bh(&inetsw_lock);
 
+#ifndef CONFIG_MIPS_BRCM
 	synchronize_net();
+#endif
 
 	return;
 
