From 44bace8c6c81bb936648cfe974a8460ab80bd46d Mon Sep 17 00:00:00 2001
From: Manuel Fombuena <mfombuena@innovara.co.uk>
Date: Sun, 28 Jan 2024 16:44:31 +0000
Subject: [PATCH] dt-bindings: arm: qcom: Add IPQ8174 ID

Signed-off-by: Manuel Fombuena <mfombuena@innovara.co.uk>
---
 include/dt-bindings/arm/qcom,ids.h | 1 +
 1 file changed, 1 insertion(+)

--- a/include/dt-bindings/arm/qcom,ids.h
+++ b/include/dt-bindings/arm/qcom,ids.h
@@ -121,6 +121,7 @@
 #define QCOM_ID_SM6125			394
 #define QCOM_ID_IPQ8070A		395
 #define QCOM_ID_IPQ8071A		396
+#define QCOM_ID_IPQ8174			399
 #define QCOM_ID_IPQ6018			402
 #define QCOM_ID_IPQ6028			403
 #define QCOM_ID_IPQ6000			421
