--- a/include/linux/netdevice.h
+++ b/include/linux/netdevice.h
@@ -99,7 +99,7 @@ struct wireless_dev;
  */
 
 #if defined(CONFIG_WLAN_80211) || defined(CONFIG_AX25) || defined(CONFIG_AX25_MODULE)
-# if defined(CONFIG_MAC80211_MESH)
+# if 1 || defined(CONFIG_MAC80211_MESH)
 #  define LL_MAX_HEADER 128
 # else
 #  define LL_MAX_HEADER 96
--- a/include/linux/skbuff.h
+++ b/include/linux/skbuff.h
@@ -378,7 +378,7 @@ struct sk_buff {
 #ifdef CONFIG_IPV6_NDISC_NODETYPE
 	__u8			ndisc_nodetype:2;
 #endif
-#if defined(CONFIG_MAC80211) || defined(CONFIG_MAC80211_MODULE)
+#if 1
 	__u8			do_not_encrypt:1;
 	__u8			requeue:1;
 #endif
