--- a/arch/mips/alchemy/mtx-1/init.c
+++ b/arch/mips/alchemy/mtx-1/init.c
@@ -32,6 +32,7 @@
 #include <linux/init.h>
 
 #include <asm/bootinfo.h>
+#include <asm/string.h>
 
 #include <prom.h>
 
@@ -49,7 +49,7 @@ void __init prom_init(void)
 	prom_argv = (char **)fw_arg1;
 	prom_envp = (char **)fw_arg2;
 
-	prom_init_cmdline();
+	strcpy(arcs_cmdline, CONFIG_CMDLINE);
 
 	memsize_str = prom_getenv("memsize");
 	if (!memsize_str)
 
