--- util-linux-2.12r.orig/fdisk/fdiskbsdlabel.h	2007-06-12 21:32:25.000000000 +0200
+++ util-linux-2.12r/fdisk/fdiskbsdlabel.h	2007-06-12 21:33:20.000000000 +0200
@@ -46,7 +46,7 @@
 
 #define BSD_LINUX_BOOTDIR "/usr/ucb/mdec"
 
-#if defined (i386) || defined (__sparc__) || defined (__arm__) || defined (__mips__) || defined (__s390__) || defined (__sh__) || defined(__x86_64__)
+#if defined (i386) || defined (__sparc__) || defined (__arm__) || defined (__mips__) || defined (__s390__) || defined (__sh__) || defined(__x86_64__) || defined (__cris__) || defined(__avr32__)
 #define BSD_LABELSECTOR   1
 #define BSD_LABELOFFSET   0
 #elif defined (__alpha__) || defined (__powerpc__) || defined (__ia64__) || defined (__hppa__)
