--- a/arch/mips/include/asm/mach-ath79/pci.h
+++ b/arch/mips/include/asm/mach-ath79/pci.h
@@ -16,13 +16,13 @@
 #if defined(CONFIG_PCI) && defined(CONFIG_SOC_AR71XX)
 int ar71xx_pcibios_init(void);
 #else
-static inline int ar71xx_pcibios_init(void) { return 0 };
+static inline int ar71xx_pcibios_init(void) { return 0; };
 #endif
 
 #if defined(CONFIG_PCI_AR724X)
 int ar724x_pcibios_init(int irq);
 #else
-static inline int ar724x_pcibios_init(int irq) { return 0 };
+static inline int ar724x_pcibios_init(int irq) { return 0; };
 #endif
 
 #endif /* __ASM_MACH_ATH79_PCI_H */
