--- a/drivers/mtd/nand/rb4xx_nand.c
+++ b/drivers/mtd/nand/rb4xx_nand.c
@@ -12,6 +12,8 @@
  *  by the Free Software Foundation.
  */
 
+#include <linux/kernel.h>
+#include <linux/module.h>
 #include <linux/init.h>
 #include <linux/mtd/nand.h>
 #include <linux/mtd/mtd.h>
@@ -22,8 +24,8 @@
 #include <linux/gpio.h>
 #include <linux/slab.h>
 
-#include <asm/mach-ar71xx/ar71xx.h>
-#include <asm/mach-ar71xx/rb4xx_cpld.h>
+#include <asm/mach-ath79/ath79.h>
+#include <asm/mach-ath79/rb4xx_cpld.h>
 
 #define DRV_NAME        "rb4xx-nand"
 #define DRV_VERSION     "0.2.0"
@@ -238,12 +240,8 @@ static int __devinit rb4xx_nand_probe(st
 		goto err_set_drvdata;
 	}
 
-#ifdef CONFIG_MTD_PARTITIONS
-	ret = add_mtd_partitions(&info->mtd, rb4xx_nand_partitions,
+	mtd_device_register(&info->mtd, rb4xx_nand_partitions,
 				ARRAY_SIZE(rb4xx_nand_partitions));
-#else
-	ret = add_mtd_device(&info->mtd);
-#endif
 	if (ret)
 		goto err_release_nand;
 
