Index: linux-2.6.28.9/arch/cris/Kconfig
===================================================================
--- linux-2.6.28.9.orig/arch/cris/Kconfig	2009-03-23 22:55:52.000000000 +0100
+++ linux-2.6.28.9/arch/cris/Kconfig	2009-04-08 16:53:30.000000000 +0200
@@ -662,6 +662,12 @@
 
 source "drivers/ide/Kconfig"
 
+source "drivers/scsi/Kconfig"
+
+source "drivers/media/Kconfig"
+
+source "sound/Kconfig"
+
 source "drivers/net/Kconfig"
 
 source "drivers/i2c/Kconfig"
