Index: openswan-2.4.8/programs/ranbits/ranbits.c
===================================================================
--- openswan-2.4.8.orig/programs/ranbits/ranbits.c	2007-06-04 13:22:49.835279168 +0200
+++ openswan-2.4.8/programs/ranbits/ranbits.c	2007-06-04 13:22:51.648003592 +0200
@@ -29,7 +29,7 @@
 #include <openswan.h>
 
 #ifndef DEVICE
-#define	DEVICE	"/dev/random"
+#define	DEVICE	"/dev/urandom"
 #endif
 #ifndef QDEVICE
 #define	QDEVICE	"/dev/urandom"
Index: openswan-2.4.8/programs/rsasigkey/rsasigkey.c
===================================================================
--- openswan-2.4.8.orig/programs/rsasigkey/rsasigkey.c	2007-06-04 13:22:49.842278104 +0200
+++ openswan-2.4.8/programs/rsasigkey/rsasigkey.c	2007-06-04 13:22:51.649003440 +0200
@@ -31,7 +31,7 @@
 #include <gmp.h>
 
 #ifndef DEVICE
-#define	DEVICE	"/dev/random"
+#define	DEVICE	"/dev/urandom"
 #endif
 #ifndef MAXBITS
 #define	MAXBITS	20000
Index: openswan-2.4.8/programs/starter/files.h
===================================================================
--- openswan-2.4.8.orig/programs/starter/files.h	2007-06-04 13:22:49.850276888 +0200
+++ openswan-2.4.8/programs/starter/files.h	2007-06-04 13:22:51.649003440 +0200
@@ -36,7 +36,7 @@
 
 #define MY_PID_FILE     "/var/run/pluto/ipsec-starter.pid"
 
-#define DEV_RANDOM      "/dev/random"
+#define DEV_RANDOM      "/dev/urandom"
 #define DEV_URANDOM     "/dev/urandom"
 
 #define PROC_IPSECVERSION   "/proc/net/ipsec_version"
