Index: openswan-2.4.8/linux/net/ipsec/sysctl_net_ipsec.c
===================================================================
--- openswan-2.4.8.orig/linux/net/ipsec/sysctl_net_ipsec.c	2007-06-04 13:22:49.815282208 +0200
+++ openswan-2.4.8/linux/net/ipsec/sysctl_net_ipsec.c	2007-06-04 13:22:51.852972432 +0200
@@ -130,7 +130,11 @@
 
 int ipsec_sysctl_register(void)
 {
-        ipsec_table_header = register_sysctl_table(ipsec_root_table, 0);
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2,6,20)
+		ipsec_table_header = register_sysctl_table(ipsec_root_table);
+#else
+		ipsec_table_header = register_sysctl_table(ipsec_root_table, 0);
+#endif
         if (!ipsec_table_header) {
                 return -ENOMEM;
 	}
