diff -urN openswan-2.4.8/linux/net/ipsec/ipsec_proc.c openswan-2.4.8.new/linux/net/ipsec/ipsec_proc.c
--- openswan-2.4.8/linux/net/ipsec/ipsec_proc.c	2006-11-15 23:21:39.000000000 +0100
+++ openswan-2.4.8.new/linux/net/ipsec/ipsec_proc.c	2007-06-13 20:00:51.000000000 +0200
@@ -27,6 +27,9 @@
 #include <linux/version.h>
 #define __NO_VERSION__
 #include <linux/module.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,0)
+#include <linux/moduleparam.h>
+#endif
 #include <linux/kernel.h> /* printk() */
 
 #include "openswan/ipsec_kversion.h"
