--- a/ld/configure.tgt
+++ b/ld/configure.tgt
@@ -92,8 +92,10 @@ cr16c-*-elf*)           targ_emul=elf32c
 cris-*-*aout*)		targ_emul=crisaout
 			targ_extra_emuls="criself crislinux"
 			targ_extra_libpath=$targ_extra_emuls ;;
-cris-*-linux-* | crisv32-*-linux-*)
-			targ_emul=crislinux ;;
+cris-*-linux-* | crisv32-*-linux-*)	targ_emul=crislinux
+			targ_extra_emuls="criself crisaout"
+			targ_extra_libpath=$targ_extra_emuls
+			;;
 cris-*-* | crisv32-*-*)	targ_emul=criself
 			targ_extra_emuls="crisaout crislinux"
 			targ_extra_libpath=$targ_extra_emuls
