--- u-boot-1.1.4.ttt/include/configs/vlink.h	2006-06-05 15:57:37.000000000 +0200
+++ u-boot-1.1.4/include/configs/vlink.h	2006-06-07 13:11:01.000000000 +0200
@@ -105,9 +105,9 @@
 
 #define CONFIG_COMMANDS		\
 		       ((CONFIG_CMD_DFL | CFG_CMD_MII |\
-			CFG_CMD_DHCP ) & \
-		      ~(CFG_CMD_BDI | \
-			CFG_CMD_IMI | \
+			CFG_CMD_DHCP | \
+			CFG_CMD_BDI ) & \
+		      ~(CFG_CMD_IMI | \
 			CFG_CMD_AUTOSCRIPT | \
 			CFG_CMD_FPGA | \
 			CFG_CMD_MISC | \
