diff -urN u-boot-1.1.4.old/lib_arm/board.c u-boot-1.1.4/lib_arm/board.c
--- u-boot-1.1.4.old/lib_arm/board.c	2007-03-23 10:53:52.000000000 +0100
+++ u-boot-1.1.4/lib_arm/board.c	2007-03-24 13:17:12.000000000 +0100
@@ -332,7 +332,7 @@
 	}
 #endif /* CONFIG_DRIVER_SMC91111 || CONFIG_DRIVER_LAN91C96 */
 
-#ifdef CONFIG_MACH_VLINK
+#ifdef CONFIG_VLINK
 	printf("Initialising MAC address\n\r");
 	eth_init(gd->bd);
 #endif
