--- u-boot-1.1.4.old/tools/Makefile	2007-06-13 13:35:59.000000000 +0200
+++ u-boot-1.1.4/tools/Makefile	2007-06-14 15:33:04.000000000 +0200
@@ -153,7 +153,7 @@
 		$(STRIP) $@
 
 ubparams.o: ubparams.c
-		$(CC) -g $(CFLAGS) -c $<
+		$(CC) -g $(CFLAGS) -DLAN_IP=$(LAN_IP) -DLAN_SERVERIP=$(LAN_SERVERIP) -c $<
 
 envcrc.o:	envcrc.c
 		$(CC) -g $(CFLAGS) -c $<
--- u-boot-1.1.4.old/tools/ubparams.c	2007-06-13 13:35:59.000000000 +0200
+++ u-boot-1.1.4/tools/ubparams.c	2007-06-14 15:31:55.000000000 +0200
@@ -44,8 +44,8 @@
 	"bootargs=setenv bootargs root=/dev/mtdblock3 ro console=/dev/ttyS0,115200,mem=32M\0"
 	"bootcmd=bootm 0xc0042000\0"
 	"ethaddr=00:30:49:00:00:01\0"
-	"ipaddr=10.0.1.73\0"
-	"serverip=10.0.1.210\0"
+	"ipaddr=" MK_STR(LAN_IP) "\0"
+	"serverip=" MK_STR(LAN_SERVERIP) "\0"
 	"serial#=MX070205484\0"
 	"\0"
 	};
