--- a/src/mps/drv_mps_vmmc_ar9.c
+++ b/src/mps/drv_mps_vmmc_ar9.c
@@ -74,7 +74,7 @@ extern mps_comm_dev *pMPSDev;
 IFX_void_t ifx_mps_release (IFX_void_t);
 extern IFX_uint32_t ifx_mps_reset_structures (mps_comm_dev * pMPSDev);
 extern IFX_int32_t ifx_mps_bufman_close (IFX_void_t);
-IFX_int32_t ifx_mps_wdog_callback (IFX_uint32_t wdog_cleared_ok_count);
+IFX_int32_t ifx_mps_wdog_callback (IFX_ulong_t wdog_cleared_ok_count);
 extern IFXOS_event_t fw_ready_evt;
 /* ============================= */
 /* Local function declaration    */
@@ -467,7 +467,7 @@ IFX_void_t ifx_mps_release (IFX_void_t)
  * \return  0        IFX_SUCCESS, cannot fail
  * \ingroup Internal
  */
-IFX_int32_t ifx_mps_wdog_callback (IFX_uint32_t wdog_cleared_ok_count)
+IFX_int32_t ifx_mps_wdog_callback (IFX_ulong_t wdog_cleared_ok_count)
 {
    IFX_uint32_t flags;
 #ifdef DEBUG
