Index: linux-2.6.23.14/drivers/usb/host/Kconfig
===================================================================
--- linux-2.6.23.14.orig/drivers/usb/host/Kconfig
+++ linux-2.6.23.14/drivers/usb/host/Kconfig
@@ -248,3 +248,6 @@ config USB_R8A66597_HCD
 	  To compile this driver as a module, choose M here: the
 	  module will be called r8a66597-hcd.
 
+config USB_ADM5120_HCD
+	tristate "ADM5120 HCD support (EXPERIMENTAL)"
+	depends on USB && MIPS_ADM5120 && EXPERIMENTAL
Index: linux-2.6.23.14/drivers/usb/host/Makefile
===================================================================
--- linux-2.6.23.14.orig/drivers/usb/host/Makefile
+++ linux-2.6.23.14/drivers/usb/host/Makefile
@@ -8,6 +8,7 @@ endif
 
 obj-$(CONFIG_PCI)		+= pci-quirks.o
 
+obj-$(CONFIG_USB_ADM5120_HCD)	+= adm5120-hcd.o
 obj-$(CONFIG_USB_EHCI_HCD)	+= ehci-hcd.o
 obj-$(CONFIG_USB_ISP116X_HCD)	+= isp116x-hcd.o
 obj-$(CONFIG_USB_OHCI_HCD)	+= ohci-hcd.o
Index: linux-2.6.23.14/drivers/usb/Makefile
===================================================================
--- linux-2.6.23.14.orig/drivers/usb/Makefile
+++ linux-2.6.23.14/drivers/usb/Makefile
@@ -16,6 +16,7 @@ obj-$(CONFIG_USB_UHCI_HCD)	+= host/
 obj-$(CONFIG_USB_SL811_HCD)	+= host/
 obj-$(CONFIG_USB_U132_HCD)	+= host/
 obj-$(CONFIG_USB_R8A66597_HCD)	+= host/
+obj-$(CONFIG_USB_ADM5120_HCD)  	+= host/
 
 obj-$(CONFIG_USB_ACM)		+= class/
 obj-$(CONFIG_USB_PRINTER)	+= class/
