--- a/arch/mips/ar7/time.c
+++ b/arch/mips/ar7/time.c
@@ -33,6 +33,8 @@ void __init plat_time_init(void)
 	/* Initialize ar7 clocks so the CPU clock frequency is correct */
 	ar7_init_clocks();
 
+	ar7_init_clocks();
+
 	cpu_clk = clk_get(NULL, "cpu");
 	if (IS_ERR(cpu_clk)) {
 		printk(KERN_ERR "unable to get cpu clock\n");
--- a/arch/mips/include/asm/mach-ar7/ar7.h
+++ b/arch/mips/include/asm/mach-ar7/ar7.h
@@ -203,4 +203,6 @@ static inline void ar7_device_off(u32 bi
 int __init ar7_gpio_init(void);
 void __init ar7_init_clocks(void);
 
+extern void __init ar7_init_clocks(void);
+
 #endif /* __AR7_H__ */
