--- a/arch/mips/amazon/pci.c
+++ b/arch/mips/amazon/pci.c
@@ -182,7 +182,7 @@ static struct pci_controller amazon_pci_
 	.io_resource = &pci_io_resource
 };
 
-int __init pcibios_map_irq(struct pci_dev *dev, u8 slot, u8 pin)
+int __init pcibios_map_irq(const struct pci_dev *dev, u8 slot, u8 pin)
 {
 	switch (slot) {
 		case 13:
@@ -240,7 +240,7 @@ int pcibios_plat_dev_init(struct pci_dev
 	return 0;
 }
 
-int amazon_pci_init(void)
+int __init amazon_pci_init(void)
 {
 	u32 temp_buffer;
 
