--- a/arch/mips/ath79/mach-ap83.c
+++ b/arch/mips/ath79/mach-ap83.c
@@ -42,41 +42,8 @@
 #define AP83_KEYS_POLL_INTERVAL		20	/* msecs */
 #define AP83_KEYS_DEBOUNCE_INTERVAL	(3 * AP83_KEYS_POLL_INTERVAL)
 
-static struct mtd_partition ap83_flash_partitions[] = {
-	{
-		.name		= "u-boot",
-		.offset		= 0,
-		.size		= 0x040000,
-		.mask_flags	= MTD_WRITEABLE,
-	}, {
-		.name		= "u-boot-env",
-		.offset		= 0x040000,
-		.size		= 0x020000,
-		.mask_flags	= MTD_WRITEABLE,
-	}, {
-		.name		= "kernel",
-		.offset		= 0x060000,
-		.size		= 0x140000,
-	}, {
-		.name		= "rootfs",
-		.offset		= 0x1a0000,
-		.size		= 0x650000,
-	}, {
-		.name		= "art",
-		.offset		= 0x7f0000,
-		.size		= 0x010000,
-		.mask_flags	= MTD_WRITEABLE,
-	}, {
-		.name		= "firmware",
-		.offset		= 0x060000,
-		.size		= 0x790000,
-	}
-};
-
 static struct physmap_flash_data ap83_flash_data = {
 	.width		= 2,
-	.parts		= ap83_flash_partitions,
-	.nr_parts	= ARRAY_SIZE(ap83_flash_partitions),
 };
 
 static struct resource ap83_flash_resources[] = {
