--- a/librt/mq_receive.c
+++ b/librt/mq_receive.c
@@ -6,7 +6,8 @@
 #include <stddef.h>
 #include <sys/syscall.h>
 #include <mqueue.h>
-#warning FIXME: hard dependency on ADVANCED REALTIME feature
+
+#ifdef __UCLIBC_HAS_ADVANCED_REALTIME__
 
 librt_hidden_proto(mq_timedreceive)
 
@@ -44,3 +45,5 @@ ssize_t mq_receive(mqd_t mqdes, char *ms
 {
 	return mq_timedreceive(mqdes, msg_ptr, msg_len, msg_prio, NULL);
 }
+
+#endif /* __UCLIBC_HAS_ADVANCED_REALTIME__ */
--- a/librt/mq_send.c
+++ b/librt/mq_send.c
@@ -6,7 +6,8 @@
 #include <stddef.h>
 #include <sys/syscall.h>
 #include <mqueue.h>
-#warning FIXME: hard dependency on ADVANCED REALTIME feature
+
+#ifdef __UCLIBC_HAS_ADVANCED_REALTIME__
 
 librt_hidden_proto(mq_timedsend)
 
@@ -43,3 +44,5 @@ int mq_send(mqd_t mqdes, const char *msg
 {
 	return mq_timedsend(mqdes, msg_ptr, msg_len, msg_prio, NULL);
 }
+
+#endif /* __UCLIBC_HAS_ADVANCED_REALTIME__ */
