--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -904,6 +904,9 @@ int __init board_register_devices(void)
 
 	bcm63xx_spi_register();
 
+	if (board.num_devs)
+		platform_add_devices(board.devs, board.num_devs);
+
 	bcm63xx_flash_register();
 
 	/* count number of LEDs defined by this device */
--- a/arch/mips/include/asm/mach-bcm63xx/board_bcm963xx.h
+++ b/arch/mips/include/asm/mach-bcm63xx/board_bcm963xx.h
@@ -49,6 +49,10 @@ struct board_info {
 
 	/* Buttons */
 	struct gpio_keys_button buttons[4];
+
+	/* Additional platform devices */
+	struct platform_device **devs;
+	unsigned int	num_devs;
 };
 
 #endif /* ! BOARD_BCM963XX_H_ */
