--- a/gdb/configure
+++ b/gdb/configure
@@ -988,7 +988,7 @@ CPP
 YACC
 YFLAGS
 XMKMF'
-ac_subdirs_all='doc testsuite
+ac_subdirs_all='doc
 gdbtk
 multi-ice
 gdbserver'
@@ -7881,7 +7881,7 @@ fi
 
 
 
-subdirs="$subdirs doc testsuite"
+subdirs="$subdirs doc"
 
 
 # Check whether to support alternative target configurations
--- a/gdb/gdbserver/configure
+++ b/gdb/gdbserver/configure
@@ -2152,7 +2152,7 @@ $as_echo "$as_me: error: \`$ac_var' was 
       ac_cache_corrupted=: ;;
     ,);;
     *)
-      if test "x$ac_old_val" != "x$ac_new_val"; then
+      if test "`echo x$ac_old_val`" != "`echo x$ac_new_val`"; then
 	# differences in whitespace do not lead to failure.
 	ac_old_val_w=`echo x $ac_old_val`
 	ac_new_val_w=`echo x $ac_new_val`
