--- a/gdb/configure
+++ b/gdb/configure
@@ -834,8 +834,7 @@ MAKEINFOFLAGS
 YACC
 YFLAGS
 XMKMF'
-ac_subdirs_all='testsuite
-gdbtk
+ac_subdirs_all='gdbtk
 multi-ice
 gdbserver'
 
@@ -5033,7 +5032,7 @@ $as_echo "$with_auto_load_safe_path" >&6
 
 
 
-subdirs="$subdirs testsuite"
+subdirs="$subdirs"
 
 
 # Check whether to support alternative target configurations
--- a/gdb/gdbserver/configure
+++ b/gdb/gdbserver/configure
@@ -2181,7 +2181,7 @@ $as_echo "$as_me: error: \`$ac_var' was
       ac_cache_corrupted=: ;;
     ,);;
     *)
-      if test "x$ac_old_val" != "x$ac_new_val"; then
+      if test "`echo x$ac_old_val`" != "`echo x$ac_new_val`"; then
 	# differences in whitespace do not lead to failure.
 	ac_old_val_w=`echo x $ac_old_val`
 	ac_new_val_w=`echo x $ac_new_val`
