--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -38,7 +38,7 @@ add_subdirectory(library)
 add_subdirectory(include)
 
 if(CMAKE_COMPILER_IS_GNUCC)
-  add_subdirectory(tests)
+#  add_subdirectory(tests)
 endif(CMAKE_COMPILER_IS_GNUCC)
 
 add_subdirectory(programs)
--- a/programs/CMakeLists.txt
+++ b/programs/CMakeLists.txt
@@ -3,6 +3,6 @@ add_subdirectory(hash)
 add_subdirectory(pkey)
 add_subdirectory(random)
 add_subdirectory(ssl)
-add_subdirectory(test)
+#add_subdirectory(test)
 add_subdirectory(x509)
 add_subdirectory(util)
