--- a/library/cipher.c
+++ b/library/cipher.c
@@ -279,15 +279,21 @@ const cipher_info_t *cipher_info_from_st
 #if defined(POLARSSL_BLOWFISH_C)
     if( !strcasecmp( "BLOWFISH-CBC", cipher_name ) )
         return cipher_info_from_type( POLARSSL_CIPHER_BLOWFISH_CBC );
+    if( !strcasecmp( "BF-CBC", cipher_name ) )
+        return cipher_info_from_type( POLARSSL_CIPHER_BLOWFISH_CBC );
 
 #if defined(POLARSSL_CIPHER_MODE_CFB)
     if( !strcasecmp( "BLOWFISH-CFB64", cipher_name ) )
         return cipher_info_from_type( POLARSSL_CIPHER_BLOWFISH_CFB64 );
+    if( !strcasecmp( "BF-CFB64", cipher_name ) )
+        return cipher_info_from_type( POLARSSL_CIPHER_BLOWFISH_CFB64 );
 #endif /* defined(POLARSSL_CIPHER_MODE_CFB) */
 
 #if defined(POLARSSL_CIPHER_MODE_CTR)
     if( !strcasecmp( "BLOWFISH-CTR", cipher_name ) )
         return cipher_info_from_type( POLARSSL_CIPHER_BLOWFISH_CTR );
+    if( !strcasecmp( "BF-CTR", cipher_name ) )
+        return cipher_info_from_type( POLARSSL_CIPHER_BLOWFISH_CTR );
 #endif /* defined(POLARSSL_CIPHER_MODE_CTR) */
 #endif
 
--- a/library/cipher_wrap.c
+++ b/library/cipher_wrap.c
@@ -643,7 +643,7 @@ const cipher_info_t blowfish_cbc_info =
     POLARSSL_CIPHER_BLOWFISH_CBC,
     POLARSSL_MODE_CBC,
     128,
-    "BLOWFISH-CBC",
+    "BF-CBC",
     8,
     8,
     &blowfish_info
@@ -654,7 +654,7 @@ const cipher_info_t blowfish_cfb64_info
     POLARSSL_CIPHER_BLOWFISH_CFB64,
     POLARSSL_MODE_CFB,
     128,
-    "BLOWFISH-CFB64",
+    "BF-CFB64",
     8,
     8,
     &blowfish_info
@@ -666,7 +666,7 @@ const cipher_info_t blowfish_ctr_info =
     POLARSSL_CIPHER_BLOWFISH_CTR,
     POLARSSL_MODE_CTR,
     128,
-    "BLOWFISH-CTR",
+    "BF-CTR",
     8,
     8,
     &blowfish_info
