--- a/drivers/bluetooth/hci_h4p/core.c
+++ b/drivers/bluetooth/hci_h4p/core.c
@@ -36,9 +36,9 @@
 #include <linux/clk.h>
 #include <linux/gpio.h>
 
-#include <mach/hardware.h>
-#include <mach/board.h>
-#include <mach/irqs.h>
+#include <plat/hardware.h>
+#include <plat/board.h>
+#include <plat/irqs.h>
 #include <plat/serial.h>
 
 #include <net/bluetooth/bluetooth.h>
--- a/drivers/bluetooth/hci_h4p/hci_h4p.h
+++ b/drivers/bluetooth/hci_h4p/hci_h4p.h
@@ -21,7 +21,7 @@
  *
  */
 
-#include <mach/board.h>
+#include <plat/board.h>
 
 #include <net/bluetooth/bluetooth.h>
 #include <net/bluetooth/hci_core.h>
--- a/drivers/bluetooth/hci_h4p/sysfs.c
+++ b/drivers/bluetooth/hci_h4p/sysfs.c
@@ -48,15 +48,8 @@ static ssize_t hci_h4p_store_bdaddr(stru
 		return -EINVAL;
 	}
 
-	//for (i = 0; i < 6; i++)
-		//info->bdaddr[i] = bdaddr[i] & 0xff;
-
-	info->bdaddr[0] = 0x00;
-	info->bdaddr[1] = 0x1D;
-	info->bdaddr[2] = 0x6E;
-	info->bdaddr[3] = 0xD4;
-	info->bdaddr[4] = 0xF0;
-	info->bdaddr[5] = 0x37;
+	for (i = 0; i < 6; i++)
+		info->bdaddr[i] = bdaddr[i] & 0xff;
 
 	return count;
 }
