---
 arch/arm/mach-omap2/board-n8x0-lcd.c |   18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

Index: linux-3.1/arch/arm/mach-omap2/board-n8x0-lcd.c
===================================================================
--- linux-3.1.orig/arch/arm/mach-omap2/board-n8x0-lcd.c	2011-10-30 00:48:48.205044041 +0200
+++ linux-3.1/arch/arm/mach-omap2/board-n8x0-lcd.c	2011-10-30 00:48:56.685038042 +0200
@@ -33,8 +33,28 @@ static void mipid_shutdown(struct mipid_
 	}
 }
 
+static int n8x0_get_backlight_level(struct mipid_platform_data *pdata)
+{
+//FIXME	return tahvo_get_backlight_level();
+return 0;
+}
+
+static int n8x0_get_max_backlight_level(struct mipid_platform_data *pdata)
+{
+//FIXME	return tahvo_get_max_backlight_level();
+return 0;
+}
+
+static void n8x0_set_backlight_level(struct mipid_platform_data *pdata, int level)
+{
+//FIXME	tahvo_set_backlight_level(level);
+}
+
 struct mipid_platform_data n8x0_mipid_platform_data = {
 	.shutdown = mipid_shutdown,
+	.get_bklight_level = n8x0_get_backlight_level,
+	.set_bklight_level = n8x0_set_backlight_level,
+	.get_bklight_max = n8x0_get_max_backlight_level,
 };
 
 void __init n8x0_mipid_init(void)
