--- a/include/linux/gpio_keys.h
+++ b/include/linux/gpio_keys.h
@@ -15,7 +15,12 @@ struct gpio_keys_button {
 struct gpio_keys_platform_data {
 	struct gpio_keys_button *buttons;
 	int nbuttons;
+	unsigned int poll_interval;     /* polling interval in msecs -
+					   for polling driver only */
 	unsigned int rep:1;		/* enable input subsystem auto repeat */
+	int (*enable)(struct device *dev);
+	void (*disable)(struct device *dev);
+	const char *name;		/* input device name */
 };
 
 #endif
