Index: libupnp-1.2.1a/ixml/src/element.c
===================================================================
--- libupnp-1.2.1a.orig/ixml/src/element.c
+++ libupnp-1.2.1a/ixml/src/element.c
@@ -454,9 +454,9 @@ ixmlElement_removeAttributeNode( IN IXML
             element->n.firstAttr = nextSib;
         }
 
-        ( IXML_Attr * ) attrNode->parentNode = NULL;
-        ( IXML_Attr * ) attrNode->prevSibling = NULL;
-        ( IXML_Attr * ) attrNode->nextSibling = NULL;
+        attrNode->parentNode = NULL;
+        attrNode->prevSibling = NULL;
+        attrNode->nextSibling = NULL;
         *rtAttr = ( IXML_Attr * ) attrNode;
         return IXML_SUCCESS;
 
