diff -urN dropbear-0.45.old/scp.c dropbear-0.45/scp.c
--- dropbear-0.45.old/scp.c	2005-03-07 05:27:02.000000000 +0100
+++ dropbear-0.45/scp.c	2005-03-25 11:28:22.000000000 +0100
@@ -249,9 +249,9 @@
 
 	args.list = NULL;
 	addargs(&args, "ssh");		/* overwritten with ssh_program */
-	addargs(&args, "-x");
-	addargs(&args, "-oForwardAgent no");
-	addargs(&args, "-oClearAllForwardings yes");
+//	addargs(&args, "-x");
+//	addargs(&args, "-oForwardAgent no");
+//	addargs(&args, "-oClearAllForwardings yes");
 
 	fflag = tflag = 0;
 	while ((ch = getopt(argc, argv, "dfl:prtvBCc:i:P:q1246S:o:F:")) != -1)
