--- a/libopkg/opkg_message.c
+++ b/libopkg/opkg_message.c
@@ -64,10 +64,10 @@ print_error_list(void)
 	struct errlist *err = error_list_head;
 
 	if (err) {
-		printf("Collected errors:\n");
+		fprintf(stderr, "Collected errors:\n");
 		/* Here we print the errors collected and free the list */
 		while (err != NULL) {
-			printf(" * %s", err->errmsg);
+			fprintf(stderr, " * %s", err->errmsg);
 			err = err->next;
 		}
 	}
