--- a/acx_func.h
+++ b/acx_func.h
@@ -714,7 +714,6 @@ int acx_e_op_set_key(struct ieee80211_hw
 int acx_e_op_config(struct ieee80211_hw *hw, u32 changed);
 void acx_e_op_bss_info_changed(struct ieee80211_hw *hw,
 		struct ieee80211_vif *vif, struct ieee80211_bss_conf *info, u32 changed);
-int acx_e_op_get_tx_stats(struct ieee80211_hw* ieee, struct ieee80211_tx_queue_stats *stats);
 int acx_e_conf_tx(struct ieee80211_hw* ieee, u16 queue,
 		const struct ieee80211_tx_queue_params *params);
 //int acx_passive_scan(struct net_device *net_dev, int state, struct ieee80211_scan_conf *conf);
--- a/common.c
+++ b/common.c
@@ -4662,24 +4662,6 @@ extern void acx_e_op_bss_info_changed(st
 	return;
 }
 
-int acx_e_op_get_tx_stats(struct ieee80211_hw *hw,
-			 struct ieee80211_tx_queue_stats *stats)
-{
-	acx_device_t *adev = ieee2adev(hw);
-	int err = -ENODEV;
-
-	FN_ENTER;
-	acx_sem_lock(adev);
-
-	stats->len = 0;
-	stats->limit = TX_CNT;
-	stats->count = 0;
-
-	acx_sem_unlock(adev);
-	FN_EXIT0;
-	return err;
-}
-
 int acx_e_conf_tx(struct ieee80211_hw *hw,
 		u16 queue, const struct ieee80211_tx_queue_params *params)
 {
--- a/mem.c
+++ b/mem.c
@@ -2321,7 +2321,6 @@ static const struct ieee80211_ops acxmem
 		.bss_info_changed = acx_e_op_bss_info_changed,
 		.set_key = acx_e_op_set_key,
 		.get_stats = acx_e_op_get_stats,
-		.get_tx_stats = acx_e_op_get_tx_stats,
 		};
 
 
--- a/pci.c
+++ b/pci.c
@@ -1482,7 +1482,6 @@ static const struct ieee80211_ops acxpci
 	.bss_info_changed = acx_e_op_bss_info_changed,
 	.set_key = acx_e_op_set_key,
 	.get_stats = acx_e_op_get_stats,
-	.get_tx_stats = acx_e_op_get_tx_stats,
 };
 
 
--- a/usb.c
+++ b/usb.c
@@ -757,7 +757,6 @@ static const struct ieee80211_ops acxusb
 	.bss_info_changed = acx_e_op_bss_info_changed,
 	.set_key = acx_e_op_set_key,
 	.get_stats = acx_e_op_get_stats,
-	.get_tx_stats = acx_e_op_get_tx_stats,
 };
 
 /***********************************************************************
