diff -urN acx-mac80211-20080805/acx_mac80211.h acx-mac80211-20080805.new/acx_mac80211.h
--- acx-mac80211-20080805/acx_mac80211.h	2009-05-07 14:32:44.000000000 +0200
+++ acx-mac80211-20080805.new/acx_mac80211.h	2009-05-07 15:59:45.000000000 +0200
@@ -36,7 +36,7 @@
         /* Current BSSID (if any). */       
         u8 *bssid;
 
-        /* Interface type. (IEEE80211_IF_TYPE_XXX) */
+        /* Interface type. (NL80211_IFTYPE_XXX) */
         int type; 
         /* Counter of active monitor interfaces. */
         int monitor;
diff -urN acx-mac80211-20080805/common.c acx-mac80211-20080805.new/common.c
--- acx-mac80211-20080805/common.c	2009-05-07 14:32:44.000000000 +0200
+++ acx-mac80211-20080805.new/common.c	2009-05-07 16:11:26.000000000 +0200
@@ -4105,28 +4105,28 @@
 
 	if (adev->interface.operating) {
 		switch (adev->interface.type) {
-			case IEEE80211_IF_TYPE_AP:
+			case NL80211_IFTYPE_AP:
 				if (adev->mode != ACX_MODE_3_AP)
 				{
 					adev->mode = ACX_MODE_3_AP;
 					changed = 1;
 				}
 				break;
-			case IEEE80211_IF_TYPE_IBSS:
+			case NL80211_IFTYPE_ADHOC:
 				if (adev->mode != ACX_MODE_0_ADHOC)
 				{
 					adev->mode = ACX_MODE_0_ADHOC;
 					changed = 1;
 				}
 				break;
-			case IEEE80211_IF_TYPE_STA:
+			case NL80211_IFTYPE_STATION:
 				if (adev->mode != ACX_MODE_2_STA)
 				{
 					adev->mode = ACX_MODE_2_STA;
 					changed = 1;
 				}
 				break;
-			case IEEE80211_IF_TYPE_WDS:
+			case NL80211_IFTYPE_WDS:
 			default:
 				if (adev->mode != ACX_MODE_OFF)
 				{
@@ -4136,7 +4136,7 @@
 			break;
 		}
 	} else {
-		if (adev->interface.type == IEEE80211_IF_TYPE_MNTR)
+		if (adev->interface.type == NL80211_IFTYPE_MONITOR)
 		{
 			if (adev->mode != ACX_MODE_MONITOR)
 			{
@@ -4180,7 +4180,7 @@
 	FN_ENTER;
 	acx_lock(adev, flags);
 
-	if (conf->type == IEEE80211_IF_TYPE_MNTR) {
+	if (conf->type == NL80211_IFTYPE_MONITOR) {
 		adev->interface.monitor++;
 	} else {
 		if (adev->interface.operating)
@@ -4226,7 +4226,7 @@
 	FN_ENTER;
 
 	acx_sem_lock(adev);
-	if (conf->type == IEEE80211_IF_TYPE_MNTR) {
+	if (conf->type == NL80211_IFTYPE_MONITOR) {
 		adev->interface.monitor--;
 //                assert(bcm->interface.monitor >= 0);
 	} else {
@@ -4380,7 +4380,7 @@
 
 	acx_lock(adev, flags);
 
-	if ((vif->type != IEEE80211_IF_TYPE_MNTR)
+	if ((vif->type != NL80211_IFTYPE_MONITOR)
 	    && (adev->vif == vif)) {
 		if (conf->bssid)
 		{
@@ -4388,7 +4388,7 @@
  	             	MAC_COPY(adev->bssid,conf->bssid);
 		}
 	}
-	if ((vif->type == IEEE80211_IF_TYPE_AP)
+	if ((vif->type == NL80211_IFTYPE_AP)
 	    && (adev->vif == vif)) {
 		if ((conf->ssid_len > 0) && conf->ssid)
 		{
