--- a/Makefile
+++ b/Makefile
@@ -29,7 +29,7 @@ ifneq ($(KERNELRELEASE),)
 	acx-mac80211-obj-$(CONFIG_ACX_MAC80211_PCI) += pci.o
 	acx-mac80211-obj-$(CONFIG_ACX_MAC80211_USB) += usb.o
 	acx-mac80211-obj-$(CONFIG_ACX_MAC80211_MEM) += mem.o
-	acx-mac80211-objs := common.o $(acx-mac80211-obj-y)
+	acx-mac80211-objs := common.o $(acx-mac80211-obj-m)
 	
 else
 # Otherwise we were called directly from the command line: the kernel build
