Index: acx-mac80211-20120422/common.c
===================================================================
--- acx-mac80211-20120422.orig/common.c	2012-04-23 17:35:13.000000000 +0200
+++ acx-mac80211-20120422/common.c	2012-04-23 17:37:02.818257747 +0200
@@ -33,6 +33,7 @@
 #include <linux/ethtool.h>
 #include <linux/nl80211.h>
 #include <linux/ieee80211.h>
+#include <linux/ratelimit.h>
 
 #include <net/mac80211.h>
 
