diff --git a/pci.c b/pci.c
index 4f5c0a7..160f3e0 100644
--- a/pci.c
+++ b/pci.c
@@ -273,10 +273,10 @@ static void acxpci_log_txbuffer(acx_device_t * adev)
  */
 
 /* OS I/O routines *always* be endianness-clean but having them doesn't hurt */
-#define acx_readl(v)	le32_to_cpu(readl((v)))
-#define acx_readw(v)	le16_to_cpu(readw((v)))
-#define acx_writew(v,r)	writew(le16_to_cpu((v)), r)
-#define acx_writel(v,r)	writel(le32_to_cpu((v)), r)
+#define acx_readl(v)	readl((v))
+#define acx_readw(v)	readw((v))
+#define acx_writew(v,r)	writew((v), r)
+#define acx_writel(v,r)	writel((v), r)
 
 INLINE_IO u32 read_reg32(acx_device_t * adev, unsigned int offset)
 {
