Index: acx-mac80211-20100422/pci.c
===================================================================
--- acx-mac80211-20100422.orig/pci.c	2010-05-02 22:06:03.000000000 +0200
+++ acx-mac80211-20100422/pci.c	2010-05-02 22:10:40.963886264 +0200
@@ -170,7 +170,7 @@
 #endif
 
 #ifdef CONFIG_VLYNQ
-int vlynq_probe(struct vlynq_device *vdev, struct vlynq_device_id *id);
+static int vlynq_probe(struct vlynq_device *vdev, struct vlynq_device_id *id);
 static void vlynq_remove(struct vlynq_device *vdev);
 #endif
 
@@ -4433,6 +4433,8 @@
 		goto fail_alloc_netdev;
 	}
 	ieee->flags &=	 ~IEEE80211_HW_RX_INCLUDES_FCS;
+	ieee->wiphy->interface_modes = BIT(NL80211_IFTYPE_STATION)
+			| BIT(NL80211_IFTYPE_ADHOC);
 	ieee->queues = 1;
 
 	adev = ieee2adev(ieee);
Index: acx-mac80211-20100422/acx_func.h
===================================================================
--- acx-mac80211-20100422.orig/acx_func.h	2010-05-02 03:07:09.000000000 +0200
+++ acx-mac80211-20100422/acx_func.h	2010-05-02 22:15:55.476748737 +0200
@@ -562,10 +562,6 @@
 
 // Driver, Module
 
-#ifdef CONFIG_VLYNQ
-int vlynq_probe(struct vlynq_device *vdev, struct vlynq_device_id *id);
-#endif
-
 int __init acxpci_e_init_module(void);
 void __exit acxpci_e_cleanup_module(void);
 
