diff -ruN asterisk-1.0.7-old/asterisk.h asterisk-1.0.7-new/asterisk.h
--- asterisk-1.0.7-old/asterisk.h	2004-09-07 17:02:53.000000000 +0200
+++ asterisk-1.0.7-new/asterisk.h	2005-05-17 09:56:13.000000000 +0200
@@ -27,7 +27,7 @@
 #define AST_LOG_DIR	ASTLOGDIR
 #define AST_AGI_DIR	ASTAGIDIR
 #define AST_KEY_DIR	ASTVARLIBDIR "/keys"
-#define AST_DB		ASTVARLIBDIR "/astdb"
+#define AST_DB		ASTSPOOLDIR "/astdb"
 #define AST_TMP_DIR	ASTSPOOLDIR "/tmp"
 
 #define AST_CONFIG_FILE ASTCONFPATH
