--- asterisk-1.0.7.old/res/Makefile	2005-04-25 20:53:11.000000000 +0200
+++ asterisk-1.0.7/res/Makefile	2005-04-25 20:53:26.000000000 +0200
@@ -11,7 +11,7 @@
 # the GNU General Public License
 #
 
-MODS=res_adsi.so res_features.so res_crypto.so res_musiconhold.so res_indications.so res_monitor.so \
+MODS=res_adsi.so res_features.so res_musiconhold.so res_indications.so res_monitor.so \
      res_agi.so
 #MODS+=$(shell if [ -f "/usr/include/odbcinst.h" ]; then echo "res_odbc.so res_config_odbc.so"; fi)
 #MODS+=$(shell if [ -f "/usr/local/include/odbcinst.h" ]; then echo "res_odbc.so res_config_odbc.so"; fi)
