diff -urN iptraf-3.0.0/support/Makefile iptraf-3.0.0.new/support/Makefile
--- iptraf-3.0.0/support/Makefile	2002-07-19 18:31:48.000000000 +0200
+++ iptraf-3.0.0.new/support/Makefile	2006-04-16 16:07:58.000000000 +0200
@@ -3,16 +3,14 @@
 OBJS			= input.o menurt.o listbox.o winops.o labels.o \
 				msgboxes.o txbox.o
 
-all: libtextbox.a
-
 libtextbox.a: $(OBJS)
 	rm -rf libtextbox.a
-	ar cq libtextbox.a $(OBJS)
-	ranlib libtextbox.a
-#	gcc -shared -o libtextbox.so $(OBJS)
+	$(AR) cq libtextbox.a $(OBJS)
+	$(RANLIB) libtextbox.a
+#	$(CC) -shared -o libtextbox.so $(OBJS)
 
 %.o: %.c *.h
-	gcc -O2 -g -Wall -fPIC $(INCLUDEDIR) -c -o $*.o $<
+	$(CC) -O2 -g -Wall -fPIC $(INCLUDEDIR) -c -o $*.o $<
 
 clean:
 	rm -rf *.o *~ libtextbox.a libtextbox.so
