Index: linux-2.6.21.5/drivers/net/Makefile
===================================================================
--- linux-2.6.21.5.orig/drivers/net/Makefile	2007-06-24 09:03:20.308277264 +0200
+++ linux-2.6.21.5/drivers/net/Makefile	2007-06-24 09:05:13.560060376 +0200
@@ -219,3 +219,5 @@
 obj-$(CONFIG_FS_ENET) += fs_enet/
 
 obj-$(CONFIG_NETXEN_NIC) += netxen/
+
+obj-$(CONFIG_AMAZON_NET_SW) += amazon_sw.o
