Index: linux-2.6.24/drivers/usb/host/adm5120-q.c
===================================================================
--- linux-2.6.24.orig/drivers/usb/host/adm5120-q.c
+++ linux-2.6.24/drivers/usb/host/adm5120-q.c
@@ -48,8 +48,7 @@ __acquires(ahcd->lock)
 	/* urb->complete() can reenter this HCD */
 	usb_hcd_unlink_urb_from_ep(admhcd_to_hcd(ahcd), urb);
 	spin_unlock(&ahcd->lock);
-	urb->status = status;
-	usb_hcd_giveback_urb(admhcd_to_hcd(ahcd), urb);
+	usb_hcd_giveback_urb(admhcd_to_hcd(ahcd), urb, status);
 	spin_lock(&ahcd->lock);
 }
 
