reverted:
--- a/drivers/net/phy/swconfig.c
+++ b/drivers/net/phy/swconfig.c
@@ -378,7 +378,7 @@ swconfig_dump_attr(struct swconfig_callb
 	int id = cb->args[0];
 	void *hdr;
 
-	hdr = genlmsg_put(msg, info->snd_portid, info->snd_seq, &switch_fam,
+	hdr = genlmsg_put(msg, info->snd_pid, info->snd_seq, &switch_fam,
 			NLM_F_MULTI, SWITCH_CMD_NEW_ATTR);
 	if (IS_ERR(hdr))
 		return -1;
@@ -801,7 +801,7 @@ swconfig_get_attr(struct sk_buff *skb, s
 	if (!msg)
 		goto error;
 
-	hdr = genlmsg_put(msg, info->snd_portid, info->snd_seq, &switch_fam,
+	hdr = genlmsg_put(msg, info->snd_pid, info->snd_seq, &switch_fam,
 			0, cmd);
 	if (IS_ERR(hdr))
 		goto nla_put_failure;
@@ -906,7 +906,7 @@ static int swconfig_dump_switches(struct
 	list_for_each_entry(dev, &swdevs, dev_list) {
 		if (++idx <= start)
 			continue;
-		if (swconfig_send_switch(skb, NETLINK_CB(cb->skb).portid,
+		if (swconfig_send_switch(skb, NETLINK_CB(cb->skb).pid,
 				cb->nlh->nlmsg_seq, NLM_F_MULTI,
 				dev) < 0)
 			break;
