--- a/drivers/pci/probe.c
+++ b/drivers/pci/probe.c
@@ -263,7 +263,8 @@ int __pci_read_base(struct pci_dev *dev,
 		if ((sizeof(pci_bus_addr_t) < 8 || sizeof(resource_size_t) < 8)
 		    && sz64 > 0x100000000ULL) {
 			res->flags |= IORESOURCE_UNSET | IORESOURCE_DISABLED;
-			resource_set_range(res, 0, 0);
+			res->start = 0;
+			res->end = 0;
 			pci_err(dev, "%s: can't handle BAR larger than 4GB (size %#010llx)\n",
 				res_name, (unsigned long long)sz64);
 			goto out;
@@ -272,7 +273,8 @@ int __pci_read_base(struct pci_dev *dev,
 		if ((sizeof(pci_bus_addr_t) < 8) && l) {
 			/* Above 32-bit boundary; try to reallocate */
 			res->flags |= IORESOURCE_UNSET;
-			resource_set_range(res, 0, sz64);
+			res->start = 0;
+			res->end = sz64 - 1;
 			pci_info(dev, "%s: can't handle BAR above 4GB (bus address %#010llx)\n",
 				 res_name, (unsigned long long)l64);
 			goto out;
