From 28f3689e0e5f4948bbd86ff2e2f3da44aa6a5bca Mon Sep 17 00:00:00 2001
From: Ahmed Naseef <naseefkm@gmail.com>
Date: Thu, 6 Aug 2026 15:50:28 -0300
Subject: [PATCH] serial: 8250: airoha: add EcoNet platform support

Signed-off-by: Matheus Sampaio Queiroga <srherobrine20@gmail.com>
---
 drivers/tty/serial/8250/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/tty/serial/8250/Kconfig b/drivers/tty/serial/8250/Kconfig
index ae199803cc95..c90809cda7be 100644
--- a/drivers/tty/serial/8250/Kconfig
+++ b/drivers/tty/serial/8250/Kconfig
@@ -358,7 +358,7 @@ config SERIAL_8250_ACORN
 
 config SERIAL_8250_AIROHA
 	tristate "Airoha UART support"
-	depends on (ARCH_AIROHA || COMPILE_TEST) && OF && SERIAL_8250
+	depends on (ARCH_AIROHA || ECONET || COMPILE_TEST) && OF && SERIAL_8250
 	help
 	  Selecting this option enables an Airoha SoC specific baud rate
 	  calculation routine on an otherwise 16550 compatible UART hardware.
-- 
2.55.0

