Index: linux-4.9.44/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts
===================================================================
--- linux-4.9.44.orig/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts	2017-09-02 08:21:11.998199513 +0800
+++ linux-4.9.44/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts	2017-09-02 08:33:09.452252584 +0800
@@ -18,7 +18,7 @@
 	};
 */
 	memory {
-		reg = <0 0x80000000 0 0x20000000>;
+		reg = <0 0x80000000 0 0x80000000>;
 	};
 
 	aliases {
