Index: linux-4.14.51/drivers/misc/mediatek/connectivity/common/conn_soc/linux/pub/wmt_chrdev_wifi.c
===================================================================
--- linux-4.14.51.orig/drivers/misc/mediatek/connectivity/common/conn_soc/linux/pub/wmt_chrdev_wifi.c
+++ linux-4.14.51/drivers/misc/mediatek/connectivity/common/conn_soc/linux/pub/wmt_chrdev_wifi.c
@@ -62,7 +62,7 @@ UINT32 gDbgLevel = WIFI_LOG_DBG;
 
 #define VERSION "1.0"
 
-#define WLAN_IFACE_NAME "wlan0"
+#define WLAN_IFACE_NAME "mtkwlan0"
 #if CFG_TC1_FEATURE
 #define LEGACY_IFACE_NAME "legacy0"
 #endif
