Index: linux-2.4.35.4/net/sched/sch_generic.c
===================================================================
--- linux-2.4.35.4.orig/net/sched/sch_generic.c	2007-12-15 05:20:10.816456638 +0100
+++ linux-2.4.35.4/net/sched/sch_generic.c	2007-12-15 05:20:16.704792197 +0100
@@ -84,6 +84,11 @@
 	struct sk_buff *skb;
 
 	/* Dequeue packet */
+	if (!q) {
+		if (net_ratelimit())
+			printk(KERN_DEBUG "HELP ME! qdisc_restart called, but no Qdisc!\n");
+		return 0;
+	}
 	if ((skb = q->dequeue(q)) != NULL) {
 		if (spin_trylock(&dev->xmit_lock)) {
 			/* Remember that the driver is grabbed by us. */
