--- a/arch/mips/mm/c-r4k.c
+++ b/arch/mips/mm/c-r4k.c
@@ -325,7 +325,7 @@ static void r4k_flush_cache_mm(struct mm
 		r4k_blast_scache();
 }
 
-static void r4k_flush_cache_page(struct vm_area_struct *vma,
+void r4k_flush_cache_page(struct vm_area_struct *vma,
 					unsigned long page)
 {
 	int exec = vma->vm_flags & VM_EXEC;
@@ -385,6 +385,7 @@ static void r4k_flush_cache_page(struct 
 			r4k_blast_icache_page_indexed(page);
 	}
 }
+EXPORT_SYMBOL(r4k_flush_cache_page); 
 
 static void r4k_flush_data_cache_page(unsigned long addr)
 {
--- a/arch/mips/mm/Makefile
+++ b/arch/mips/mm/Makefile
@@ -11,7 +11,7 @@ USE_STANDARD_AS_RULE := true
 O_TARGET := mm.o
 
 export-objs			:= cache.o ioremap.o loadmmu.o remap.o \
-				   tlb-r4k.o tlb-sb1.o
+				   tlb-r4k.o tlb-sb1.o c-r4k.o
 obj-y				+= cache.o extable.o init.o ioremap.o fault.o \
 				   loadmmu.o
 
