Index: linux-2.6.23/drivers/serial/Kconfig
===================================================================
--- linux-2.6.23.orig/drivers/serial/Kconfig	2007-12-10 21:38:39.000000000 +0100
+++ linux-2.6.23/drivers/serial/Kconfig	2007-12-10 21:38:42.000000000 +0100
@@ -1259,4 +1259,10 @@
 	  Currently, only 8250 compatible ports are supported, but
 	  others can easily be added.
 
+config SERIAL_DANUBE
+	bool "Danube serial driver"
+	depends on DANUBE
+	help
+	  Driver for the danubes built in ASC hardware
+
 endmenu
Index: linux-2.6.23/drivers/serial/Makefile
===================================================================
--- linux-2.6.23.orig/drivers/serial/Makefile	2007-12-10 21:38:39.000000000 +0100
+++ linux-2.6.23/drivers/serial/Makefile	2007-12-10 21:38:42.000000000 +0100
@@ -64,3 +64,4 @@
 obj-$(CONFIG_SERIAL_NETX) += netx-serial.o
 obj-$(CONFIG_SERIAL_OF_PLATFORM) += of_serial.o
 obj-$(CONFIG_SERIAL_KS8695) += serial_ks8695.o
+obj-$(CONFIG_SERIAL_DANUBE) += danube_asc.o
Index: linux-2.6.23/drivers/mtd/maps/Makefile
===================================================================
--- linux-2.6.23.orig/drivers/mtd/maps/Makefile	2007-12-10 21:38:39.000000000 +0100
+++ linux-2.6.23/drivers/mtd/maps/Makefile	2007-12-10 21:38:42.000000000 +0100
@@ -71,3 +71,4 @@
 obj-$(CONFIG_MTD_OMAP_NOR)	+= omap_nor.o
 obj-$(CONFIG_MTD_MTX1)		+= mtx-1_flash.o
 obj-$(CONFIG_MTD_TQM834x)	+= tqm834x.o
+obj-$(CONFIG_MTD_DANUBE)	+= danube.o
Index: linux-2.6.23/drivers/net/Kconfig
===================================================================
--- linux-2.6.23.orig/drivers/net/Kconfig	2007-12-10 21:38:42.000000000 +0100
+++ linux-2.6.23/drivers/net/Kconfig	2007-12-10 21:38:42.000000000 +0100
@@ -339,6 +339,18 @@
 
 source "drivers/net/arm/Kconfig"
 
+config DANUBE_MII0
+	tristate "Infineon Danube eth0 driver"
+	depends on DANUBE
+	help
+	  Support for the MII0 inside the Danube SOC
+
+config DANUBE_MII1
+	tristate "Infineon Danube eth1 driver"
+	depends on DANUBE
+	help
+	  Support for the MII1 inside the Danube SOC
+
 config AX88796
 	tristate "ASIX AX88796 NE2000 clone support"
 	depends on ARM || MIPS
Index: linux-2.6.23/drivers/net/Makefile
===================================================================
--- linux-2.6.23.orig/drivers/net/Makefile	2007-12-10 21:38:42.000000000 +0100
+++ linux-2.6.23/drivers/net/Makefile	2007-12-10 21:38:42.000000000 +0100
@@ -208,6 +208,7 @@
 obj-$(CONFIG_FEC_8XX) += fec_8xx/
 obj-$(CONFIG_PASEMI_MAC) += pasemi_mac.o
 obj-$(CONFIG_MLX4_CORE) += mlx4/
+obj-$(CONFIG_DANUBE_MII0) += danube_mii0.o
 
 obj-$(CONFIG_MACB) += macb.o
 
Index: linux-2.6.23/drivers/char/watchdog/Makefile
===================================================================
--- linux-2.6.23.orig/drivers/char/watchdog/Makefile	2007-12-10 21:41:51.000000000 +0100
+++ linux-2.6.23/drivers/char/watchdog/Makefile	2007-12-10 21:42:30.000000000 +0100
@@ -90,6 +90,7 @@
 obj-$(CONFIG_INDYDOG) += indydog.o
 obj-$(CONFIG_WDT_MTX1)	+= mtx-1_wdt.o
 obj-$(CONFIG_WDT_RM9K_GPI) += rm9k_wdt.o
+obj-$(CONFIG_DANUBE_WDT) += danube_wdt.o
 
 # PARISC Architecture
 
