--- a/arch/arm64/boot/dts/mediatek/mt7622-bananapi-bpi-r64.dts
+++ b/arch/arm64/boot/dts/mediatek/mt7622-bananapi-bpi-r64.dts
@@ -108,10 +108,6 @@
 	status = "disabled";
 };
 
-&btif {
-	status = "okay";
-};
-
 &cir {
 	pinctrl-names = "default";
 	pinctrl-0 = <&irrx_pins>;
--- a/arch/arm64/boot/dts/mediatek/mt7622-rfb1.dts
+++ b/arch/arm64/boot/dts/mediatek/mt7622-rfb1.dts
@@ -89,10 +89,6 @@
 	status = "disabled";
 };
 
-&btif {
-	status = "okay";
-};
-
 &cir {
 	pinctrl-names = "default";
 	pinctrl-0 = <&irrx_pins>;
