Index: linux-2.6.32.13/arch/mips/bcm63xx/boards/board_bcm963xx.c
===================================================================
--- linux-2.6.32.13.orig/arch/mips/bcm63xx/boards/board_bcm963xx.c	2010-05-27 21:48:23.000000000 -0400
+++ linux-2.6.32.13/arch/mips/bcm63xx/boards/board_bcm963xx.c	2010-05-27 21:48:31.000000000 -0400
@@ -1322,6 +1322,8 @@
 		if (BCMCPU_IS_6348())
 			val |= GPIO_MODE_6348_G3_EXT_MII |
 				GPIO_MODE_6348_G0_EXT_MII;
+		else if (BCMCPU_IS_6358())
+			val |= GPIO_MODE_6358_ENET1_MII_CLK_INV;
 	}
 
 	bcm_gpio_writel(val, GPIO_MODE_REG);
Index: linux-2.6.32.13/arch/mips/include/asm/mach-bcm63xx/bcm63xx_regs.h
===================================================================
--- linux-2.6.32.13.orig/arch/mips/include/asm/mach-bcm63xx/bcm63xx_regs.h	2010-05-27 21:48:23.000000000 -0400
+++ linux-2.6.32.13/arch/mips/include/asm/mach-bcm63xx/bcm63xx_regs.h	2010-05-27 21:48:31.000000000 -0400
@@ -402,6 +402,8 @@
 #define GPIO_MODE_6358_EXTRA_SPI_SS	(1 << 7)
 #define GPIO_MODE_6358_SERIAL_LED	(1 << 10)
 #define GPIO_MODE_6358_UTOPIA		(1 << 12)
+#define GPIO_MODE_6358_ENET0_MII_CLK_INV (1 << 30)
+#define GPIO_MODE_6358_ENET1_MII_CLK_INV (1 << 31)
 
 
 /*************************************************************************
