--- vsftpd-2.0.4-orig/tunables.c	2006-01-07 20:32:44.000000000 +0100
+++ vsftpd-2.0.4/tunables.c	2006-01-11 18:25:15.000000000 +0100
@@ -92,7 +92,7 @@
 unsigned int tunable_max_per_ip = 0;
 unsigned int tunable_trans_chunk_size = 0;
 
-const char* tunable_secure_chroot_dir = "/usr/share/empty";
+const char* tunable_secure_chroot_dir = "/var/run/vsftpd";
 const char* tunable_ftp_username = "ftp";
 const char* tunable_chown_username = "root";
 const char* tunable_xferlog_file = "/var/log/xferlog";
