--- linux-2.6.26/include/linux/pim.h	2008-07-13 23:51:29.000000000 +0200
+++ linux-2.6.26.new/include/linux/pim.h	2008-08-05 20:52:25.000000000 +0200
@@ -7,10 +7,10 @@
 struct pim {
 #if defined(__LITTLE_ENDIAN_BITFIELD)
 	__u8	pim_type:4,		/* PIM message type */
-		pim_ver:4;		/* PIM version */
+	__u8	pim_ver:4;		/* PIM version */
 #elif defined(__BIG_ENDIAN_BITFIELD)
 	__u8	pim_ver:4;		/* PIM version */
-		pim_type:4;		/* PIM message type */
+	__u8	pim_type:4;		/* PIM message type */
 #endif
 	__u8	pim_rsv;		/* Reserved */
 	__be16	pim_cksum;		/* Checksum */
