--- a/crypto/Kconfig
+++ b/crypto/Kconfig
@@ -22,20 +22,20 @@ if CRYPTO
 comment "Crypto core or helper"
 
 config CRYPTO_ALGAPI
-	tristate
+	tristate "ALG API"
 	help
 	  This option provides the API for cryptographic algorithms.
 
 config CRYPTO_AEAD
-	tristate
+	tristate "AEAD"
 	select CRYPTO_ALGAPI
 
 config CRYPTO_BLKCIPHER
-	tristate
+	tristate "Block cipher"
 	select CRYPTO_ALGAPI
 
 config CRYPTO_HASH
-	tristate
+	tristate "HASH"
 	select CRYPTO_ALGAPI
 
 config CRYPTO_MANAGER
