--- a/sysklogd/syslogd.c
+++ b/sysklogd/syslogd.c
@@ -278,7 +278,7 @@ static void parse_syslogdcfg(const char 
 	parser_t *parser;
 
 	parser = config_open2(file ? file : "/etc/syslog.conf",
-				file ? xfopen_for_read : fopen_or_warn_stdin);
+				file ? xfopen_for_read : fopen_for_read);
 	if (!parser)
 		/* didn't find default /etc/syslog.conf */
 		/* proceed as if we built busybox without config support */
@@ -678,7 +678,7 @@ static void timestamp_and_log(int pri, c
 	if (LOG_PRI(pri) < G.logLevel) {
 #if ENABLE_FEATURE_IPC_SYSLOG
 		if ((option_mask32 & OPT_circularlog) && G.shbuf) {
-			log_to_shmem(msg);
+			log_to_shmem(G.printbuf);
 			return;
 		}
 #endif
