--- a/networking/httpd_indexcgi.c
+++ b/networking/httpd_indexcgi.c
@@ -315,7 +315,7 @@ int main(int argc, char *argv[])
 		if (S_ISREG(cdir->dl_mode))
 			fmt_ull(cdir->dl_size);
 		fmt_str("<td class=dt>");
-		tm = gmtime(&cdir->dl_mtime);
+		ptm = gmtime(&cdir->dl_mtime);
 		fmt_04u(1900 + ptm->tm_year); *dst++ = '-';
 		fmt_02u(ptm->tm_mon + 1); *dst++ = '-';
 		fmt_02u(ptm->tm_mday); *dst++ = ' ';
