Index: busybox-1.4.2/Makefile.flags
===================================================================
--- busybox-1.4.2.orig/Makefile.flags	2007-06-04 13:21:32.387053088 +0200
+++ busybox-1.4.2/Makefile.flags	2007-06-04 13:21:32.871979368 +0200
@@ -16,7 +16,7 @@
 	-D"BB_VER=KBUILD_STR($(BB_VER))" -DBB_BT=AUTOCONF_TIMESTAMP
 
 CFLAGS += \
-	-Wall -Wstrict-prototypes -Wshadow -Werror -Wundef \
+	-Wall -Wstrict-prototypes -Wshadow -Wundef \
 	-funsigned-char -fno-builtin-strlen -finline-limit=0 -static-libgcc \
 	-Os -falign-functions=1 -falign-jumps=1 -falign-loops=1 \
 	-fomit-frame-pointer -ffunction-sections -fdata-sections
