diff -urN busybox-1.4.1/networking/wget.c busybox-1.4.1-wgetSEGV/networking/wget.c
--- busybox-1.4.1/networking/wget.c	2007-01-24 22:34:34.000000000 +0100
+++ busybox-1.4.1-wgetSEGV/networking/wget.c	2007-02-11 17:21:18.000000000 +0100
@@ -543,7 +543,9 @@
 	p = strchr(h->host, '?'); if (!sp || (p && sp > p)) sp = p;
 	p = strchr(h->host, '#'); if (!sp || (p && sp > p)) sp = p;
 	if (!sp) {
-		h->path = "";
+		/* must be writable because of bb_get_last_path_component() */
+		static char nullstr[] = "";
+		h->path = nullstr;
 	} else if (*sp == '/') {
 		*sp = '\0';
 		h->path = sp + 1;
