--- a/miscutils/crond.c
+++ b/miscutils/crond.c
@@ -24,7 +24,7 @@
 
 
 #ifndef CRONTABS
-#define CRONTABS        "/var/spool/cron/crontabs"
+#define CRONTABS        "/etc/crontabs"
 #endif
 #ifndef TMPDIR
 #define TMPDIR          "/var/spool/cron"
--- a/miscutils/crontab.c
+++ b/miscutils/crontab.c
@@ -13,7 +13,7 @@
 #include "libbb.h"
 
 #ifndef CRONTABS
-#define CRONTABS        "/var/spool/cron/crontabs"
+#define CRONTABS        "/etc/crontabs"
 #endif
 #ifndef CRONUPDATE
 #define CRONUPDATE      "cron.update"
