Index: busybox-1.7.2/miscutils/crond.c
===================================================================
--- busybox-1.7.2.orig/miscutils/crond.c	2007-10-30 15:35:00.000000000 -0500
+++ busybox-1.7.2/miscutils/crond.c	2007-10-30 15:35:00.000000000 -0500
@@ -15,7 +15,7 @@
 #include "libbb.h"
 
 #ifndef CRONTABS
-#define CRONTABS        "/var/spool/cron/crontabs"
+#define CRONTABS        "/etc/crontabs"
 #endif
 #ifndef TMPDIR
 #define TMPDIR          "/var/spool/cron"
Index: busybox-1.7.2/miscutils/crontab.c
===================================================================
--- busybox-1.7.2.orig/miscutils/crontab.c	2007-10-30 15:35:00.000000000 -0500
+++ busybox-1.7.2/miscutils/crontab.c	2007-10-30 15:35:00.000000000 -0500
@@ -13,7 +13,7 @@
 #include "libbb.h"
 
 #ifndef CRONTABS
-#define CRONTABS        "/var/spool/cron/crontabs"
+#define CRONTABS        "/etc/crontabs"
 #endif
 #ifndef TMPDIR
 #define TMPDIR          "/var/spool/cron"
