Index: busybox-1.7.2/networking/udhcp/dhcpc.c
===================================================================
--- busybox-1.7.2.orig/networking/udhcp/dhcpc.c	2007-10-04 14:36:41.521873204 +0200
+++ busybox-1.7.2/networking/udhcp/dhcpc.c	2007-10-04 14:38:28.231954268 +0200
@@ -309,6 +309,12 @@
  jump_in:
 		tv.tv_usec = 0;
 
+		/* When running on a bridge, the ifindex may have changed (e.g. if
+		 * member interfaces were added/removed or if the status of the
+		 * bridge changed).
+		 * Workaround: refresh it here before processing the next packet */
+		read_interface(client_config.interface, &client_config.ifindex, NULL, client_config.arp);
+
 		if (listen_mode != LISTEN_NONE && sockfd < 0) {
 			if (listen_mode == LISTEN_KERNEL)
 				sockfd = listen_socket(/*INADDR_ANY,*/ CLIENT_PORT, client_config.interface);
