Index: busybox-1.4.2/include/platform.h
===================================================================
--- busybox-1.4.2.orig/include/platform.h	2007-06-04 13:21:32.055103552 +0200
+++ busybox-1.4.2/include/platform.h	2007-06-04 13:21:35.387596936 +0200
@@ -128,9 +128,11 @@
 # include <netinet/in.h>
 #endif
 
+#ifndef __APPLE__
 #ifndef __socklen_t_defined
 typedef int socklen_t;
 #endif
+#endif
 
 /* ---- Compiler dependent settings ------------------------- */
 #if (defined __digital__ && defined __unix__)
@@ -161,7 +163,7 @@
 #define HAVE_STDINT_H
 #else
 /* Largest integral types.  */
-#if __BIG_ENDIAN__
+#if __BIG_ENDIAN__ && !__APPLE__
 typedef long int                intmax_t;
 typedef unsigned long int       uintmax_t;
 #else
Index: busybox-1.4.2/include/libbb.h
===================================================================
--- busybox-1.4.2.orig/include/libbb.h	2007-06-04 13:21:32.062102488 +0200
+++ busybox-1.4.2/include/libbb.h	2007-06-04 13:21:35.388596784 +0200
@@ -17,7 +17,9 @@
 #include <errno.h>
 #include <fcntl.h>
 #include <inttypes.h>
+#ifndef __APPLE__
 #include <malloc.h>
+#endif
 #include <netdb.h>
 #include <setjmp.h>
 #include <signal.h>
@@ -30,7 +32,9 @@
 #include <sys/mman.h>
 #include <sys/socket.h>
 #include <sys/stat.h>
+#ifndef __APPLE__
 #include <sys/statfs.h>
+#endif
 #include <sys/time.h>
 #include <sys/types.h>
 #include <sys/wait.h>
