Index: busybox-1.15.2/networking/ping.c
===================================================================
--- busybox-1.15.2.orig/networking/ping.c	2009-11-24 22:57:29.000000000 +0100
+++ busybox-1.15.2/networking/ping.c	2009-11-24 22:58:58.000000000 +0100
@@ -769,6 +769,7 @@
 int ping6_main(int argc, char **argv) MAIN_EXTERNALLY_VISIBLE;
 int ping6_main(int argc UNUSED_PARAM, char **argv)
 {
+	argv[-1] = argv[0];
 	argv[0] = (char*)"-6";
 	return ping_main(0 /* argc+1 - but it's unused anyway */,
 			argv - 1);
