Index: busybox-1.8.2/archival/libipkg/pkg.c
===================================================================
--- busybox-1.8.2.orig/archival/libipkg/pkg.c	2008-01-02 21:53:12.616453611 +0100
+++ busybox-1.8.2/archival/libipkg/pkg.c	2008-01-02 21:54:41.462489150 +0100
@@ -587,13 +587,14 @@
 	          return NULL;
                }
                temp[0]='\0';
-               strncpy(temp, "Conffiles:\n", 12);
+               strncpy(temp, "Conffiles: ", 12);
 	       for (iter = pkg->conffiles.head; iter; iter = iter->next) {
 		    if (iter->data->name && iter->data->value) {
-                         snprintf(line_str, LINE_LEN, "%s %s\n", iter->data->name, iter->data->value);
+                         snprintf(line_str, LINE_LEN, "%s %s", iter->data->name, iter->data->value);
                          strncat(temp, line_str, strlen(line_str));           
 		    }
 	       }
+	       strcat(temp, "\n");
 	  } else if (strcasecmp(field, "Conflicts") == 0) {
 	       int i;
 
