diff -Naur busybox-1.19.3-old/include/platform.h busybox-1.19.3/include/platform.h
--- busybox-1.19.3-old/include/platform.h	2011-10-29 05:43:01.000000000 -0600
+++ busybox-1.19.3/include/platform.h	2011-12-18 15:53:49.817278374 -0700
@@ -443,6 +443,13 @@
 # undef HAVE_NET_ETHERNET_H
 #endif
 
+#if defined(__musl__)
+# undef HAVE_SETBIT
+# include <stddef.h>
+# include <termios.h>
+# include <sys/ioctl.h>
+#endif
+
 /*
  * Now, define prototypes for all the functions defined in platform.c
  * These must come after all the HAVE_* macros are defined (or not)
