diff -Nur privoxy-3.0.3-stable/configure.in privoxy-3.0.3-stable.patched/configure.in
--- privoxy-3.0.3-stable/configure.in	2004-01-30 10:26:03.000000000 +0100
+++ privoxy-3.0.3-stable.patched/configure.in	2006-03-13 20:50:57.000000000 +0100
@@ -1092,7 +1092,8 @@
 AC_PROG_GCC_TRADITIONAL
 dnl uncommenting does not work for swa. suse linux
 dnl AC_FUNC_MALLOC
-AC_FUNC_SETPGRP
+dnl uncommenting does not work for openwrt
+dnl AC_FUNC_SETPGRP
 AC_TYPE_SIGNAL
 dnl uncommenting does not work for swa. suse linux
 dnl AC_FUNC_STAT
