--- a/sierra_net.c
+++ b/sierra_net.c
@@ -858,10 +858,7 @@ static int sierra_net_bind(struct usbnet
 	/* test whether firmware supports DHCP */
 	if (!(status == sizeof(fwattr) && (cpu_to_le16(fwattr) & SWI_GET_FW_ATTR_MASK))) {
 		/* found incompatible firmware version */
-		dev_err(&dev->udev->dev, "Incompatible driver and firmware"
-			" versions\n");
-		kfree(priv);
-		return -ENODEV;
+		dev_err(&dev->udev->dev, "Warning: Firmware does not have DHCP support\n");
 	}
 	/* prepare sync message from template */
 	memcpy(priv->sync_msg, sync_tmplate, sizeof(priv->sync_msg));
