--- a/src/ap/beacon.c
+++ b/src/ap/beacon.c
@@ -614,6 +614,10 @@ void handle_probe_req(struct hostapd_dat
 		return;
 	}
 
+	if (!sta && hapd->num_sta >= hapd->conf->max_num_sta)
+		wpa_printf(MSG_MSGDUMP, "Probe Request from " MACSTR " ignored,"
+			   " too many connected stations.", MAC2STR(mgmt->sa));
+
 #ifdef CONFIG_INTERWORKING
 	if (hapd->conf->interworking &&
 	    elems.interworking && elems.interworking_len >= 1) {
