Index: util-linux-2.12r/fdisk/llseek.c
===================================================================
--- util-linux-2.12r.orig/fdisk/llseek.c	2007-06-04 13:22:28.258559328 +0200
+++ util-linux-2.12r/fdisk/llseek.c	2007-06-04 13:22:28.339547016 +0200
@@ -29,16 +29,10 @@
 static int _llseek (unsigned int, unsigned long,
 		   unsigned long, long long *, unsigned int);
 
-#ifdef __NR__llseek
-
-static _syscall5(int,_llseek,unsigned int,fd,unsigned long,offset_high,
-		 unsigned long, offset_low,long long *,result,
-		 unsigned int, origin)
-
-#else
+#ifndef __NR__llseek
 
 /* no __NR__llseek on compilation machine - might give it explicitly */
-static int _llseek (unsigned int fd, unsigned long oh,
+static int __llseek (unsigned int fd, unsigned long oh,
 		    unsigned long ol, long long *result,
 		    unsigned int origin) {
 	errno = ENOSYS;
@@ -47,17 +41,8 @@
 
 #endif
 
-static long long my_llseek (unsigned int fd, long long offset,
-		unsigned int origin)
-{
-	long long result;
-	int retval;
-
-	retval = _llseek (fd, ((unsigned long long) offset) >> 32,
-			((unsigned long long) offset) & 0xffffffff,
-			&result, origin);
-	return (retval == -1 ? (long long) retval : result);
-}
+extern long long llseek(unsigned int fd, long long offset, unsigned int origin);
+#define my_llseek llseek
 
 #endif /* __alpha__ */
 
Index: util-linux-2.12r/fdisk/sfdisk.c
===================================================================
--- util-linux-2.12r.orig/fdisk/sfdisk.c	2007-06-04 13:22:28.265558264 +0200
+++ util-linux-2.12r/fdisk/sfdisk.c	2007-06-04 13:22:28.338547168 +0200
@@ -176,12 +176,6 @@
 #define use_lseek
 #endif
 
-#ifndef use_lseek
-static __attribute__used
-_syscall5(int,  _llseek,  unsigned int,  fd, ulong, hi, ulong, lo,
-       loff_t *, res, unsigned int, wh);
-#endif
-
 static int
 sseek(char *dev, unsigned int fd, unsigned long s) {
     loff_t in, out;
Index: util-linux-2.12r/partx/partx.c
===================================================================
--- util-linux-2.12r.orig/partx/partx.c	2007-06-04 13:22:28.272557200 +0200
+++ util-linux-2.12r/partx/partx.c	2007-06-04 13:22:28.339547016 +0200
@@ -339,9 +339,6 @@
 
 #ifdef NEED__llseek
 #include <linux/unistd.h>       /* _syscall */
-static
-_syscall5(int,  _llseek,  uint,  fd, ulong, hi, ulong, lo,
-	  long long *, res, uint, wh);
 #endif
 
 static int
