--- a/arch/mips/bcm63xx/boards/board_common.c
+++ b/arch/mips/bcm63xx/boards/board_common.c
@@ -12,6 +12,7 @@
 #include <linux/string.h>
 #include <linux/platform_device.h>
 #include <linux/ssb/ssb.h>
+#include <linux/spi/spi.h>
 #include <asm/addrspace.h>
 #include <bcm63xx_board.h>
 #include <bcm63xx_cpu.h>
@@ -210,6 +211,9 @@ int __init board_register_devices(void)
 	if (board.num_devs)
 		platform_add_devices(board.devs, board.num_devs);
 
+	if (board.num_spis)
+		spi_register_board_info(board.spis, board.num_spis);
+
 	bcm63xx_flash_register();
 
 	bcm63xx_led_data.num_leds = ARRAY_SIZE(board.leds);
--- a/arch/mips/include/asm/mach-bcm63xx/board_bcm963xx.h
+++ b/arch/mips/include/asm/mach-bcm63xx/board_bcm963xx.h
@@ -57,6 +57,10 @@ struct board_info {
 	/* Additional platform devices */
 	struct platform_device **devs;
 	unsigned int	num_devs;
+
+	/* Additional platform devices */
+	struct spi_board_info *spis;
+	unsigned int	num_spis;
 };
 
 #endif /* ! BOARD_BCM963XX_H_ */
