--- a/arch/mips/ath79/Kconfig
+++ b/arch/mips/ath79/Kconfig
@@ -75,6 +75,16 @@ config ATH79_MACH_ANTMINER_S1
 	select ATH79_DEV_USB
 	select ATH79_DEV_WMAC
 
+config ATH79_MACH_ANTMINER_S3
+        bool "Bitmain Antminer S3 support"
+        select SOC_AR933X
+        select ATH79_DEV_ETH
+        select ATH79_DEV_GPIO_BUTTONS
+        select ATH79_DEV_LEDS_GPIO
+        select ATH79_DEV_M25P80
+        select ATH79_DEV_USB
+        select ATH79_DEV_WMAC
+
 config ATH79_MACH_AP113
 	bool "Atheros AP113 board support"
 	select SOC_AR724X
--- a/arch/mips/ath79/Makefile
+++ b/arch/mips/ath79/Makefile
@@ -44,6 +44,7 @@ obj-$(CONFIG_ATH79_MACH_ALFA_NX)	+= mach
 obj-$(CONFIG_ATH79_MACH_ALL0258N)	+= mach-all0258n.o
 obj-$(CONFIG_ATH79_MACH_ALL0315N)	+= mach-all0315n.o
 obj-$(CONFIG_ATH79_MACH_ANTMINER_S1)+= mach-antminer-s1.o
+obj-$(CONFIG_ATH79_MACH_ANTMINER_S3)+= mach-antminer-s3.o
 obj-$(CONFIG_ATH79_MACH_AP113)		+= mach-ap113.o
 obj-$(CONFIG_ATH79_MACH_AP121)		+= mach-ap121.o
 obj-$(CONFIG_ATH79_MACH_AP132)		+= mach-ap132.o
--- a/arch/mips/ath79/machtypes.h
+++ b/arch/mips/ath79/machtypes.h
@@ -23,6 +23,7 @@ enum ath79_mach_type {
 	ATH79_MACH_ALL0305,		/* Allnet ALL0305 */
 	ATH79_MACH_ALL0315N,		/* Allnet ALL0315N */
 	ATH79_MACH_ANTMINER_S1,	/* Antminer-S1 */
+	ATH79_MACH_ANTMINER_S3, /* Antminer-S3 */
 	ATH79_MACH_AP113,		/* Atheros AP113 reference board */
 	ATH79_MACH_AP121,		/* Atheros AP121 reference board */
 	ATH79_MACH_AP121_MINI,		/* Atheros AP121-MINI reference board */
