--- a/arch/mips/ath79/Kconfig
+++ b/arch/mips/ath79/Kconfig
@@ -741,7 +741,7 @@ config ATH79_MACH_WNR2000
 	select ATH79_DEV_WMAC
 
 config ATH79_MACH_WNR2000_V3
-	bool "NETGEAR WNR2000 V3/WNR612 v2/WNR1000 v2 board support"
+	bool "NETGEAR WNR2000 V3/WNR612 v2/WNR1000 v2/WPN824N board support"
 	select SOC_AR724X
 	select ATH79_DEV_AP9X_PCI if PCI
 	select ATH79_DEV_ETH
--- a/arch/mips/ath79/machtypes.h
+++ b/arch/mips/ath79/machtypes.h
@@ -204,6 +204,7 @@ enum ath79_mach_type {
 	ATH79_MACH_WNR2200,		/* NETGEAR WNR2200 */
 	ATH79_MACH_WNR612_V2,		/* NETGEAR WNR612 v2 */
 	ATH79_MACH_WNR1000_V2,		/* NETGEAR WNR1000 v2 */
+	ATH79_MACH_WPN824N,		/* NETGEAR WPN824N */
 	ATH79_MACH_WP543,		/* Compex WP543 */
 	ATH79_MACH_WPE72,		/* Compex WPE72 */
 	ATH79_MACH_WPJ344,		/* Compex WPJ344 */
