Index: linux-2.6.21.1/drivers/leds/Makefile
===================================================================
--- linux-2.6.21.1.orig/drivers/leds/Makefile    2007-07-10 20:56:30.000000000 +0200
+++ linux-2.6.21.1/drivers/leds/Makefile       2007-07-19 16:43:07.000000000 +0200
@@ -16,6 +16,7 @@
 obj-$(CONFIG_LEDS_WRAP)			+= leds-wrap.o
 obj-$(CONFIG_LEDS_H1940)		+= leds-h1940.o
 obj-$(CONFIG_LEDS_COBALT)		+= leds-cobalt.o
+obj-$(CONFIG_LEDS_RB500)                += leds-rb500.o
 
 # LED Triggers
 obj-$(CONFIG_LEDS_TRIGGER_TIMER)	+= ledtrig-timer.o
Index: linux-2.6.21.1/drivers/leds/Kconfig
===================================================================
--- linux-2.6.21.1.orig/drivers/leds/Kconfig     2007-07-10 20:56:30.000000000 +0200
+++ linux-2.6.21.1/drivers/leds/Kconfig 2007-07-19 16:42:55.000000000 +0200
@@ -95,6 +95,12 @@
 	help
 	  This option enables support for the front LED on Cobalt Server
 
+config LEDS_RB500
+	tristate "LED Support for RB5xx boards"
+	depends LEDS_CLASS && MIKROTIK_RB500
+	help
+	  This option enables support for the yellow user LED on RB5xx boards.
+
 comment "LED Triggers"
 
 config LEDS_TRIGGERS
