diff -urN linux-2.6.23.12/drivers/leds/Kconfig linux-2.6.23.12.new/drivers/leds/Kconfig
--- linux-2.6.23.12/drivers/leds/Kconfig	2007-12-18 22:55:57.000000000 +0100
+++ linux-2.6.23.12.new/drivers/leds/Kconfig	2008-01-11 09:01:19.000000000 +0100
@@ -101,6 +101,12 @@
 	  outputs. To be useful the particular board must have LEDs
 	  and they must be connected to the GPIO lines.
 
+config LEDS_RB500
+	tristate "LED Support for RB5xx boards"
+	depends LEDS_CLASS && MIKROTIK_RB500
+	help
+	  This option enables support for the yellow user LED on RB5xx boards.
+
 comment "LED Triggers"
 
 config LEDS_TRIGGERS
diff -urN linux-2.6.23.12/drivers/leds/Makefile linux-2.6.23.12.new/drivers/leds/Makefile
--- linux-2.6.23.12/drivers/leds/Makefile	2007-12-18 22:55:57.000000000 +0100
+++ linux-2.6.23.12.new/drivers/leds/Makefile	2008-01-11 09:00:49.000000000 +0100
@@ -17,6 +17,7 @@
 obj-$(CONFIG_LEDS_H1940)		+= leds-h1940.o
 obj-$(CONFIG_LEDS_COBALT)		+= leds-cobalt.o
 obj-$(CONFIG_LEDS_GPIO)			+= leds-gpio.o
+obj-$(CONFIG_LEDS_RB500)		+= leds-rb500.o
 
 # LED Triggers
 obj-$(CONFIG_LEDS_TRIGGER_TIMER)	+= ledtrig-timer.o
