Index: linux-2.6.23.17/arch/mips/kernel/time.c
===================================================================
--- linux-2.6.23.17.orig/arch/mips/kernel/time.c
+++ linux-2.6.23.17/arch/mips/kernel/time.c
@@ -479,3 +479,5 @@ EXPORT_SYMBOL(rtc_lock);
 EXPORT_SYMBOL(to_tm);
 EXPORT_SYMBOL(rtc_mips_set_time);
 EXPORT_SYMBOL(rtc_mips_get_time);
+EXPORT_SYMBOL(mips_hpt_frequency);
+
Index: linux-2.6.23.17/drivers/char/watchdog/Kconfig
===================================================================
--- linux-2.6.23.17.orig/drivers/char/watchdog/Kconfig
+++ linux-2.6.23.17/drivers/char/watchdog/Kconfig
@@ -647,6 +647,19 @@ config WATCHDOG_RTAS
 	  To compile this driver as a module, choose M here. The module
 	  will be called wdrtas.
 
+config RC32434_WDT
+	tristate "IDT Interprise 79RC32434 SoC hardware watchdog"
+	depends on WATCHDOG && MIKROTIK_RB500
+	help
+	  This is a driver for hardware watchdog integrated in IDT Interprise
+	  79RC32434 SoC. This watchdog simply watches your kernel to make sure
+	  it doesn't freeze, and if it does, it reboots your computer after a
+	  certain amount of time.
+
+	  To compile this driver as a module, choose M here: the module will be
+	  called rc32434_wdt.
+
+
 # S390 Architecture
 
 config ZVM_WATCHDOG
Index: linux-2.6.23.17/drivers/char/watchdog/Makefile
===================================================================
--- linux-2.6.23.17.orig/drivers/char/watchdog/Makefile
+++ linux-2.6.23.17/drivers/char/watchdog/Makefile
@@ -90,6 +90,7 @@ obj-$(CONFIG_SBC_EPX_C3_WATCHDOG) += sbc
 obj-$(CONFIG_INDYDOG) += indydog.o
 obj-$(CONFIG_WDT_MTX1)	+= mtx-1_wdt.o
 obj-$(CONFIG_WDT_RM9K_GPI) += rm9k_wdt.o
+obj-$(CONFIG_RC32434_WDT) += rc32434_wdt.o
 
 # PARISC Architecture
 
