Index: gcc-4.2.3/gcc/config/avr32/linux-elf.h
===================================================================
--- gcc-4.2.3.orig/gcc/config/avr32/linux-elf.h	2008-05-21 13:45:54.173288052 +0200
+++ gcc-4.2.3/gcc/config/avr32/linux-elf.h	2008-05-21 13:45:57.862288662 +0200
@@ -122,11 +122,6 @@
 	builtin_define ("__AVR32_HAS_BRANCH_PRED__");		\
       if (TARGET_FAST_FLOAT)                                    \
         builtin_define ("__AVR32_FAST_FLOAT__");                \
-      if (flag_pic)						\
-	{							\
-	  builtin_define ("__PIC__");				\
-	  builtin_define ("__pic__");				\
-	}							\
     }								\
   while (0)
 
