--- a/config/override.m4
+++ b/config/override.m4
@@ -32,31 +32,6 @@ dnl Redefine AC_CONFIG_SUBDIRS so acloca
 dnl when needed.
 
 ifdef([m4_PACKAGE_VERSION],
-[dnl AC_DEFUN a commonly used macro so this file is picked up.
-m4_copy([AC_PREREQ], [_AC_PREREQ])
-AC_DEFUN([AC_PREREQ], [frob])
-m4_copy_force([_AC_PREREQ], [AC_PREREQ])
-
-
-dnl Ensure exactly this Autoconf version is used
-m4_ifndef([_GCC_AUTOCONF_VERSION],
-  [m4_define([_GCC_AUTOCONF_VERSION], [2.64])])
-
-dnl Test for the exact version when AC_INIT is expanded.
-dnl This allows to update the tree in steps (for testing)
-dnl by putting
-dnl   m4_define([_GCC_AUTOCONF_VERSION], [X.Y])
-dnl in configure.ac before AC_INIT,
-dnl without rewriting this file.
-dnl Or for updating the whole tree at once with the definition above.
-AC_DEFUN([_GCC_AUTOCONF_VERSION_CHECK],
-[m4_if(m4_defn([_GCC_AUTOCONF_VERSION]),
-  m4_defn([m4_PACKAGE_VERSION]), [],
-  [m4_fatal([Please use exactly Autoconf ]_GCC_AUTOCONF_VERSION[ instead of ]m4_defn([m4_PACKAGE_VERSION])[.])])
-])
-m4_define([AC_INIT], m4_defn([AC_INIT])[
-_GCC_AUTOCONF_VERSION_CHECK
-])
 
 
 dnl Turn AC_DISABLE_OPTION_CHECKING into a no-op if not defined.
--- a/gas/aclocal.m4
+++ b/gas/aclocal.m4
@@ -998,7 +998,6 @@ m4_include([../config/nls.m4])
 m4_include([../config/override.m4])
 m4_include([../config/po.m4])
 m4_include([../config/progtest.m4])
-m4_include([../libtool.m4])
 m4_include([../ltoptions.m4])
 m4_include([../ltsugar.m4])
 m4_include([../ltversion.m4])
--- a/bfd/aclocal.m4
+++ b/bfd/aclocal.m4
@@ -978,7 +978,6 @@ m4_include([../config/plugins.m4])
 m4_include([../config/po.m4])
 m4_include([../config/progtest.m4])
 m4_include([../config/stdint.m4])
-m4_include([../libtool.m4])
 m4_include([../ltoptions.m4])
 m4_include([../ltsugar.m4])
 m4_include([../ltversion.m4])
--- a/binutils/aclocal.m4
+++ b/binutils/aclocal.m4
@@ -1003,7 +1003,6 @@ m4_include([../config/override.m4])
 m4_include([../config/plugins.m4])
 m4_include([../config/po.m4])
 m4_include([../config/progtest.m4])
-m4_include([../libtool.m4])
 m4_include([../ltoptions.m4])
 m4_include([../ltsugar.m4])
 m4_include([../ltversion.m4])
--- a/ld/aclocal.m4
+++ b/ld/aclocal.m4
@@ -999,7 +999,6 @@ m4_include([../config/nls.m4])
 m4_include([../config/override.m4])
 m4_include([../config/po.m4])
 m4_include([../config/progtest.m4])
-m4_include([../libtool.m4])
 m4_include([../ltoptions.m4])
 m4_include([../ltsugar.m4])
 m4_include([../ltversion.m4])
--- a/opcodes/aclocal.m4
+++ b/opcodes/aclocal.m4
@@ -978,7 +978,6 @@ m4_include([../config/nls.m4])
 m4_include([../config/override.m4])
 m4_include([../config/po.m4])
 m4_include([../config/progtest.m4])
-m4_include([../libtool.m4])
 m4_include([../ltoptions.m4])
 m4_include([../ltsugar.m4])
 m4_include([../ltversion.m4])
