diff -ruN ulogd-1.23-old/ulogd.conf.in.rej ulogd-1.23-new/ulogd.conf.in.rej
--- ulogd-1.23-old/ulogd.conf.in.rej	2005-04-18 16:19:06.000000000 +0200
+++ ulogd-1.23-new/ulogd.conf.in.rej	1970-01-01 01:00:00.000000000 +0100
@@ -1,16 +0,0 @@
-***************
-*** 62,68 ****
-  user="laforge"
-  db="ulogd"
-  host="localhost"
-- use_ssl=0
-  
-  [PGSQL]
-  table="ulog"
---- 62,67 ----
-  user="laforge"
-  db="ulogd"
-  host="localhost"
-  
-  [PGSQL]
-  table="ulog"
