--- a/drivers/mtd/maps/Kconfig
+++ b/drivers/mtd/maps/Kconfig
@@ -167,6 +167,12 @@ config MTD_RDC3210_BUSWIDTH
 	  Number of bytes addressed on the RDC-3210 flash device before
 	  addressing the same chip again
 
+config MTD_R8610
+	tristate "CFI flash device mapped on R8610"
+	depends on X86 && MTD_CFI && MTD_PARTITIONS
+	help
+	  Flash support for the RDC R8610 evaluation board.
+
 config MTD_SC520CDP
 	tristate "CFI Flash device mapped on AMD SC520 CDP"
 	depends on X86 && MTD_CFI && MTD_CONCAT
--- a/drivers/mtd/maps/Makefile
+++ b/drivers/mtd/maps/Makefile
@@ -28,6 +28,7 @@ obj-$(CONFIG_MTD_PMC_MSP_EVM)   += pmcms
 obj-$(CONFIG_MTD_PMC_MSP_RAMROOT)+= pmcmsp-ramroot.o
 obj-$(CONFIG_MTD_PCMCIA)	+= pcmciamtd.o
 obj-$(CONFIG_MTD_RDC3210)	+= rdc3210.o
+obj-$(CONFIG_MTD_R8610)		+= r8610.o
 obj-$(CONFIG_MTD_RPXLITE)	+= rpxlite.o
 obj-$(CONFIG_MTD_TQM8XXL)	+= tqm8xxl.o
 obj-$(CONFIG_MTD_SA1100)	+= sa1100-flash.o
