Index: linux-2.6.30.10/scripts/Makefile.lib
===================================================================
--- linux-2.6.30.10.orig/scripts/Makefile.lib	2010-05-15 16:46:57.000000000 +0200
+++ linux-2.6.30.10/scripts/Makefile.lib	2010-05-15 18:53:02.000000000 +0200
@@ -204,4 +204,4 @@
 # ---------------------------------------------------------------------------
 
 quiet_cmd_lzma = LZMA    $@
-cmd_lzma = lzma e $< $@ -lc1 -lp2 -pb2 -eos
+cmd_lzma = (lzma e $< -lc8 -eos -so && $(size_append) $<) > $@ || (rm -f $@ ; false)
Index: linux-2.6.30.10/arch/x86/include/asm/boot.h
===================================================================
--- linux-2.6.30.10.orig/arch/x86/include/asm/boot.h	2010-05-15 16:45:57.000000000 +0200
+++ linux-2.6.30.10/arch/x86/include/asm/boot.h	2010-05-15 16:46:58.000000000 +0200
@@ -13,7 +13,7 @@
 				+ (CONFIG_PHYSICAL_ALIGN - 1)) \
 				& ~(CONFIG_PHYSICAL_ALIGN - 1))
 
-#ifdef CONFIG_KERNEL_BZIP2
+#if defined(CONFIG_KERNEL_BZIP2) || defined(CONFIG_KERNEL_LZMA)
 #define BOOT_HEAP_SIZE             0x400000
 #else /* !CONFIG_KERNEL_BZIP2 */
 
