diff -ur busybox.orig/networking/Config.in busybox/networking/Config.in
--- busybox.orig/networking/Config.in	2004-09-23 22:08:46.000000000 +0200
+++ busybox/networking/Config.in	2004-10-31 20:09:54.622922408 +0100
@@ -492,7 +492,7 @@
 config CONFIG_TELNETD
 	bool "telnetd"
 	default n
-	select CONFIG_LOGIN
+	#select CONFIG_LOGIN
 	help
 	  A daemon for the TELNET protocol, allowing you to log onto the host
 	  running the daemon.  Please keep in mind that the TELNET protocol
