--- a/hostapd/hw_features.c
+++ b/hostapd/hw_features.c
@@ -329,7 +329,7 @@ static int ieee80211n_check_40mhz_5g(str
 	if (!match) {
 		for (i = 0; i < scan_res->num; i++) {
 			struct wpa_scan_res *bss = scan_res->res[i];
-			ieee80211n_get_pri_sec_chan(bss, &pri_chan, &sec_chan);
+			ieee80211n_get_pri_sec_chan(bss, &bss_pri_chan, &bss_sec_chan);
 			if (pri_chan == bss_sec_chan &&
 			    sec_chan == bss_pri_chan) {
 				wpa_printf(MSG_INFO, "Switch own primary and "
