Index: madwifi-ng-r2568-20070710/hal/public/mips-be-elf.inc
===================================================================
--- madwifi-ng-r2568-20070710.orig/hal/public/mips-be-elf.inc	2007-10-20 20:51:37.000000000 +0200
+++ madwifi-ng-r2568-20070710/hal/public/mips-be-elf.inc	2007-10-20 20:52:09.000000000 +0200
@@ -71,5 +71,5 @@
 
 LDOPTS=	-EB
 COPTS+=	-DAH_BYTE_ORDER=AH_BIG_ENDIAN
-COPTS+=	-G 0 -EB -mno-abicalls -fno-pic -mips2 -Wa,--trap \
+COPTS+=	-G 0 -mno-abicalls -fno-pic -mips32 -Wa,--trap \
 	-fno-strict-aliasing -fno-common -fomit-frame-pointer -mlong-calls
Index: madwifi-ng-r2568-20070710/hal/public/mips-le-elf.inc
===================================================================
--- madwifi-ng-r2568-20070710.orig/hal/public/mips-le-elf.inc	2007-10-20 20:51:37.000000000 +0200
+++ madwifi-ng-r2568-20070710/hal/public/mips-le-elf.inc	2007-10-20 20:52:09.000000000 +0200
@@ -68,5 +68,5 @@
 
 LDOPTS=	-EL
 COPTS+=	-DAH_BYTE_ORDER=AH_LITTLE_ENDIAN
-COPTS+=	-G 0 -EL -mno-abicalls -fno-pic -mips2 -Wa,--trap \
+COPTS+=	-G 0 -mno-abicalls -fno-pic -mips32 -Wa,--trap \
 	-fno-strict-aliasing -fno-common -fomit-frame-pointer -mlong-calls
Index: madwifi-ng-r2568-20070710/hal/public/xscale-be-elf.inc
===================================================================
--- madwifi-ng-r2568-20070710.orig/hal/public/xscale-be-elf.inc	2007-10-20 20:51:37.000000000 +0200
+++ madwifi-ng-r2568-20070710/hal/public/xscale-be-elf.inc	2007-10-20 20:52:09.000000000 +0200
@@ -77,5 +77,5 @@
 
 LDOPTS=	-EB
 COPTS+=	-DAH_BYTE_ORDER=AH_BIG_ENDIAN
-COPTS+= -march=armv4 -mbig-endian -fno-strict-aliasing -fno-common -mapcs-32 \
+COPTS+= -march=armv5te -mbig-endian -fno-strict-aliasing -fno-common  \
 	-mtune=xscale  -mshort-load-bytes
