diff -urN madwifi-ng-refcount-r2313-20070505.old/ath/if_ath.c madwifi-ng-refcount-r2313-20070505.dev/ath/if_ath.c
--- madwifi-ng-refcount-r2313-20070505.old/ath/if_ath.c	2007-05-13 18:17:55.278165480 +0200
+++ madwifi-ng-refcount-r2313-20070505.dev/ath/if_ath.c	2007-05-13 18:17:55.287164112 +0200
@@ -401,7 +401,7 @@
 	HAL_STATUS status;
 	int error = 0;
 	unsigned int i;
-	int autocreatemode = IEEE80211_M_STA;
+	int autocreatemode = -1;
 	u_int8_t csz;
 
 	sc->devid = devid;
