Index: madwifi-ng-r2834-20071106/ath/if_ath.c
===================================================================
--- madwifi-ng-r2834-20071106.orig/ath/if_ath.c	2007-11-07 14:01:59.713354541 +0100
+++ madwifi-ng-r2834-20071106/ath/if_ath.c	2007-11-07 14:02:00.533401273 +0100
@@ -426,7 +426,7 @@
 	HAL_STATUS status;
 	int error = 0;
 	unsigned int i;
-	int autocreatemode = IEEE80211_M_STA;
+	int autocreatemode = -1;
 	u_int8_t csz;
 
 	sc->devid = devid;
