Index: madwifi-ng-r2834-20071106/net80211/ieee80211_rate.c
===================================================================
--- madwifi-ng-r2834-20071106.orig/net80211/ieee80211_rate.c	2007-11-07 14:01:59.685352943 +0100
+++ madwifi-ng-r2834-20071106/net80211/ieee80211_rate.c	2007-11-07 14:02:00.745413352 +0100
@@ -100,8 +100,18 @@
 		ieee80211_load_module(buf);
 
 	if (!ratectls[id].attach) {
-		printk(KERN_ERR "Error loading module \"%s\"\n", buf);
-		return NULL;
+		/* pick the first available rate control module */
+		printk(KERN_INFO "Rate control module \"%s\" not available\n", buf);
+		for (id = 0; id < IEEE80211_RATE_MAX; id++) {
+			if (ratectls[id].attach)
+				break;
+		}
+		if (!ratectls[id].attach) {
+			printk(KERN_ERR "No rate control module available");
+			return NULL;
+		} else {
+			printk(KERN_INFO "Using \"%s\" instead.\n", module_names[id]);
+		}
 	}
 
 	ctl = ratectls[id].attach(sc);
