Index: trunk/ath/if_ath.c
===================================================================
--- trunk/ath/if_ath.c (revision 1751)
+++ trunk/ath/if_ath.c (revision 1752)
@@ -1600,5 +1600,9 @@
  */
 irqreturn_t
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,19)
+ath_intr(int irq, void *dev_id)
+#else
 ath_intr(int irq, void *dev_id, struct pt_regs *regs)
+#endif
 {
 	struct net_device *dev = dev_id;
Index: trunk/ath/if_athvar.h
===================================================================
--- trunk/ath/if_athvar.h (revision 1726)
+++ trunk/ath/if_athvar.h (revision 1752)
@@ -727,5 +727,9 @@
 void ath_suspend(struct net_device *);
 void ath_shutdown(struct net_device *);
-irqreturn_t ath_intr(int, void *, struct pt_regs *);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,19)
+irqreturn_t ath_intr(int, void *);
+#else
+irqreturn_t ath_intr(int, void *, struct pt_regs *regs);
+#endif
 int ath_ioctl_ethtool(struct ath_softc *, int, void __user *);
 void bus_read_cachesize(struct ath_softc *, u_int8_t *);
