Index: madwifi-ng-r2799-20071030/net80211/ieee80211_scan.c
===================================================================
--- madwifi-ng-r2799-20071030.orig/net80211/ieee80211_scan.c	2007-10-31 14:04:46.717283745 +0100
+++ madwifi-ng-r2799-20071030/net80211/ieee80211_scan.c	2007-10-31 14:04:48.993413455 +0100
@@ -291,7 +291,8 @@
 	struct ieee80211com *ic = vap->iv_ic;
 	int delay;
 
-	ieee80211_sta_pwrsave(vap, 1);
+	if (vap->iv_opmode != IEEE80211_M_IBSS)
+	    ieee80211_sta_pwrsave(vap, 1);
 	/*
 	 * Use an initial 1ms delay to ensure the null
 	 * data frame has a chance to go out.
