diff -ur madwifi.old/net80211/ieee80211_input.c madwifi.dev/net80211/ieee80211_input.c
--- madwifi.old/net80211/ieee80211_input.c	2007-06-01 16:44:25.174897520 +0200
+++ madwifi.dev/net80211/ieee80211_input.c	2007-06-01 17:43:52.153634328 +0200
@@ -685,6 +685,7 @@
  				IEEE80211_NODE_STAT(ni, rx_decap);
  				goto err;
  			}
+			skb1 = skb_clone(skb, GFP_ATOMIC);
 
 			/* get to the tunneled headers */
 			ath_hdr = (struct athl2p_tunnel_hdr *)
