Index: madwifi-ng-r2420-20070602/ath/if_ath.c
===================================================================
--- madwifi-ng-r2420-20070602.orig/ath/if_ath.c	2007-06-04 13:21:55.808492488 +0200
+++ madwifi-ng-r2420-20070602/ath/if_ath.c	2007-06-04 13:21:56.713354928 +0200
@@ -271,7 +271,7 @@
 static char *ratectl = DEF_RATE_CTL;
 static int rfkill = 0;
 #ifdef ATH_CAP_TPC
-static int tpc = -1;
+static int tpc = 0;
 #endif
 static int countrycode = -1;
 static int outdoor = -1;
