Index: madwifi-ng-r2834-20071106/ath/if_athvar.h
===================================================================
--- madwifi-ng-r2834-20071106.orig/ath/if_athvar.h	2007-11-07 14:02:02.861533944 +0100
+++ madwifi-ng-r2834-20071106/ath/if_athvar.h	2007-11-07 14:02:03.557573608 +0100
@@ -128,6 +128,11 @@
 #define	NETDEV_TX_BUSY	1
 #endif
 
+#ifndef NETDEV_TX_OK
+#define NETDEV_TX_OK    0
+#define NETDEV_TX_BUSY  1
+#endif
+
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2,4,23)
 static inline struct net_device *_alloc_netdev(int sizeof_priv, const char *mask,
 					       void (*setup)(struct net_device *))
