Index: madwifi-dfs-r3053/ath/if_ath.c
===================================================================
--- madwifi-dfs-r3053.orig/ath/if_ath.c	2007-12-13 05:25:12.854357994 +0100
+++ madwifi-dfs-r3053/ath/if_ath.c	2007-12-13 05:25:13.122373270 +0100
@@ -6410,8 +6410,9 @@
 			/*
 			 * Reject error frames if we have no vaps that
 			 * are operating in monitor mode.
+			 * Reject empty frames as well
 			 */
-			if (sc->sc_nmonvaps == 0)
+			if ((sc->sc_nmonvaps == 0) || (rs->rs_datalen == 0))
 				goto rx_next;
 		}
 rx_accept:
