Index: madwifi-ng-r2756-20071018/ath/if_ath.c
===================================================================
--- madwifi-ng-r2756-20071018.orig/ath/if_ath.c	2007-10-24 06:20:19.097777599 +0200
+++ madwifi-ng-r2756-20071018/ath/if_ath.c	2007-10-24 06:20:20.001829115 +0200
@@ -5834,8 +5834,9 @@
 			/*
 			 * Reject error frames if we have no vaps that 
 			 * are operating in monitor mode.
+			 * Reject empty frames as well
 			 */
-			if (sc->sc_nmonvaps == 0)
+			if ((sc->sc_nmonvaps == 0) || (rs->rs_datalen == 0))
 				goto rx_next;
 		}
 rx_accept:
