Index: madwifi-ng-r2756-20071018/ath/if_ath.c
===================================================================
--- madwifi-ng-r2756-20071018.orig/ath/if_ath.c	2007-10-24 06:20:20.001829115 +0200
+++ madwifi-ng-r2756-20071018/ath/if_ath.c	2007-10-24 06:20:20.213841201 +0200
@@ -4162,6 +4162,8 @@
 	}
 	bf = avp->av_bcbuf;
 	ni = bf->bf_node;
+	if (!ni)
+		return NULL;
 
 #ifdef ATH_SUPERG_DYNTURBO
 	/* 
