Index: madwifi-ng-r2799-20071030/ath/if_ath.c
===================================================================
--- madwifi-ng-r2799-20071030.orig/ath/if_ath.c	2007-10-31 14:04:50.329489593 +0100
+++ madwifi-ng-r2799-20071030/ath/if_ath.c	2007-10-31 14:04:50.577503726 +0100
@@ -4162,6 +4162,8 @@
 	}
 	bf = avp->av_bcbuf;
 	ni = bf->bf_node;
+	if (!ni)
+		return NULL;
 
 #ifdef ATH_SUPERG_DYNTURBO
 	/* 
