diff -ur madwifi.old/ath/if_ath.c madwifi.dev/ath/if_ath.c
--- madwifi.old/ath/if_ath.c	2007-02-20 22:50:52.277096688 +0100
+++ madwifi.dev/ath/if_ath.c	2007-02-20 22:53:50.476006336 +0100
@@ -72,7 +72,7 @@
 #include <net80211/if_llc.h>
 #endif
 
-#define	AR_DEBUG
+#undef	AR_DEBUG
 
 #include "if_athrate.h"
 #include "net80211/if_athproto.h"
@@ -347,7 +347,7 @@
 		ath_keyprint(__func__, ix, hk, mac);		\
 } while (0)
 #else /* defined(AR_DEBUG) */
-#define	IFF_DUMPPKTS(sc, _m)	netif_msg_dumppkts(&sc->sc_ic)
+#define	IFF_DUMPPKTS(sc, _m)
 #define	DPRINTF(sc, _m, _fmt, ...)
 #define	KEYPRINTF(sc, k, ix, mac)
 #endif /* defined(AR_DEBUG) */
diff -ur madwifi.old/ath_rate/amrr/amrr.c madwifi.dev/ath_rate/amrr/amrr.c
--- madwifi.old/ath_rate/amrr/amrr.c	2007-02-20 22:50:52.278096536 +0100
+++ madwifi.dev/ath_rate/amrr/amrr.c	2007-02-20 22:50:31.652232144 +0100
@@ -67,7 +67,7 @@
 
 #include "amrr.h"
 
-#define	AMRR_DEBUG
+#undef	AMRR_DEBUG
 #ifdef AMRR_DEBUG
 #define	DPRINTF(sc, _fmt, ...) do {					\
 	if (sc->sc_debug & 0x10)					\
diff -ur madwifi.old/ath_rate/onoe/onoe.c madwifi.dev/ath_rate/onoe/onoe.c
--- madwifi.old/ath_rate/onoe/onoe.c	2007-02-20 22:50:52.278096536 +0100
+++ madwifi.dev/ath_rate/onoe/onoe.c	2007-02-20 22:50:31.653231992 +0100
@@ -63,7 +63,7 @@
 
 #include "onoe.h"
 
-#define	ONOE_DEBUG
+#undef	ONOE_DEBUG
 #ifdef ONOE_DEBUG
 enum {
 	ATH_DEBUG_RATE	= 0x00000010,	/* rate control */
diff -ur madwifi.old/ath_rate/sample/sample.c madwifi.dev/ath_rate/sample/sample.c
--- madwifi.old/ath_rate/sample/sample.c	2007-02-20 22:50:52.278096536 +0100
+++ madwifi.dev/ath_rate/sample/sample.c	2007-02-20 22:50:31.653231992 +0100
@@ -65,7 +65,7 @@
 
 #include "sample.h"
 
-#define	SAMPLE_DEBUG
+#undef	SAMPLE_DEBUG
 #ifdef SAMPLE_DEBUG
 enum {
 	ATH_DEBUG_RATE		= 0x00000010	/* rate control */
diff -ur madwifi.old/net80211/ieee80211_var.h madwifi.dev/net80211/ieee80211_var.h
--- madwifi.old/net80211/ieee80211_var.h	2007-02-20 22:50:52.279096384 +0100
+++ madwifi.dev/net80211/ieee80211_var.h	2007-02-20 22:50:31.654231840 +0100
@@ -37,7 +37,7 @@
 /*
  * Definitions for IEEE 802.11 drivers.
  */
-#define	IEEE80211_DEBUG
+#undef	IEEE80211_DEBUG
 #undef	IEEE80211_DEBUG_REFCNT			/* node refcnt stuff */
 
 #include <net80211/ieee80211_linux.h>
diff -ur madwifi.old/tools/do_multi.c madwifi.dev/tools/do_multi.c
--- madwifi.old/tools/do_multi.c	2007-02-20 22:50:52.279096384 +0100
+++ madwifi.dev/tools/do_multi.c	2007-02-20 22:50:31.654231840 +0100
@@ -9,16 +9,20 @@
 
     progname = basename(argv[0]);
 
+#if 0
     if(strcmp(progname, "80211debug") == 0)
 	ret = a80211debug_init(argc, argv);
+#endif
     if(strcmp(progname, "80211stats") == 0)
 	ret = a80211stats_init(argc, argv);
     if(strcmp(progname, "athchans") == 0)
 	ret = athchans_init(argc, argv);
     if(strcmp(progname, "athctrl") == 0)
 	ret =  athctrl_init(argc, argv);
+#if 0
     if(strcmp(progname, "athdebug") == 0)
 	ret =  athdebug_init(argc, argv);
+#endif
     if(strcmp(progname, "athkey") == 0)
 	ret =  athkey_init(argc, argv);
     if(strcmp(progname, "athstats") == 0)
diff -ur madwifi.old/tools/Makefile madwifi.dev/tools/Makefile
--- madwifi.old/tools/Makefile	2007-02-20 22:50:52.279096384 +0100
+++ madwifi.dev/tools/Makefile	2007-02-20 22:50:31.654231840 +0100
@@ -52,7 +52,7 @@
 	
 ifdef DOMULTI
 OBJS=	do_multi.o athstats.o 80211stats.o athkey.o athchans.o athctrl.o \
-	athdebug.o 80211debug.o wlanconfig.o 
+	wlanconfig.o 
 ALL=	${OBJS} madwifi_multi
 endif
 
@@ -85,7 +85,7 @@
 	${CC} -c -o do_multi.o -DDOMULTI=1 ${ALL_CFLAGS} do_multi.c
 madwifi_multi:
 	${CC} -o madwifi_multi ${LDFLAGS} ${OBJS}
-	for i in athstats 80211stats athkey athchans athctrl athdebug wlanconfig 80211debug; do \
+	for i in athstats 80211stats athkey athchans athctrl wlanconfig; do \
 	ln -s -f madwifi_multi $$i; \
 	done
 
