Index: madwifi-dfs-r3246/ath/if_ath.c
===================================================================
--- madwifi-dfs-r3246.orig/ath/if_ath.c	2008-01-25 18:17:16.252409313 +0100
+++ madwifi-dfs-r3246/ath/if_ath.c	2008-01-25 18:24:41.981809984 +0100
@@ -42,7 +42,6 @@
  * This software is derived from work of Atsushi Onoe; his contribution
  * is greatly appreciated.
  */
-#define	AR_DEBUG
 #include "if_ath_debug.h"
 #include "opt_ah.h"
 
@@ -522,9 +521,11 @@
 	u_int8_t csz;
 
 	sc->devid = devid;
+#ifdef AR_DEBUG
 	ath_debug_global = (ath_debug & ATH_DEBUG_GLOBAL);
 	sc->sc_debug 	 = (ath_debug & ~ATH_DEBUG_GLOBAL);
 	DPRINTF(sc, ATH_DEBUG_ANY, "%s: devid 0x%x\n", __func__, devid);
+#endif
 
 	/* Allocate space for dynamically determined maximum VAP count */
 	sc->sc_bslot = 
@@ -1301,12 +1302,14 @@
 		/* If no default VAP debug flags are passed, allow a few to
 		 * transfer down from the driver to new VAPs so we can have load
 		 * time debugging for VAPs too. */
+#ifdef AR_DEBUG
 		vap->iv_debug = 0 |
 			((sc->sc_debug & ATH_DEBUG_RATE) ? IEEE80211_MSG_XRATE  : 0) | 
 			((sc->sc_debug & ATH_DEBUG_XMIT) ? IEEE80211_MSG_OUTPUT : 0) | 
 			((sc->sc_debug & ATH_DEBUG_RECV) ? IEEE80211_MSG_INPUT  : 0) |
 			0
 			;
+#endif
 	}
 	ic->ic_debug = (sc->sc_default_ieee80211_debug & IEEE80211_MSG_IC);
 
@@ -10524,9 +10527,11 @@
 				/* XXX validate? */
 				sc->sc_ledpin = val;
 				break;
+#ifdef AR_DEBUG
 			case ATH_DEBUG:
 				sc->sc_debug 	 = (val & ~ATH_DEBUG_GLOBAL);
 				ath_debug_global = (val &  ATH_DEBUG_GLOBAL);
+#endif
 				break;
 			case ATH_TXANTENNA:
 				/*
@@ -10946,9 +10951,11 @@
 	}
 
 	/* initialize values */
+#ifdef AR_DEBUG
 	ath_debug_global = (ath_debug & ATH_DEBUG_GLOBAL);
 	sc->sc_debug 	 = (ath_debug & ~ATH_DEBUG_GLOBAL);
 	sc->sc_default_ieee80211_debug = ieee80211_debug;
+#endif
 	sc->sc_txantenna = 0;		/* default to auto-selection */
 	sc->sc_txintrperiod = ATH_TXQ_INTR_PERIOD;
 }
Index: madwifi-dfs-r3246/ath_rate/amrr/amrr.c
===================================================================
--- madwifi-dfs-r3246.orig/ath_rate/amrr/amrr.c	2008-01-25 18:17:16.260409771 +0100
+++ madwifi-dfs-r3246/ath_rate/amrr/amrr.c	2008-01-25 18:17:18.304526256 +0100
@@ -70,7 +70,9 @@
 
 #include "amrr.h"
 
+#ifdef AR_DEBUG
 #define	AMRR_DEBUG
+#endif
 #ifdef AMRR_DEBUG
 #define	DPRINTF(sc, _fmt, ...) do {					\
 	if (sc->sc_debug & 0x10)					\
Index: madwifi-dfs-r3246/ath_rate/minstrel/minstrel.c
===================================================================
--- madwifi-dfs-r3246.orig/ath_rate/minstrel/minstrel.c	2008-01-25 18:17:16.268410227 +0100
+++ madwifi-dfs-r3246/ath_rate/minstrel/minstrel.c	2008-01-25 18:17:18.316526944 +0100
@@ -117,7 +117,9 @@
 
 #include "minstrel.h"
 
+#ifdef AR_DEBUG
 #define	MINSTREL_DEBUG
+#endif
 #ifdef MINSTREL_DEBUG
 enum {
 		ATH_DEBUG_RATE		= 0x00000010	/* rate control */
Index: madwifi-dfs-r3246/ath_rate/onoe/onoe.c
===================================================================
--- madwifi-dfs-r3246.orig/ath_rate/onoe/onoe.c	2008-01-25 18:17:16.276410682 +0100
+++ madwifi-dfs-r3246/ath_rate/onoe/onoe.c	2008-01-25 18:17:18.320527171 +0100
@@ -66,7 +66,9 @@
 
 #include "onoe.h"
 
+#ifdef AR_DEBUG
 #define	ONOE_DEBUG
+#endif
 #ifdef ONOE_DEBUG
 enum {
 	ATH_DEBUG_RATE	= 0x00000010,	/* rate control */
Index: madwifi-dfs-r3246/ath_rate/sample/sample.c
===================================================================
--- madwifi-dfs-r3246.orig/ath_rate/sample/sample.c	2008-01-25 18:17:16.284411140 +0100
+++ madwifi-dfs-r3246/ath_rate/sample/sample.c	2008-01-25 18:17:18.324527396 +0100
@@ -68,7 +68,9 @@
 
 #include "sample.h"
 
-#define	SAMPLE_DEBUG
+#ifdef AR_DEBUG
+#define SAMPLE_DEBUG
+#endif
 #ifdef SAMPLE_DEBUG
 enum {
 	ATH_DEBUG_RATE		= 0x00000010,	/* rate control */
Index: madwifi-dfs-r3246/tools/do_multi.c
===================================================================
--- madwifi-dfs-r3246.orig/tools/do_multi.c	2008-01-25 18:17:16.292411592 +0100
+++ madwifi-dfs-r3246/tools/do_multi.c	2008-01-25 18:24:59.262794773 +0100
@@ -9,16 +9,20 @@
 
     progname = basename(argv[0]);
 
+#ifdef AR_DEBUG
     if(strcmp(progname, "80211debug") == 0)
 	ret = a80211debug_init(argc, argv);
+#endif
     if(strcmp(progname, "80211stats") == 0)
 	ret = a80211stats_init(argc, argv);
     if(strcmp(progname, "athchans") == 0)
 	ret = athchans_init(argc, argv);
     if(strcmp(progname, "athctrl") == 0)
 	ret =  athctrl_init(argc, argv);
+#ifdef AR_DEBUG
     if(strcmp(progname, "athdebug") == 0)
 	ret =  athdebug_init(argc, argv);
+#endif
     if(strcmp(progname, "athkey") == 0)
 	ret =  athkey_init(argc, argv);
     if(strcmp(progname, "athstats") == 0)
Index: madwifi-dfs-r3246/tools/Makefile
===================================================================
--- madwifi-dfs-r3246.orig/tools/Makefile	2008-01-25 18:17:16.300412050 +0100
+++ madwifi-dfs-r3246/tools/Makefile	2008-01-25 18:17:29.737177767 +0100
@@ -46,9 +46,10 @@
 HAL=   $(TOP)/hal
 endif
 
+DEBUG=-DAR_DEBUG
 
 ALLPROGS=	athstats 80211stats athkey athchans athctrl \
-	athdebug 80211debug wlanconfig ath_info
+	$(if $(DEBUG),80211debug athdebug) wlanconfig ath_info
 
 OBJS=	$(patsubst %,%.o,$(ALLPROGS) do_multi)
 ALL=	${OBJS} madwifi_multi
@@ -57,7 +58,7 @@
 
 INCS=	-I. -I$(HAL) -I$(TOP) -I$(ATH_HAL)
 CFLAGS=	-g -O2 -Wall
-ALL_CFLAGS= $(CFLAGS) $(INCS)
+ALL_CFLAGS= $(CFLAGS) $(INCS) $(DEBUG)
 LDFLAGS=
 
 all:	$(ALL)
@@ -84,7 +85,7 @@
 	${CC} -c -o do_multi.o -DDOMULTI=1 ${ALL_CFLAGS} do_multi.c
 madwifi_multi: $(OBJS)
 	${CC} -o madwifi_multi ${LDFLAGS} ${OBJS}
-	for i in athstats 80211stats athkey athchans athctrl athdebug wlanconfig 80211debug; do \
+	for i in $(ALLPROGS); do \
 	ln -s -f madwifi_multi $$i; \
 	done
 
Index: madwifi-dfs-r3246/net80211/ieee80211_linux.h
===================================================================
--- madwifi-dfs-r3246.orig/net80211/ieee80211_linux.h	2008-01-25 18:17:16.308412507 +0100
+++ madwifi-dfs-r3246/net80211/ieee80211_linux.h	2008-01-25 18:17:18.324527396 +0100
@@ -29,8 +29,6 @@
 #ifndef _NET80211_IEEE80211_LINUX_H_
 #define _NET80211_IEEE80211_LINUX_H_
 
-#define	IEEE80211_DEBUG
-#define	IEEE80211_DEBUG_REFCNT			/* Node reference count debugging */
 /* #define ATH_DEBUG_SPINLOCKS */		/* announce before spinlocking */
 
 #include <linux/wireless.h>
Index: madwifi-dfs-r3246/Makefile.inc
===================================================================
--- madwifi-dfs-r3246.orig/Makefile.inc	2008-01-25 18:17:16.316412961 +0100
+++ madwifi-dfs-r3246/Makefile.inc	2008-01-25 18:17:18.328527625 +0100
@@ -148,7 +148,8 @@
 TOOLS=  $(TOP)/tools 
 
 WARNINGS = -Werror
-COPTS+= $(WARNINGS)
+DEBUG = -DAR_DEBUG -DIEEE80211_DEBUG
+COPTS+= $(WARNINGS) $(DEBUG)
 INCS=	-include $(TOP)/include/compat.h -I$(TOP)/include
 
 # TARGET defines the target platform architecture. It must match one of
