Index: madwifi-trunk-r3314/ath/if_ath.c
===================================================================
--- madwifi-trunk-r3314.orig/ath/if_ath.c	2008-02-20 21:57:55.057877957 +0100
+++ madwifi-trunk-r3314/ath/if_ath.c	2008-02-20 21:57:58.494073773 +0100
@@ -1351,7 +1351,7 @@
 		TAILQ_FOREACH(v, &ic->ic_vaps, iv_next)
 			id_mask |= (1 << ATH_GET_VAP_ID(v->iv_myaddr));
 
-		for (id = 1; id < ath_maxvaps; id++) {
+		for (id = 0; id < ath_maxvaps; id++) {
 			/* get the first available slot */
 			if ((id_mask & (1 << id)) == 0) {
 				ATH_SET_VAP_BSSID(vap->iv_myaddr, id);
