Index: madwifi-trunk-r3280/net80211/ieee80211_input.c
===================================================================
--- madwifi-trunk-r3280.orig/net80211/ieee80211_input.c	2008-01-28 17:36:32.052771162 +0100
+++ madwifi-trunk-r3280/net80211/ieee80211_input.c	2008-01-28 17:36:55.186089457 +0100
@@ -1135,7 +1135,7 @@
 		if (ETHER_IS_MULTICAST(eh->ether_dhost)) {
 			/* Create a SKB for the BSS to send out. */
 			skb1 = skb_copy(skb, GFP_ATOMIC);
-			ieee80211_skb_copy_noderef(skb, skb1);
+			SKB_CB(skb1)->ni = ieee80211_ref_node(vap->iv_bss);
 		}
 		else {
 			/*
