Index: madwifi-dfs-r3280/ath/if_ath.c
===================================================================
--- madwifi-dfs-r3280.orig/ath/if_ath.c	2008-01-28 17:18:52.497966105 +0100
+++ madwifi-dfs-r3280/ath/if_ath.c	2008-01-28 17:19:06.982791546 +0100
@@ -526,7 +526,8 @@
 
 	/* Allocate space for dynamically determined maximum VAP count */
 	sc->sc_bslot = 
-		kzalloc(ath_maxvaps * sizeof(struct ieee80211vap), GFP_KERNEL);
+		kmalloc(ath_maxvaps * sizeof(struct ieee80211vap), GFP_KERNEL);
+	memset(sc->sc_bslot, 0, ath_maxvaps * sizeof(struct ieee80211vap));
 
 	/*
 	 * Cache line size is used to size and align various
Index: madwifi-dfs-r3280/ath/if_ath_radar.c
===================================================================
--- madwifi-dfs-r3280.orig/ath/if_ath_radar.c	2008-01-28 17:14:16.411947654 +0100
+++ madwifi-dfs-r3280/ath/if_ath_radar.c	2008-01-28 17:19:07.290809101 +0100
@@ -1539,9 +1539,11 @@
 
 	ath_rp_clear(sc);
 
-	sc->sc_rp = (struct ath_rp *)kzalloc(
-			sizeof(struct ath_rp) * 
+	sc->sc_rp = (struct ath_rp *)kmalloc(
+			sizeof(struct ath_rp) *
 			ATH_RADAR_PULSE_NR, GFP_KERNEL);
+	memset(sc->sc_rp, 0, sizeof(struct ath_rp) *
+			ATH_RADAR_PULSE_NR);
 
 	if (sc->sc_rp == NULL)
 		return;
