Index: madwifi-dfs-r3280/ath/if_ath.c
===================================================================
--- madwifi-dfs-r3280.orig/ath/if_ath.c	2008-01-28 17:19:06.982791546 +0100
+++ madwifi-dfs-r3280/ath/if_ath.c	2008-01-28 17:19:28.448014778 +0100
@@ -3269,10 +3269,6 @@
 		/* If the clone works, bump the reference count for our copy. */
 		SKB_CB(skb)->ni = ieee80211_ref_node(SKB_CB(skb_orig)->ni);
 		ieee80211_dev_kfree_skb(&skb_orig);
-	} else {
-		if (SKB_CB(skb)->ni != NULL) 
-			ieee80211_unref_node(&SKB_CB(skb)->ni);
-		skb_orphan(skb);
 	}
 
 	eh = (struct ether_header *)skb->data;
