Index: madwifi-trunk-r3314/ath/if_ath_ahb.c
===================================================================
--- madwifi-trunk-r3314.orig/ath/if_ath_ahb.c
+++ madwifi-trunk-r3314/ath/if_ath_ahb.c
@@ -245,6 +245,8 @@
 	num_activesc++;
 	/* Ready to process interrupts */
 
+	sc->aps_sc.sc_softled = 1; /* SoftLED over GPIO */
+	sc->aps_sc.sc_ledpin = config->board->sysLedGpio;
 	sc->aps_sc.sc_invalid = 0;
 	return 0;
 
