--- a/net80211/ieee80211_node.c
+++ b/net80211/ieee80211_node.c
@@ -1592,10 +1592,6 @@ found:
 				ieee80211_ref_node(ni);
 #endif
 			}
-		} else {
-			IEEE80211_NOTE_MAC(vap, IEEE80211_MSG_OUTPUT, mac,
-				"no node, discard frame (%s)", __func__);
-			vap->iv_stats.is_tx_nonode++;
 		}
 	}
 	return ni;
--- a/net80211/ieee80211_output.c
+++ b/net80211/ieee80211_output.c
@@ -208,6 +208,9 @@ ieee80211_hardstart(struct sk_buff *skb,
 	ni = ieee80211_find_txnode(vap, eh->ether_dhost);
 	if (ni == NULL) {
 		/* NB: ieee80211_find_txnode does stat+msg */
+		IEEE80211_NOTE_MAC(vap, IEEE80211_MSG_OUTPUT, mac,
+			"no node, discard frame (%s)", __func__);
+		vap->iv_stats.is_tx_nonode++;
 		goto bad;
 	}
 
