--- a/ath/if_athvar.h
+++ b/ath/if_athvar.h
@@ -974,7 +974,7 @@ typedef void (*ath_callback) (struct ath
 #endif
 
 /* Protects the device from concurrent accesses */
-#define	ATH_LOCK_INIT(_sc)		init_MUTEX(&(_sc)->sc_lock)
+#define	ATH_LOCK_INIT(_sc)		sema_init(&(_sc)->sc_lock, 1)
 #define	ATH_LOCK_DESTROY(_sc)
 #define	ATH_LOCK(_sc)			down(&(_sc)->sc_lock)
 #define	ATH_UNLOCK(_sc)			up(&(_sc)->sc_lock)
