diff -ruN matrixssl-1.1.2-old/src/Makefile matrixssl-1.1.2-new/src/Makefile
--- matrixssl-1.1.2-old/src/Makefile	2004-06-10 21:03:25.000000000 +0200
+++ matrixssl-1.1.2-new/src/Makefile	2004-07-09 22:58:06.000000000 +0200
@@ -46,8 +46,8 @@
 #	Compile options
 #
 SHARED	= -shared
-CFLAGS  = $(DFLAGS) -DLINUX
-LDFLAGS = -nostdlib -lc -lpthread
+CFLAGS  = $(DFLAGS) -DLINUX -fPIC
+LDFLAGS = -lc -lpthread -Wl,-soname,libmatrixssl.so.1.2
 
 #
 #	Override variables for compilation on Mac OS X (Darwin)
