--- a/Makefile
+++ b/Makefile
@@ -30,7 +30,7 @@ obj-$(CONFIG_COMPAT_WIRELESS_MODULES) +=
 obj-$(CONFIG_COMPAT_NET_USB_MODULES) += drivers/net/usb/
 
 obj-$(CONFIG_COMPAT_NETWORK_MODULES) += drivers/net/
-obj-$(CONFIG_COMPAT_VAR_MODULES) +=  drivers/ssb/ drivers/misc/eeprom/
+obj-$(CONFIG_COMPAT_VAR_MODULES) +=  drivers/misc/eeprom/
 endif
 
 obj-$(CONFIG_COMPAT_BLUETOOTH) += net/bluetooth/
--- a/config.mk
+++ b/config.mk
@@ -9,7 +9,6 @@ ifeq ($(wildcard $(KLIB_BUILD)/.config),
  CONFIG_PCI=y
  CONFIG_USB=y
  CONFIG_PCMCIA=y
- CONFIG_SSB=m
 else
 include $(KLIB_BUILD)/.config
 endif
@@ -240,21 +239,6 @@ CONFIG_IPW2200_QOS=y
 #
 # % echo 1 > /sys/bus/pci/drivers/ipw2200/*/rtap_iface
 
-ifneq ($(CONFIG_SSB),)
-# Sonics Silicon Backplane
-CONFIG_SSB_SPROM=y
-# CONFIG_SSB_DEBUG=y
-
-CONFIG_SSB_BLOCKIO=y
-CONFIG_SSB_PCIHOST=y
-CONFIG_SSB_B43_PCI_BRIDGE=y
-ifneq ($(CONFIG_PCMCIA),)
-CONFIG_SSB_PCMCIAHOST=y
-endif
-# CONFIG_SSB_DEBUG=y
-CONFIG_SSB_DRIVER_PCICORE=y
-endif
-
 CONFIG_P54_PCI=m
 
 # CONFIG_B44=m
@@ -418,7 +402,6 @@ endif # end of SPI driver list
 
 ifneq ($(CONFIG_MMC),)
 
-CONFIG_SSB_SDIOHOST=y
 CONFIG_B43_SDIO=y
 CONFIG_WL1251_SDIO=m
 
