--- a/net/wireless/compat-2.6.28.c
+++ b/net/wireless/compat-2.6.28.c
@@ -12,10 +12,11 @@
 
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,28))
 
-#include <linux/usb.h>
-
 /* 2.6.28 compat code goes here */
 
+#ifdef CONFIG_AR9170_USB
+#include <linux/usb.h>
+
 /*
  * Compat-wireless notes for USB backport stuff:
  *
@@ -206,6 +207,7 @@ int usb_anchor_empty(struct usb_anchor *
 }
 
 EXPORT_SYMBOL_GPL(usb_anchor_empty);
+#endif /* CONFIG_AR9170_USB */
 
 
 void __iomem *pci_ioremap_bar(struct pci_dev *pdev, int bar)
--- a/include/net/compat-2.6.28.h
+++ b/include/net/compat-2.6.28.h
@@ -9,7 +9,6 @@
 
 #include <linux/skbuff.h>
 #include <linux/if_ether.h>
-#include <linux/usb.h>
 
 #ifndef ETH_P_PAE
 #define ETH_P_PAE 0x888E      /* Port Access Entity (IEEE 802.1X) */
@@ -37,6 +36,8 @@
 #define pcmcia_parse_tuple(tuple, parse) pccard_parse_tuple(tuple, parse)
 #endif
 
+#ifdef CONFIG_AR9170_USB
+#include <linux/usb.h>
 #if 0
 extern void usb_poison_urb(struct urb *urb);
 #endif
@@ -49,6 +50,7 @@ extern void usb_poison_anchored_urbs(str
 extern struct urb *usb_get_from_anchor(struct usb_anchor *anchor);
 extern void usb_scuttle_anchored_urbs(struct usb_anchor *anchor);
 extern int usb_anchor_empty(struct usb_anchor *anchor);
+#endif /* CONFIG_AR9170_USB */
 
 
 void __iomem *pci_ioremap_bar(struct pci_dev *pdev, int bar);
--- a/net/wireless/compat-2.6.29.c
+++ b/net/wireless/compat-2.6.29.c
@@ -12,6 +12,7 @@
 
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,29))
 
+#ifdef CONFIG_AR9170_USB
 #include <linux/usb.h>
 
 /**
@@ -34,6 +35,7 @@ void usb_unpoison_anchored_urbs(struct u
 	spin_unlock_irqrestore(&anchor->lock, flags);
 }
 EXPORT_SYMBOL_GPL(usb_unpoison_anchored_urbs);
+#endif /* CONFIG_AR9170_USB */
 
 
 #endif /* LINUX_VERSION_CODE < KERNEL_VERSION(2,6,29) */
--- a/include/net/compat-2.6.29.h
+++ b/include/net/compat-2.6.29.h
@@ -41,7 +41,9 @@ static inline struct sk_buff *skb_queue_
 	return skb->prev;
 }
 
+#ifdef CONFIG_AR9170_USB
 extern void usb_unpoison_anchored_urbs(struct usb_anchor *anchor);
+#endif /* CONFIG_AR9170_USB */
 
 #endif /* (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,29)) */
 
