--- a/drivers/net/wireless/b43/main.c~	2011-07-23 00:33:46.573306410 +0200
+++ b/drivers/net/wireless/b43/main.c	2011-07-23 00:36:14.657726075 +0200
@@ -4955,7 +4955,7 @@
 static void b43_wireless_core_detach(struct b43_wldev *dev)
 {
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,31)
-	if (dev->sdev->bus->bustype != SSB_BUSTYPE_SDIO)
+	if (dev->dev->sdev->bus->bustype != SSB_BUSTYPE_SDIO)
 		compat_destroy_threaded_irq(&dev->irq_compat);
 #endif
 	/* We release firmware that late to not be required to re-request
