--- a/compat/Makefile
+++ b/compat/Makefile
@@ -2,9 +2,9 @@ obj-m += compat.o
 #compat-objs :=
 
 obj-$(CONFIG_COMPAT_FIRMWARE_CLASS) += compat_firmware_class.o
-obj-$(CONFIG_COMPAT_KFIFO) += kfifo.o
 
 compat-y += main.o
+compat-$(CONFIG_COMPAT_KFIFO) += kfifo.o
 
 # Compat kernel compatibility code
 compat-$(CONFIG_COMPAT_KERNEL_14) += compat-2.6.14.o
--- a/config.mk
+++ b/config.mk
@@ -90,7 +90,7 @@ endif #CONFIG_FW_LOADER
 endif #CONFIG_COMPAT_KERNEL_33
 
 ifdef CONFIG_COMPAT_KERNEL_36
-CONFIG_COMPAT_KFIFO=m
+CONFIG_COMPAT_KFIFO=y
 endif #CONFIG_COMPAT_KERNEL_36
 
 
