--- a/net/mac80211/ieee80211_i.h
+++ b/net/mac80211/ieee80211_i.h
@@ -642,6 +642,7 @@ enum queue_stop_reason {
 	IEEE80211_QUEUE_STOP_REASON_SKB_ADD,
 };
 
+#ifdef CONFIG_MAC80211_LEDS
 struct tpt_led_trigger {
 	struct led_trigger trig;
 	char name[32];
@@ -652,6 +653,7 @@ struct tpt_led_trigger {
 	unsigned long prev_traffic;
 	unsigned long tx_bytes, rx_bytes;
 };
+#endif
 
 /**
  * mac80211 scan flags - currently active scan mode
