--- a/net/mac80211/rx.c
+++ b/net/mac80211/rx.c
@@ -2555,7 +2555,6 @@ static void ieee80211_rx_handlers(struct
 		 * same TID from the same station
 		 */
 		rx->skb = skb;
-		rx->flags = 0;
 
 		CALL_RXH(ieee80211_rx_h_decrypt)
 		CALL_RXH(ieee80211_rx_h_check_more_data)
@@ -2626,6 +2625,7 @@ void ieee80211_release_reorder_timeout(s
 		.sdata = sta->sdata,
 		.local = sta->local,
 		.queue = tid,
+		.flags = 0,
 	};
 	struct tid_ampdu_rx *tid_agg_rx;
 
