Index: compat-wireless-2011-02-25/net/mac80211/main.c
===================================================================
--- compat-wireless-2011-02-25.orig/net/mac80211/main.c	2011-03-07 12:58:14.996968980 +0100
+++ compat-wireless-2011-02-25/net/mac80211/main.c	2011-03-07 13:03:26.732273903 +0100
@@ -384,6 +384,9 @@ void ieee80211_restart_hw(struct ieee802
 
 	trace_api_restart_hw(local);
 
+	wiphy_info(hw->wiphy,
+		   "Hardware restart was requested\n");
+
 	/* use this reason, ieee80211_reconfig will unblock it */
 	ieee80211_stop_queues_by_reason(hw,
 		IEEE80211_QUEUE_STOP_REASON_SUSPEND);
