--- a/rules/command.c
+++ b/rules/command.c
@@ -374,9 +374,9 @@ RULES_COMMAND_F(cmd_firmware) {
 	if (firmware == NULL)
 		return -1;
 
-	if (snprintf(sysfs_path_loading, PATH_MAX, "/sysfs%s/loading", devpath) >= PATH_MAX)
+	if (snprintf(sysfs_path_loading, PATH_MAX, "/sys%s/loading", devpath) >= PATH_MAX)
 		return -1;
-	if (snprintf(sysfs_path_data, PATH_MAX, "/sysfs%s/data", devpath) >= PATH_MAX)
+	if (snprintf(sysfs_path_data, PATH_MAX, "/sys%s/data", devpath) >= PATH_MAX)
 		return -1;
 	if (snprintf(firmware_path, PATH_MAX, "%s/%s", argv[0], firmware) >= PATH_MAX)
 		return -1;
