--- a/workers/worker_fork.c
+++ b/workers/worker_fork.c
@@ -396,7 +396,7 @@ static int worker_fork_process(void *in_
 		worker_fork_update_children(ctx);
 
 		child = NULL;
-		for (i = 0; i < ctx->children_count; i++) {
+		for (i = 0; i < ctx->children_count && i < ctx->max_children; i++) {
 			if (ctx->children[i]->busy == 0) {
 				child = ctx->children[i];
 				break;
@@ -426,7 +426,8 @@ static int worker_fork_process(void *in_
 			/*
 			 * We have to fork off a new child.
 			 */
-			if (ctx->children_count < ctx->max_children)
+			if (ctx->children_count < ctx->max_children || 
+			(ruleset_flags(&ctx->settings->rules, uevent) & FLAG_SLOW))
 				child = worker_fork_spawn(ctx);
 
 			for (i = 0; i < uevent->env_vars_c; i++) {
