diff -Nur openssl-0.9.7f/Makefile.org openssl-0.9.7f.new/Makefile.org
--- openssl-0.9.7f/Makefile.org	2005-03-15 10:46:13.000000000 +0100
+++ openssl-0.9.7f.new/Makefile.org	2005-04-03 20:15:26.000000000 +0200
@@ -175,7 +175,7 @@
 # we might set SHLIB_MARK to '$(SHARED_LIBS)'.
 SHLIB_MARK=
 
-DIRS=   crypto fips ssl $(SHLIB_MARK) sigs apps test tools
+DIRS=   crypto ssl apps
 SHLIBDIRS= crypto ssl
 
 # dirs in crypto to build
