--- a/arch/powerpc/boot/dts/fsl/mpc8568si-post.dtsi
+++ b/arch/powerpc/boot/dts/fsl/mpc8568si-post.dtsi
@@ -134,17 +134,8 @@
 
 	};
 
-	duart-sleep-nexus {
-		#address-cells = <1>;
-		#size-cells = <1>;
-		compatible = "simple-bus";
-		sleep = <&pmc 0x00000002>;
-		ranges;
-
 /include/ "pq3-duart-0.dtsi"
 
-	};
-
 	L2: l2-cache-controller@20000 {
 		compatible = "fsl,mpc8568-l2-cache-controller";
 		reg = <0x20000 0x1000>;
diff --git a/arch/powerpc/boot/dts/mpc8568mds.dts b/arch/powerpc/boot/dts/mpc8568mds.dts
index bead2b6..99ffa57 100644
--- a/arch/powerpc/boot/dts/mpc8568mds.dts
+++ b/arch/powerpc/boot/dts/mpc8568mds.dts
@@ -309,6 +309,9 @@
 			gpios = <&bcsr5 3 0>;
 		};
 	};
+	chosen {
+		linux,stdout-path = "/soc8568@e0000000/serial@4500";
+	};
 };
 
 /include/ "fsl/mpc8568si-post.dtsi"
