diff -urN linux-2.6.19.2/drivers/leds/Kconfig linux-2.6.19.2.new/drivers/leds/Kconfig
--- linux-2.6.19.2/drivers/leds/Kconfig	2007-01-10 20:10:37.000000000 +0100
+++ linux-2.6.19.2.new/drivers/leds/Kconfig	2007-04-16 22:09:40.000000000 +0200
@@ -76,6 +76,12 @@
 	  This option enables support for the Soekris net4801 and net4826 error
 	  LED.
 
+config LEDS_RDC3211
+	tristate "LED Support for RDC3211 boards"
+	depends on LEDS_CLASS && X86_RDC
+	help
+	  This option enables support for the RDC3211 various LEDs.
+
 comment "LED Triggers"
 
 config LEDS_TRIGGERS
diff -urN linux-2.6.21.1/drivers/leds/Makefile linux-2.6.21.1.new/drivers/leds/Makefile
--- linux-2.6.21.1/drivers/leds/Makefile	2007-04-27 23:49:26.000000000 +0200
+++ linux-2.6.21.1.new/drivers/leds/Makefile	2007-05-25 12:19:13.000000000 +0200
@@ -16,6 +16,7 @@
 obj-$(CONFIG_LEDS_WRAP)			+= leds-wrap.o
 obj-$(CONFIG_LEDS_H1940)		+= leds-h1940.o
 obj-$(CONFIG_LEDS_COBALT)		+= leds-cobalt.o
+obj-$(CONFIG_LEDS_RDC3211)		+= leds-rdc3211.o
 
 # LED Triggers
 obj-$(CONFIG_LEDS_TRIGGER_TIMER)	+= ledtrig-timer.o
