--- fuse-2.2.1-orig/configure	2005-03-08 15:48:57.000000000 +0100
+++ fuse-2.2.1-1/configure	2005-04-06 14:39:18.000000000 +0200
@@ -9354,7 +9354,7 @@
 
 fi
 
-CFLAGS="-Wall -W -g -O2"
+[ -z "$CFLAGS" ] && CFLAGS="-Wall -W -g -O2"
 CPPFLAGS="$CPPFLAGS -D_FILE_OFFSET_BITS=64 -D_REENTRANT -DFUSE_USE_VERSION=22"
 
 # Check whether --enable-kernel-module or --disable-kernel-module was given.
--- fuse-2.2.1-orig/configure.in	2005-03-08 15:39:23.000000000 +0100
+++ fuse-2.2.1-1/configure.in	2005-04-06 14:39:20.000000000 +0200
@@ -13,7 +13,7 @@
 	AC_SUBST(mkdir_p)
 fi
 
-CFLAGS="-Wall -W -g -O2"
+[ -z "$CFLAGS" ] && CFLAGS="-Wall -W -g -O2"
 CPPFLAGS="$CPPFLAGS -D_FILE_OFFSET_BITS=64 -D_REENTRANT -DFUSE_USE_VERSION=22"
 
 AC_ARG_ENABLE(kernel-module,
