Index: linux-2.6.21.7/arch/mips/kernel/head.S
===================================================================
--- linux-2.6.21.7.orig/arch/mips/kernel/head.S
+++ linux-2.6.21.7/arch/mips/kernel/head.S
@@ -129,11 +129,15 @@
 #endif
 	.endm
 
+
+	j kernel_entry
+	nop
+
 	/*
 	 * Reserved space for exception handlers.
 	 * Necessary for machines which link their kernels at KSEG0.
 	 */
-	.fill	0x400
+	.align 10
 
 EXPORT(stext)					# used for profiling
 EXPORT(_stext)
