--- a/plat/mediatek/common/drivers/nandx/core/nand/device_spi.c
+++ b/plat/mediatek/common/drivers/nandx/core/nand/device_spi.c
@@ -150,6 +150,21 @@ static struct device_spi spi_nand[] = {
 		&spi_extend_cmds, 0xff, 0xff
 	},
 	{
+		NAND_DEVICE("FM35X1GA",
+			    NAND_PACK_ID(0xe5, 0x71, 0, 0, 0, 0, 0, 0),
+			    2, 0, 3, 3,
+			    1, 1, 1, 1024, KB(128), KB(2), 64, 1,
+			    &spi_cmds, &spi_addressing, &spi_status[0],
+			    &spi_endurance, &spi_array_timing),
+		{
+			NAND_SPI_PROTECT(0xa0, 1, 2, 6),
+			NAND_SPI_CONFIG(0xb0, 4, 6, 1),
+			NAND_SPI_STATUS(0xc0, 4, 5),
+			NAND_SPI_CHARACTER(0xff, 0xff, 0xff, 0xff)
+		},
+		&spi_extend_cmds, 0xff, 0xff
+	},
+	{
 		NAND_DEVICE("NO-DEVICE",
 			    NAND_PACK_ID(0, 0, 0, 0, 0, 0, 0, 0), 0, 0, 0, 0,
 			    0, 0, 0, 0, 0, 0, 0, 1,
