Index: linux-4.3/arch/arm/boot/dts/imx6dl-wandboard.dts
===================================================================
--- linux-4.3.orig/arch/arm/boot/dts/imx6dl-wandboard.dts	2015-11-01 16:05:25.000000000 -0800
+++ linux-4.3/arch/arm/boot/dts/imx6dl-wandboard.dts	2015-12-18 10:39:44.923158318 -0800
@@ -19,4 +19,8 @@
 	memory {
 		reg = <0x10000000 0x40000000>;
 	};
+
+	chosen {
+		bootargs = "console=ttymxc0,115200";
+	};
 };
