Index: linux-4.3/drivers/net/phy/Kconfig
===================================================================
--- linux-4.3.orig/drivers/net/phy/Kconfig	2015-12-18 10:39:44.371158315 -0800
+++ linux-4.3/drivers/net/phy/Kconfig	2015-12-18 10:39:44.951158318 -0800
@@ -309,6 +309,14 @@
 
 source "drivers/net/phy/b53/Kconfig"
 
+config GATEWORKS_GW16083
+	tristate "Gateworks GW16083 Ethernet Expansion Mezzanine"
+	---help---
+	  The Gateworks GW16083 Ethernet Expansion Mezzanine connects to a
+	  Gateworks Ventana baseboard and provides a 7-port GbE managed
+	  Ethernet switch with 4 dedicated GbE RJ45 ports, and 2 Gbe/SFP
+	  ports"
+
 endif # PHYLIB
 
 config MICREL_KS8995MA
Index: linux-4.3/drivers/net/phy/Makefile
===================================================================
--- linux-4.3.orig/drivers/net/phy/Makefile	2015-12-18 10:39:44.371158315 -0800
+++ linux-4.3/drivers/net/phy/Makefile	2015-12-18 10:39:44.951158318 -0800
@@ -44,6 +44,7 @@
 obj-$(CONFIG_DP83867_PHY)	+= dp83867.o
 obj-$(CONFIG_STE10XP)		+= ste10Xp.o
 obj-$(CONFIG_MICREL_PHY)	+= micrel.o
+obj-$(CONFIG_GATEWORKS_GW16083)	+= gw16083.o
 obj-$(CONFIG_MDIO_OCTEON)	+= mdio-octeon.o
 obj-$(CONFIG_MICREL_KS8995MA)	+= spi_ks8995.o
 obj-$(CONFIG_AT803X_PHY)	+= at803x.o
