diff -ruN old/arch/arm/boot/dts/sun7i-a20-pcduino3.dts new/arch/arm/boot/dts/sun7i-a20-pcduino3.dts
--- old/arch/arm/boot/dts/sun7i-a20-pcduino3.dts	2014-08-01 21:28:06.000000000 +0200
+++ new/arch/arm/boot/dts/sun7i-a20-pcduino3.dts	2014-08-03 16:52:51.621238166 +0200
@@ -12,8 +12,7 @@
 
 /dts-v1/;
 /include/ "sun7i-a20.dtsi"
-/include/ "sunxi-ahci-reg.dtsi"
-/include/ "sun4i-a10-usb-vbus-reg.dtsi"
+/include/ "sunxi-common-regulators.dtsi"
 #include <dt-bindings/input/input.h>
 
 / {
