--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -130,28 +130,28 @@ static struct board_info __initdata boar
 
 	.leds = {
 		{
-			.name		= "adsl",
+			.name		= "96338GW:green:adsl",
 			.gpio		= 3,
 			.active_low	= 1,
 		},
 		{
-			.name		= "ses",
+			.name		= "96338GW:green:ses",
 			.gpio		= 5,
 			.active_low	= 1,
 		},
 		{
-			.name		= "ppp-fail",
+			.name		= "96338GW:green:ppp-fail",
 			.gpio		= 4,
 			.active_low	= 1,
 		},
 		{
-			.name		= "power",
+			.name		= "96338GW:green:power",
 			.gpio		= 0,
 			.active_low	= 1,
 			.default_trigger = "default-on",
 		},
 		{
-			.name		= "stop",
+			.name		= "96338GW:green:stop",
 			.gpio		= 1,
 			.active_low	= 1,
 		}
@@ -171,28 +171,28 @@ static struct board_info __initdata boar
 
 	.leds = {
 		{
-			.name		= "adsl",
+			.name		= "96338W:green:adsl",
 			.gpio		= 3,
 			.active_low	= 1,
 		},
 		{
-			.name		= "ses",
+			.name		= "96338W:green:ses",
 			.gpio		= 5,
 			.active_low	= 1,
 		},
 		{
-			.name		= "ppp-fail",
+			.name		= "96338W:green:ppp-fail",
 			.gpio		= 4,
 			.active_low	= 1,
 		},
 		{
-			.name		= "power",
+			.name		= "96338W:green:power",
 			.gpio		= 0,
 			.active_low	= 1,
 			.default_trigger = "default-on",
 		},
 		{
-			.name		= "stop",
+			.name		= "96338W:green:stop",
 			.gpio		= 1,
 			.active_low	= 1,
 		},
@@ -231,29 +231,29 @@ static struct board_info __initdata boar
 
 	.leds = {
 		{
-			.name		= "adsl-fail",
+			.name		= "96348R:green:adsl-fail",
 			.gpio		= 2,
 			.active_low	= 1,
 		},
 		{
-			.name		= "ppp",
+			.name		= "96348R:green:ppp",
 			.gpio		= 3,
 			.active_low	= 1,
 		},
 		{
-			.name		= "ppp-fail",
+			.name		= "96348R:green:ppp-fail",
 			.gpio		= 4,
 			.active_low	= 1,
 		},
 		{
-			.name		= "power",
+			.name		= "96348R:green:power",
 			.gpio		= 0,
 			.active_low	= 1,
 			.default_trigger = "default-on",
 
 		},
 		{
-			.name		= "stop",
+			.name		= "96348R:green:stop",
 			.gpio		= 1,
 			.active_low	= 1,
 		},
@@ -292,28 +292,28 @@ static struct board_info __initdata boar
 
 	.leds = {
 		{
-			.name		= "adsl-fail",
+			.name		= "96348GW-10:green:adsl-fail",
 			.gpio		= 2,
 			.active_low	= 1,
 		},
 		{
-			.name		= "ppp",
+			.name		= "96348GW-10:green:ppp",
 			.gpio		= 3,
 			.active_low	= 1,
 		},
 		{
-			.name		= "ppp-fail",
+			.name		= "96348GW-10:green:ppp-fail",
 			.gpio		= 4,
 			.active_low	= 1,
 		},
 		{
-			.name		= "power",
+			.name		= "96348GW-10:green:power",
 			.gpio		= 0,
 			.active_low	= 1,
 			.default_trigger = "default-on",
 		},
 		{
-			.name		= "stop",
+			.name		= "96348GW-10:green:stop",
 			.gpio		= 1,
 			.active_low	= 1,
 		},
@@ -346,28 +346,28 @@ static struct board_info __initdata boar
 
 	.leds = {
 		{
-			.name		= "adsl-fail",
+			.name		= "96348GW-11:green:adsl-fail",
 			.gpio		= 2,
 			.active_low	= 1,
 		},
 		{
-			.name		= "ppp",
+			.name		= "96348GW-11:green:ppp",
 			.gpio		= 3,
 			.active_low	= 1,
 		},
 		{
-			.name		= "ppp-fail",
+			.name		= "96348GW-11:green:ppp-fail",
 			.gpio		= 4,
 			.active_low	= 1,
 		},
 		{
-			.name		= "power",
+			.name		= "96348GW-11:green:power",
 			.gpio		= 0,
 			.active_low	= 1,
 			.default_trigger = "default-on",
 		},
 		{
-			.name		= "stop",
+			.name		= "96348GW-11:green:stop",
 			.gpio		= 1,
 			.active_low	= 1,
 		},
@@ -414,28 +414,28 @@ static struct board_info __initdata boar
 
 	.leds = {
 		{
-			.name		= "adsl-fail",
+			.name		= "96348GW:green:adsl-fail",
 			.gpio		= 2,
 			.active_low	= 1,
 		},
 		{
-			.name		= "ppp",
+			.name		= "96348GW:green:ppp",
 			.gpio		= 3,
 			.active_low	= 1,
 		},
 		{
-			.name		= "ppp-fail",
+			.name		= "96348GW:green:ppp-fail",
 			.gpio		= 4,
 			.active_low	= 1,
 		},
 		{
-			.name		= "power",
+			.name		= "96348GW:green:power",
 			.gpio		= 0,
 			.active_low	= 1,
 			.default_trigger = "default-on",
 		},
 		{
-			.name		= "stop",
+			.name		= "96348GW:green:stop",
 			.gpio		= 1,
 			.active_low	= 1,
 		},
@@ -567,27 +567,27 @@ static struct board_info __initdata boar
 
 	.leds = {
 		{
-			.name		= "adsl-fail",
+			.name		= "96358VW:green:adsl-fail",
 			.gpio		= 15,
 			.active_low	= 1,
 		},
 		{
-			.name		= "ppp",
+			.name		= "96358VW:green:ppp",
 			.gpio		= 22,
 			.active_low	= 1,
 		},
 		{
-			.name		= "ppp-fail",
+			.name		= "96358VW:green:ppp-fail",
 			.gpio		= 23,
 			.active_low	= 1,
 		},
 		{
-			.name		= "power",
+			.name		= "96358VW:green:power",
 			.gpio		= 4,
 			.default_trigger = "default-on",
 		},
 		{
-			.name		= "stop",
+			.name		= "96358VW:green:stop",
 			.gpio		= 5,
 		},
 	},
@@ -619,22 +619,22 @@ static struct board_info __initdata boar
 
 	.leds = {
 		{
-			.name		= "adsl",
+			.name		= "96358VW2:green:adsl",
 			.gpio		= 22,
 			.active_low	= 1,
 		},
 		{
-			.name		= "ppp-fail",
+			.name		= "96358VW2:green:ppp-fail",
 			.gpio		= 23,
 		},
 		{
-			.name		= "power",
+			.name		= "96358VW2:green:power",
 			.gpio		= 5,
 			.active_low	= 1,
 			.default_trigger = "default-on",
 		},
 		{
-			.name		= "stop",
+			.name		= "96358VW2:green:stop",
 			.gpio		= 4,
 			.active_low	= 1,
 		},
