diff -urN mrd6-0.9.5/src/Makefile mrd6-0.9.5.new/src/Makefile
--- mrd6-0.9.5/src/Makefile	2005-12-18 19:15:04.000000000 +0100
+++ mrd6-0.9.5.new/src/Makefile	2006-02-11 18:30:44.000000000 +0100
@@ -43,7 +43,7 @@
 	CFLAGS += -DNO_INET6_OPTION
 endif
 
-DEST_PREFIX = $(DESTDIR)$(PREFIX)
+DEST_PREFIX = $(DESTDIR)/$(PREFIX)
 
 ifeq ($(OPTIMIZE),yes)
 	ifeq ($(SPACE_OPTIMIZE),yes)
@@ -97,7 +97,7 @@
 	endif
 endif
 
-LDFLAGS += -lm
+LDFLAGS += -lm -lnotimpl -L$(STAGING_DIR)/usr/lib
 ifeq ($(PLATFORM),OS_LINUX)
 	LDFLAGS += -ldl
 endif
