--- a/backport-include/asm/barrier.h
+++ b/backport-include/asm/barrier.h
@@ -1,9 +1,9 @@
-#ifndef __BACKPORT_ASM_GENERIC_BARRIER_H
-#define __BACKPORT_ASM_GENERIC_BARRIER_H
+#ifndef __BACKPORT_ASM_BARRIER_H
+#define __BACKPORT_ASM_BARRIER_H
 #include_next <asm/barrier.h>
 
 #ifndef dma_rmb
 #define dma_rmb()	rmb()
 #endif
 
-#endif /* __BACKPORT_ASM_GENERIC_BARRIER_H */
+#endif /* __BACKPORT_ASM_BARRIER_H */
