From: Felix Fietkau <nbd@openwrt.org>
Date: Mon, 5 Oct 2015 17:41:25 +0200
Subject: [PATCH] mac80211: initialize tid field in struct ieee80211_txq

Signed-off-by: Felix Fietkau <nbd@openwrt.org>
---

--- a/net/mac80211/util.c
+++ b/net/mac80211/util.c
@@ -3323,9 +3323,11 @@ void ieee80211_init_tx_queue(struct ieee
 	if (sta) {
 		txqi->txq.sta = &sta->sta;
 		sta->sta.txq[tid] = &txqi->txq;
+		txqi->txq.tid = tid;
 		txqi->txq.ac = ieee802_1d_to_ac[tid & 7];
 	} else {
 		sdata->vif.txq = &txqi->txq;
+		txqi->txq.tid = 0;
 		txqi->txq.ac = IEEE80211_AC_BE;
 	}
 }
