--- a/include/linux/union.h
+++ b/include/linux/union.h
@@ -77,7 +77,7 @@
 #define __union_copyup(x, y, z)		({ BUG(); (0); })
 #define union_copyup(x, y)		({ (0); })
 #define union_copyup_dir(x)		({ BUG(); (0); })
-#define union_permission(x, y)		inode_permission(x->dentry->d_inode, y)
+#define union_permission(x, y)		inode_permission((x)->dentry->d_inode, y)
 
 #endif	/* CONFIG_UNION_MOUNT */
 #endif	/* __KERNEL__ */
