--- a/arch/mips/ath79/Kconfig
+++ b/arch/mips/ath79/Kconfig
@@ -483,6 +483,17 @@ config ATH79_MACH_WRT400N
 	select ATH79_DEV_LEDS_GPIO
 	select ATH79_DEV_M25P80
 
+config ATH79_MACH_R6100
+	bool "NETGEAR R6100 board support"
+	select SOC_AR934X
+	select ATH79_DEV_AP9X_PCI if PCI
+	select ATH79_DEV_ETH
+	select ATH79_DEV_GPIO_BUTTONS
+	select ATH79_DEV_LEDS_GPIO
+	select ATH79_DEV_NFC
+	select ATH79_DEV_USB
+	select ATH79_DEV_WMAC
+
 config ATH79_MACH_RB4XX
 	bool "MikroTik RouterBOARD 4xx series support"
 	select SOC_AR71XX
--- a/arch/mips/ath79/Makefile
+++ b/arch/mips/ath79/Makefile
@@ -88,6 +88,7 @@ obj-$(CONFIG_ATH79_MACH_PB42)		+= mach-p
 obj-$(CONFIG_ATH79_MACH_PB44)		+= mach-pb44.o
 obj-$(CONFIG_ATH79_MACH_PB92)		+= mach-pb92.o
 obj-$(CONFIG_ATH79_MACH_QIHOO_C301)	+= mach-qihoo-c301.o
+obj-$(CONFIG_ATH79_MACH_R6100)		+= mach-r6100.o
 obj-$(CONFIG_ATH79_MACH_RB4XX)		+= mach-rb4xx.o
 obj-$(CONFIG_ATH79_MACH_RB750)		+= mach-rb750.o
 obj-$(CONFIG_ATH79_MACH_RB91X)		+= mach-rb91x.o
--- a/arch/mips/ath79/machtypes.h
+++ b/arch/mips/ath79/machtypes.h
@@ -84,6 +84,7 @@ enum ath79_mach_type {
 	ATH79_MACH_PB42,		/* Atheros PB42 */
 	ATH79_MACH_PB92,		/* Atheros PB92 */
 	ATH79_MACH_QIHOO_C301,		/* Qihoo 360 C301 */
+	ATH79_MACH_R6100,		/* NETGEAR R6100 */
 	ATH79_MACH_RB_411,		/* MikroTik RouterBOARD 411/411A/411AH */
 	ATH79_MACH_RB_411U,		/* MikroTik RouterBOARD 411U */
 	ATH79_MACH_RB_433,		/* MikroTik RouterBOARD 433/433AH */
