diff -ur madwifi-svn/net80211/ieee80211_linux.h madwifi-svn.openwrt/net80211/ieee80211_linux.h
--- madwifi-svn/net80211/ieee80211_linux.h	2005-10-22 18:16:58.000000000 +0200
+++ madwifi-svn.openwrt/net80211/ieee80211_linux.h	2005-10-22 18:31:50.000000000 +0200
@@ -341,7 +341,7 @@
 }
 #endif
 /* XXX not in 2.6.6 but is in 2.6.10; might have appeared earlier */
-#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,10) && LINUX_VERSION_CODE < KERNEL_VERSION(2,4,31)
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,10) && LINUX_VERSION_CODE < KERNEL_VERSION(2,4,30)
 static __inline unsigned long msecs_to_jiffies(const unsigned int m)
 {
 	return (m * HZ + 999) / 1000;
