--- a/drivers/mtd/devices/m25p80.c
+++ b/drivers/mtd/devices/m25p80.c
@@ -138,30 +138,6 @@ static ssize_t m25p80_read(struct spi_no
 	/* convert the dummy cycles to the number of bytes */
 	dummy /= 8;
 
-	if (spi_flash_read_supported(spi)) {
-		struct spi_flash_read_message msg;
-
-		memset(&msg, 0, sizeof(msg));
-
-		msg.buf = buf;
-		msg.from = from;
-		msg.len = len;
-		msg.read_opcode = nor->read_opcode;
-		msg.addr_width = nor->addr_width;
-		msg.dummy_bytes = dummy;
-		/* TODO: Support other combinations */
-		msg.opcode_nbits = SPI_NBITS_SINGLE;
-		msg.addr_nbits = SPI_NBITS_SINGLE;
-		msg.data_nbits = m25p80_rx_nbits(nor);
-
-		ret = spi_flash_read(spi, &msg);
-		if (ret != -EOPNOTSUPP) {
-			if (ret < 0)
-				return ret;
-			return msg.retlen;
-		}
-	}
-
 	spi_message_init(&m);
 	memset(t, 0, (sizeof t));
 
