Index: linux-4.9.120/drivers/net/usb/cdc_ether.c
===================================================================
--- linux-4.9.120.orig/drivers/net/usb/cdc_ether.c
+++ linux-4.9.120/drivers/net/usb/cdc_ether.c
@@ -53,12 +53,19 @@ static int is_wireless_rndis(struct usb_
 		desc->bInterfaceSubClass == 1 &&
 		desc->bInterfaceProtocol == 3);
 }
+static int is_novatel_rndis(struct usb_interface_descriptor *desc)
+{
+	return (desc->bInterfaceClass == USB_CLASS_MISC &&
+			desc->bInterfaceSubClass == 4 &&
+			desc->bInterfaceProtocol == 1);
+}
 
 #else
 
 #define is_rndis(desc)		0
 #define is_activesync(desc)	0
 #define is_wireless_rndis(desc)	0
+#define is_novatel_rndis(desc)	0
 
 #endif
 
@@ -150,7 +157,8 @@ int usbnet_generic_cdc_bind(struct usbne
 	 */
 	rndis = (is_rndis(&intf->cur_altsetting->desc) ||
 		 is_activesync(&intf->cur_altsetting->desc) ||
-		 is_wireless_rndis(&intf->cur_altsetting->desc));
+		 is_wireless_rndis(&intf->cur_altsetting->desc) ||
+		 is_novatel_rndis(&intf->cur_altsetting->desc));
 
 	memset(info, 0, sizeof(*info));
 	info->control = intf;
@@ -819,7 +827,7 @@ static const struct usb_device_id	produc
 	USB_VENDOR_AND_INTERFACE_INFO(HUAWEI_VENDOR_ID, USB_CLASS_COMM,
 				      USB_CDC_SUBCLASS_ETHERNET, 255),
 	.driver_info = (unsigned long)&wwan_info,
-},
+},	
 	{ },		/* END */
 };
 MODULE_DEVICE_TABLE(usb, products);
Index: linux-4.9.120/drivers/net/usb/rndis_host.c
===================================================================
--- linux-4.9.120.orig/drivers/net/usb/rndis_host.c
+++ linux-4.9.120/drivers/net/usb/rndis_host.c
@@ -631,6 +631,10 @@ static const struct usb_device_id	produc
 	/* RNDIS for tethering */
 	USB_INTERFACE_INFO(USB_CLASS_WIRELESS_CONTROLLER, 1, 3),
 	.driver_info = (unsigned long) &rndis_info,
+},{
+	/* Novatel Verizon USB730L */
+	USB_INTERFACE_INFO(USB_CLASS_MISC, 4, 1),
+		.driver_info = (unsigned long) &rndis_info,
 },
 	{ },		// END
 };
