Index: lib_ifxos-1.5.12/src/linux/ifxos_linux_select_drv.c
===================================================================
--- lib_ifxos-1.5.12.orig/src/linux/ifxos_linux_select_drv.c	2010-04-24 11:33:55.000000000 +0200
+++ lib_ifxos-1.5.12/src/linux/ifxos_linux_select_drv.c	2010-04-24 11:34:21.000000000 +0200
@@ -33,6 +33,7 @@
 #include <linux/fs.h>
 #include <linux/wait.h>
 #include <linux/poll.h>
+#include <linux/sched.h>
 
 #include "ifx_types.h"
 #include "ifxos_rt_if_check.h"
