--- a/src/include/linux/ifxos_linux_common.h
+++ b/src/include/linux/ifxos_linux_common.h
@@ -73,7 +73,7 @@
 #  define __BYTE_ORDER                       __BIG_ENDIAN
    /** set the common IFXOS byte order for BIG endian */
 #  define IFXOS_BYTE_ORDER                   IFXOS_BIG_ENDIAN
-#elif
+#else
 #  error "missing endian definiton"
 #endif
 
--- a/src/linux/ifxos_linux_thread_drv.c
+++ b/src/linux/ifxos_linux_thread_drv.c
@@ -34,7 +34,9 @@
 #include <linux/sched.h>
 #include <linux/version.h>
 #include <linux/completion.h>
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,33))
 #include <linux/smp_lock.h>
+#endif
 #include <linux/signal.h>
 
 
