--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -562,6 +562,59 @@ static struct board_info __initdata boar
  * known 6358 boards
  */
 #ifdef CONFIG_BCM63XX_CPU_6358
+
+static struct board_info __initdata board_96358gw = {
+	.name				= "96358GW",
+	.expected_cpu_id		= 0x6358,
+
+	.has_uart0			= 1,
+	.has_enet1			= 1,
+	.has_pci			= 1,
+
+	.enet1 = {
+		.force_speed_100	= 1,
+		.force_duplex_full	= 1,
+	},
+
+	.leds = {
+		{
+			.name		= "power:green",
+			.gpio		= 5,
+			.active_low	= 1,
+			.default_trigger = "default-on",
+		},
+		{
+			.name		= "power:red",
+			.gpio		= 4,
+			.active_low	= 1,
+		},
+		{
+			.name		= "adsl",
+			.gpio		= 9,
+			.active_low	= 1,
+		},
+		{
+			.name		= "internet:green",
+			.gpio		= 2,
+		},
+		{
+			.name		= "internet:red",
+			.gpio		= 10,
+		},
+	},
+
+	.buttons = {
+		{
+			.desc		= "reset",
+			.gpio		= 34,
+			.active_low	= 1,
+			.type		= EV_KEY,
+			.code		= KEY_RESTART,
+			.threshold	= 3,
+		},
+	},
+};
+
 static struct board_info __initdata board_96358vw = {
 	.name				= "96358VW",
 	.expected_cpu_id		= 0x6358,
@@ -1413,6 +1466,7 @@ static const struct board_info __initdat
 #endif
 
 #ifdef CONFIG_BCM63XX_CPU_6358
+	&board_96358gw,
 	&board_96358vw,
 	&board_96358vw2,
 	&board_AGPFS0,
