Index: linux-2.6.24.2/arch/arm/mach-ixp4xx/avila-setup.c
===================================================================
--- linux-2.6.24.2.orig/arch/arm/mach-ixp4xx/avila-setup.c
+++ linux-2.6.24.2/arch/arm/mach-ixp4xx/avila-setup.c
@@ -132,12 +132,42 @@ static struct platform_device avila_pata
 	.resource		= avila_pata_resources,
 };
 
+/* Built-in 10/100 Ethernet MAC interfaces */
+static struct eth_plat_info avila_plat_eth[] = {
+	{
+		.phy		= 0,
+		.rxq		= 3,
+		.txreadyq	= 20,
+	}, {
+		.phy		= 1,
+		.rxq		= 4,
+		.txreadyq	= 21,
+	}
+};
+
+static struct platform_device avila_eth[] = {
+	{
+		.name			= "ixp4xx_eth",
+		.id			= IXP4XX_ETH_NPEB,
+		.dev.platform_data	= avila_plat_eth,
+	}, {
+		.name			= "ixp4xx_eth",
+		.id			= IXP4XX_ETH_NPEC,
+		.dev.platform_data	= avila_plat_eth + 1,
+	}
+};
+
 static struct platform_device *avila_devices[] __initdata = {
 	&avila_i2c_gpio,
 	&avila_flash,
 	&avila_uart
 };
 
+static struct platform_device *avila_eth_devices[] = {
+	&avila_eth[0],
+	&avila_eth[1]
+};
+
 static void __init avila_init(void)
 {
 	ixp4xx_sys_init();
@@ -159,6 +189,7 @@ static void __init avila_init(void)
 
 	platform_device_register(&avila_pata);
 
+	platform_add_devices(avila_eth_devices, ARRAY_SIZE(avila_eth_devices));
 }
 
 MACHINE_START(AVILA, "Gateworks Avila Network Platform")
