Index: linux-2.6.23.14/include/asm-arm/arch-ixp4xx/avila.h
===================================================================
--- linux-2.6.23.14.orig/include/asm-arm/arch-ixp4xx/avila.h	2008-01-31 17:40:36.000000000 +0100
+++ linux-2.6.23.14/include/asm-arm/arch-ixp4xx/avila.h	2008-01-31 17:40:42.000000000 +0100
@@ -25,7 +25,7 @@
 /*
  * AVILA PCI IRQs
  */
-#define AVILA_PCI_MAX_DEV	4
+#define AVILA_PCI_MAX_DEV	6
 #define LOFT_PCI_MAX_DEV    6
 #define AVILA_PCI_IRQ_LINES	4
 
