--- a/drivers/net/Kconfig
+++ b/drivers/net/Kconfig
@@ -362,6 +362,13 @@ config AX88796_93CX6
 	help
 	  Select this if your platform comes with an external 93CX6 eeprom.
 
+config KORINA
+	tristate "Korina Local Ethernet support"
+	depends on NET_ETHERNET && ( IDT_EB434 || MIKROTIK_RB500)
+	help
+	IDT RC32434 has one local ethernet port. Say Y here to enable it.
+	To compile this driver as a module, choose M here.
+
 config MACE
 	tristate "MACE (Power Mac ethernet) support"
 	depends on PPC_PMAC && PPC32
--- a/drivers/net/Makefile
+++ b/drivers/net/Makefile
@@ -30,6 +30,8 @@ ucc_geth_driver-objs := ucc_geth.o ucc_g
 #
 obj-$(CONFIG_PLIP) += plip.o
 
+obj-$(CONFIG_KORINA) += korina.o
+
 obj-$(CONFIG_ROADRUNNER) += rrunner.o
 
 obj-$(CONFIG_HAPPYMEAL) += sunhme.o
