--- a/arch/mips/lib/iomap-pci.c
+++ b/arch/mips/lib/iomap-pci.c
@@ -32,7 +32,7 @@ static void __iomem *ioport_map_pci(stru
 		       "report it to linux-mips@linux-mips.org or your "
 		       "vendor.\n", name);
 #ifdef CONFIG_PCI_DOMAINS
-		panic("To avoid data corruption io_map_base MUST be set with "
+		printk(KERN_WARNING "To avoid data corruption io_map_base MUST be set with "
 		      "multiple PCI domains.");
 #endif
 	}
