--- a/src/SDCCicode.c
+++ b/src/SDCCicode.c
@@ -1130,6 +1130,9 @@ iCode *getBuiltinParms (iCode *ic, int *
     return ic;
 }
 
+#pragma GCC push_options
+#pragma GCC optimize ("O0")
+
 /*-----------------------------------------------------------------*/
 /* operandOperation - performs operations on operands             */
 /*-----------------------------------------------------------------*/
@@ -1405,6 +1408,7 @@ operandOperation (operand * left, operan
 
   return retval;
 }
+#pragma GCC pop_options
 
 
 /*-----------------------------------------------------------------*/
