diff --git a/sdcc/sdas/as6808/Makefile.in b/sdcc/sdas/as6808/Makefile.in
index 5041265..0b8732e 100644
--- a/sdas/as6808/Makefile.in
+++ b/sdas/as6808/Makefile.in
@@ -36,7 +36,8 @@ VPATH           = @srcdir@
 
 CPPFLAGS        = @CPPFLAGS@ -DSDCDB -DNOICE -DINDEXLIB -I. -I$(srcdir) -I$(srcdir)/../asxxsrc -I$(srcdir)/../../support/Util
 CFLAGS          = @CFLAGS@ @WALL_FLAG@ $(CPPFLAGS)
-LDFLAGS         = @LDFLAGS@ -lm
+LDFLAGS         = @LDFLAGS@
+LIBS            = -lm
 
 OBJDIR = obj
 
@@ -65,7 +66,7 @@ transform       = @program_transform_name@
 all: checkconf $(ASX)
 
 $(ASX): $(ASOBJECTS)
-	$(CC) $(CFLAGS) $(LDFLAGS) -o $@ $(ASOBJECTS)
+	$(CC) $(CFLAGS) $(LDFLAGS) -o $@ $(ASOBJECTS) $(LIBS)
 
 # Compiling and installing everything and running test
 # ----------------------------------------------------
diff --git a/sdcc/sdas/as8051/Makefile.in b/sdcc/sdas/as8051/Makefile.in
index 8fa5946..fde3001 100644
--- a/sdas/as8051/Makefile.in
+++ b/sdas/as8051/Makefile.in
@@ -36,7 +36,8 @@ VPATH           = @srcdir@
 
 CPPFLAGS        = @CPPFLAGS@ -DSDCDB -DNOICE -DINDEXLIB -I. -I$(srcdir) -I$(srcdir)/../asxxsrc -I$(srcdir)/../../support/Util
 CFLAGS          = @CFLAGS@ @WALL_FLAG@ $(CPPFLAGS)
-LDFLAGS         = @LDFLAGS@ -lm
+LDFLAGS         = @LDFLAGS@
+LIBS            = -lm
 
 OBJDIR = obj
 
@@ -65,7 +66,7 @@ transform       = @program_transform_name@
 all: checkconf $(ASX)
 
 $(ASX): $(ASOBJECTS)
-	$(CC) $(CFLAGS) $(LDFLAGS) -o $@ $(ASOBJECTS)
+	$(CC) $(CFLAGS) $(LDFLAGS) -o $@ $(ASOBJECTS) $(LIBS)
 
 # Compiling and installing everything and running test
 # ----------------------------------------------------
diff --git a/sdcc/sdas/asgb/Makefile.in b/sdcc/sdas/asgb/Makefile.in
index 2763b3f..f9afb23 100644
--- a/sdas/asgb/Makefile.in
+++ b/sdas/asgb/Makefile.in
@@ -36,7 +36,8 @@ VPATH           = @srcdir@
 
 CPPFLAGS        = @CPPFLAGS@ -DSDCDB -DNOICE -DINDEXLIB -I. -I$(srcdir) -I$(srcdir)/../asxxsrc -I$(srcdir)/../../support/Util
 CFLAGS          = @CFLAGS@ @WALL_FLAG@ $(CPPFLAGS)
-LDFLAGS         = @LDFLAGS@ -lm
+LDFLAGS         = @LDFLAGS@
+LIBS            = -lm
 
 OBJDIR = obj
 
@@ -65,7 +66,7 @@ transform       = @program_transform_name@
 all: checkconf $(ASX)
 
 $(ASX): $(ASOBJECTS)
-	$(CC) $(CFLAGS) $(LDFLAGS) -o $@ $(ASOBJECTS)
+	$(CC) $(CFLAGS) $(LDFLAGS) -o $@ $(ASOBJECTS) $(LIBS)
 
 # Compiling and installing everything and running test
 # ----------------------------------------------------
diff --git a/sdcc/sdas/asz80/Makefile.in b/sdcc/sdas/asz80/Makefile.in
index d2bf1ac..89bd91c 100644
--- a/sdas/asz80/Makefile.in
+++ b/sdas/asz80/Makefile.in
@@ -36,7 +36,8 @@ VPATH           = @srcdir@
 
 CPPFLAGS        = @CPPFLAGS@ -DSDCDB -DNOICE -DINDEXLIB -I. -I$(srcdir) -I$(srcdir)/../asxxsrc -I$(srcdir)/../../support/Util
 CFLAGS          = @CFLAGS@ @WALL_FLAG@ $(CPPFLAGS)
-LDFLAGS         = @LDFLAGS@ -lm
+LDFLAGS         = @LDFLAGS@
+LIBS            = -lm
 
 OBJDIR = obj
 
@@ -65,7 +66,7 @@ transform       = @program_transform_name@
 all: checkconf $(ASX)
 
 $(ASX): $(ASOBJECTS)
-	$(CC) $(CFLAGS) $(LDFLAGS) -o $@ $(ASOBJECTS)
+	$(CC) $(CFLAGS) $(LDFLAGS) -o $@ $(ASOBJECTS) $(LIBS)
 
 # Compiling and installing everything and running test
 # ----------------------------------------------------
