diff -ur libvorbis-1.1.2/Makefile.in libvorbis-1.1.2-owrt/Makefile.in
--- libvorbis-1.1.2/Makefile.in	2005-11-28 06:43:44.000000000 +0100
+++ libvorbis-1.1.2-owrt/Makefile.in	2006-01-29 21:55:23.000000000 +0100
@@ -217,7 +217,7 @@
 target_os = @target_os@
 target_vendor = @target_vendor@
 AUTOMAKE_OPTIONS = foreign dist-zip
-SUBDIRS = lib include doc examples vq
+SUBDIRS = lib include doc vq
 m4datadir = $(datadir)/aclocal
 m4data_DATA = vorbis.m4
 pkgconfigdir = $(libdir)/pkgconfig
