--- a/drivers/mtd/nand/spi/core.c
+++ b/drivers/mtd/nand/spi/core.c
@@ -20,6 +20,7 @@
 #include <linux/spi/spi.h>
 #include <linux/spi/spi-mem.h>
 #include <linux/delay.h>
+#include <linux/mtd/mtk_bmt.h>
 
 static int spinand_read_reg_op(struct spinand_device *spinand, u8 reg, u8 *val)
 {
@@ -1622,6 +1623,7 @@ static int spinand_probe(struct spi_mem
 	if (ret)
 		return ret;
 
+	mtk_bmt_attach(mtd);
 	ret = mtd_device_register(mtd, NULL, 0);
 	if (ret)
 		goto err_spinand_cleanup;
@@ -1629,6 +1631,7 @@ static int spinand_probe(struct spi_mem
 	return 0;
 
 err_spinand_cleanup:
+	mtk_bmt_detach(mtd);
 	spinand_cleanup(spinand);
 
 	return ret;
@@ -1647,6 +1650,7 @@ static int spinand_remove(struct spi_mem
 	if (ret)
 		return ret;
 
+	mtk_bmt_detach(mtd);
 	spinand_cleanup(spinand);
 
 	return 0;
