From dba8578e06aedf1e67312ebfc6162e2fadc9448d Mon Sep 17 00:00:00 2001
From: John Crispin <blogic@openwrt.org>
Date: Thu, 7 Aug 2014 18:32:12 +0200
Subject: [PATCH 36/36] owrt: generic dtb image hack

Signed-off-by: John Crispin <blogic@openwrt.org>
---
 arch/mips/kernel/head.S |    3 +++
 1 file changed, 3 insertions(+)

--- a/arch/mips/kernel/head.S
+++ b/arch/mips/kernel/head.S
@@ -86,6 +86,9 @@ EXPORT(__image_cmdline)
 	.fill	0x400
 #endif /* CONFIG_IMAGE_CMDLINE_HACK */
 
+	.ascii  "OWRTDTB:"
+	EXPORT(__image_dtb)
+	.fill   0x4000
 	__REF
 
 NESTED(kernel_entry, 16, sp)			# kernel entry point
--- a/arch/mips/lantiq/Kconfig
+++ b/arch/mips/lantiq/Kconfig
@@ -32,7 +32,6 @@ choice
 config DT_EASY50712
 	bool "Easy50712"
 	depends on SOC_XWAY
-	select BUILTIN_DTB
 endchoice
 
 config PCI_LANTIQ
