--- a/libopkg/opkg_cmd.c
+++ b/libopkg/opkg_cmd.c
@@ -735,49 +735,6 @@
 }
 
 static int
-opkg_flag_cmd(int argc, char **argv)
-{
-     int i;
-     pkg_t *pkg;
-     const char *flags = argv[0];
-    
-     signal(SIGINT, sigint_handler);
-
-     for (i=1; i < argc; i++) {
-	  if (conf->restrict_to_default_dest) {
-	       pkg = pkg_hash_fetch_installed_by_name_dest(argv[i],
-							   conf->default_dest);
-	  } else {
-	       pkg = pkg_hash_fetch_installed_by_name(argv[i]);
-	  }
-
-	  if (pkg == NULL) {
-	       opkg_msg(ERROR, "Package %s is not installed.\n", argv[i]);
-	       continue;
-	  }
-          if (( strcmp(flags,"hold")==0)||( strcmp(flags,"noprune")==0)||
-              ( strcmp(flags,"user")==0)||( strcmp(flags,"ok")==0)) {
-	      pkg->state_flag = pkg_state_flag_from_str(flags);
-          }
-
-	  /* 
-	   * Useful if a package is installed in an offline_root, and
-	   * should be configured by opkg-cl configure at a later date.
-	   */
-          if (( strcmp(flags,"installed")==0)||( strcmp(flags,"unpacked")==0)){
-	      pkg->state_status = pkg_state_status_from_str(flags);
-          }
-
-	  opkg_state_changed++;
-	  opkg_msg(NOTICE, "Setting flags for package %s to %s.\n",
-		       pkg->name, flags);
-     }
-
-     write_status_files_if_changed();
-     return 0;
-}
-
-static int
 opkg_files_cmd(int argc, char **argv)
 {
      pkg_t *pkg;
@@ -1159,7 +1116,6 @@
      {"list_upgradable", 0, (opkg_cmd_fun_t)opkg_list_upgradable_cmd, PFM_SOURCE},
      {"list-upgradable", 0, (opkg_cmd_fun_t)opkg_list_upgradable_cmd, PFM_SOURCE},
      {"info", 0, (opkg_cmd_fun_t)opkg_info_cmd, 0},
-     {"flag", 1, (opkg_cmd_fun_t)opkg_flag_cmd, PFM_DESCRIPTION|PFM_SOURCE},
      {"status", 0, (opkg_cmd_fun_t)opkg_status_cmd, PFM_DESCRIPTION|PFM_SOURCE},
      {"install", 1, (opkg_cmd_fun_t)opkg_install_cmd, PFM_DESCRIPTION|PFM_SOURCE},
      {"remove", 1, (opkg_cmd_fun_t)opkg_remove_cmd, PFM_DESCRIPTION|PFM_SOURCE},
