From 9a86b6c16abc11b1090fbf4e102b4eed1d474d96 Mon Sep 17 00:00:00 2001
From: arokux <arokux@gmail.com>
Date: Wed, 18 Sep 2013 00:30:40 +0200
Subject: [PATCH] ARM: sun4i: dt: Add EHCI bindings to the Mele A1000

Signed-off-by: Hans de Goede <hdegoede@redhat.com>
---
 arch/arm/boot/dts/sun4i-a10-a1000.dts | 46 +++++++++++++++++++++++++++++++++++
 1 file changed, 46 insertions(+)

--- a/arch/arm/boot/dts/sun4i-a10-a1000.dts
+++ b/arch/arm/boot/dts/sun4i-a10-a1000.dts
@@ -73,6 +73,20 @@
 				allwinner,drive = <0>;
 				allwinner,pull = <0>;
 			};
+
+			usb1_vbus_pin: usb1_vbus_pin@0 {
+					allwinner,pins = "PH6";
+					allwinner,function = "gpio_out";
+					allwinner,drive = <0>;
+					allwinner,pull = <2>;
+			};
+
+			usb2_vbus_pin: usb2_vbus_pin@0 {
+					allwinner,pins = "PH3";
+					allwinner,function = "gpio_out";
+					allwinner,drive = <0>;
+					allwinner,pull = <2>;
+			};
 		};
 
 		uart0: serial@01c28000 {
@@ -86,6 +100,16 @@
 			pinctrl-0 = <&i2c0_pins_a>;
 			status = "okay";
 		};
+
+		ehci0: ehci0@0x01c14000 {
+			vbus-supply = <&reg_usb1_vbus>;
+			status = "okay";
+		};
+
+		ehci1: ehci1@0x01c1c000 {
+			vbus-supply = <&reg_usb2_vbus>;
+			status = "okay";
+		};
 	};
 
 	leds {
@@ -117,5 +141,27 @@
 			enable-active-high;
 			gpio = <&pio 7 15 0>;
 		};
+
+		reg_usb1_vbus: usb1-vbus {
+			compatible = "regulator-fixed";
+			pinctrl-names = "default";
+			pinctrl-0 = <&usb1_vbus_pin>;
+			regulator-name = "usb1-vbus";
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3300000>;
+			enable-active-high;
+			gpio = <&pio 7 6 0>;
+		};
+
+		reg_usb2_vbus: usb2-vbus {
+			compatible = "regulator-fixed";
+			pinctrl-names = "default";
+			pinctrl-0 = <&usb2_vbus_pin>;
+			regulator-name = "usb2-vbus";
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3300000>;
+			enable-active-high;
+			gpio = <&pio 7 3 0>;
+		};
 	};
 };
