diff --git a/dlib/interposition.c b/dlib/interposition.c
index c61e9ce..471ad4f 100644
--- a/dlib/interposition.c
+++ b/dlib/interposition.c
@@ -93,13 +93,13 @@ write$NOCANCEL(HAVE_PROT_WRITE_1, HAVE_PROT_WRITE_2, HAVE_PROT_WRITE_3);
 
 #endif /* HAVE_DARWIN */
 
-#ifndef __USE_GNU
-#define __USE_GNU /* XXX for RTLD_NEXT on Linux */
-#endif /* !__USE_GNU */
-
 
 #include <dlfcn.h>
 
+#ifndef RTLD_NEXT
+#define RTLD_NEXT  ((void *) -1l)
+#endif
+
 #ifdef __COVERITY__
 /*
  * Coverity naturally has no idea what the function sys_foo calls does,
diff --git a/lib/address.c b/lib/address.c
index 89d95c8..4eb505c 100644
--- a/lib/address.c
+++ b/lib/address.c
@@ -48,11 +48,12 @@
 
 #include "upnp.h"
 
-#ifndef __USE_GNU
-#define __USE_GNU /* XXX for RTLD_NEXT on Linux */
-#endif /* !__USE_GNU */
 #include <dlfcn.h>
 
+#ifndef RTLD_NEXT
+#define RTLD_NEXT  ((void *) -1l)
+#endif
+
 static const char rcsid[] =
 "$Id: address.c,v 1.288.4.4.6.4 2020/11/11 17:02:23 karls Exp $";
 
