diff -ruN eglibc-2.7-orig/ports/sysdeps/unix/sysv/linux/arm/nptl/lowlevellock.h eglibc-2.7/ports/sysdeps/unix/sysv/linux/arm/nptl/lowlevellock.h
--- eglibc-2.7-orig/ports/sysdeps/unix/sysv/linux/arm/nptl/lowlevellock.h	2007-09-12 18:26:54.000000000 +0200
+++ eglibc-2.7/ports/sysdeps/unix/sysv/linux/arm/nptl/lowlevellock.h	2008-11-04 17:11:21.000000000 +0100
@@ -25,6 +25,7 @@
 #include <atomic.h>
 #include <sysdep.h>
 #include <kernel-features.h>
+#include <tls.h>
 
 #define FUTEX_WAIT		0
 #define FUTEX_WAKE		1
