--- mkelfimage-2.7.old//Makefile	2006-03-28 02:44:59.000000000 +0300
+++ mkelfimage-2.7/Makefile	2011-01-17 14:15:13.000000000 +0200
@@ -14,7 +14,7 @@
 pkglibdir = $(libdir)/$(PACKAGE)
 pkgincludedir = $(includedir)/$(PACKAGE)
 
-CPPFLAGS=-DVERSION='"$(VERSION)"' -DRELEASE_DATE='"$(RELEASE_DATE)"' -I include
+CPPFLAGS=-DVERSION='"$(VERSION)"' -DRELEASE_DATE='"$(RELEASE_DATE)"' -I include -fno-stack-protector
 HOST_CPPFLAGS=$(CPPFLAGS)
 I386_CPPFLAGS=$(CPPFLAGS) -I arch/i386/include
 IA64_CPPFLAGS=$(CPPFLAGS) -I arch/ia64/include
