--- a/libc/sysdeps/linux/common/bits/mathcalls.h
+++ b/libc/sysdeps/linux/common/bits/mathcalls.h
@@ -229,7 +229,7 @@
 
 
 /* Return nonzero if VALUE is not a number.  */
-__MATHDECL_PRIV (int,__isnan,, (_Mdouble_ __value), (__const__));
+__MATHDECL_PRIV (int,isnan,, (_Mdouble_ __value), (__const__));
 
 #if defined __USE_MISC || defined __USE_XOPEN
 /* Return nonzero if VALUE is not a number.  */
