Index: linux-2.6.25/drivers/gpio/Kconfig
===================================================================
--- linux-2.6.25.orig/drivers/gpio/Kconfig	2008-04-27 23:35:57.000000000 +0100
+++ linux-2.6.25/drivers/gpio/Kconfig	2008-04-27 23:42:52.000000000 +0100
@@ -2,6 +2,13 @@
 # GPIO infrastructure and expanders
 #
 
+config GPIO_DEVICE
+	tristate "GPIO device support"
+	depends on GENERIC_GPIO
+	help
+	  Say Y to enable Linux GPIO device support.  This allows control of
+	  GPIO pins using a character device
+
 config HAVE_GPIO_LIB
 	bool
 	help
Index: linux-2.6.25/drivers/gpio/Makefile
===================================================================
--- linux-2.6.25.orig/drivers/gpio/Makefile	2008-04-27 23:42:59.000000000 +0100
+++ linux-2.6.25/drivers/gpio/Makefile	2008-04-27 23:43:41.000000000 +0100
@@ -7,3 +7,4 @@ obj-$(CONFIG_HAVE_GPIO_LIB)	+= gpiolib.o
 obj-$(CONFIG_GPIO_MCP23S08)	+= mcp23s08.o
 obj-$(CONFIG_GPIO_PCA953X)	+= pca953x.o
 obj-$(CONFIG_GPIO_PCF857X)	+= pcf857x.o
+obj-$(CONFIG_GPIO_DEVICE)	+= gpio_dev.o
