diff -Nru linux-2.6.25.10-old/crypto/ocf/cryptosoft.c linux-2.6.25.10/crypto/ocf/cryptosoft.c
--- linux-2.6.25.10-old/crypto/ocf/cryptosoft.c	2008-07-08 23:38:58.114980680 +0200
+++ linux-2.6.25.10/crypto/ocf/cryptosoft.c	2008-07-08 23:31:15.866980267 +0200
@@ -48,6 +48,7 @@
 #include <linux/skbuff.h>
 #include <linux/random.h>
 #include <asm/scatterlist.h>
+#include <linux/scatterlist.h>
 
 #include <cryptodev.h>
 #include <uio.h>
