Index: strongswan-2.8.2/programs/starter/Makefile
===================================================================
--- strongswan-2.8.2.orig/programs/starter/Makefile	2007-06-04 13:23:04.711017712 +0200
+++ strongswan-2.8.2/programs/starter/Makefile	2007-06-04 13:23:06.227787128 +0200
@@ -16,7 +16,6 @@
 FREESWANSRCDIR?=$(shell cd ../..; pwd)
 include ${FREESWANSRCDIR}/Makefile.inc
 
-LD=$(CC)
 RM=rm
 LEX=flex
 BISON=bison
@@ -59,7 +58,7 @@
 all:	starter
 
 starter:	$(OBJS) $(FREESWANLIB)
-		$(LD) $(LDFLAGS) -o starter $(OBJS) $(LIBS)
+		$(CC) $(LDFLAGS) -o starter $(OBJS) $(LIBS)
 
 lex.yy.c:	parser.tab.c parser.l parser.y parser.h
 		$(LEX) parser.l
