Index: strongswan-2.8.2/Makefile.inc
===================================================================
--- strongswan-2.8.2.orig/Makefile.inc	2007-06-04 13:23:04.604033976 +0200
+++ strongswan-2.8.2/Makefile.inc	2007-06-04 13:23:06.855691672 +0200
@@ -123,7 +123,7 @@
 # With a non-null DESTDIR, INC_RCDEFAULT will be used unless one of the
 # INC_RCDIRS directories has been pre-created under DESTDIR.
 INC_RCDIRS=/etc/rc.d/init.d /etc/rc.d /etc/init.d /sbin/init.d
-INC_RCDEFAULT=/etc/rc.d/init.d
+INC_RCDEFAULT=/etc/init.d
 
 # RCDIR is where boot/shutdown scripts go; FINALRCDIR is where they think
 # will finally be (so utils/Makefile can create a symlink in BINDIR to the
Index: strongswan-2.8.2/programs/showhostkey/showhostkey.in
===================================================================
--- strongswan-2.8.2.orig/programs/showhostkey/showhostkey.in	2007-06-04 13:23:04.612032760 +0200
+++ strongswan-2.8.2/programs/showhostkey/showhostkey.in	2007-06-04 13:23:06.855691672 +0200
@@ -62,7 +62,7 @@
 	exit 1
 fi
 
-host="`hostname --fqdn`"
+host="`cat /proc/sys/kernel/hostname`"
 
 awk '	BEGIN {
 		inkey = 0
