--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -77,6 +77,7 @@ MOSTLYCLEANFILES = yacc
 
 yacc:
 	echo '#! /bin/sh' >$@
+	echo 'test -n "$$STAGING_DIR" && exec "$$STAGING_DIR/../host/bin/bison" -y "$$@"' >>$@
 	echo "exec '$(bindir)/bison' -y "'"$$@"' >>$@
 	chmod a+x $@
 
--- a/src/Makefile.in
+++ b/src/Makefile.in
@@ -1352,6 +1352,7 @@ uninstall-am: uninstall-binPROGRAMS unin
 
 yacc:
 	echo '#! /bin/sh' >$@
+	echo 'test -n "$$STAGING_DIR" && exec "$$STAGING_DIR/../host/bin/bison" -y "$$@"' >>$@
 	echo "exec '$(bindir)/bison' -y "'"$$@"' >>$@
 	chmod a+x $@
 
