diff -urN linux-2.6.19.2.orig/drivers/net/wireless/zd1211rw/zd_usb.c linux-2.6.19.2/drivers/net/wireless/zd1211rw/zd_usb.c
--- linux-2.6.19.2.orig/drivers/net/wireless/zd1211rw/zd_usb.c	2007-06-13 23:14:44.000000000 +0200
+++ linux-2.6.19.2/drivers/net/wireless/zd1211rw/zd_usb.c	2007-06-13 23:19:51.000000000 +0200
@@ -47,11 +47,13 @@
 	{ USB_DEVICE(0x0586, 0x3402), .driver_info = DEVICE_ZD1211 },
 	{ USB_DEVICE(0x0b3b, 0x5630), .driver_info = DEVICE_ZD1211 },
 	{ USB_DEVICE(0x0b05, 0x170c), .driver_info = DEVICE_ZD1211 },
+	{ USB_DEVICE(0x0586, 0x3401), .driver_info = DEVICE_ZD1211 },
 	/* ZD1211B */
 	{ USB_DEVICE(0x0ace, 0x1215), .driver_info = DEVICE_ZD1211B },
 	{ USB_DEVICE(0x157e, 0x300d), .driver_info = DEVICE_ZD1211B },
 	{ USB_DEVICE(0x079b, 0x0062), .driver_info = DEVICE_ZD1211B },
 	{ USB_DEVICE(0x1582, 0x6003), .driver_info = DEVICE_ZD1211B },
+	
 	/* "Driverless" devices that need ejecting */
 	{ USB_DEVICE(0x0ace, 0x2011), .driver_info = DEVICE_INSTALLER },
 	{}
Binary files linux-2.6.19.2.orig/drivers/net/wireless/zd1211rw/.zd_usb.c.swp and linux-2.6.19.2/drivers/net/wireless/zd1211rw/.zd_usb.c.swp differ
