--- uClibc-0.9.27-old/libc/sysdeps/linux/common/bits/in.h	2005-01-12 08:59:21.000000000 +0100
+++ uClibc-0.9.27-new/libc/sysdeps/linux/common/bits/in.h	2005-08-23 02:12:47.000000000 +0200
@@ -115,6 +115,9 @@
 #define IPV6_MTU_DISCOVER	23
 #define IPV6_MTU		24
 #define IPV6_RECVERR		25
+#define IPV6_V6ONLY             26
+#define IPV6_JOIN_ANYCAST       27
+#define IPV6_LEAVE_ANYCAST      28
 
 #define SCM_SRCRT		IPV6_RXSRCRT
 
