--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,6 @@
 #
-#   Makefile - build and install the comgt package
+#   Makefile - build and install the gmon package
 #   Copyright (C) 2005  Martin Gregorie
-#   Copyright (C) 2006  Paul Hardwick
 #
 #   This program is free software; you can redistribute it and/or modify
 #   it under the terms of the GNU General Public License as published by
@@ -17,63 +16,41 @@
 #    along with this program; if not, write to the Free Software
 #    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 #
-#    martin@gregorie.org, paul@peck.org.uk
+#    martin@gregorie.org
 #
-#    $Id: Makefile,v 1.4 2006/10/20 14:30:19 pharscape Exp $
+#	$Id$
 #
-#
-
-LIB     = -L/usr/local/lib
-INC     = -I/usr/local/include
 EXE	= /usr/local/bin
 MAN     = /usr/share/man/man1
-CPROG	= comgt
-SCRIPTPATH = /etc/comgt/
-SCRIPTSRC = ./scripts/
-BIN     = $(CPROG) 
-MANP	= comgt.1 sigmon.1
-
-CFLAGS  = -c
-LDFLAGS =
+CPROG	= gcom
+SCRIPT  = sigmon
+BIN     = $(CPROG) $(SCRIPT)
+MANP	= gcom.1 sigmon.1
 
 all: $(BIN)
 
 install:
+	mv ($BIN} {$BIN2}
 	chmod a-w $(BIN)
 	chmod u+rw $(BIN)
 	chmod a+x $(BIN)
-	cp $(BIN) $(EXE)
+	cp $(BIN) $(EXE)/gcom
 	chmod a-wx $(MANP)
 	chmod u+rw $(MANP)
 	chmod a+r $(MANP)
 	cp $(MANP) $(MAN)
-	-mkdir $(SCRIPTPATH)
-	chmod a-w $(SCRIPTPATH)
-	chmod u+rw $(SCRIPTPATH)
-	chmod a+x $(SCRIPTPATH)
-	cp -f $(SCRIPTSRC)* $(SCRIPTPATH)
-	chmod a-w $(SCRIPTPATH)*
-	chmod u+rw $(SCRIPTPATH)*
-	chmod a+x $(SCRIPTPATH)*
-
-
-
 
 uninstall:
 	cd $(EXE); rm $(BIN)
 	cd $(MAN); rm $(MANP)
-	-rm -r $(SCRIPTPATH)
 
 clean:
-	-rm *.o 
-	-rm $(CPROG) 
-	-rm *~
-	-rm $(SCRIPTSRC)*~
+	rm *.o $(CPROG)
 
 
-comgt: comgt.o
-	cc comgt.o $(LDFLAGS) -o comgt
+gcom: gcom.o
+	$(CC) gcom.o $(LDFLAGS) -o gcom
 
-comgt.o: comgt.c comgt.h
-	cc comgt.c $(CFLAGS) 
+gcom.o: comgt.c comgt.h
+	$(CC) -c $(CFLAGS) comgt.c -o gcom.o
 
