--- a/include/osdep_service_linux.h
+++ b/include/osdep_service_linux.h
@@ -163,7 +163,11 @@ typedef void*		_thread_hdl_;
 typedef int		thread_return;
 typedef void*	thread_context;
 
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5,17,0))
+#define thread_exit() kthread_complete_and_exit(NULL, 0)
+#else
 #define thread_exit() complete_and_exit(NULL, 0)
+#endif
 
 typedef void timer_hdl_return;
 typedef void* timer_hdl_context;
