--- a/drivers/net/ethernet/lantiq_xrx200.c
+++ b/drivers/net/ethernet/lantiq_xrx200.c
@@ -143,6 +143,7 @@
 #define PMAC_IPG_MASK		0xf
 #define PMAC_HD_CTL_AS		0x0008
 #define PMAC_HD_CTL_AC		0x0004
+#define PMAC_HD_CTL_RC		0x0010
 #define PMAC_HD_CTL_RXSH	0x0040
 #define PMAC_HD_CTL_AST		0x0080
 #define PMAC_HD_CTL_RST		0x0100
@@ -1502,12 +1503,12 @@ static void xrx200_hw_init(struct xrx200
 #ifdef SW_ROUTING
 	/* enable status header, enable CRC */
 	ltq_pmac_w32_mask(0,
-		PMAC_HD_CTL_RST | PMAC_HD_CTL_AST | PMAC_HD_CTL_RXSH | PMAC_HD_CTL_AS | PMAC_HD_CTL_AC,
+		PMAC_HD_CTL_RST | PMAC_HD_CTL_AST | PMAC_HD_CTL_RXSH | PMAC_HD_CTL_AS | PMAC_HD_CTL_AC | PMAC_HD_CTL_RC,
 		PMAC_HD_CTL);
 #else
 	/* disable status header, enable CRC */
 	ltq_pmac_w32_mask(PMAC_HD_CTL_AST | PMAC_HD_CTL_RXSH | PMAC_HD_CTL_AS,
-		PMAC_HD_CTL_AC,
+		PMAC_HD_CTL_AC | PMAC_HD_CTL_RC,
 		PMAC_HD_CTL);
 #endif
 
