diff -ruN curl-7.14.0-old/configure curl-7.14.0-new/configure
--- curl-7.14.0-old/configure	2005-05-14 00:23:47.000000000 +0200
+++ curl-7.14.0-new/configure	2005-09-02 02:03:38.000000000 +0200
@@ -26132,10 +26132,10 @@
 
   fi
 
-  if test "$OPENSSL_ENABLED" = "1"; then
-                LD_LIBRARY_PATH="$LD_LIBRARY_PATH:$EXTRA_SSL/lib$libsuff"
-    export LD_LIBRARY_PATH
-  fi
+#  if test "$OPENSSL_ENABLED" = "1"; then
+#                LD_LIBRARY_PATH="$LD_LIBRARY_PATH:$EXTRA_SSL/lib$libsuff"
+#    export LD_LIBRARY_PATH
+#  fi
 
 fi
 
@@ -26325,8 +26325,8 @@
 echo "$as_me: detected GnuTLS version $version" >&6;}
 
 
-        LD_LIBRARY_PATH="$LD_LIBRARY_PATH:$gtlsprefix/lib$libsuff"
-        export LD_LIBRARY_PATH
+#        LD_LIBRARY_PATH="$LD_LIBRARY_PATH:$gtlsprefix/lib$libsuff"
+#        export LD_LIBRARY_PATH
       fi
 
     fi
diff -ruN curl-7.14.0-old/configure.ac curl-7.14.0-new/configure.ac
--- curl-7.14.0-old/configure.ac	2005-05-13 23:19:21.000000000 +0200
+++ curl-7.14.0-new/configure.ac	2005-09-02 02:03:45.000000000 +0200
@@ -917,13 +917,13 @@
 
   fi
 
-  if test "$OPENSSL_ENABLED" = "1"; then
-    dnl when the ssl shared libs were found in a path that the run-time linker
-    dnl doesn't search through, we need to add it to LD_LIBRARY_PATH to
-    dnl prevent further configure tests to fail due to this
-    LD_LIBRARY_PATH="$LD_LIBRARY_PATH:$EXTRA_SSL/lib$libsuff"
-    export LD_LIBRARY_PATH
-  fi
+#  if test "$OPENSSL_ENABLED" = "1"; then
+#    dnl when the ssl shared libs were found in a path that the run-time linker
+#    dnl doesn't search through, we need to add it to LD_LIBRARY_PATH to
+#    dnl prevent further configure tests to fail due to this
+#    LD_LIBRARY_PATH="$LD_LIBRARY_PATH:$EXTRA_SSL/lib$libsuff"
+#    export LD_LIBRARY_PATH
+#  fi
 
 fi
 
@@ -1021,8 +1021,8 @@
         dnl LD_LIBRARY_PATH to prevent further configure tests to fail
         dnl due to this
 
-        LD_LIBRARY_PATH="$LD_LIBRARY_PATH:$gtlsprefix/lib$libsuff"
-        export LD_LIBRARY_PATH
+#        LD_LIBRARY_PATH="$LD_LIBRARY_PATH:$gtlsprefix/lib$libsuff"
+#        export LD_LIBRARY_PATH
       fi
 
     fi
