Index: ipsec-tools-0.6.6/configure.ac
===================================================================
--- ipsec-tools-0.6.6.orig/configure.ac	2007-06-04 13:22:36.540300312 +0200
+++ ipsec-tools-0.6.6/configure.ac	2007-06-04 13:22:36.841254560 +0200
@@ -183,7 +183,7 @@
 
 if test "x$crypto_dir" != "x"; then
 	LIBS="$LIBS -L${crypto_dir}/lib"
-	CPPFLAGS="-I${crypto_dir}/include $CPPLAGS"
+	CPPFLAGS="-I${crypto_dir}/include $CPPFLAGS"
 fi
 AC_MSG_CHECKING(openssl version)
 
Index: ipsec-tools-0.6.6/configure
===================================================================
--- ipsec-tools-0.6.6.orig/configure	2007-06-04 13:22:36.547299248 +0200
+++ ipsec-tools-0.6.6/configure	2007-06-04 13:22:36.850253192 +0200
@@ -23687,7 +23687,7 @@
 
 if test "x$crypto_dir" != "x"; then
 	LIBS="$LIBS -L${crypto_dir}/lib"
-	CPPFLAGS="-I${crypto_dir}/include $CPPLAGS"
+	CPPFLAGS="-I${crypto_dir}/include $CPPFLAGS"
 fi
 echo "$as_me:$LINENO: checking openssl version" >&5
 echo $ECHO_N "checking openssl version... $ECHO_C" >&6
