--- linux-2.6.22.1.orig/arch/mips/au1000/mtx-1/board_setup.c	2007-07-29 18:57:35.000000000 +0200
+++ linux-2.6.22.1/arch/mips/au1000/mtx-1/board_setup.c	2007-07-29 18:59:48.000000000 +0200
@@ -54,7 +54,7 @@
 
 void __init board_setup(void)
 {
-#ifdef CONFIG_USB_OHCI
+#ifdef CONFIG_USB_OHCI_HCD || CONFIG_USB_OHCI_HCD_MODULE 
 	// enable USB power switch
 	au_writel( au_readl(GPIO2_DIR) | 0x10, GPIO2_DIR );
 	au_writel( 0x100000, GPIO2_OUTPUT );
