Index: shfs-0.35/shfs/Linux-2.6/dir.c
===================================================================
--- shfs-0.35.orig/shfs/Linux-2.6/dir.c	2007-06-04 13:22:58.041031704 +0200
+++ shfs-0.35/shfs/Linux-2.6/dir.c	2007-06-04 13:22:58.822912840 +0200
@@ -19,6 +19,8 @@
 #include "shfs_debug.h"
 #include "proc.h"
 
+static struct dentry_operations shfs_dentry_operations;
+	
 static int
 shfs_dir_open(struct inode *inode, struct file *filp)
 {
Index: shfs-0.35/shfs/Linux-2.6/shfs_fs.h
===================================================================
--- shfs-0.35.orig/shfs/Linux-2.6/shfs_fs.h	2007-06-04 13:22:58.398977288 +0200
+++ shfs-0.35/shfs/Linux-2.6/shfs_fs.h	2007-06-04 13:22:58.823912688 +0200
@@ -56,7 +56,6 @@
 #define ROUND_TO_MINS(x) do { (x).tv_sec = ((x).tv_sec / 60) * 60; (x).tv_nsec = 0; } while (0)
 
 /* shfs/dir.c */
-extern struct dentry_operations shfs_dentry_operations;
 extern struct file_operations shfs_dir_operations;
 extern struct inode_operations shfs_dir_inode_operations;
 extern void shfs_new_dentry(struct dentry *dentry);
