Index: shfs-0.35/shfs/Linux-2.6/dir.c
===================================================================
--- shfs-0.35.orig/shfs/Linux-2.6/dir.c	2007-06-17 04:08:32.449815896 +0200
+++ shfs-0.35/shfs/Linux-2.6/dir.c	2007-06-17 04:08:41.766399560 +0200
@@ -13,6 +13,7 @@
 #include <asm/uaccess.h>
 #include <linux/smp_lock.h>
 #include <linux/stat.h>
+#include <linux/jiffies.h>
 
 #include "shfs_fs.h"
 #include "shfs_fs_i.h"
Index: shfs-0.35/shfs/Linux-2.6/inode.c
===================================================================
--- shfs-0.35.orig/shfs/Linux-2.6/inode.c	2007-06-17 04:09:01.961329464 +0200
+++ shfs-0.35/shfs/Linux-2.6/inode.c	2007-06-17 04:13:08.501849608 +0200
@@ -15,6 +15,9 @@
 #include <asm/uaccess.h>
 #include <linux/file.h>
 #include <linux/smp_lock.h>
+#include <linux/jiffies.h>
+#include <linux/sched.h>
+#include <asm/current.h>
 
 #include "shfs_fs.h"
 #include "shfs_fs_sb.h"
