--- a/arch/mips/lantiq/xway/devices.c
+++ b/arch/mips/lantiq/xway/devices.c
@@ -121,6 +121,16 @@
 	}
 }
 
+/* madwifi */
+int lantiq_emulate_madwifi_eep = 0;
+EXPORT_SYMBOL(lantiq_emulate_madwifi_eep);
+
+void __init
+ltq_register_madwifi_eep(void)
+{
+	lantiq_emulate_madwifi_eep = 1;
+}
+
 static struct resource ltq_spi_resources[] = {
 	{
 		.start  = LTQ_SSC_BASE_ADDR,
--- a/arch/mips/lantiq/xway/devices.h
+++ b/arch/mips/lantiq/xway/devices.h
@@ -19,5 +19,6 @@
 extern void ltq_register_etop(struct ltq_eth_data *eth);
 extern void ltq_register_spi(struct ltq_spi_platform_data *pdata,
 	struct spi_board_info const *info, unsigned n);
+extern void ltq_register_madwifi_eep(void);
 
 #endif
