Index: linux-3.0.3/drivers/mtd/maps/lantiq-flash.c
===================================================================
--- linux-3.0.3.orig/drivers/mtd/maps/lantiq-flash.c	2011-08-17 19:57:16.000000000 +0200
+++ linux-3.0.3/drivers/mtd/maps/lantiq-flash.c	2011-09-29 20:45:14.785132132 +0200
@@ -20,6 +20,8 @@
 #include <linux/platform_device.h>
 #include <linux/mtd/physmap.h>
 
+#include "../mtdcore.h"
+
 #include <lantiq_soc.h>
 #include <lantiq_platform.h>
 
