diff -urN orig/dhcp6-1.0/client/Makefile dhcp6-1.0/client/Makefile
--- orig/dhcp6-1.0/client/Makefile	2025-06-28 23:21:40.000000000 +0200
+++ dhcp6-1.0/client/Makefile	2005-12-13 13:43:11.000000000 +0100
@@ -1,42 +1,40 @@
-
-CC = gcc
 STD_HEADERS = stdhead.h head.h struct.h options_type.h message_type.h states.h timer_val.h constants.h macros.h status_codes.h
 OBJECT_FILES = client.o solicit.o clilib.o parse.o request.o decline.o renew.o rebind.o release.o
 HEADER_FILES = solicit.h clilib.h parse.h request.h decline.h renew.h rebind.h release.h
-
+IFLAGS = -I$(STAGING_DIR)/usr/include
 
 dhcpv6_client : dhcpv6_client.c cli
-	$(CC) -g3 dhcpv6_client.c -o dhcpv6_client
+	$(CC) $(IFLAGS) dhcpv6_client.c -o dhcpv6_client
 
 cli : $(OBJECT_FILES)
-	$(CC) -g3 $(OBJECT_FILES) -o cli -lncurses
+	$(CC) $(IFLAGS) $(OBJECT_FILES) -o cli -L$(STAGING_DIR)/usr/lib -lncurses
 
 client.o : client.c $(STD_HEADERS) $(HEADER_FILES)
-	$(CC) -g3 -c client.c -o client.o
+	$(CC) $(IFLAGS) -c client.c -o client.o
 
 release.o : release.c release.h clilib.h
-	$(CC) -g3 -c release.c -o release.o
+	$(CC) $(IFLAGS) -c release.c -o release.o
 
 rebind.o : rebind.c rebind.h clilib.h
-	$(CC) -g3 -c rebind.c -o rebind.o
+	$(CC)  $(IFLAGS) -c rebind.c -o rebind.o
 
 renew.o : renew.c renew.h clilib.h
-	$(CC) -g3 -c renew.c -o renew.o
+	$(CC) $(IFLAGS) -c renew.c -o renew.o
 
 request.o : request.c request.h clilib.h
-	$(CC) -g3 -c request.c -o request.o
+	$(CC) $(IFLAGS) -c request.c -o request.o
 
 decline.o : decline.c decline.h clilib.h parse.h solicit.h
-	$(CC) -g3 -c decline.c -o decline.o
+	$(CC) $(IFLAGS) -c decline.c -o decline.o
 
 solicit.o : solicit.c solicit.h clilib.h parse.h
-	$(CC) -g3 -c solicit.c -o solicit.o
+	$(CC) $(IFLAGS) -c solicit.c -o solicit.o
 
 clilib.o : clilib.c clilib.h parse.h
-	$(CC) -g3 -c clilib.c -o clilib.o
+	$(CC) $(IFLAGS) -c clilib.c -o clilib.o
 
 parse.o : parse.c parse.h clilib.h
-	$(CC) -g3 -c parse.c -o parse.o
+	$(CC) $(IFLAGS) -c parse.c -o parse.o
 
 clean :
 	rm -f *.o cli dhcpv6_client
diff -urN orig/dhcp6-1.0/client/clilib.c dhcp6-1.0/client/clilib.c
--- orig/dhcp6-1.0/client/clilib.c	2025-06-28 23:21:40.000000000 +0200
+++ dhcp6-1.0/client/clilib.c	2005-12-13 13:42:43.000000000 +0100
@@ -554,7 +554,7 @@
 
 void generate_trans_id (u_int32_t *trans_id)
 {
-    extern u_int32_t g_trans_id;
+    extern int g_trans_id;
     time_t t;
     srand (time (&t));
     *trans_id = 0;
@@ -890,9 +890,7 @@
 		    break;
 
 		default :
-#if DEBUG == 3
 		    printf ("Unrecognized DUID type\n");
-#endif
 	    }
 
 	    next_opt = (struct OPTIONS *) malloc (sizeof (struct OPTIONS));
diff -urN orig/dhcp6-1.0/server/Makefile dhcp6-1.0/server/Makefile
--- orig/dhcp6-1.0/server/Makefile	2002-06-28 14:50:04.000000000 +0200
+++ dhcp6-1.0/server/Makefile	2005-12-13 11:35:53.000000000 +0100
@@ -1,27 +1,26 @@
-CC = gcc
 STD_HEADERS = stdhead.h head.h struct.h options_type.h message_type.h macros.h status_codes.h constants.h
 OBJECT_FILES = server.o lib.o advertise.o parse.o leases.o reply.o
 
 serv: $(OBJECT_FILES)
-	$(CC) -g3 $(OBJECT_FILES) -o serv
+	$(CC) $(OBJECT_FILES) -o serv
 
 lib.o: lib.c lib.h $(STD_HEADERS)
-	$(CC) -g3 -c lib.c -o lib.o
+	$(CC) -c lib.c -o lib.o
 
 parse.o: parse.c parse.h $(STD_HEADERS)
-	$(CC) -g3 -c parse.c -o parse.o
+	$(CC) -c parse.c -o parse.o
 
 leases.o: leases.c leases.h $(STD_HEADERS)
-	$(CC) -g3 -c leases.c -o leases.o
+	$(CC) -c leases.c -o leases.o
 
 advertise.o: advertise.c advertise.h $(STD_HEADERS)
-	$(CC) -g3 -c advertise.c -o advertise.o
+	$(CC) -c advertise.c -o advertise.o
 
 reply.o: reply.c reply.h $(STD_HEADERS)
-	$(CC) -g3 -c reply.c -o reply.o
+	$(CC) -c reply.c -o reply.o
 
 server.o: server.c lib.h advertise.h reply.h leases.h parse.h $(STD_HEADERS)
-	$(CC) -g3 -c server.c -o server.o
+	$(CC) -c server.c -o server.o
 
 clean: 
 	rm -f *.o serv
