--- a/Makefile
+++ b/Makefile
@@ -21,34 +21,17 @@
 # MA 02111-1307 USA
 #
 
-include $(TOPDIR)/config.mk
-
-HOSTSRCS := $(SRCTREE)/lib/crc32.c  fw_env.c  fw_env_main.c
+SRCS := crc32.c fw_env.c fw_env_main.c
 HEADERS	:= fw_env.h
 
-# Compile for a hosted environment on the target
-HOSTCPPFLAGS  = -idirafter $(SRCTREE)/include \
-		-idirafter $(OBJTREE)/include2 \
-		-idirafter $(OBJTREE)/include \
-		-DUSE_HOSTCC
-
-ifeq ($(MTD_VERSION),old)
-HOSTCPPFLAGS += -DMTD_OLD
-endif
-
-all:	$(obj)fw_printenv
-
-# Some files complain if compiled with -pedantic, use HOSTCFLAGS_NOPED
-$(obj)fw_printenv:	$(HOSTSRCS) $(HEADERS)
-	$(HOSTCC) $(HOSTCFLAGS_NOPED) $(HOSTLDFLAGS) -o $@ $(HOSTSRCS)
+CPPFLAGS := -Wall $(CFLAGS)
 
-clean:
-	rm -f $(obj)fw_printenv
+all: fw_printenv
 
-#########################################################################
+fw_printenv: $(SRCS) $(HEADERS)
+	$(CC) $(CPPFLAGS) $(SRCS) -o fw_printenv
 
-include $(TOPDIR)/rules.mk
-
-sinclude $(obj).depend
+clean:
+	rm -f fw_printenv
 
 #########################################################################
