--- a/bin/quilt.in
+++ b/bin/quilt.in
@@ -7,9 +7,15 @@
 #  See the COPYING and AUTHORS files for more details.
 
 export TEXTDOMAIN=quilt
-export TEXTDOMAINDIR=@LOCALEDIR@
 
-: ${QUILT_DIR=@QUILT_DIR@} ${QUILT_LIB=@QUILT_LIB@}
+if test -n "$STAGING_DIR"; then
+	export TEXTDOMAINDIR="$STAGING_DIR/../host/share/locale"
+	: ${QUILT_DIR=$STAGING_DIR/../host/share/quilt} ${QUILT_LIB=$STAGING_DIR/../host/lib/quilt}
+else
+	export TEXTDOMAINDIR=@LOCALEDIR@
+	: ${QUILT_DIR=@QUILT_DIR@} ${QUILT_LIB=@QUILT_LIB@}
+fi
+
 export QUILT_DIR QUILT_LIB
 
 if [ -z "$QUILTRC" ]
