--- a/Makefile.in
+++ b/Makefile.in
@@ -266,13 +266,10 @@ $(patsubst %.in,%,$(wildcard bin/*.in qu
 	@$(if $(filter $@,$(NON_EXEC_IN)),,chmod +x $@)
 
 configure : configure.ac aclocal.m4
-	autoconf
-	@echo "Please run ./configure"
-	@false
+	@touch $@
 
 Makefile : Makefile.in configure
-	@echo "Please run ./configure"
-	@false
+	@touch $@
 
 compat_leftover := $(filter-out $(COMPAT),$(shell $(FIND) compat -maxdepth 1 -type f -perm +111))
 
