--- /tmp/ipkg-build	2008-08-01 09:36:41.000000000 +0200
+++ ipkg-utils-1.7/ipkg-build	2008-08-06 15:04:52.000000000 +0200
@@ -10,7 +10,8 @@
 set -e
 
 version=1.0
-
+FIND="$(which find)"
+FIND="${FIND:-$(which gfind)}"
 TAR="${TAR:-$(which tar)}"
 
 ipkg_extract_value() {
@@ -49,7 +50,7 @@
 
 	PKG_ERROR=0
 
-	cvs_dirs=`find . -name 'CVS'`
+	cvs_dirs=`$FIND . -name 'CVS'`
 	if [ -n "$cvs_dirs" ]; then
 	    if [ "$noclean" = "1" ]; then
 		echo "*** Warning: The following CVS directories where found.
@@ -62,7 +63,7 @@
 	    fi
 	fi
 
-	tilde_files=`find . -name '*~'`
+	tilde_files=`$FIND . -name '*~'`
 	if [ -n "$tilde_files" ]; then
 	    if [ "$noclean" = "1" ]; then
 		echo "*** Warning: The following files have names ending in '~'.
@@ -75,7 +76,7 @@
 	    fi
 	fi
 
-	large_uid_files=`find . -uid +99 || true`
+	large_uid_files=`$FIND . -uid +99 || true`
 
 	if [ "$ogargs" = "" ]  && [ -n "$large_uid_files" ]; then
 		echo "*** Warning: The following files have a UID greater than 99.
