--- a/arch/mips/ath79/Kconfig
+++ b/arch/mips/ath79/Kconfig
@@ -579,7 +579,7 @@ config ATH79_MACH_TL_WDR4300
 	select ATH79_DEV_WMAC
 
 config ATH79_MACH_TL_WR703N
-	bool "TP-LINK TL-WR703N/TL-MR10U support"
+	bool "TP-LINK TL-WR703N/TL-WR710N/TL-MR10U support"
 	select SOC_AR933X
 	select ATH79_DEV_ETH
 	select ATH79_DEV_GPIO_BUTTONS
--- a/arch/mips/ath79/machtypes.h
+++ b/arch/mips/ath79/machtypes.h
@@ -98,6 +98,7 @@ enum ath79_mach_type {
 	ATH79_MACH_TL_WR1043ND,		/* TP-LINK TL-WR1043ND */
 	ATH79_MACH_TL_WR2543N,		/* TP-LINK TL-WR2543N/ND */
 	ATH79_MACH_TL_WR703N,		/* TP-LINK TL-WR703N */
+	ATH79_MACH_TL_WR710N,		/* TP-LINK TL-WR710N */
 	ATH79_MACH_TL_WR720N_V3,	/* TP-LINK TL-WR720N v3 */
 	ATH79_MACH_TL_WR741ND,		/* TP-LINK TL-WR741ND */
 	ATH79_MACH_TL_WR741ND_V4,	/* TP-LINK TL-WR741ND  v4*/
