--- a/arch/mips/ath79/machtypes.h
+++ b/arch/mips/ath79/machtypes.h
@@ -127,6 +127,7 @@ enum ath79_mach_type {
 	ATH79_MACH_WNDR4300,		/* NETGEAR WNDR4300 */
 	ATH79_MACH_WNR2000,		/* NETGEAR WNR2000 */
 	ATH79_MACH_WNR2000_V3,		/* NETGEAR WNR2000 v3 */
+	ATH79_MACH_WNR612_V2,		/* NETGEAR WNR612 v2 */
 	ATH79_MACH_WP543,		/* Compex WP543 */
 	ATH79_MACH_WPE72,		/* Compex WPE72 */
 	ATH79_MACH_WRT160NL,		/* Linksys WRT160NL */
--- a/arch/mips/ath79/Kconfig
+++ b/arch/mips/ath79/Kconfig
@@ -419,7 +419,7 @@ config ATH79_MACH_WNR2000
 	select ATH79_DEV_WMAC
 
 config ATH79_MACH_WNR2000_V3
-	bool "NETGEAR WNR2000 V3 board support"
+	bool "NETGEAR WNR2000 V3/WNR612 v2 board support"
 	select SOC_AR724X
 	select ATH79_DEV_AP9X_PCI if PCI
 	select ATH79_DEV_ETH
