--- a/arch/mips/ath79/machtypes.h
+++ b/arch/mips/ath79/machtypes.h
@@ -81,6 +81,7 @@ enum ath79_mach_type {
 	ATH79_MACH_RB_751,		/* MikroTik RouterBOARD 751 */
 	ATH79_MACH_RB_751G,		/* Mikrotik RouterBOARD 751G */
 	ATH79_MACH_RB_951G,		/* Mikrotik RouterBOARD 951G */
+	ATH79_MACH_RB_951U,		/* Mikrotik RouterBOARD 951Ui-2HnD */
 	ATH79_MACH_RB_2011G,		/* Mikrotik RouterBOARD 2011UAS-2HnD */
 	ATH79_MACH_RB_2011L,		/* Mikrotik RouterBOARD 2011L */
 	ATH79_MACH_RB_2011US,		/* Mikrotik RouterBOARD 2011UAS */
