--- linux-2.6.37.6.org/Makefile	2012-03-28 17:10:58.297558774 +0200
+++ linux-2.6.37.6.mod/Makefile	2012-03-28 17:15:00.161564521 +0200
@@ -354,7 +354,8 @@
 		   -fno-strict-aliasing -fno-common \
 		   -Werror-implicit-function-declaration \
 		   -Wno-format-security \
-		   -fno-delete-null-pointer-checks
+		   -fno-delete-null-pointer-checks \
+		   -Wno-error=unused-but-set-variable
 KBUILD_AFLAGS_KERNEL :=
 KBUILD_CFLAGS_KERNEL :=
 KBUILD_AFLAGS   := -D__ASSEMBLY__
