--- a/arch/arm/mach-orion5x/wrt350n-v2-setup.c
+++ b/arch/arm/mach-orion5x/wrt350n-v2-setup.c
@@ -58,19 +58,11 @@ static struct mtd_partition wrt350n_v2_n
 	{
 		.name		= "kernel",
 		.offset		= 0x00000000,
-		.size		= 0x00760000,
+		.size		= 0x00100000,
 	}, {
 		.name		= "rootfs",
-		.offset		= 0x001a0000,
-		.size		= 0x005c0000,
-	}, {
-		.name		= "lang",
-		.offset		= 0x00760000,
-		.size		= 0x00040000,
-	}, {
-		.name		= "nvram",
-		.offset		= 0x007a0000,
-		.size		= 0x00020000,
+		.offset		= 0x00100000,
+		.size		= 0x00650000,
 	}, {
 		.name		= "u-boot",
 		.offset		= 0x007c0000,
