diff -Nur ueagle-atm-1.0d1/driver/ueagle.c ueagle-atm-1.0d1.patched/driver/ueagle.c
--- ueagle-atm-1.0d1/driver/ueagle.c	2005-11-19 18:36:04.000000000 +0100
+++ ueagle-atm-1.0d1.patched/driver/ueagle.c	2006-03-29 02:03:43.000000000 +0200
@@ -1652,7 +1652,6 @@
  * USB driver descriptor
  */
 static struct usb_driver uea_driver = {
-	.owner = THIS_MODULE,
 	.name = "ueagle-atm",
 	.id_table = uea_ids,
 	.probe = uea_probe,
