--- linux-2.6.16.4/arch/mips/sibyte/sb1250/setup.c	2006-04-21 16:16:18.000000000 -0700
+++ linux-2.6.16.4/arch/mips/sibyte/sb1250/setup.c	2006-04-21 16:15:22.000000000 -0700
@@ -140,6 +140,14 @@
 		periph_rev = 3;
 		pass_str = "A2";
 		break;
+	case K_SYS_REVISION_BCM112x_A3:
+		periph_rev = 3;
+		pass_str = "A3";
+		break;
+	case K_SYS_REVISION_BCM112x_A4:
+		periph_rev = 3;
+		pass_str = "A4";
+		break;
 	default:
 		prom_printf("Unknown %s rev %x\n", soc_str, soc_pass);
 		ret = 1;
