From 9f90b78f834e3d18520c03af4644ff239a2e891b Mon Sep 17 00:00:00 2001
From: warmcat <andy@warmcat.com>
Date: Sun, 13 Apr 2008 07:25:46 +0100
Subject: [PATCH] local-config-ext2.patch

---
 defconfig-2.6.24 |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/defconfig-2.6.24 b/defconfig-2.6.24
index 270c4f9..833e15e 100644
--- a/defconfig-2.6.24
+++ b/defconfig-2.6.24
@@ -1565,7 +1565,7 @@ CONFIG_RTC_DRV_S3C=m
 #
 # File systems
 #
-CONFIG_EXT2_FS=m
+CONFIG_EXT2_FS=y
 # CONFIG_EXT2_FS_XATTR is not set
 # CONFIG_EXT2_FS_XIP is not set
 CONFIG_EXT3_FS=y
-- 
1.5.6.5

