From a320183ac1d4bd8a430ef9f3e3b957ed4177454b Mon Sep 17 00:00:00 2001
From: Willie <willie_chen@openmoko.com>
Date: Sun, 13 Apr 2008 07:25:50 +0100
Subject: [PATCH] use gpio control leds

---
 drivers/leds/leds-neo1973-gta02.c |    9 ++++-----
 1 files changed, 4 insertions(+), 5 deletions(-)

diff --git a/drivers/leds/leds-neo1973-gta02.c b/drivers/leds/leds-neo1973-gta02.c
index bf1d540..e442e48 100644
--- a/drivers/leds/leds-neo1973-gta02.c
+++ b/drivers/leds/leds-neo1973-gta02.c
@@ -127,22 +127,22 @@ static int __init gta02led_probe(struct platform_device *pdev)
 
 		switch (lp->gpio) {
 		case S3C2410_GPB0:
-			lp->has_pwm = 1;
+			lp->has_pwm = 0;
 			lp->pwm.timerid = PWM0;
 			s3c2410_gpio_cfgpin(lp->gpio, S3C2410_GPB0_TOUT0);
 			break;
 		case S3C2410_GPB1:
-			lp->has_pwm = 1;
+			lp->has_pwm = 0;
 			lp->pwm.timerid = PWM1;
 			s3c2410_gpio_cfgpin(lp->gpio, S3C2410_GPB1_TOUT1);
 			break;
 		case S3C2410_GPB2:
-			lp->has_pwm = 1;
+			lp->has_pwm = 0;
 			lp->pwm.timerid = PWM2;
 			s3c2410_gpio_cfgpin(lp->gpio, S3C2410_GPB2_TOUT2);
 			break;
 		case S3C2410_GPB3:
-			lp->has_pwm = 1;
+			lp->has_pwm = 0;
 			lp->pwm.timerid = PWM3;
 			s3c2410_gpio_cfgpin(lp->gpio, S3C2410_GPB3_TOUT3);
 			break;
@@ -167,7 +167,6 @@ static int __init gta02led_probe(struct platform_device *pdev)
 			break;
 		default:
 			break;
-		}
 
 		mutex_init(&lp->mutex);
 		rc = led_classdev_register(&pdev->dev, &lp->cdev);
-- 
1.5.6.5

