From 421ed0bdd5b468d9de3ba571f49e7b840a8032e3 Mon Sep 17 00:00:00 2001
From: No Attribution <dunno@dunno.org>
Date: Tue, 3 Jun 2008 10:49:55 +0100
Subject: [PATCH] Remove GTA01 PM debug

---
 defconfig-gta01 |    6 ++----
 1 files changed, 2 insertions(+), 4 deletions(-)

diff --git a/defconfig-gta01 b/defconfig-gta01
index f574da5..ab84e8b 100644
--- a/defconfig-gta01
+++ b/defconfig-gta01
@@ -159,7 +159,7 @@ CONFIG_CPU_LLSERIAL_S3C2440=y
 #
 # Power management
 #
-CONFIG_S3C2410_PM_DEBUG=y
+# CONFIG_S3C2410_PM_DEBUG is not set
 # CONFIG_S3C2410_PM_CHECK is not set
 CONFIG_S3C_LOWLEVEL_UART_PORT=0
 
@@ -1693,9 +1693,7 @@ CONFIG_FORCED_INLINING=y
 # CONFIG_SAMPLES is not set
 # CONFIG_DEBUG_USER is not set
 CONFIG_DEBUG_ERRORS=y
-CONFIG_DEBUG_LL=y
-# CONFIG_DEBUG_ICEDCC is not set
-CONFIG_DEBUG_S3C_PORT=y
+# CONFIG_DEBUG_LL is not set
 CONFIG_DEBUG_S3C_UART=0
 
 #
-- 
1.5.6.5

