From 88bf43840b9df0eb0a077a1394eb564be80a412e Mon Sep 17 00:00:00 2001
From: Andy Green <andy@openmoko.com>
Date: Mon, 11 Aug 2008 20:50:05 +0100
Subject: [PATCH] clean-checkpatch-gsm-resume-dep.patch

Signed-off-by: Andy Green <andy@openmoko.com>
---
 drivers/mfd/glamo/glamo-core.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/drivers/mfd/glamo/glamo-core.c b/drivers/mfd/glamo/glamo-core.c
index e310382..b30ec9f 100644
--- a/drivers/mfd/glamo/glamo-core.c
+++ b/drivers/mfd/glamo/glamo-core.c
@@ -1249,7 +1249,8 @@ void glamo_register_resume_dependency(struct resume_dependency *
 	register_resume_dependency(&glamo_handle->resume_dependency,
 							     resume_dependency);
 	if (glamo_handle->is_suspended)
-		activate_all_resume_dependencies(&glamo_handle->resume_dependency);
+		activate_all_resume_dependencies(
+					      &glamo_handle->resume_dependency);
 }
 EXPORT_SYMBOL_GPL(glamo_register_resume_dependency);
 
-- 
1.5.6.5

