--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -624,6 +624,8 @@ static struct board_info __initdata boar
 	.has_uart0			= 1,
 	.has_enet0			= 1,
 	.enet0 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -667,6 +669,8 @@ static struct board_info __initdata boar
 	.has_uart0			= 1,
 	.has_enet0			= 1,
 	.enet0 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -952,6 +956,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1025,6 +1031,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1254,6 +1262,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1429,6 +1439,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1543,6 +1555,12 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
+		.has_phy		= 1,
+		.phy_id			= 0,
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1918,6 +1936,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -1970,6 +1990,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -2110,6 +2132,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -2227,6 +2251,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
