--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -14,6 +14,7 @@
 #include <linux/ssb/ssb.h>
 #include <linux/gpio_keys.h>
 #include <linux/input.h>
+#include <linux/export.h>
 #include <linux/spi/spi.h>
 #include <linux/spi/spi_gpio.h>
 #include <linux/spi/74x164.h>
@@ -4310,7 +4311,7 @@ static const struct board_info __initcon
  * bcm4318 WLAN work
  */
 #ifdef CONFIG_SSB_PCIHOST
-static struct ssb_sprom bcm63xx_sprom = {
+struct ssb_sprom bcm63xx_sprom = {
 	.revision		= 0x02,
 	.board_rev		= 0x17,
 	.country_code		= 0x0,
@@ -4330,6 +4331,7 @@ static struct ssb_sprom bcm63xx_sprom =
 	.boardflags_lo		= 0x2848,
 	.boardflags_hi		= 0x0000,
 };
+EXPORT_SYMBOL(bcm63xx_sprom);
 
 int bcm63xx_get_fallback_sprom(struct ssb_bus *bus, struct ssb_sprom *out)
 {
