--- a/drivers/net/phy/sfp.c	2023-06-13 22:42:45.849605297 +0200
+++ b/drivers/net/phy/sfp.c	2023-06-13 22:39:12.506921843 +0200
@@ -383,6 +383,13 @@
 		.modes = sfp_quirk_2500basex,
 		.fixup = sfp_fixup_ignore_tx_fault,
 	}, {
+		// GPON-ONU-34-20BI can operate at 2500base-X, but report 1.2GBd
+		// NRZ in their EEPROM
+		.vendor = "FS",
+		.part = "GPON-ONU-34-20BI",
+		.modes = sfp_quirk_2500basex,
+		.fixup = sfp_fixup_ignore_tx_fault,
+	}, {
 		// OEM SFP-GE-T is 1000Base-T module
 		.vendor = "OEM",
 		.part = "SFP-GE-T",
