--- a/Makefile
+++ b/Makefile
@@ -21,12 +21,10 @@
 # MA 02111-1307 USA
 #
 
-include $(TOPDIR)/config.mk
-
 SRCS	:= $(obj)crc32.c  fw_env.c  fw_env_main.c
 HEADERS	:= fw_env.h
 
-CPPFLAGS := -Wall -DUSE_HOSTCC -I$(SRCTREE)/include
+CPPFLAGS := -Wall $(CFLAGS)
 
 ifeq ($(MTD_VERSION),old)
 CPPFLAGS += -DMTD_OLD
@@ -35,18 +33,7 @@ endif
 all:	$(obj)fw_printenv
 
 $(obj)fw_printenv:	$(SRCS) $(HEADERS)
-	$(CROSS_COMPILE)gcc $(CPPFLAGS) $(SRCS) -o $(obj)fw_printenv
+	$(CC) $(CPPFLAGS) $(SRCS) -o $(obj)fw_printenv
 
 clean:
 	rm -f $(obj)fw_printenv $(obj)crc32.c
-
-$(obj)crc32.c:
-	ln -s $(src)../../lib_generic/crc32.c $(obj)crc32.c
-
-#########################################################################
-
-include $(TOPDIR)/rules.mk
-
-sinclude $(obj).depend
-
-#########################################################################
