--- a/ath10k-6.10/mac.c
+++ b/ath10k-6.10/mac.c
@@ -1675,7 +1675,7 @@ static void ath10k_recalc_radar_detectio
 		 * by indicating that radar was detected.
 		 */
 		ath10k_warn(ar, "failed to start CAC: %d\n", ret);
-		ieee80211_radar_detected(ar->hw);
+		ieee80211_radar_detected(ar->hw, NULL);
 	}
 }
 
@@ -6238,7 +6238,7 @@ err:
 	return ret;
 }
 
-static void ath10k_stop(struct ieee80211_hw *hw)
+static void ath10k_stop(struct ieee80211_hw *hw, bool suspend)
 {
 	struct ath10k *ar = hw->priv;
 	u32 opt;
--- a/ath10k-6.10/debug.c
+++ b/ath10k-6.10/debug.c
@@ -3319,7 +3319,7 @@ static ssize_t ath10k_write_simulate_rad
 	if (!arvif->is_started)
 		return -EINVAL;
 
-	ieee80211_radar_detected(ar->hw);
+	ieee80211_radar_detected(ar->hw, NULL);
 
 	return count;
 }
--- a/ath10k-6.10/wmi.c
+++ b/ath10k-6.10/wmi.c
@@ -4402,7 +4402,7 @@ static void ath10k_radar_detected(struct
 	if (ar->dfs_block_radar_events)
 		ath10k_info(ar, "DFS Radar detected, but ignored as requested\n");
 	else
-		ieee80211_radar_detected(ar->hw);
+		ieee80211_radar_detected(ar->hw, NULL);
 }
 
 static void ath10k_radar_confirmation_work(struct work_struct *work)
