Index: drv_dsl_cpe_api-3.24.4.4/src/include/drv_dsl_cpe_os_linux.h
===================================================================
--- drv_dsl_cpe_api-3.24.4.4.orig/src/include/drv_dsl_cpe_os_linux.h	2015-02-08 09:59:02.424259834 +0100
+++ drv_dsl_cpe_api-3.24.4.4/src/include/drv_dsl_cpe_os_linux.h	2015-02-08 10:02:19.580740173 +0100
@@ -218,7 +218,7 @@
 #define DSL_DRV_WAKEUP_WAKELIST(queue)       wake_up_interruptible(&(queue))
 #define DSL_DRV_INIT_EVENT(name,ev)          init_waitqueue_head(&(ev))
 /* wait for an event, timeout is measured in ms */
-#define DSL_DRV_WAIT_EVENT_TIMEOUT(ev,t)     interruptible_sleep_on_timeout(&(ev), (t) * HZ / 1000)
+#define DSL_DRV_WAIT_EVENT_TIMEOUT(ev,t)     wait_event_interruptible_timeout((ev), 0 == 1, (t) * HZ / 1000)
 #define DSL_DRV_WAIT_EVENT(ev)               interruptible_sleep_on(&(ev))
 #define DSL_DRV_WAKEUP_EVENT(ev)             wake_up_interruptible(&(ev))
 #define DSL_DRV_TimeMSecGet()                DSL_DRV_ElapsedTimeMSecGet(0)
