diff -urN gd-2.0.33.old/configure gd-2.0.33.dev/configure
--- gd-2.0.33.old/configure	2004-11-04 00:14:47.000000000 +0100
+++ gd-2.0.33.dev/configure	2005-05-05 15:06:26.000000000 +0200
@@ -7217,8 +7217,6 @@
 shlibpath_overrides_runpath=unknown
 version_type=none
 dynamic_linker="$host_os ld.so"
-sys_lib_dlsearch_path_spec="/lib /usr/lib"
-sys_lib_search_path_spec="/lib /usr/lib /usr/local/lib"
 
 case $host_os in
 aix3*)
@@ -7481,8 +7479,6 @@
   *)
     ;;
   esac
-  sys_lib_dlsearch_path_spec="/lib${libsuff} /usr/lib${libsuff}"
-  sys_lib_search_path_spec="/lib${libsuff} /usr/lib${libsuff} /usr/local/lib${libsuff}"
   ;;
 
 netbsd*)
@@ -11032,14 +11028,14 @@
 fi
 
   if test -n "$LIBPNG12_CONFIG"; then
-    libpng_CPPFLAGS=`libpng12-config --cflags`
+    libpng_CPPFLAGS=`$LIBPNG12_CONFIG --cflags`
     # should be --ldopts, but it's currently broken
-    libpng_LDFLAGS=`libpng12-config --ldflags`
+    libpng_LDFLAGS=`$LIBPNG12_CONFIG --ldflags`
     libpng_LDFLAGS=`echo " $libpng_LDFLAGS" | sed 's/ -l[^ ][^ ]*//g'`
   elif test -n "$LIBPNG_CONFIG"; then
-    libpng_CPPFLAGS=`libpng-config --cflags`
+    libpng_CPPFLAGS=`$LIBPNG_CONFIG --cflags`
     # should be --ldopts, but it's currently broken
-    libpng_LDFLAGS=`libpng-config --ldflags`
+    libpng_LDFLAGS=`$LIBPNG_CONFIG --ldflags`
     libpng_LDFLAGS=`echo " $libpng_LDFLAGS" | sed 's/ -l[^ ][^ ]*//g'`
   elif test -d "$withval"; then
     libpng_CPPFLAGS="-I$withval/include"
