diff -ruN lua-5.0.2-orig/config lua-5.0.2-1/config
--- lua-5.0.2-orig/config	2003-04-11 16:00:41.000000000 +0200
+++ lua-5.0.2-1/config	2005-05-25 11:23:35.000000000 +0200
@@ -25,15 +25,15 @@
 # interface (e.g., Linux, Solaris, IRIX, BSD, AIX, HPUX, and probably others),
 # uncomment the next two lines.
 #
-#LOADLIB= -DUSE_DLOPEN=1
-#DLLIB= -ldl
+LOADLIB= -DUSE_DLOPEN=1
+DLLIB= -ldl
 #
 # In Linux with gcc, you should also uncomment the next definition for
 # MYLDFLAGS, which passes -E (= -export-dynamic) to the linker. This option
 # allows dynamic libraries to link back to the `lua' program, so that they do
 # not need the Lua libraries. (Other systems may have an equivalent facility.)
 #
-#MYLDFLAGS= -Wl,-E
+MYLDFLAGS= -Wl,-E
 #
 # On Windows systems. support for dynamic loading is enabled by default.
 # To disable this support, uncomment the next line.
@@ -142,7 +142,7 @@
 
 # This should work in all Unix systems, but you may want to add options.
 #
-STRIP= strip
+STRIP= /bin/true
 
 # ------------------------------------------------------------------ install
 
