Index: linux-2.6.21.7/arch/arm/mach-at91/board-vlink.c
===================================================================
--- linux-2.6.21.7.orig/arch/arm/mach-at91/board-vlink.c
+++ linux-2.6.21.7/arch/arm/mach-at91/board-vlink.c
@@ -175,6 +175,7 @@ static void __init vlink_board_init(void
 /* USB Device control */
        at91_set_gpio_input(AT91_PIN_PB27, 1);  // UDB_CNX
        at91_set_gpio_output(AT91_PIN_PB28, 1); // UDB_PUP
+			 at91_set_multi_drive(AT91_PIN_PB28, 1); // Set to multi-drive
 
 }
 
