Index: linux-2.6.21.7/drivers/mtd/devices/at91_dataflash.c
===================================================================
--- linux-2.6.21.7.orig/drivers/mtd/devices/at91_dataflash.c
+++ linux-2.6.21.7/drivers/mtd/devices/at91_dataflash.c
@@ -161,12 +161,12 @@ static struct mtd_partition static_parti
 		.mask_flags	= MTD_WRITEABLE,	/* read-only */
 	},
 	{
-		.name		= "kernel",
+		.name		= "knlroot",
 		.offset		= MTDPART_OFS_NXTBLK,
-		.size		= 5 * 32 * 8 * 1056,	/* 5 sectors */
+		.size		= 0x320400,	/* 5 sectors */
 	},
 	{
-		.name		= "filesystem",
+		.name		= "jffs2",
 		.offset		= MTDPART_OFS_NXTBLK,
 		.size		= MTDPART_SIZ_FULL,	/* rest = 26 sectors */
 	}
