Index: iptables-1.3.7/extensions/libipt_iprange.c
===================================================================
--- iptables-1.3.7.orig/extensions/libipt_iprange.c	2007-06-04 13:21:43.288395832 +0200
+++ iptables-1.3.7/extensions/libipt_iprange.c	2007-06-04 13:21:44.343235472 +0200
@@ -6,6 +6,7 @@
 #include <getopt.h>
 
 #include <iptables.h>
+#include <linux/types.h>
 #include <linux/netfilter_ipv4/ipt_iprange.h>
 
 /* Function which prints out usage message. */
