diff -urN sangam_atm-D7.03.01.00.orig/tn7atm.c sangam_atm-D7.03.01.00/tn7atm.c
--- sangam_atm-D7.03.01.00.orig/tn7atm.c	2008-01-05 03:11:23.000000000 +0100
+++ sangam_atm-D7.03.01.00/tn7atm.c	2008-01-05 03:11:34.000000000 +0100
@@ -566,7 +566,7 @@
  *  Description: tnetd73xx SAR interrupt.
  *
  *~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
-static void tn7atm_sar_irq (int irq, void *voiddev, struct pt_regs *regs)
+static irqreturn_t tn7atm_sar_irq (int irq, void *voiddev)
 {
   struct atm_dev *atmdev;
   Tn7AtmPrivate *priv;
@@ -593,6 +593,7 @@
 #ifdef TIATM_INST_SUPP
   psp_trace_par (ATM_DRV_SAR_ISR_EXIT, retval);
 #endif
+  return IRQ_HANDLED;
 }
 
 /*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
@@ -602,7 +603,7 @@
  *  Description: tnetd73xx DSL interrupt.
  *
  *~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
-static void tn7atm_dsl_irq (int irq, void *voiddev, struct pt_regs *regs)
+static irqreturn_t tn7atm_dsl_irq (int irq, void *voiddev)
 {
   struct atm_dev *atmdev;
   Tn7AtmPrivate *priv;
@@ -624,6 +625,8 @@
 #ifdef TIATM_INST_SUPP
   psp_trace_par (ATM_DRV_DSL_ISR_EXIT, retval);
 #endif
+
+  return IRQ_HANDLED;
 }
 
 /*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
