--- a/arch/arm/mach-cns3xxx/pcie.c
+++ b/arch/arm/mach-cns3xxx/pcie.c
@@ -161,12 +161,12 @@ static int cns3xxx_pci_setup(int nr, str
 	return 1;
 }
 
-static struct pci_ops cns3xxx_pcie_ops = {
+struct pci_ops cns3xxx_pcie_ops = {
 	.read = cns3xxx_pci_read_config,
 	.write = cns3xxx_pci_write_config,
 };
 
-static struct pci_bus *cns3xxx_pci_scan_bus(int nr, struct pci_sys_data *sys)
+struct pci_bus * __devinit cns3xxx_pci_scan_bus(int nr, struct pci_sys_data *sys)
 {
 	return pci_scan_root_bus(NULL, sys->busnr, &cns3xxx_pcie_ops, sys,
 				 &sys->resources);
