Index: linux-3.14.18/arch/arm/boot/dts/Makefile
===================================================================
--- linux-3.14.18.orig/arch/arm/boot/dts/Makefile
+++ linux-3.14.18/arch/arm/boot/dts/Makefile
@@ -233,7 +233,10 @@ dtb-$(CONFIG_ARCH_OMAP2PLUS) += omap2420
 	am43x-epos-evm.dtb \
 	dra7-evm.dtb
 dtb-$(CONFIG_ARCH_ORION5X) += orion5x-lacie-ethernet-disk-mini-v2.dtb
-dtb-$(CONFIG_ARCH_OXNAS) += ox820-pogoplug-pro.dtb
+dtb-$(CONFIG_ARCH_OXNAS) += ox820-pogoplug-pro.dtb \
+	ox820-pogoplug-v3.dtb \
+	ox820-stg212.dtb \
+	ox820-kd20.dtb
 dtb-$(CONFIG_ARCH_PRIMA2) += prima2-evb.dtb
 dtb-$(CONFIG_ARCH_U8500) += ste-snowball.dtb \
 	ste-hrefprev60-stuib.dtb \
